From 76ad3dd2c30c16cec8cbc79db3240418e5d26614 Mon Sep 17 00:00:00 2001
From: Alex Gaynor <alex.gaynor@gmail.com>
Date: Sun, 19 Apr 2026 16:38:24 -0400
Subject: [PATCH] Make OpenSSL ASN.1 string types opaque for OpenSSL 4 compat
 (#14698)

Upstream-Status: Backport [https://github.com/pyca/cryptography/commit/76ad3dd2c]
Signed-off-by: Alex Gaynor <alex.gaynor@gmail.com>

--- a/src/_cffi_src/openssl/asn1.py
+++ b/src/_cffi_src/openssl/asn1.py
@@ -13,18 +13,11 @@ typedef int... time_t;

 typedef ... ASN1_INTEGER;

-struct asn1_string_st {
-    int length;
-    int type;
-    unsigned char *data;
-    long flags;
-};
-
-typedef struct asn1_string_st ASN1_OCTET_STRING;
-typedef struct asn1_string_st ASN1_IA5STRING;
-typedef struct asn1_string_st ASN1_TIME;
+typedef ... ASN1_OCTET_STRING;
+typedef ... ASN1_IA5STRING;
+typedef ... ASN1_TIME;
 typedef ... ASN1_OBJECT;
-typedef struct asn1_string_st ASN1_STRING;
+typedef ... ASN1_STRING;
 typedef ... ASN1_GENERALIZEDTIME;
 typedef ... ASN1_ENUMERATED;

