--- a/arch/arm64/boot/dts/amlogic/dreamone.dts	2021-10-29 12:55:38.000000000 +0200
+++ b/arch/arm64/boot/dts/amlogic/dreamone.dts	2026-07-04 22:06:09.000000000 +0200
@@ -77,14 +77,14 @@
 		logo_reserved:linux,meson-fb {
 			compatible = "shared-dma-pool";
 			reusable;
-			size = <0x0 0x800000>;
+			reg = <0x0 0x7f800000 0x0 0x800000
+			       0x0 0x7f000000 0x0 0x800000>;
 			alignment = <0x0 0x400000>;
-			alloc-ranges = <0x0 0x7f800000 0x0 0x800000>;
 		};
 		ion_cma_reserved:linux,ion-dev {
 			compatible = "shared-dma-pool";
 			reusable;
-			size = <0x0 0x20000000>;
+			size = <0x0 0x30000000>;
 			alignment = <0x0 0x400000>;
 		};
 
@@ -890,10 +890,12 @@
 &meson_fb {
 	status = "okay";
 	display_size_default = <1920 1080 1920 2160 32>;
-	mem_size = <0x00800000 0x1800000 0x100000 0x100000 0x800000>;
+	mem_size = <0x00800000 0x4b80000 0x100000 0x100000 0x800000>;
 	logo_addr = "0x7f800000";
 	mem_alloc = <1>;
 	pxp_mode = <0>; /** 0:normal mode 1:pxp mode */
+	4k2k_fb = <1>;
+	scale_mode = <1>;
 };
 
 &pwm_ab {
