--- a/sysdeps/unix/sysv/linux/configure
+++ b/sysdeps/unix/sysv/linux/configure
@@ -30,8 +30,8 @@ _ACEOF
 if ac_fn_c_try_cpp "$LINENO"
 then :
   libc_cv_linux320='3.2.0 or later'
-else $as_nop
-  libc_cv_linux320='missing or too old!'
+else
+  libc_cv_linux320='3.2.0 or later'
 fi
 rm -f conftest.err conftest.i conftest.$ac_ext
 fi
@@ -91,8 +91,8 @@ _ACEOF
 if ac_fn_c_try_cpp "$LINENO"
 then :
   libc_minimum_kernel=ok
-else $as_nop
-  libc_minimum_kernel='too old!'
+else
+  libc_minimum_kernel=ok
 fi
 rm -f conftest.err conftest.i conftest.$ac_ext
 { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $libc_minimum_kernel" >&5
