Index: m2crypto-0.18.2/setup.py
===================================================================
--- m2crypto-0.18.2.orig/setup.py
+++ m2crypto-0.18.2/setup.py
@@ -38,7 +38,7 @@ class _M2CryptoBuildExt(build_ext.build_
             self.openssl = 'c:\\pkg'
         else:
             self.libraries = ['ssl', 'crypto']
-            self.openssl = '/usr'
+            self.openssl = os.environ.get( "STAGING_DIR" )
        
     
     def finalize_options(self):
@@ -47,8 +47,8 @@ class _M2CryptoBuildExt(build_ext.build_
 
         build_ext.build_ext.finalize_options(self)
 
-        opensslIncludeDir = os.path.join(self.openssl, 'include')
-        opensslLibraryDir = os.path.join(self.openssl, 'lib')
+        opensslIncludeDir = os.environ.get( "STAGING_INCDIR" )
+        opensslLibraryDir = os.environ.get( "STAGING_LIBDIR" )
         
         self.swig_opts = ['-I%s' % i for i in self.include_dirs + \
                           [opensslIncludeDir]]
