Index: amplayer/Makefile.mk
===================================================================
--- amplayer.orig/Makefile.mk
+++ amplayer/Makefile.mk
@@ -1,20 +1,20 @@
 
 ROOTFS?=$(PREFIX)
 
-CROSS=$(CC:%gcc=%)
+#CROSS=$(CC:%gcc=%)
 
 QUITE_CMD?=no
 
 INSTALL_FLAGS=-m 755
 
-CPP=$(CROSS)g++
-AS=$(CROSS)as
-AR=$(CROSS)ar
-LD=$(CROSS)ld
-NM=$(CROSS)nm
-STRIP=$(CROSS)strip
-OBJCOPY=$(CROSS)objcopy
-export CC CPP AS AR LD NM STRIP OBJCOPY
+#CPP=$(CROSS)g++
+#AS=$(CROSS)as
+#AR=$(CROSS)ar
+#LD=$(CROSS)ld
+#NM=$(CROSS)nm
+#STRIP=$(CROSS)strip
+#OBJCOPY=$(CROSS)objcopy
+#export CC CPP AS AR LD NM STRIP OBJCOPY
 
 TOPDIR=$(shell pwd)
 SRC?=$(TOPDIR)
