--- a/container/container_ffmpeg.c
+++ b/container/container_ffmpeg.c
@@ -972,7 +972,7 @@ static void FFMPEGThread(Context_t *context)
                 }
                 
                 pcmPrivateData_t pcmExtradata;
-                pcmExtradata.channels              = get_codecpar(audioTrack->stream)->channels;
+                pcmExtradata.channels              = get_codecpar(audioTrack->stream)->ch_layout.nb_channels;
                 pcmExtradata.bits_per_coded_sample = get_codecpar(audioTrack->stream)->bits_per_coded_sample;
                 pcmExtradata.sample_rate           = get_codecpar(audioTrack->stream)->sample_rate;
                 pcmExtradata.bit_rate              = get_codecpar(audioTrack->stream)->bit_rate;
@@ -1112,13 +1112,13 @@ static void FFMPEGThread(Context_t *context)
                             }
 
                             swr = swr_alloc();
-                            out_channels = c->channels;
+                            out_channels = c->ch_layout.nb_channels;
 
-                            if (c->channel_layout == 0)
+                            if (&c->ch_layout == 0)
                             {
-                                c->channel_layout = av_get_default_channel_layout( c->channels );
+                                av_channel_layout_default(&c->ch_layout,  c->ch_layout.nb_channels );
                             }
-                            out_channel_layout = c->channel_layout;
+                            out_channel_layout = (uint64_t)&c->ch_layout;
                             
                             uint8_t downmix = stereo_software_decoder && out_channels > 2 ? 1 : 0;
 #ifdef __sh__
@@ -1134,7 +1134,7 @@ static void FFMPEGThread(Context_t *context)
                                 out_channels = 2;
                             }
                             
-                            av_opt_set_int(swr, "in_channel_layout",	c->channel_layout,	0);
+                            av_opt_set_int(swr, "in_channel_layout",	(uint64_t)&c->ch_layout,0);
                             av_opt_set_int(swr, "out_channel_layout",	out_channel_layout,	0);
                             av_opt_set_int(swr, "in_sample_rate",		c->sample_rate,		0);
                             av_opt_set_int(swr, "out_sample_rate",		out_sample_rate,	0);
@@ -1173,7 +1173,7 @@ static void FFMPEGThread(Context_t *context)
                         
                         //////////////////////////////////////////////////////////////////////
                         // Update pcmExtradata according to decode parameters
-                        pcmExtradata.channels              = av_get_channel_layout_nb_channels(out_channel_layout);
+                        pcmExtradata.channels              = av_popcount64(out_channel_layout);
                         pcmExtradata.bits_per_coded_sample = 16;
                         pcmExtradata.sample_rate           = out_sample_rate;
                         // The data described by the sample format is always in native-endian order
@@ -2336,7 +2336,7 @@ int32_t container_ffmpeg_update_tracks(Context_t *context, char *filename, int32
 
                             int32_t object_type = 2; // LC
                             int32_t sample_index = aac_get_sample_rate_index(get_codecpar(stream)->sample_rate);
-                            int32_t chan_config = get_chan_config(get_codecpar(stream)->channels);
+                            int32_t chan_config = get_chan_config(get_codecpar(stream)->ch_layout.nb_channels);
                             ffmpeg_printf(1,"aac object_type %d\n", object_type);
                             ffmpeg_printf(1,"aac sample_index %d\n", sample_index);
                             ffmpeg_printf(1,"aac chan_config %d\n", chan_config);
