diff --git a/arch/mips/bmips/setup.c b/arch/mips/bmips/setup.c
index 3269f4f..c5a2a90 100644
--- a/arch/mips/bmips/setup.c
+++ b/arch/mips/bmips/setup.c
@@ -121,7 +121,7 @@ void __init prom_free_prom_memory(void)
 
 const char *get_system_type(void)
 {
-	return "Generic BMIPS kernel";
+	return "BCM7xxx Generic BMIPS kernel";
 }
 
 #define REG_BCM7346_TIMER	((void __iomem *)CKSEG1ADDR(0x104067c0))
