From fd087e008342ae400403a1b5017bccc40bbcce8b Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Sat, 5 May 2018 07:19:51 +0200
Subject: [PATCH] rmap


diff --git a/mm/rmap.c b/mm/rmap.c
index 068522d8..8da644c2 100644
--- a/mm/rmap.c
+++ b/mm/rmap.c
@@ -526,7 +526,7 @@ __vma_address(struct page *page, struct vm_area_struct *vma)
 inline unsigned long
 vma_address(struct page *page, struct vm_area_struct *vma)
 {
-	unsigned long address = __vma_address(page, vma);
+	unsigned long address = vma_address(page, vma);
 
 	/* page should be within @vma mapping range */
 	VM_BUG_ON(address < vma->vm_start || address >= vma->vm_end);
-- 
2.17.0

