From 491f2a6f824aefe21a83dcd82ac43213e224069c Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Sat, 5 May 2018 11:25:09 +0200
Subject: [PATCH] rtnetlink


diff --git a/net/core/rtnetlink.c b/net/core/rtnetlink.c
index cf67144d..8f8fae9e 100644
--- a/net/core/rtnetlink.c
+++ b/net/core/rtnetlink.c
@@ -1144,11 +1144,11 @@ static const struct nla_policy ifla_info_policy[IFLA_INFO_MAX+1] = {
 	[IFLA_INFO_DATA]	= { .type = NLA_NESTED },
 };
 
-static const struct nla_policy ifla_vfinfo_policy[IFLA_VF_INFO_MAX+1] = {
+const struct nla_policy ifla_vfinfo_policy[IFLA_VF_INFO_MAX+1] = {
 	[IFLA_VF_INFO]		= { .type = NLA_NESTED },
 };
 
-static const struct nla_policy ifla_vf_policy[IFLA_VF_MAX+1] = {
+const struct nla_policy ifla_vf_policy[IFLA_VF_MAX+1] = {
 	[IFLA_VF_MAC]		= { .type = NLA_BINARY,
 				    .len = sizeof(struct ifla_vf_mac) },
 	[IFLA_VF_VLAN]		= { .type = NLA_BINARY,
-- 
2.17.0

