From 960786bfb797d0b6587a010829741b1c75a50c36 Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Sat, 5 May 2018 12:05:59 +0200
Subject: [PATCH] fscache object


diff --git a/fs/fscache/object.c b/fs/fscache/object.c
index 53d35c50..b2e23144 100644
--- a/fs/fscache/object.c
+++ b/fs/fscache/object.c
@@ -80,7 +80,7 @@ static WORK_STATE(INIT_OBJECT,		"INIT", fscache_initialise_object);
 static WORK_STATE(PARENT_READY,		"PRDY", fscache_parent_ready);
 static WORK_STATE(ABORT_INIT,		"ABRT", fscache_abort_initialisation);
 static WORK_STATE(LOOK_UP_OBJECT,	"LOOK", fscache_look_up_object);
-static WORK_STATE(CREATE_OBJECT,	"CRTO", fscache_look_up_object);
+WORK_STATE(CREATE_OBJECT,	"CRTO", fscache_look_up_object);
 static WORK_STATE(OBJECT_AVAILABLE,	"AVBL", fscache_object_available);
 static WORK_STATE(JUMPSTART_DEPS,	"JUMP", fscache_jumpstart_dependents);
 
-- 
2.17.0

