From f98a47f32f40803b479882387f7ea0b0b09a8c14 Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Sat, 5 May 2018 14:17:42 +0200
Subject: [PATCH] xattr


diff --git a/fs/hfsplus/xattr.c b/fs/hfsplus/xattr.c
index 3c6136f9..ab85884d 100644
--- a/fs/hfsplus/xattr.c
+++ b/fs/hfsplus/xattr.c
@@ -139,7 +139,7 @@ static void hfsplus_init_header_node(struct inode *attr_file,
 	char *bmp;
 	u32 used_nodes;
 	u32 used_bmp_bytes;
-	loff_t tmp;
+	u64 tmp;
 
 	hfs_dbg(ATTR_MOD, "init_hdr_attr_file: clump %u, node_size %u\n",
 				clump_size, node_size);
-- 
2.17.0

