From de8b35f7a7011daec2d3c04535d6d0d73e9620c7 Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Sat, 5 May 2018 12:25:43 +0200
Subject: [PATCH] net-sysfs


diff --git a/net/core/net-sysfs.c b/net/core/net-sysfs.c
index f3edf963..91dd2dbd 100644
--- a/net/core/net-sysfs.c
+++ b/net/core/net-sysfs.c
@@ -27,7 +27,7 @@
 
 #ifdef CONFIG_SYSFS
 static const char fmt_hex[] = "%#x\n";
-static const char fmt_long_hex[] = "%#lx\n";
+const char fmt_long_hex[] = "%#lx\n";
 static const char fmt_dec[] = "%d\n";
 static const char fmt_udec[] = "%u\n";
 static const char fmt_ulong[] = "%lu\n";
-- 
2.17.0

