From 187b5e4ac61fbc1e504bde7caee6f4d2b21364b7 Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Sat, 5 May 2018 15:24:22 +0200
Subject: [PATCH] arp


diff --git a/net/ipv4/arp.c b/net/ipv4/arp.c
index 7808093c..c034e852 100644
--- a/net/ipv4/arp.c
+++ b/net/ipv4/arp.c
@@ -149,6 +149,7 @@ static const struct neigh_ops arp_direct_ops = {
 	.connected_output =	neigh_direct_output,
 };
 
+#if 0
 static const struct neigh_ops arp_broken_ops = {
 	.family =		AF_INET,
 	.solicit =		arp_solicit,
@@ -156,6 +157,7 @@ static const struct neigh_ops arp_broken_ops = {
 	.output =		neigh_compat_output,
 	.connected_output =	neigh_compat_output,
 };
+#endif
 
 struct neigh_table arp_tbl = {
 	.family		= AF_INET,
-- 
2.17.0

