From 24ce6a020fbbb3399d1be87d06d66a5987e6b8a2 Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Sat, 5 May 2018 16:06:43 +0200
Subject: [PATCH] ehci ohci brcm usb_hcd_brcm_match


diff --git a/drivers/usb/host/ehci-brcm.c b/drivers/usb/host/ehci-brcm.c
index 5ce4ec3b..75c2dd69 100644
--- a/drivers/usb/host/ehci-brcm.c
+++ b/drivers/usb/host/ehci-brcm.c
@@ -289,10 +289,12 @@ static int ehci_hcd_brcm_remove(struct platform_device *pdev)
 	return brcm_usb_remove(pdev);
 }
 
+#if 0
 static const struct of_device_id usb_hcd_brcm_match[] = {
 	{ .compatible = "brcm,ehci-brcm" },
 	{},
 };
+#endif
 
 static struct platform_driver ehci_hcd_brcm_driver = {
 	.probe = ehci_hcd_brcm_probe,
diff --git a/drivers/usb/host/ohci-brcm.c b/drivers/usb/host/ohci-brcm.c
index 9e65d601..e88051d5 100644
--- a/drivers/usb/host/ohci-brcm.c
+++ b/drivers/usb/host/ohci-brcm.c
@@ -169,10 +169,12 @@ static int ohci_hcd_brcm_remove(struct platform_device *pdev)
 	return brcm_usb_remove(pdev);
 }
 
+#if 0
 static const struct of_device_id usb_hcd_brcm_match[] = {
 	{ .compatible = "brcm,ohci-brcm" },
 	{},
 };
+#endif
 
 static struct platform_driver ohci_hcd_brcm_driver = {
 	.probe		= ohci_hcd_brcm_probe,
-- 
2.17.0

