From ef5160b76bf68fed4c4c5493215ef3ad33d93ad6 Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Thu, 3 May 2018 11:56:36 +0200
Subject: [PATCH] sortextable define relocs_size


diff --git a/scripts/sortextable.h b/scripts/sortextable.h
index 8fac3fd6..ba870042 100644
--- a/scripts/sortextable.h
+++ b/scripts/sortextable.h
@@ -103,7 +103,7 @@ do_func(Elf_Ehdr *ehdr, char const *const fname, table_sort_t custom_sort)
 	Elf_Sym *sort_needed_sym;
 	Elf_Shdr *sort_needed_sec;
 	Elf_Rel *relocs = NULL;
-	int relocs_size;
+	int relocs_size = 0;
 	uint32_t *sort_done_location;
 	const char *secstrtab;
 	const char *strtab;
-- 
2.17.0

