From 374c4a7a29641bcc8eb363047bf4c0061ee949b3 Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Wed, 9 May 2018 07:08:46 +0200
Subject: [PATCH] mac80211_hwsim


diff --git a/drivers/net/wireless/mac80211_hwsim.c b/drivers/net/wireless/mac80211_hwsim.c
index a1b32ee9..ddd24cb6 100644
--- a/drivers/net/wireless/mac80211_hwsim.c
+++ b/drivers/net/wireless/mac80211_hwsim.c
@@ -450,11 +450,16 @@ static void mac80211_hwsim_set_tsf(struct ieee80211_hw *hw,
 	struct mac80211_hwsim_data *data = hw->priv;
 	u64 now = mac80211_hwsim_get_tsf(hw, vif);
 	u32 bcn_int = data->beacon_int;
-	s64 delta = tsf - now;
+	u64 delta = abs64(tsf - now);
 
-	data->tsf_offset += delta;
 	/* adjust after beaconing with new timestamp at old TBTT */
+	if (tsf > now) {
+	data->tsf_offset += delta;
 	data->bcn_delta = do_div(delta, bcn_int);
+	} else {
+	data->tsf_offset -= delta;
+	data->bcn_delta = -do_div(delta, bcn_int);
+	}
 }
 
 static void mac80211_hwsim_monitor_rx(struct ieee80211_hw *hw,
-- 
2.17.0

