From 6c15814254bf57460d6ce1099263a88a6f0e964d Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Wed, 9 May 2018 07:45:14 +0200
Subject: [PATCH 4/6] r820t


diff --git a/drivers/media/tuners/r820t.c b/drivers/media/tuners/r820t.c
index d9ee43fa..cbec91e3 100644
--- a/drivers/media/tuners/r820t.c
+++ b/drivers/media/tuners/r820t.c
@@ -342,11 +342,11 @@ static int r820t_xtal_capacitor[][2] = {
  *	http://steve-m.de/projects/rtl-sdr/gain_measurement/r820t/
  */
 
-static const int r820t_lna_gain_steps[]  = {
+const int r820t_lna_gain_steps[]  = {
 	0, 9, 13, 40, 38, 13, 31, 22, 26, 31, 26, 14, 19, 5, 35, 13
 };
 
-static const int r820t_mixer_gain_steps[]  = {
+const int r820t_mixer_gain_steps[]  = {
 	0, 5, 10, 10, 19, 9, 10, 25, 17, 10, 8, 16, 13, 6, 3, -8
 };
 
-- 
2.17.0

