From 4ed80c390b9f360e35ffd2ff23f64567455ecb4f Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Wed, 9 May 2018 08:10:36 +0200
Subject: [PATCH] tbs-usb


diff --git a/drivers/media/usb/dvb-usb/tbs-usb.c b/drivers/media/usb/dvb-usb/tbs-usb.c
index f142be3b..3b6b112d 100644
--- a/drivers/media/usb/dvb-usb/tbs-usb.c
+++ b/drivers/media/usb/dvb-usb/tbs-usb.c
@@ -739,7 +739,7 @@ static int tbsusbci_init(struct dvb_usb_adapter *a)
 	return 0;
 
 err:	err("Cannot initialize CI: Error %d.", ret);
-	memset(&state->ca, 0, sizeof(state->ca));
+//	memset(&state->ca, 0, sizeof(state->ca));
 	return ret;
 }
 
-- 
2.17.0

