From 66fe5886d6c819f8a464b02208d6563a3486e11e Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Wed, 9 May 2018 08:22:03 +0200
Subject: [PATCH 3/3] usbvision-core


diff --git a/drivers/media/usb/usbvision/usbvision-core.c b/drivers/media/usb/usbvision/usbvision-core.c
index 816b1cff..bff586ab 100644
--- a/drivers/media/usb/usbvision/usbvision-core.c
+++ b/drivers/media/usb/usbvision/usbvision-core.c
@@ -88,10 +88,10 @@ MODULE_PARM_DESC(adjust_y_offset, "adjust Y offset display [core]");
 #define DBG_SCRATCH	(1 << 4)
 #define DBG_FUNC	(1 << 5)
 
-static const int max_imgwidth = MAX_FRAME_WIDTH;
-static const int max_imgheight = MAX_FRAME_HEIGHT;
-static const int min_imgwidth = MIN_FRAME_WIDTH;
-static const int min_imgheight = MIN_FRAME_HEIGHT;
+const int max_imgwidth = MAX_FRAME_WIDTH;
+const int max_imgheight = MAX_FRAME_HEIGHT;
+const int min_imgwidth = MIN_FRAME_WIDTH;
+const int min_imgheight = MIN_FRAME_HEIGHT;
 
 /* The value of 'scratch_buf_size' affects quality of the picture
  * in many ways. Shorter buffers may cause loss of data when client
-- 
2.17.0

