From 87717aaf12f38c2230d844e96cfeec5d85fe14bf Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Wed, 9 May 2018 11:37:48 +0200
Subject: [PATCH] ieee80211


diff --git a/drivers/staging/rtl8712/ieee80211.c b/drivers/staging/rtl8712/ieee80211.c
index cc68d974..3b479835 100644
--- a/drivers/staging/rtl8712/ieee80211.c
+++ b/drivers/staging/rtl8712/ieee80211.c
@@ -78,7 +78,7 @@ uint r8712_is_cckrates_included(u8 *rate)
 		if ((((rate[i]) & 0x7f) == 2) || (((rate[i]) & 0x7f) == 4) ||
 		    (((rate[i]) & 0x7f) == 11) || (((rate[i]) & 0x7f) == 22))
 			return true;
-			i++;
+		i++;
 		}
 		return false;
 }
-- 
2.17.0

