--- 2011_0107_RT3070_RT3370_Linux_STA_v2.3.0.4/include/rtmp_def.h.orig	2011-24-07 15:00:00.000000000 +0200
+++ 2011_0107_RT3070_RT3370_Linux_STA_v2.3.0.4/include/rtmp_def.h	2011-24-07 15:00:00.000000000 +0200
@@ -1497,8 +1497,8 @@
 #define SET_ENTRY_CLIWDS(_x)   CLIENT_STATUS_SET_FLAG((_x), fCLIENT_STATUS_CLI_WDS)
 #endif // CLIENT_WDS //
 
-#define INF_MAIN_DEV_NAME		"ra"
-#define INF_MBSSID_DEV_NAME		"ra"
+#define INF_MAIN_DEV_NAME		"wlan"
+#define INF_MBSSID_DEV_NAME		"wlan"
 #define INF_WDS_DEV_NAME		"wds"
 #define INF_APCLI_DEV_NAME		"apcli"
 #define INF_MESH_DEV_NAME		"mesh"