From b8b52ac137fda5f5e1dbdf13434629431120ea05 Mon Sep 17 00:00:00 2001
From: Andy Blackburn <andrew.blackburn644@btinternet.com>
Date: Fri, 12 Apr 2013 19:34:25 +0100

---
 Makefile.am | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 09343ff1..15891a7f 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -22,7 +22,6 @@ SUBDIRS = \
 	dreammediathek \
 	dvdbackup \
 	easyinfo \
-	easymedia \
 	ecasa \
 	eibox \
 	emailclient \
@@ -59,13 +58,11 @@ SUBDIRS = \
 	ofdb \
 	orfat \
 	orfteletext \
-	partnerbox \
 	permanentclock \
 	pipservicerelation \
 	podcast \
 	porncenter \
 	pushservice \
-	quickbutton \
 	reconstructapsc \
 	reftopiconname \
 	remotetimer \
@@ -129,7 +126,6 @@ SUBDIRS += \
 	mytube \
 	yttrailer \
 	webinterface \
-	menusort \
 	pluginhider \
 	pluginsort 
 endif
