From 06bf0b8866a8abd0b582000dff08776599509682 Mon Sep 17 00:00:00 2001
From: Captain <captain.onboard@web.de>
Date: Mon, 22 Jun 2015 18:42:14 +0200
Subject: [PATCH] [WETEK] add support AIR mouse and switch to FB2

---
 tuxtxt/tuxtxt.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/tuxtxt.c b/tuxtxt.c
index 92ef7a2..adb4e39 100644
--- a/tuxtxt.c
+++ b/tuxtxt.c
@@ -204,9 +204,9 @@ int tuxtxt_run_ui(int pid, int demux)
 	tuxtxt_cache.vtxtpid = pid;
 
 	/* open Framebuffer */
-	if ((renderinfo.fb=open("/dev/fb/0", O_RDWR)) == -1)
+	if ((renderinfo.fb=open("/dev/fb/2", O_RDWR)) == -1)
 	{
-		perror("TuxTxt <open /dev/fb/0>");
+		perror("TuxTxt <open /dev/fb/2>");
 		return 0;
 	}
 	rc[0] = rc[1] =-1;
@@ -225,7 +225,7 @@ int tuxtxt_run_ui(int pid, int demux)
 		}
 		if (ioctl(rc[rc_num], EVIOCGNAME(128), tmp) < 0)
 			perror("EVIOCGNAME failed");
-		if (!strstr(tmp, "remote control"))
+		if (!strstr(tmp, "HBGIC"))
 		{
 			close(rc[rc_num]);
 			rc[rc_num] = -1;
-- 
1.9.5.msysgit.0

