diff --git a/plugin/plugin.py b/plugin/plugin.py
index 1bb9a38..5284053 100644
--- a/plugin/plugin.py
+++ b/plugin/plugin.py
@@ -44,6 +44,12 @@
 			self.setOSDAlpha(config.osd.alpha_teletext)
 		except:
 			pass
+		try:
+			from Components.UsageConfig import patchTuxtxtConfFile
+			if config.usage.tuxtxt_ConfFileHasBeenPatched.value == False:
+				patchTuxtxtConfFile(config.usage.tuxtxt_ConfFileHasBeenPatched)
+		except:
+			pass
 		self.skin = ShellStarter.skin
 		Screen.__init__(self, session)
 		self.session = session
