diff --git a/plugin/plugin.py b/plugin/plugin.py
index 29c89a5..1b959da 100644
--- a/plugin/plugin.py
+++ b/plugin/plugin.py
@@ -46,6 +46,12 @@ class ShellStarter(Screen):
 			self.setOSDAlpha(config.osd.alpha_teletext)
 		except:
 			pass
+		try:
+			from Components.UsageConfig import patchTuxtxtConfFile
+			if config.usage.tuxtxt_ConfFileHasBeenPatched.value == False:
+				patchTuxtxtConfFile(config.usage.tuxtxt_ConfFileHasBeenPatched)
+		except:
+			pass
 		self.skin = ShellStarter.skin
 		Screen.__init__(self, session)
 		self.session = session
