diff -ruN a/addons/skin.estouchy/xml/Home.xml b/addons/skin.estouchy/xml/Home.xml
--- a/addons/skin.estouchy/xml/Home.xml	2019-08-31 14:05:38.238692000 +0200
+++ b/addons/skin.estouchy/xml/Home.xml	2019-10-23 18:43:01.274036000 +0200
@@ -183,7 +183,7 @@
 				<content>
 					<item>
 						<label>13012</label>
-						<onclick>Shutdown</onclick>
+						<onclick>XBMC.Quit()</onclick>
 						<icon>icon_button_shutdown.png</icon>
 					</item>
 					<item>
diff -ruN a/addons/skin.estuary/xml/Home.xml b/addons/skin.estuary/xml/Home.xml
--- a/addons/skin.estuary/xml/Home.xml	2019-10-24 12:05:17.617794000 +0200
+++ b/addons/skin.estuary/xml/Home.xml	2019-10-24 12:07:16.962378000 +0200
@@ -1005,7 +1005,7 @@
 					<align>justify</align>
 					<include content="IconButton">
 						<param name="control_id" value="804" />
-						<param name="onclick" value="ActivateWindow(shutdownmenu)" />
+						<param name="onclick" value="XBMC.Quit()" />
 						<param name="icon" value="icons/power.png" />
 						<param name="label" value="$LOCALIZE[33060]" />
 					</include>
diff -ruN a/system/settings/settings.xml b/system/settings/settings.xml
--- a/system/settings/settings.xml	2019-08-31 14:05:38.902696000 +0200
+++ b/system/settings/settings.xml	2019-10-23 18:49:42.064354000 +0200
@@ -493,7 +493,7 @@
           <control type="spinner" format="integer" delayed="true"/>
         </setting>
         <setting id="subtitles.font" type="string" label="14089" help="36185">
-          <level>1</level>
+          <level>3</level>
           <default>arial.ttf</default>
           <constraints>
             <options>fonts</options>
@@ -501,7 +501,7 @@
           <control type="list" format="string" />
         </setting>
         <setting id="subtitles.charset" type="string" parent="subtitles.font" label="735" help="36189">
-          <level>1</level>
+          <level>3</level>
           <default>DEFAULT</default>
           <constraints>
             <options>charsets</options>
@@ -2444,7 +2444,7 @@
         <setting id="audiooutput.processquality" type="integer" label="13505" help="36169">
           <requirement>HAS_AE_QUALITY_LEVELS</requirement>
           <level>2</level>
-          <default>30</default> <!-- AE_QUALITY_MID -->
+          <default>20</default> <!-- AE_QUALITY_MID -->
           <constraints>
             <options>aequalitylevels</options>
           </constraints>
@@ -2462,7 +2462,7 @@
         </setting>
         <setting id="audiooutput.samplerate" type="integer" label="458" help="36523">
           <level>2</level>
-          <default>48000</default>
+          <default>44100</default>
           <dependencies>
             <dependency type="visible">
               <and>
@@ -2474,10 +2474,6 @@
           <constraints>
             <options>
               <option label="34124">44100</option>
-              <option label="34125">48000</option>
-              <option label="34126">88200</option>
-              <option label="34127">96000</option>
-              <option label="34128">192000</option>
               <option label="34129">352800</option>
               <option label="34130">384000</option>
             </options>
@@ -3247,7 +3243,7 @@
       <group id="2" label="14270">
         <setting id="locale.country" type="string" label="14271" help="36115">
           <level>0</level>
-          <default>USA (12h)</default>
+          <default>Central Europe</default>
           <constraints>
             <options>regions</options>
           </constraints>
diff -ruN a/tools/depends/target/libdvdcss/Makefile b/tools/depends/target/libdvdcss/Makefile
--- a/tools/depends/target/libdvdcss/Makefile	2019-08-31 14:05:38.998697000 +0200
+++ b/tools/depends/target/libdvdcss/Makefile	2019-10-23 18:57:55.747119000 +0200
@@ -3,7 +3,7 @@
 DEPS= DVDCSS-VERSION Makefile
 
 ifeq ($(CROSS_COMPILING), yes)
-  DEPS += ../../Makefile.include
+#  DEPS += ../../Makefile.include
 else
   ROOT_DIR := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
   PREFIX := $(ROOT_DIR)/dvdcss-install
@@ -19,7 +19,7 @@
 ARCHIVE_TOOL := tar --strip-components=1 -xf
 
 # configuration settings
-config = --prefix=$(PREFIX) --disable-shared --enable-static --with-pic
+config = --build=$(BUILD_SYS) --host=$(HOST_SYS) --prefix=$(PREFIX) --disable-shared --enable-static --with-pic
 
 GITREV=$(shell git ls-remote $(BASE_URL) $(VERSION) | awk '{print $$1}')
 ifeq ($(GITREV),)
diff -ruN a/tools/depends/target/libdvdnav/Makefile b/tools/depends/target/libdvdnav/Makefile
--- a/tools/depends/target/libdvdnav/Makefile	2019-08-31 14:05:38.998697000 +0200
+++ b/tools/depends/target/libdvdnav/Makefile	2019-10-23 18:59:03.343493000 +0200
@@ -3,11 +3,11 @@
 DEPS = DVDNAV-VERSION Makefile
 
 # configuration settings
-config = --prefix=$(PREFIX) --disable-shared --enable-static --with-pic
+config = --build=$(BUILD_SYS) --host=$(HOST_SYS) --prefix=$(PREFIX) --disable-shared --enable-static --with-pic
 EXTRA_CFLAGS = -D_XBMC -DHAVE_DVDCSS_DVDCSS_H
 
 ifeq ($(CROSS_COMPILING), yes)
-  DEPS += ../../Makefile.include
+#  DEPS += ../../Makefile.include
 else
   ROOT_DIR := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
   PREFIX := $(ROOT_DIR)/dvdnav-install
diff -ruN a/tools/depends/target/libdvdread/Makefile b/tools/depends/target/libdvdread/Makefile
--- a/tools/depends/target/libdvdread/Makefile	2019-08-31 14:05:38.998697000 +0200
+++ b/tools/depends/target/libdvdread/Makefile	2019-10-23 19:00:36.024006000 +0200
@@ -3,11 +3,12 @@
 DEPS = DVDREAD-VERSION Makefile
 
 # configuration settings
-config = --prefix=$(PREFIX) --disable-shared --enable-static --with-pic --with-libdvdcss
+#config = --prefix=$(PREFIX) --disable-shared --enable-static --with-pic --with-libdvdcss
+config = --build=$(BUILD_SYS) --host=$(HOST_SYS) --prefix=$(PREFIX) --disable-shared --enable-static --with-pic
 EXTRA_CFLAGS = -D_XBMC -DHAVE_DVDCSS_DVDCSS_H
 
 ifeq ($(CROSS_COMPILING), yes)
-  DEPS += ../../Makefile.include
+#  DEPS += ../../Makefile.include
 else
   ROOT_DIR := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
   PREFIX := $(ROOT_DIR)/dvdread-install
diff -ruN a/tools/Linux/kodi.sh.in b/tools/Linux/kodi.sh.in
--- a/tools/Linux/kodi.sh.in	2019-10-24 12:08:58.610873000 +0200
+++ b/tools/Linux/kodi.sh.in	2019-11-27 22:58:04.742961285 +0100
@@ -26,16 +26,44 @@
 datarootdir="@datarootdir@"
 LIBDIR="@libdir@"
 APP_BINARY=$LIBDIR/${bin_name}/@APP_BINARY@
-CRASHLOG_DIR=${CRASHLOG_DIR:-$HOME}
-KODI_DATA=${KODI_DATA:-"${HOME}/.${bin_name}"} # mapped to special://home/
+#CRASHLOG_DIR=${CRASHLOG_DIR:-$HOME}
+#KODI_DATA=${KODI_DATA:-"${HOME}/.${bin_name}"} # mapped to special://home/
 WINDOWING=${WINDOWING:-"auto"}
 
+export KODI_TEMP="/tmp/kodi"
+CRASHLOG_DIR=${CRASHLOG_DIR:-$KODI_TEMP}
+
 # Workaround for high CPU load with nvidia GFX
 export __GL_YIELD=USLEEP
 
 # Fix wasting RAM due to fragmentation
 export MALLOC_MMAP_THRESHOLD_=131072
 
+check_media_hdd() {
+    [ -e /proc/mounts ] || return 0
+
+    exec 9<&0 < /proc/mounts
+    while read DEV MTPT REST; do
+        case $DEV in
+            /dev/sd[a-z][0-9])
+            if [ "$MTPT" == "/media/hdd" ]; then
+                export HOME="$MTPT"
+                break
+            fi
+            ;;
+	    /dev/mmcblk[0-9]p[0-9])
+            if [ "$MTPT" == "/media/mmc" ]; then
+            	export HOME="$MTPT"
+            fi
+	    ;;
+        esac
+    done
+    exec 0<&9 9<&-
+}
+
+check_media_hdd
+KODI_DATA="${HOME}/.${bin_name}"
+
 # Check for some options used by this script
 while [ "$#" -gt "0" ]
 do
@@ -209,13 +237,13 @@
   LOOP=0
   ${KODI_BINARY} $SAVED_ARGS
   RET=$?
-  if [ $(( $RET == 65 )) = "1" ]
-  then # User requested to restart app
-    LOOP=1
-  elif [ $(( ($RET >= 131 && $RET <= 136) || $RET == 139 )) = "1" ]
-  then # Crashed with core dump
-    print_crash_report
-  fi
+#  if [ $(( $RET == 65 )) = "1" ]
+#  then # User requested to restart app
+#    LOOP=1
+#  elif [ $(( ($RET >= 131 && $RET <= 136) || $RET == 139 )) = "1" ]
+#  then # Crashed with core dump
+#    print_crash_report
+#  fi
 done
 
 exit $RET
diff -ruN a/xbmc/guilib/guiinfo/SystemGUIInfo.cpp b/xbmc/guilib/guiinfo/SystemGUIInfo.cpp
--- a/xbmc/guilib/guiinfo/SystemGUIInfo.cpp	2019-08-31 14:05:39.210698000 +0200
+++ b/xbmc/guilib/guiinfo/SystemGUIInfo.cpp	2019-10-23 19:09:16.841459000 +0200
@@ -57,6 +57,11 @@
 #if defined(TARGET_POSIX)
     g_cpuInfo.getTemperature(m_cpuTemp);
     m_gpuTemp = GetGPUTemperature();
+#if defined(TARGET_STB_EXTEND)
+    m_gpuTemp = m_cpuTemp;
+#else
+     m_gpuTemp = GetGPUTemperature();
+#endif
 #endif
   }
 
diff -ruN a/xbmc/input/InputManager.cpp b/xbmc/input/InputManager.cpp
--- a/xbmc/input/InputManager.cpp	2019-08-31 14:05:39.214698000 +0200
+++ b/xbmc/input/InputManager.cpp	2019-10-23 19:14:00.814560000 +0200
@@ -558,6 +558,18 @@
           useKeyboard = true;
       }
     }
+#if defined(TARGET_STB)
+    switch(key.GetVKey()) {
+      case 0x0d:
+      case 0x80:
+      case 0x81:
+      case 0x82:
+      case 0x83:
+        if (iWin == WINDOW_DIALOG_KEYBOARD)
+          useKeyboard = false;
+        break;
+    }
+#endif
     if (useKeyboard)
     {
       // use the virtualkeyboard section of the keymap, and send keyboard-specific or navigation
diff -ruN a/xbmc/utils/CPUInfo.cpp b/xbmc/utils/CPUInfo.cpp
--- a/xbmc/utils/CPUInfo.cpp	2019-08-31 14:05:39.362699000 +0200
+++ b/xbmc/utils/CPUInfo.cpp	2019-11-13 17:50:42.067459700 +0100
@@ -277,8 +277,16 @@
     m_fProcTemperature = fopen("/sys/class/hwmon/hwmon0/temp1_input", "r");
   if (m_fProcTemperature == NULL)
     m_fProcTemperature = fopen("/sys/class/thermal/thermal_zone0/temp", "r");  // On Raspberry PIs
+  if (m_fProcTemperature == NULL)
+    m_fProcTemperature = fopen("/proc/stb/fp/temp_sensor_avs", "r"); // On STBs
 
   m_fCPUFreq = fopen ("/sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq", "r");
+  
+  if (m_fCPUFreq == NULL)
+    m_fCPUFreq = fopen ("/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_cur_freq", "r");
+  if (m_fCPUFreq == NULL)  
+    m_fCPUFreq = fopen ("/sys/devices/platform/brcmstb/cpu_khz", "r");
+
   if (!m_fCPUFreq)
   {
     m_cpuInfoForFreq = true;
@@ -420,6 +428,30 @@
           }
         }
       }
+#ifdef TARGET_STB_EXTEND
+      else if (strncmp(buffer, "system type", strlen("system type"))==0)
+      {
+        char *needle = strchr(buffer, ':');
+        if (needle && strlen(needle)>3)
+        {
+          needle+=2;
+          m_cpuModel = needle;
+          StringUtils::Trim(m_cpuModel);
+          StringUtils::ToUpper(m_cpuModel);
+          m_cpuModel.insert(0, "Broadcom ");
+        }
+      }
+      else if (strncmp(buffer, "cpu model", strlen("cpu model"))==0)
+      {
+        char *needle = strchr(buffer, ':');
+        if (needle && strlen(needle)>3)
+        {
+          needle+=2;
+          m_cores[nCurrId].m_strModel = needle;
+          StringUtils::Trim(m_cores[nCurrId].m_strModel);
+        }
+      }
+#endif
     }
     fclose(fCPUInfo);
     // new socs use the sysfs soc interface to describe the hardware
@@ -628,6 +660,15 @@
 #if defined(TARGET_DARWIN_OSX)
   value = SMCGetTemperature(SMC_KEY_CPU_TEMP);
   scale = 'c';
+#elif defined(TARGET_XCORE)
+  value = 0;
+  scale = 'c';
+  FILE        *p    = fopen("/proc/stb/fp/temp_sensor_avs", "r");
+  if (p)
+  {
+    fscanf(p, "%d", &value);
+    fclose(p);
+  }
 #else
   int         ret   = 0;
   FILE        *p    = NULL;
@@ -661,7 +702,9 @@
     if (!ret)
     {
       ret = fscanf(m_fProcTemperature, "%d", &value);
+#if !defined(TARGET_EDISION)
       value = value / 1000;
+#endif
       scale = 'c';
       ret++;
     }
diff -ruN a/xbmc/view/ViewStateSettings.cpp b/xbmc/view/ViewStateSettings.cpp
--- a/xbmc/view/ViewStateSettings.cpp	2019-08-31 14:05:39.406700000 +0200
+++ b/xbmc/view/ViewStateSettings.cpp	2019-10-23 22:01:03.501161000 +0200
@@ -114,7 +114,7 @@
     if (XMLUtils::GetInt(pElement, XML_SETTINGLEVEL, settingLevel, static_cast<int>(SettingLevel::Basic), static_cast<int>(SettingLevel::Expert)))
       m_settingLevel = (SettingLevel)settingLevel;
     else
-      m_settingLevel = SettingLevel::Standard;
+      m_settingLevel = SettingLevel::Expert;
 
     const TiXmlNode* pEventLogNode = pElement->FirstChild(XML_EVENTLOG);
     if (pEventLogNode != NULL)
