--- Makefile.am	2017-01-02 20:43:58.000000000 +0100
+++ Makefile.am	2017-01-02 21:21:37.000000000 +0100
@@ -63,6 +63,7 @@ SUBDIRS = \
 	partnerbox \
 	permanentclock \
 	pipservicerelation \
+	pluginhider \
 	podcast \
 	porncenter \
 	pushservice \
@@ -90,8 +91,6 @@ SUBDIRS = \
 	weatherplugin \
 	weathercomponent \
 	weathercomponenthandler \
-	webadmin \
-	webbouqueteditor \
 	webcamviewer \
 	werbezapper \
 	zaphistorybrowser \
@@ -101,8 +100,6 @@ SUBDIRS = \
 	esame \
 	schiffe \
 	sudoku \
-	vierg \
-	bmediacenter \
 	advancedmovieselection
 
 if DREAMBOX
@@ -130,16 +127,14 @@ if TPM
 SUBDIRS += \
 	mytube \
 	yttrailer \
-	webinterface \
 	menusort \
-	pluginhider \
 	pluginsort 
 endif
 
 if PLI
 SUBDIRS += networkbrowserpli epgsearchpli
 else
-SUBDIRS += networkbrowser epgsearch
+SUBDIRS += epgsearch
 endif
 
 metadir = $(datadir)/meta
