diff --git a/Makefile.am b/Makefile.am
index d0f54e5..e535329 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -60,9 +60,9 @@ SUBDIRS = \
 	ofdb \
 	orfat \
 	orfteletext \
-	partnerbox \
 	permanentclock \
 	pipservicerelation \
+	pluginhider \
 	podcast \
 	porncenter \
 	pushservice \
@@ -86,12 +86,9 @@ SUBDIRS = \
	valixdcontrol \
	videocolorspace \
-	vlcplayer \
	vps \
	weatherplugin \
 	weathercomponent \
 	weathercomponenthandler \
-	webadmin \
-	webbouqueteditor \
 	webcamviewer \
 	werbezapper \
 	zaphistorybrowser \
@@ -102,7 +99,6 @@ SUBDIRS = \
 	schiffe \
 	sudoku \
 	vierg \
-	bmediacenter \
 	advancedmovieselection
 
 if DREAMBOX
@@ -130,9 +126,7 @@ if TPM
 SUBDIRS += \
 	mytube \
 	yttrailer \
-	webinterface \
 	menusort \
-	pluginhider \
 	pluginsort 
 endif
