--- a/drivers/mtd/devices/block2mtd.c	2013-03-08 15:04:13.272322463 +0100
+++ b/drivers/mtd/devices/block2mtd.c	2013-03-08 15:04:13.272322463 +0100
@@ -21,6 +21,7 @@
 #define ERROR(fmt, args...) printk(KERN_ERR "block2mtd: " fmt "\n" , ## args)
 #define INFO(fmt, args...) printk(KERN_INFO "block2mtd: " fmt "\n" , ## args)
 
+extern int del_mtd_device(struct mtd_info *mtd);
 
 /* Info for the block device */
 struct block2mtd_dev {
@@ -210,7 +211,7 @@
 
 
 /* FIXME: ensure that mtd->size % erase_size == 0 */
-static struct block2mtd_dev *add_device(char *devname, int erase_size)
+static struct block2mtd_dev *add_device(char *devname, int erase_size, int write_size)
 {
 	const fmode_t mode = FMODE_READ | FMODE_WRITE | FMODE_EXCL;
 	struct block_device *bdev;
@@ -261,7 +262,7 @@
 
 	dev->mtd.size = dev->blkdev->bd_inode->i_size & PAGE_MASK;
 	dev->mtd.erasesize = erase_size;
-	dev->mtd.writesize = 1;
+	dev->mtd.writesize = write_size;
 	dev->mtd.writebufsize = PAGE_SIZE;
 	dev->mtd.type = MTD_RAM;
 	dev->mtd.flags = MTD_CAP_RAM;
@@ -277,9 +278,10 @@
 		goto devinit_err;
 	}
 	list_add(&dev->list, &blkmtd_device_list);
-	INFO("mtd%d: [%s] erase_size = %dKiB [%d]", dev->mtd.index,
+	INFO("mtd%d: [%s] erase_size = %dKiB [%d] write_size = %dKiB [%d]", dev->mtd.index,
 			dev->mtd.name + strlen("block2mtd: "),
-			dev->mtd.erasesize >> 10, dev->mtd.erasesize);
+			dev->mtd.erasesize >> 10, dev->mtd.erasesize,
+			dev->mtd.writesize >> 10, dev->mtd.writesize);
 	return dev;
 
 devinit_err:
@@ -346,17 +348,45 @@
 
 #ifndef MODULE
 static int block2mtd_init_called = 0;
-static char block2mtd_paramline[80 + 12]; /* 80 for device, 12 for erase size */
+static char block2mtd_paramline[80 + 12 + 12]; /* 80 for device, 12 for erase size, 12 for writesize */
 #endif
 
+static void remove_device_by_name(const char *name)
+{
+	struct list_head *pos, *next;
+	int name_offset = strlen("block2mtd: ");
+
+	list_for_each_safe(pos, next, &blkmtd_device_list) {
+		struct block2mtd_dev *dev = list_entry(pos, typeof(*dev), list);
+		if (!strcmp(name, dev->mtd.name + name_offset)) {
+			int err;
+			block2mtd_sync(&dev->mtd);
+			err = del_mtd_device(&dev->mtd);
+			if (err == 0) {
+				INFO("mtd%d: [%s] removed",
+						dev->mtd.index, name);
+				list_del(&dev->list);
+				block2mtd_free_device(dev);
+			} else
+				INFO("mtd%d: [%s] cannot remove: %d",
+						dev->mtd.index, name, err);
+
+			return;
+		}
+	}
+	INFO("no such device: %s", name);
+}
+
+
 
 static int block2mtd_setup2(const char *val)
 {
-	char buf[80 + 12]; /* 80 for device, 12 for erase size */
+	char buf[80 + 12 + 12]; /* 80 for device, 12 for erase size, 12 for writesize */
 	char *str = buf;
-	char *token[2];
+	char *token[3];
 	char *name;
 	size_t erase_size = PAGE_SIZE;
+	size_t write_size = 1;
 	int i, ret;
 
 	if (strnlen(val, sizeof(buf)) >= sizeof(buf))
@@ -365,7 +394,7 @@
 	strcpy(str, val);
 	kill_final_newline(str);
 
-	for (i = 0; i < 2; i++)
+	for (i = 0; i < 3; i++)
 		token[i] = strsep(&str, ",");
 
 	if (str)
@@ -379,13 +408,22 @@
 		parse_err("device name too long");
 
 	if (token[1]) {
+		if (!strcmp("remove", token[1])) {
+			remove_device_by_name(name);
+			return 0;
+		}
 		ret = parse_num(&erase_size, token[1]);
 		if (ret) {
 			parse_err("illegal erase size");
 		}
 	}
-
-	add_device(name, erase_size);
+	if (token[2]) {
+		ret = parse_num(&write_size, token[2]);
+		if (ret) {
+			pr_err("illegal write size");
+		}
+	}
+	add_device(name, erase_size, write_size);
 
 	return 0;
 }
@@ -419,7 +457,7 @@
 
 
 module_param_call(block2mtd, block2mtd_setup, NULL, NULL, 0200);
-MODULE_PARM_DESC(block2mtd, "Device to use. \"block2mtd=<dev>[,<erasesize>]\"");
+MODULE_PARM_DESC(block2mtd, "Device to use. \"block2mtd=<dev>[,<erasesize>,<writesize>]\"");
 
 static int __init block2mtd_init(void)
 {
--- a/drivers/mtd/ubi/build.c	2025-05-09 19:09:52.704087443 +0000
+++ b/drivers/mtd/ubi/build.c	2025-05-09 19:10:03.646094305 +0000
@@ -830,18 +830,18 @@
 		vtbl_rec = ubi->vtbl[vol_id];
 		err = ubi_change_vtbl_record(ubi, vol_id, &vtbl_rec);
 		if (err)
-			ubi_err("cannot clean auto-resize flag for volume %d",
+			ubi_warn("cannot clean auto-resize flag for volume %d",
 				vol_id);
 	} else {
 		desc.vol = vol;
 		err = ubi_resize_volume(&desc,
 					old_reserved_pebs + ubi->avail_pebs);
 		if (err)
-			ubi_err("cannot auto-resize volume %d", vol_id);
+			ubi_warn("cannot auto-resize volume %d", vol_id);
 	}
 
 	if (err)
-		return err;
+		return 0;
 
 	ubi_msg("volume %d (\"%s\") re-sized from %d to %d LEBs", vol_id,
 		vol->name, old_reserved_pebs, vol->reserved_pebs);
--- a/drivers/mtd/ubi/eba.c	2025-05-09 17:48:00.827322274 +0000
+++ b/drivers/mtd/ubi/eba.c	2025-05-09 17:49:11.203361267 +0000
@@ -1366,17 +1366,18 @@
 		}
 	}
 
-	if (ubi->avail_pebs < EBA_RESERVED_PEBS) {
-		ubi_err("no enough physical eraseblocks (%d, need %d)",
+	int eba_reserved_pebs = EBA_RESERVED_PEBS;
+
+	if (ubi->avail_pebs < eba_reserved_pebs) {
+		ubi_warn("not enough physical eraseblocks (%d < %d), continuing anyway!",
 			ubi->avail_pebs, EBA_RESERVED_PEBS);
 		if (ubi->corr_peb_count)
-			ubi_err("%d PEBs are corrupted and not used",
+			ubi_warn("%d PEBs are corrupted and not used",
 				ubi->corr_peb_count);
-		err = -ENOSPC;
-		goto out_free;
+		eba_reserved_pebs = ubi->avail_pebs;
 	}
-	ubi->avail_pebs -= EBA_RESERVED_PEBS;
-	ubi->rsvd_pebs += EBA_RESERVED_PEBS;
+	ubi->avail_pebs -= eba_reserved_pebs;
+	ubi->rsvd_pebs += eba_reserved_pebs;
 
 	if (ubi->bad_allowed) {
 		ubi_calculate_reserved(ubi);
--- a/drivers/mtd/ubi/vtbl.c	2025-05-09 13:31:09.241144808 +0000
+++ b/drivers/mtd/ubi/vtbl.c	2025-05-09 13:30:58.345129078 +0000
@@ -243,7 +243,7 @@
 			ubi_err("too large reserved_pebs %d, good PEBs %d",
 				reserved_pebs, ubi->good_peb_count);
 			err = 9;
-			goto bad;
+			//goto bad;
 		}
 
 		if (name_len > UBI_VOL_NAME_MAX) {
--- a/drivers/mtd/ubi/wl.c	2025-05-09 18:35:49.586538897 +0000
+++ b/drivers/mtd/ubi/wl.c	2025-05-09 18:36:31.043598587 +0000
@@ -1973,12 +1973,13 @@
 #endif
 
 	if (ubi->avail_pebs < reserved_pebs) {
-		ubi_err("no enough physical eraseblocks (%d, need %d)",
+		ubi_warn("not enough physical eraseblocks (%d < %d), continuing anyway!",
 			ubi->avail_pebs, reserved_pebs);
 		if (ubi->corr_peb_count)
-			ubi_err("%d PEBs are corrupted and not used",
+			ubi_warn("%d PEBs are corrupted and not used",
 				ubi->corr_peb_count);
-		goto out_free;
+		//goto out_free;
+		reserved_pebs = ubi->avail_pebs;
 	}
 	ubi->avail_pebs -= reserved_pebs;
 	ubi->rsvd_pebs += reserved_pebs;
 
