diff --git a/arch/mips/brcmstb/board.c b/arch/mips/brcmstb/board.c
index d90d2b1..ca960e1 100644
--- a/arch/mips/brcmstb/board.c
+++ b/arch/mips/brcmstb/board.c
@@ -863,21 +863,29 @@ static struct mtd_partition fixed_partition_map[] = {
 };
 
 #else
+#define SMALLEST_FLASH_SIZE     (16<<20)
+#define DEFAULT_RESERVED_SIZE   (10<<20)
+#define DEFAULT_ECM_SIZE        (0)     
+
+#define DEFAULT_ROOTFS_SIZE (MTDPART_SIZ_FULL - DEFAULT_RESERVED_SIZE - DEFAULT_ECM_SIZE)
+
 static struct mtd_partition fixed_partition_map[] = {
-        /* name                 offset          size */
-        { name: "rootfs",       offset: 0,                      size: (128-4-4-2-3-1)<<20 },   /* rootfs is total nand size - 6 M Bytes. referr to cfe. bcm97335_devs.c */
-        { name: "kernel",       offset: 0x07200000,             size: 4<<20 },
-        { name: "boot",         offset: 0x07600000,             size: 4<<20 },
-        { name: "splash",       offset: 0x07a00000,             size: 2<<20 },
-        { name: "cfe",          offset: 0x07c00000,             size: 1<<20 },
-        { name: "mac",          offset: 0x07d00000,             size: 1<<19 },
-        { name: "env",          offset: 0x07d80000,             size: 1<<19 },
-        { name: "nvm",          offset: 0x07e00000,             size: 1<<20 },
+        /* name                 offset          size */ 
+        { name: "rootfs",               offset: 0,              size:0x1F200000 /* DEFAULT_ROOTFS_SIZE*/ },   /* rootfs is total nand size - 6 M Bytes. referr to cfe. bcm97335_devs.c */
+        { name: "kernel",       offset: 0x1F200000,             size: 4<<20 },
+        { name: "boot",         offset: 0x1F600000,             size: 4<<20 },
+        { name: "splash",       offset: 0x1FA00000,             size: 2<<20 },
+        { name: "cfe",          offset: 0x1FC00000,             size: 1<<20 },
+        { name: "mac",          offset: 0x1FD00000,             size: 1<<19 },
+        { name: "env",          offset: 0x1FD80000,             size: 1<<19 },
+        { name: "nvm",          offset: 0x1FE00000,             size: 1<<20 },
+        { name: "data",         offset: 0x20000000,             size: 0x1FC00000 },
         /* BBT 1MB not mountable by anyone */
 /*        { name: "data",         offset: 0x20000000,             size: 0 },*/
 /* Add 1 extra place-holder partition for splash, and a safety guard element */
 /*        {name: NULL, offset: 0, size: 0},*/
 };
+
 #endif
 
 
