diff -Naur org/src/zdusb.c patched/src/zdusb.c
--- org/src/zdusb.c	2007-11-06 13:06:36.000000000 +0100
+++ patched/src/zdusb.c	2008-08-19 12:44:48.000000000 +0200
@@ -371,6 +371,8 @@
 	g_dev = net;  //save this for CBs use
 	//macp = net->priv; //kernel 2.4
 	net->priv = macp;   //kernel 2.6
+  /* ath_desc: use /dev/wlanX as device node */
+  strcpy(net->name, "wlan%d");	
 	macp->device = net;
 	macp->usb = dev;
 	SET_MODULE_OWNER(net);
