--- a/umountfs	2017-06-13 10:08:37.765557695 +0200
+++ b/umountfs	2017-06-13 10:12:04.297229284 +0200
@@ -11,14 +11,96 @@
 
 PATH=/sbin:/bin:/usr/sbin:/usr/bin
 
-echo "Deactivating swap..."
-[ -x /sbin/swapoff ] && swapoff -a
+exec 9<&0 </proc/mounts
 
-# We leave /proc mounted.
-echo "Unmounting local filesystems..."
-grep -q /mnt/ram /proc/mounts && mount -o remount,ro /mnt/ram
-mount -o remount,ro /
+PROTECTED_MOUNTS="$(sed -n '0,/^\/[^ ]* \/ /p' /proc/mounts)"
+WEAK_MTPTS="" # be gentle, don't use force
+REG_MTPTS=""
+TMPFS_MTPTS=""
+NETFS_MTPTS=""
 
-umount -f -a -r > /dev/null 2>&1
+while read -r DEV MTPT FSTYPE REST
+do
+	echo "$PROTECTED_MOUNTS" | grep -qs "^$DEV $MTPT " && continue
+	case "$MTPT" in
+	  /|/proc|/dev|/.dev|/dev/pts|/dev/shm|/dev/.static/dev|/proc/*|/sys|/lib/init/rw)
+		continue
+		;;
+	  /var/run)
+		continue
+		;;
+	  /var/lock)
+		continue
+		;;
+	esac
+	case "$FSTYPE" in 
+	  proc|procfs|linprocfs|devfs|sysfs|usbfs|usbdevfs|devpts)
+		continue
+		;;
+	  tmpfs)
+		TMPFS_MTPTS="$MTPT $TMPFS_MTPTS"
+		;;
+	  nfs|nfs4|smbfs|cifs)
+		NETFS_MTPTS="${DEV%:*}"="$MTPT $NETFS_MTPTS"
+		;;
+	  *)
+		if echo "$PROTECTED_MOUNTS" | grep -qs "^$DEV "; then
+			WEAK_MTPTS="$MTPT $WEAK_MTPTS"
+		else
+			REG_MTPTS="$MTPT $REG_MTPTS"
+		fi
+		;;
+	esac
+done
+
+exec 0<&9 9<&-
+
+#
+# Make sure tmpfs file systems are umounted before turning off
+# swap, to avoid running out of memory if the tmpfs filesystems
+# use a lot of space.
+#
+if [ "$TMPFS_MTPTS" ]
+then
+	echo "Unmounting temporary filesystems"
+	umount $TMPFS_MTPTS
+fi
+
+#
+# Deactivate swap
+#
+echo "Deactivating swap"
+echo "swapoff -a >/dev/null"
+
+#
+# Unmount local filesystems
+#
+if [ "$WEAK_MTPTS" ]; then
+	echo "Unmounting weak filesystems"
+	umount -r $WEAK_MTPTS
+fi
+if [ "$REG_MTPTS" ]
+then
+	echo "Unmounting local filesystems"
+	umount -f -r $REG_MTPTS
+fi
+
+#
+# Unmount network filesystems
+# If there is no response from the ping, do not umount,
+# because umount freezes on a non working share
+#
+if [ "$NETFS_MTPTS" ]
+then
+	echo "Unmounting network filesystems"
+	for DEV in $NETFS_MTPTS
+	do
+		ping -c1 -W3 ${DEV%=*}
+		if [ $? -eq 0 ]
+		then
+			umount -f -r ${DEV#*=}
+		fi
+	done
+fi
 
 : exit 0
--- a/halt	2017-06-13 10:07:00.053937415 +0200
+++ b/halt	2017-06-13 10:14:07.968038646 +0200
@@ -24,6 +24,15 @@
 	hddown=""
 fi
 
-halt SED_HALTARGS -p $hddown
+#green led off
+/usr/bin/fp_control -l 1 0
+
+#red led on
+/usr/bin/fp_control -l 0 1
+
+/usr/bin/fp_control -s `date +"%H:%M:%S %d-%m-%Y"`
+/usr/bin/fp_control -e
+echo "0" > /proc/progress
+halt -d -f -p $hddown
 
 : exit 0
