diff -urN a/arch/mips/brcmstb/prom.c b/arch/mips/brcmstb/prom.c
--- a/arch/mips/brcmstb/prom.c	2014-05-07 21:14:28.000000000 +0200
+++ b/arch/mips/brcmstb/prom.c	2014-07-07 06:54:26.687296000 +0200
@@ -224,6 +224,10 @@
 	FETCH("LINUX_FFS_SIZE", parse_hex, &brcm_mtd_rootfs_len);
 	FETCH("LINUX_PART_STARTAD", parse_hex, &brcm_mtd_kernel_start);
 	FETCH("LINUX_PART_SIZE", parse_hex, &brcm_mtd_kernel_len);
+	FETCH("LINUX_MACADDR_STARTAD", parse_hex, &brcm_mtd_macaddr_start);
+	FETCH("LINUX_MACADDR_SIZE", parse_hex, &brcm_mtd_macaddr_len);
+	FETCH("MACADR_PART_STARTAD", parse_hex, &brcm_mtd_macadr_start);
+	FETCH("MACADR_PART_SIZE", parse_hex, &brcm_mtd_macadr_len);
 	FETCH("OCAP_PART_STARTAD", parse_hex, &brcm_mtd_ocap_start);
 	FETCH("OCAP_PART_SIZE", parse_hex, &brcm_mtd_ocap_len);
 	FETCH("FLASH_SIZE", parse_ulong, &brcm_mtd_flash_size_mb);
diff -urN a/drivers/brcmstb/board.c b/drivers/brcmstb/board.c
--- a/drivers/brcmstb/board.c	2014-05-07 21:14:29.000000000 +0200
+++ b/drivers/brcmstb/board.c	2014-07-22 22:39:07.995495349 +0200
@@ -34,6 +34,10 @@
 unsigned long brcm_mtd_rootfs_len;
 unsigned long brcm_mtd_kernel_start;
 unsigned long brcm_mtd_kernel_len;
+unsigned long brcm_mtd_macaddr_start;
+unsigned long brcm_mtd_macaddr_len;
+unsigned long brcm_mtd_macadr_start;
+unsigned long brcm_mtd_macadr_len;
 unsigned long brcm_mtd_ocap_start;
 unsigned long brcm_mtd_ocap_len;
 unsigned long brcm_mtd_flash_size_mb;
@@ -258,7 +262,7 @@
 		brcm_moca_rf_band = MOCA_BAND_HIGHRF;
 
 #elif defined(CONFIG_BCM7346)
-
+#if 0
 	PINMUX(15, gpio_068, 2);	/* MoCA link */
 	PINMUX(16, gpio_069, 1);	/* MoCA activity */
 
@@ -287,10 +291,10 @@
 	PINMUX(18, vo_656_clk, 1);	/* SDIO_CLK */
 	PINMUX(18, vo_656_7, 1);	/* SDIO_CMD */
 	PINMUX(18, vo_656_6, 2);	/* SDIO_WPROT */
-
+#endif
 #elif defined(CONFIG_BCM7358) || defined(CONFIG_BCM7552) || \
 	defined(CONFIG_BCM7360) || defined(CONFIG_BCM7362)
-
+#if 0
 	PINMUX(11, gpio_89, 1);		/* UARTB TX */
 	PINMUX(11, gpio_90, 1);		/* UARTB RX */
 	PINMUX(11, gpio_91, 1);		/* UARTC TX */
@@ -318,7 +322,7 @@
 	AON_PADCTRL(1, aon_gpio_18_pad_ctrl, 2);
 	AON_PADCTRL(1, aon_gpio_19_pad_ctrl, 2);
 	AON_PADCTRL(1, aon_gpio_20_pad_ctrl, 2);
-
+#endif
 #elif defined(CONFIG_BCM7425)
 
 	if (sdio_0_disabled < 0)
@@ -745,6 +749,10 @@
 	nr_parts = 2;
 	if (brcm_mtd_kernel_len != 0)
 		nr_parts++;
+	if (brcm_mtd_macaddr_len != 0)
+		nr_parts++;
+	if (brcm_mtd_macadr_len != 0)
+		nr_parts++;
 
 	ret = kzalloc(nr_parts * sizeof(struct mtd_partition), GFP_KERNEL);
 	if (!ret)
@@ -754,15 +762,25 @@
 	ret[0].size = brcm_mtd_rootfs_len;
 	ret[0].name = "rootfs";
 
-	ret[1].offset = 0;
-	ret[1].size = MTDPART_SIZ_FULL;
-	ret[1].name = "entire_device";
+	ret[1].offset = brcm_mtd_rootfs_start;
+	ret[1].size = brcm_mtd_rootfs_len;
+	ret[1].name = "rootfs(redundant)";
 
 	if (brcm_mtd_kernel_len != 0) {
 		ret[2].offset = brcm_mtd_kernel_start;
 		ret[2].size = brcm_mtd_kernel_len;
 		ret[2].name = "kernel";
 	}
+	if (brcm_mtd_macaddr_len != 0) {
+		ret[3].offset = brcm_mtd_macaddr_start;
+		ret[3].size = brcm_mtd_macaddr_len;
+		ret[3].name = "mac";
+	}
+	if (brcm_mtd_macadr_len != 0) {
+		ret[3].offset = brcm_mtd_macadr_start;
+		ret[3].size = brcm_mtd_macadr_len;
+		ret[3].name = "mac";
+	}
 
 	*p = ret;
 	return nr_parts;
diff -urN a/include/linux/brcmstb/brcmstb.h b/include/linux/brcmstb/brcmstb.h
--- a/include/linux/brcmstb/brcmstb.h	2014-05-07 21:14:47.000000000 +0200
+++ b/include/linux/brcmstb/brcmstb.h	2014-07-07 06:54:45.334829000 +0200
@@ -699,6 +699,10 @@
 extern unsigned long brcm_mtd_rootfs_len;
 extern unsigned long brcm_mtd_kernel_start;
 extern unsigned long brcm_mtd_kernel_len;
+extern unsigned long brcm_mtd_macaddr_start;
+extern unsigned long brcm_mtd_macaddr_len;
+extern unsigned long brcm_mtd_macadr_start;
+extern unsigned long brcm_mtd_macadr_len;
 extern unsigned long brcm_mtd_ocap_start;
 extern unsigned long brcm_mtd_ocap_len;
 extern unsigned long brcm_mtd_flash_size_mb;
