diff --git a/drivers/net/ethernet/broadcom/genet/bcmmii.c b/drivers/net/ethernet/broadcom/genet/bcmmii.c
index f8349a6..d4153be 100644
--- a/drivers/net/ethernet/broadcom/genet/bcmmii.c
+++ b/drivers/net/ethernet/broadcom/genet/bcmmii.c
@@ -302,6 +302,7 @@ int bcmgenet_mii_init(struct net_device *dev)
 		/* fall through */
 	case BRCM_PHY_TYPE_EXT_RGMII:
 		phy_name = "external RGMII";
+		id_mode_dis |= (0xe << 12);
 		GENET_RGMII_OOB_CTRL(pDevCtrl) |= RGMII_MODE_EN | id_mode_dis;
 		pDevCtrl->sys->sys_port_ctrl = PORT_MODE_EXT_GPHY;
 		/*
