From: arn354 <arn354@email.de>
Date: Sun, 27 Jul 2014 20:24:03 +0200
Subject: [PATCH] vlan_depth

---
 net/core/dev.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/core/dev.c b/net/core/dev.c
index d52b172..497e572 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -2296,7 +2296,7 @@ EXPORT_SYMBOL(skb_checksum_help);
 __be16 skb_network_protocol(struct sk_buff *skb, int *depth)
 {
 	__be16 type = skb->protocol;
-	int vlan_depth = ETH_HLEN;
+	int vlan_depth = skb->mac_len;
 
 	/* Tunnel gso handlers can set protocol to ethernet. */
 	if (type == htons(ETH_P_TEB)) {
-- 
1.8.3.msysgit.0

