From f5ec343373ce07ef8c0034913056796d25d31ad6 Mon Sep 17 00:00:00 2001
From: Captain <captain.onboard@web.de>
Date: Sun, 13 May 2018 15:29:52 +0200
Subject: [PATCH] fix cfg80211 build

---
 os_dep/linux/ioctl_cfg80211.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/os_dep/linux/ioctl_cfg80211.c b/os_dep/linux/ioctl_cfg80211.c
index 26988ce..a1b6eee 100644
--- a/os_dep/linux/ioctl_cfg80211.c
+++ b/os_dep/linux/ioctl_cfg80211.c
@@ -897,7 +897,11 @@ void rtw_cfg80211_indicate_disconnect(_adapter *padapter)
 		#else
 
 		if(check_fwstate(&padapter->mlmepriv, _FW_LINKED))		
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 2, 0)
 			cfg80211_disconnected(padapter->pnetdev, 0, NULL, 0, 0, GFP_ATOMIC);
+#else
+			cfg80211_disconnected(padapter->pnetdev, 0, NULL, 0, GFP_ATOMIC);
+#endif
 		else
 			cfg80211_connect_result(padapter->pnetdev, NULL, NULL, 0, NULL, 0, 
 				WLAN_STATUS_UNSPECIFIED_FAILURE, GFP_ATOMIC/*GFP_KERNEL*/);
-- 
2.17.0.windows.1

