--- Module/rtmp_main.c_orig	2008-05-06 07:26:44.000000000 +0200
+++ Module/rtmp_main.c	2008-08-12 09:19:55.000000000 +0200
@@ -987,7 +987,7 @@
 
 		for (i = 0; i < 8; i++)
 		{
-			sprintf(slot_name, "rausb%d", i);
+			sprintf(slot_name, "wlan%d", i);
 			
 			read_lock_bh(&dev_base_lock); // avoid multiple init
 //Benson modify -->
@@ -1010,7 +1010,7 @@
 			return NULL;
 		}
 
-		sprintf(pAd->net_dev->name, "rausb%d", i);
+		sprintf(pAd->net_dev->name, "wlan%d", i);
 		DBGPRINT(RT_DEBUG_ERROR, "usb device name %s\n",pAd->net_dev->name);
 
         /* get Max Packet Size from usb_dev endpoint */
@@ -1191,7 +1191,7 @@
         
 		for (i = 0; i < 8; i++)
 		{
-			sprintf(slot_name, "rausb%d", i);
+			sprintf(slot_name, "wlan%d", i);
 
 #if 1          
 //--> Benson add 20080505, for Linux_kernel > 2,6,24 
@@ -1231,7 +1231,7 @@
 			return res;
 		}
 
-		sprintf(pAd->net_dev->name, "rausb%d", i);
+		sprintf(pAd->net_dev->name, "wlan%d", i);
 		DBGPRINT(RT_DEBUG_ERROR, "usb device name %s\n", pAd->net_dev->name);
 
 
