diff --git a/src/backend/dab_decoder.h b/src/backend/dab_decoder.h
index 3bd3545..9f26b90 100644
--- a/src/backend/dab_decoder.h
+++ b/src/backend/dab_decoder.h
@@ -26,7 +26,6 @@
 #include <string>
 #include <vector>
 
-#define MPG123_NO_LARGENAME	// disable large file API here
 #include "mpg123.h"
 #if MPG123_API_VERSION < 36
 #	error "At least version 1.14.0 (API version 36) of mpg123 is required!"
