Index: linux-3.2.2/block/cfq-iosched.c
===================================================================
--- linux-3.2.2.orig/block/cfq-iosched.c
+++ linux-3.2.2/block/cfq-iosched.c
@@ -30,8 +30,8 @@ static const int cfq_back_penalty = 2;
 static const int cfq_slice_sync = HZ / 10;
 static int cfq_slice_async = HZ / 25;
 static const int cfq_slice_async_rq = 2;
-static int cfq_slice_idle = HZ / 125;
-static int cfq_group_idle = HZ / 125;
+static int cfq_slice_idle = HZ / 1000; /* 1 ms */
+static int cfq_group_idle = HZ / 250;  /* 4 ms, not used */
 static const int cfq_target_latency = HZ * 3/10; /* 300 ms */
 static const int cfq_hist_divisor = 4;
 
