--- a/drivers/mtd/ubi/build.c	2025-05-09 19:09:52.704087443 +0000
+++ b/drivers/mtd/ubi/build.c	2025-05-09 19:10:03.646094305 +0000
@@ -839,18 +839,18 @@
 		       sizeof(struct ubi_vtbl_record));
 		err = ubi_change_vtbl_record(ubi, vol_id, &vtbl_rec);
 		if (err)
-			ubi_err("cannot clean auto-resize flag for volume %d",
+			ubi_warn("cannot clean auto-resize flag for volume %d",
 				vol_id);
 	} else {
 		desc.vol = vol;
 		err = ubi_resize_volume(&desc,
 					old_reserved_pebs + ubi->avail_pebs);
 		if (err)
-			ubi_err("cannot auto-resize volume %d", vol_id);
+			ubi_warn("cannot auto-resize volume %d", vol_id);
 	}
 
 	if (err)
-		return err;
+		return 0;
 
 	ubi_msg("volume %d (\"%s\") re-sized from %d to %d LEBs", vol_id,
 		vol->name, old_reserved_pebs, vol->reserved_pebs);
--- a/drivers/mtd/ubi/vtbl.c	2025-05-09 13:31:09.241144808 +0000
+++ b/drivers/mtd/ubi/vtbl.c	2025-05-09 13:30:58.345129078 +0000
@@ -248,7 +248,7 @@
 			dbg_err("too large reserved_pebs %d, good PEBs %d",
 				reserved_pebs, ubi->good_peb_count);
 			err = 9;
-			goto bad;
+			//goto bad;
 		}
 
 		if (name_len > UBI_VOL_NAME_MAX) {
--- a/drivers/mtd/ubi/eba.c	2025-05-09 17:48:00.827322274 +0000
+++ b/drivers/mtd/ubi/eba.c	2025-05-09 17:49:11.203361267 +0000
@@ -1266,17 +1266,18 @@
 		}
 	}
 
-	if (ubi->avail_pebs < EBA_RESERVED_PEBS) {
-		ubi_err("no enough physical eraseblocks (%d, need %d)",
+	int eba_reserved_pebs = EBA_RESERVED_PEBS;
+
+	if (ubi->avail_pebs < eba_reserved_pebs) {
+		ubi_warn("not enough physical eraseblocks (%d < %d), continuing anyway!",
 			ubi->avail_pebs, EBA_RESERVED_PEBS);
 		if (ubi->corr_peb_count)
-			ubi_err("%d PEBs are corrupted and not used",
+			ubi_warn("%d PEBs are corrupted and not used",
 				ubi->corr_peb_count);
-		err = -ENOSPC;
-		goto out_free;
+		eba_reserved_pebs = ubi->avail_pebs;
 	}
-	ubi->avail_pebs -= EBA_RESERVED_PEBS;
-	ubi->rsvd_pebs += EBA_RESERVED_PEBS;
+	ubi->avail_pebs -= eba_reserved_pebs;
+	ubi->rsvd_pebs += eba_reserved_pebs;
 
 	if (ubi->bad_allowed) {
 		ubi_calculate_reserved(ubi);
--- a/drivers/mtd/ubi/wl.c	2025-05-09 18:35:49.586538897 +0000
+++ b/drivers/mtd/ubi/wl.c	2025-05-09 18:36:31.043598587 +0000
@@ -1507,17 +1507,20 @@
 		}
 	}
 
-	if (ubi->avail_pebs < WL_RESERVED_PEBS) {
-		ubi_err("no enough physical eraseblocks (%d, need %d)",
-			ubi->avail_pebs, WL_RESERVED_PEBS);
+	int reserved_pebs = WL_RESERVED_PEBS;
+
+	if (ubi->avail_pebs < reserved_pebs) {
+		ubi_warn("not enough physical eraseblocks (%d < %d), continuing anyway!",
+			ubi->avail_pebs, reserved_pebs);
 		if (ubi->corr_peb_count)
-			ubi_err("%d PEBs are corrupted and not used",
+			ubi_warn("%d PEBs are corrupted and not used",
 				ubi->corr_peb_count);
-		err = -ENOSPC;
-		goto out_free;
+		// = -ENOSPC;
+		//goto out_free;
+		reserved_pebs = ubi->avail_pebs;
 	}
-	ubi->avail_pebs -= WL_RESERVED_PEBS;
-	ubi->rsvd_pebs += WL_RESERVED_PEBS;
+	ubi->avail_pebs -= reserved_pebs;
+	ubi->rsvd_pebs += reserved_pebs;
 
 	/* Schedule wear-leveling if needed */
 	err = ensure_wear_leveling(ubi);
