From a36aec5d3b2f182cd110d31d1337fe161acfe502 Mon Sep 17 00:00:00 2001
From: Armin Kuster <akuster808@gmail.com>
Date: Sat, 7 May 2016 12:30:29 -0700
Subject: [PATCH] main: fix inline issue

Signed-off-by: Armin Kuster <akuster808@gmail.com>
---
 common/main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/common/main.c b/common/main.c
index 5312e7d..ff000b7 100644
--- a/common/main.c
+++ b/common/main.c
@@ -63,7 +63,7 @@ extern int __g_nTStep_4BC722B3__;
 /*
  * Board-specific Platform code can reimplement show_boot_progress () if needed
  */
-void inline __show_boot_progress (int val) {}
+void static inline __show_boot_progress (int val) {}
 void show_boot_progress (int val) __attribute__((weak, alias("__show_boot_progress")));
 
 #if defined(CONFIG_UPDATE_TFTP)
-- 
1.9.1

