From bde456a5cee7720fbef24ebba76cf0e8e698c262 Mon Sep 17 00:00:00 2001
From: Armin Kuster <akuster808@gmail.com>
Date: Sat, 7 May 2016 12:47:12 -0700
Subject: [PATCH] usb: use define not func

Signed-off-by: Armin Kuster <akuster808@gmail.com>
---
 common/usb.c  | 4 ++--
 include/usb.h | 4 +++-
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/common/usb.c b/common/usb.c
index ea62c30..08f592f 100644
--- a/common/usb.c
+++ b/common/usb.c
@@ -90,12 +90,12 @@ static int hub_port_reset(struct usb_device *dev, int port,
  * wait_ms
  */
 
-inline void wait_ms(unsigned long ms)
+/*static inline void wait_ms(unsigned long ms)
 {
 	while (ms-- > 0)
 		udelay(1000);
 }
-
+*/
 /***************************************************************************
  * Init USB Device
  */
diff --git a/include/usb.h b/include/usb.h
index 97ea37b..0f97021 100644
--- a/include/usb.h
+++ b/include/usb.h
@@ -197,7 +197,9 @@ int usb_submit_int_msg(struct usb_device *dev, unsigned long pipe,
 			void *buffer, int transfer_len, int interval);
 void usb_disable_asynch(int disable);
 int usb_maxpacket(struct usb_device *dev, unsigned long pipe);
-inline void wait_ms(unsigned long ms);
+//static inline void wait_ms(unsigned long ms);
+
+#define wait_ms(a) udelay(a*1000);
 int usb_get_configuration_no(struct usb_device *dev, unsigned char *buffer,
 				int cfgno);
 int usb_get_report(struct usb_device *dev, int ifnum, unsigned char type,
-- 
1.9.1

