diff --git a/arch/arm/cpu/aml_meson/config.mk b/arch/arm/cpu/aml_meson/config.mk
index 151ef9a..5a0071d 100755
--- a/arch/arm/cpu/aml_meson/config.mk
+++ b/arch/arm/cpu/aml_meson/config.mk
@@ -1,6 +1,6 @@
 CROSS_COMPILE=arm-none-eabi-
 ARM_CPU=cortex-a9
-PLATFORM_CPPFLAGS += $(call cc-option,-mcpu=cortex-a9 -ffixed-r8 -mno-long-calls  -Wall -fPIC )
+PLATFORM_CPPFLAGS += $(call cc-option,-mcpu=cortex-a9 -ffixed-r8 -mno-long-calls  -Wall -fPIC -mfloat-abi=hard -mfpu=vfpv3 )
 PF_CPPFLAGS_NO_UNALIGNED := $(call cc-option, -mno-unaligned-access,)
 PLATFORM_CPPFLAGS += $(PF_CPPFLAGS_NO_UNALIGNED)
 #USE_PRIVATE_LIBGCC=yes
