--- 2009_0525_RT3070_Linux_STA_v2.1.1.0/os/linux/config.mk	2009-05-20 20:33:56.000000000 -0700
+++ 2009_0525_RT3070_Linux_STA_v2.1.1.0/os/linux/config.mk	2009-08-22 19:42:55.000000000 -0700
@@ -4,15 +4,11 @@
 # Support 28xx QA ATE function
 HAS_28xx_QA=n
 
-#ifdef WPA_SUPPLICANT_SUPPORT
 # Support Wpa_Supplicant
-HAS_WPA_SUPPLICANT=n
-#endif // WPA_SUPPLICANT_SUPPORT //
+HAS_WPA_SUPPLICANT=y
 
-#ifdef NATIVE_WPA_SUPPLICANT_SUPPORT
 # Support Native WpaSupplicant for Network Maganger
-HAS_NATIVE_WPA_SUPPLICANT_SUPPORT=n
-#endif // NATIVE_WPA_SUPPLICANT_SUPPORT //
+HAS_NATIVE_WPA_SUPPLICANT_SUPPORT=y
 
 #Support Net interface block while Tx-Sw queue full
 HAS_BLOCK_NET_IF=n
