diff --git a/drivers/media/common/tuners/tda18218.c b/drivers/media/common/tuners/tda18218.c
index aacfe23..255d965 100644
--- a/drivers/media/common/tuners/tda18218.c
+++ b/drivers/media/common/tuners/tda18218.c
@@ -145,7 +145,7 @@ static int tda18218_set_params(struct dvb_frontend *fe,
 		break;
 	case BANDWIDTH_7_MHZ:
 		LP_Fc = 1;
-		LO_Frac = params->frequency + 3500000;
+		LO_Frac = params->frequency + 4000000;
 		break;
 	case BANDWIDTH_8_MHZ:
 	default:
