diff --git a/arch/mips/boot/dts/brcm/xc7346.dts b/arch/mips/boot/dts/brcm/xc7346.dts
index 3621ce06f5dc..74e2a934b08c 100644
--- a/arch/mips/boot/dts/brcm/xc7346.dts
+++ b/arch/mips/boot/dts/brcm/xc7346.dts
@@ -8,7 +8,7 @@
 
 	memory@0 {
 		device_type = "memory";
-		reg = <0x00000000 0x04000000>, <0x30000000 0x20000000>;
+		reg = <0x00000000 0x09000000>, <0x30000000 0x20000000>;
 	};
 
 	chosen {
