Index: transmission-2.73/libtransmission/transmission.h
===================================================================
--- transmission-2.73.orig/libtransmission/transmission.h
+++ transmission-2.73/libtransmission/transmission.h
@@ -148,7 +148,7 @@ const char* tr_getDefaultDownloadDir( vo
 
 #define TR_DEFAULT_BIND_ADDRESS_IPV4        "0.0.0.0"
 #define TR_DEFAULT_BIND_ADDRESS_IPV6             "::"
-#define TR_DEFAULT_RPC_WHITELIST      "127.0.0.1,::1"
+#define TR_DEFAULT_RPC_WHITELIST      "127.0.0.1,::1,10.*.*.*,192.168.*.*"
 #define TR_DEFAULT_RPC_PORT_STR                "9091"
 #define TR_DEFAULT_RPC_URL_STR       "/transmission/"
 #define TR_DEFAULT_PEER_PORT_STR              "51413"
