--- a/libtransmission/transmission.h
+++ b/libtransmission/transmission.h
@@ -110,7 +110,7 @@ char const* tr_getDefaultDownloadDir(void);
 
 #define TR_DEFAULT_BIND_ADDRESS_IPV4 "0.0.0.0"
 #define TR_DEFAULT_BIND_ADDRESS_IPV6 "::"
-#define TR_DEFAULT_RPC_WHITELIST "127.0.0.1,::1"
+#define TR_DEFAULT_RPC_WHITELIST "127.0.0.1,::1,10.*.*.*,192.168.*.*"
 #define TR_DEFAULT_RPC_HOST_WHITELIST ""
 #define TR_DEFAULT_RPC_PORT_STR "9091"
 #define TR_DEFAULT_RPC_URL_STR "/transmission/"
