From 062cf74b498c1d12fd76f9d9bca926ac806e0b99 Mon Sep 17 00:00:00 2001
From: Changqing Li <changqing.li@windriver.com>
Date: Thu, 20 Jun 2019 14:15:18 +0800
Subject: [PATCH] glibc only fix

Signed-off-by: Changqing Li <changqing.li@windriver.com>
---
 ctdb/tests/src/test_mutex_raw.c | 2 ++
 1 file changed, 2 insertions(+)

diff -Nur a/ctdb/tests/src/test_mutex_raw.c b/ctdb/tests/src/test_mutex_raw.c
--- a/ctdb/tests/src/test_mutex_raw.c	2019-11-02 14:35:36.949719710 +0100
+++ b/ctdb/tests/src/test_mutex_raw.c	2019-11-02 14:53:43.697345661 +0100
@@ -291,8 +291,10 @@
 					"failed\n");
 				exit (EXIT_FAILURE);
 			}
+#ifdef __GLIBC__
 		} else if (ret == EBUSY) {
 			printf("INFO: pid=%u\n", mutex->__data.__owner);
+#endif
 		} else if (ret == 0) {
 			pthread_mutex_unlock(mutex);
 		}
