--- a/source3/utils/status.c
+++ b/source3/utils/status.c
@@ -803,7 +803,6 @@ int main(int argc, const char *argv[])
 			TALLOC_FREE(db_path);
 			exit(0);
 		} else {
-			TALLOC_FREE(db);
 			TALLOC_FREE(db_path);
 		}
 
@@ -813,6 +812,8 @@ int main(int argc, const char *argv[])
 			goto done;
 		}
 
+		TALLOC_FREE(db);
+
 		result = share_entry_forall(print_share_mode, &resolve_uids);
 
 		if (result == 0) {
@@ -827,7 +828,7 @@ int main(int argc, const char *argv[])
 			brl_forall(print_brl, NULL);
 		}
 
-		locking_end();
+//		locking_end();
 	}
 
 	if (show_notify) {
