--- a/src/udev-builtin-input_id.c	2012-01-29 01:15:46.000000000 +0100
+++ b/src/udev-builtin-input_id.c	2012-08-04 19:58:29.255349890 +0200
@@ -30,6 +30,10 @@
 
 #include "udev.h"
 
+#ifndef BTN_TRIGGER_HAPPY
+#define BTN_TRIGGER_HAPPY               0x2c0
+#endif
+
 /* we must use this kernel-compatible implementation */
 #define BITS_PER_LONG (sizeof(unsigned long) * 8)
 #define NBITS(x) ((((x)-1)/BITS_PER_LONG)+1)
