diff --git a/Makefile b/Makefile
index 002da9a..03831b6 100644
--- a/Makefile
+++ b/Makefile
@@ -193,7 +193,7 @@ SUBARCH := $(shell uname -m | sed -e s/i.86/i386/ -e s/sun4u/sparc64/ \
 # Default value for CROSS_COMPILE is not to prefix executables
 # Note: Some architectures assign CROSS_COMPILE in their arch/*/Makefile
 export KBUILD_BUILDHOST := $(SUBARCH)
-ARCH		?= $(SUBARCH)
+ARCH		?= mips
 CROSS_COMPILE	?= $(CONFIG_CROSS_COMPILE:"%"=%)
 
 # Architecture as present in compile.h
diff --git a/drivers/brcmstb/bchip.c b/drivers/brcmstb/bchip.c
index b4c5b0b..4f871cd 100644
--- a/drivers/brcmstb/bchip.c
+++ b/drivers/brcmstb/bchip.c
@@ -450,9 +450,14 @@ void __init bchip_early_setup(void)
 		genet_pdata[0].phy_id = BRCM_PHY_ID_AUTO;
 	}
 #else
+#if defined(CONFIG_BCMGENET_0_GPHY)
+        genet_pdata[0].phy_type = BRCM_PHY_TYPE_EXT_RGMII;
+        genet_pdata[0].phy_id = BRCM_PHY_ID_AUTO;
+#else
 	genet_pdata[0].phy_type = BRCM_PHY_TYPE_INT;
 	genet_pdata[0].phy_id = 1;
 #endif
+#endif
 
 #endif
 
diff --git a/drivers/net/ethernet/broadcom/genet/Kconfig b/drivers/net/ethernet/broadcom/genet/Kconfig
index c1ed960..8e87443 100644
--- a/drivers/net/ethernet/broadcom/genet/Kconfig
+++ b/drivers/net/ethernet/broadcom/genet/Kconfig
@@ -1,5 +1,27 @@
 menu "BCMGENET options"
 
+config BCMGENET_0_GPHY
+        bool "GENET0 is connected to GPHY via RGMII"
+        depends on BCMGENET && BRCM_HAS_GENET_0
+        default n
+        help
+          Say Y to use an external GPHY for GENET_0.  By default this will
+          use RGMII mode and disable the internal 10/100 PHY (if present);
+          the pinmux must be set up properly in arch/mips/brcmstb/board.c .
+
+          If unsure, say N.
+
+config BCMGENET_1_GPHY
+        bool "GENET1 is connected to GPHY via RGMII"
+        depends on BCMGENET && BRCM_HAS_GENET_1
+        default n
+        help
+          Say Y to use an external GPHY for GENET_1.  By default this will
+          use RGMII mode and disable the internal 10/100 PHY (if present);
+          the pinmux must be set up properly in arch/mips/brcmstb/board.c .
+
+          If unsure, say N.
+
 config BCMGENET_RX_DESC_THROTTLE 
 	bool "use hardware rx descriptor throttling"
 	depends on BCMGENET
