diff --git a/arch/mips/brcmstb/prom.c b/arch/mips/brcmstb/prom.c
index c575425..b9dfff2 100644
--- a/arch/mips/brcmstb/prom.c
+++ b/arch/mips/brcmstb/prom.c
@@ -224,6 +224,8 @@ static void __init __maybe_unused cfe_read_configuration(void)
 	FETCH("LINUX_FFS_SIZE", parse_hex, &brcm_mtd_rootfs_len);
 	FETCH("LINUX_PART_STARTAD", parse_hex, &brcm_mtd_kernel_start);
 	FETCH("LINUX_PART_SIZE", parse_hex, &brcm_mtd_kernel_len);
+	FETCH("LINUX_MACADDR_STARTAD", parse_hex, &brcm_mtd_macaddr_start);
+	FETCH("LINUX_MACADDR_SIZE", parse_hex, &brcm_mtd_macaddr_len);
 	FETCH("OCAP_PART_STARTAD", parse_hex, &brcm_mtd_ocap_start);
 	FETCH("OCAP_PART_SIZE", parse_hex, &brcm_mtd_ocap_len);
 	FETCH("FLASH_SIZE", parse_ulong, &brcm_mtd_flash_size_mb);
