From 1ad4b471387cd0fd8167a9db33a875e97dc8f862 Mon Sep 17 00:00:00 2001
From: Captain <openatv@gmail.com>
Date: Sat, 18 May 2024 05:10:43 +0200
Subject: [PATCH 08/14] reinstate-system-h

---
 xbmc/system.h | 12 ++++++++++++
 1 file changed, 12 insertions(+)
 create mode 100644 xbmc/system.h

diff --git a/xbmc/system.h b/xbmc/system.h
new file mode 100644
index 0000000000..a0a15f58b8
--- /dev/null
+++ b/xbmc/system.h
@@ -0,0 +1,12 @@
+/*
+ *  Copyright (C) 2005-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#define SAFE_DELETE(p)       do { delete (p);     (p)=NULL; } while (0)
+#define SAFE_RELEASE(p)      do { if(p) { (p)->Release(); (p)=NULL; } } while (0)
-- 
2.45.1.windows.1

