From b10cce4af5fbd2e9d6af2b6446bc49e5f06aa716 Mon Sep 17 00:00:00 2001
From: NickerSK <nickersk@gmail.com>
Date: Thu, 11 Nov 2021 08:55:33 +0100
Subject: [PATCH] kodi 20 - updated to latest git rev, updated patches

---
 cmake/scripts/common/ArchSetup.cmake | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/cmake/scripts/common/ArchSetup.cmake b/cmake/scripts/common/ArchSetup.cmake
index 97a09b4433..41652b63ff 100644
--- a/cmake/scripts/common/ArchSetup.cmake
+++ b/cmake/scripts/common/ArchSetup.cmake
@@ -156,6 +156,10 @@ if(NOT DEFINED NEON OR NEON)
   endif()
 endif()
 
+if(PLATFORM_DEFINES)
+  add_options(ALL_LANGUAGES ALL_BUILDS ${PLATFORM_DEFINES})
+endif()
+
 if(NOT MSVC)
   # these options affect all code built by cmake including external projects.
   add_options(ALL_LANGUAGES ALL_BUILDS
