diff -r -u gitold/xbmc/windowing/egl/WinEventsSTB.cpp gitnew/xbmc/windowing/egl/WinEventsSTB.cpp
--- gitold/xbmc/windowing/egl/WinEventsSTB.cpp	2024-01-25 11:00:52.000000000 +0100
+++ gitnew/xbmc/windowing/egl/WinEventsSTB.cpp	2024-01-25 11:29:23.563224181 +0100
@@ -23,7 +23,7 @@
 #include "WinEventsSTB.h"
 #include "windowing/WinEvents.h"
 #include "windowing/XBMC_events.h"
-#include "input/XBMC_keysym.h"
+#include "input/keyboard/XBMC_keysym.h"
 #include "application/Application.h"
 #include "input/mouse/MouseStat.h"
 #include "utils/log.h"
diff -r -u gitold/xbmc/platform/linux/input/LinuxInputDevices.cpp gitnew/xbmc/platform/linux/input/LinuxInputDevices.cpp
--- gitold/xbmc/platform/linux/input/LinuxInputDevices.cpp	2024-01-25 11:00:52.000000000 +0100
+++ gitnew/xbmc/platform/linux/input/LinuxInputDevices.cpp	2024-01-25 11:32:51.950074363 +0100
@@ -105,7 +105,7 @@
 #include <stdio.h>
 
 #include "windowing/GraphicContext.h"
-#include "input/XBMC_keysym.h"
+#include "input/keyboard/XBMC_keysym.h"
 #include "LinuxInputDevices.h"
 #include "input/mouse/MouseStat.h"
 #include "utils/log.h"
diff -r -u gitold/xbmc/platform/linux/input/LinuxInputDevices.h gitnew/xbmc/platform/linux/input/LinuxInputDevices.h
--- gitold/xbmc/platform/linux/input/LinuxInputDevices.h	2024-01-25 11:00:52.000000000 +0100
+++ gitnew/xbmc/platform/linux/input/LinuxInputDevices.h	2024-01-25 11:32:02.905399372 +0100
@@ -24,7 +24,7 @@
 #include <string>
 #include <deque>
 #include "windowing/XBMC_events.h"
-#include "input/XBMC_keyboard.h"
+#include "input/keyboard/XBMC_keyboard.h"
 #include "threads/SingleLock.h"
 #include "input/touch/ITouchInputHandler.h"
 #include "input/touch/generic/IGenericTouchGestureDetector.h"
 --- gitold/tools/Linux/kodi.sh.in       2024-01-25 11:00:52.000000000 +0100
+++ gitnew/tools/Linux/kodi.sh.in       2024-01-27 19:22:19.813998135 +0100
@@ -40,24 +40,9 @@
 
 check_media_hdd() {
     [ -e /proc/mounts ] || return 0
+     [ -d /media/mmc ] && export HOME="/media/mmc"
+     [ -d /media/hdd ] && export HOME="/media/hdd"
 
-    exec 9<&0 < /proc/mounts
-    while read DEV MTPT REST; do
-        case $DEV in
-            /dev/sd[a-z][0-9])
-            if [ "$MTPT" == "/media/hdd" ]; then
-                export HOME="$MTPT"
-                break
-            fi
-            ;;
-           /dev/mmcblk[0-9]p[0-9])
-            if [ "$MTPT" == "/media/mmc" ]; then
-               export HOME="$MTPT"
-            fi
-           ;;
-        esac
-    done
-    exec 0<&9 9<&-
 }
 
 check_media_hdd

