diff -Nur a/python2/httplib2/__init__.py b/httplib2-0.9.1/python2/httplib2/__init__.py
--- a/python2/httplib2/__init__.py	2015-04-11 15:56:56.000000000 +0200
+++ b/python2/httplib2/__init__.py	2015-06-21 10:20:36.986816754 +0200
@@ -938,7 +938,7 @@
     """
     def __init__(self, host, port=None, key_file=None, cert_file=None,
                  strict=None, timeout=None, proxy_info=None,
-                 ca_certs=None, disable_ssl_certificate_validation=False):
+                 ca_certs=None, disable_ssl_certificate_validation=True):
         httplib.HTTPSConnection.__init__(self, host, port=port,
                                          key_file=key_file,
                                          cert_file=cert_file, strict=strict)
@@ -1114,7 +1114,7 @@
         """
         def __init__(self, host, port=None, key_file=None, cert_file=None,
                      strict=None, timeout=None, proxy_info=None, ca_certs=None,
-                     disable_ssl_certificate_validation=False):
+                     disable_ssl_certificate_validation=True):
             httplib.HTTPConnection.__init__(self, host, port=port,
                                             strict=strict, timeout=timeout)
 
@@ -1122,7 +1122,7 @@
         """Same as AppEngineHttpConnection, but for HTTPS URIs."""
         def __init__(self, host, port=None, key_file=None, cert_file=None,
                      strict=None, timeout=None, proxy_info=None, ca_certs=None,
-                     disable_ssl_certificate_validation=False):
+                     disable_ssl_certificate_validation=True):
             httplib.HTTPSConnection.__init__(self, host, port=port,
                                              key_file=key_file,
                                              cert_file=cert_file, strict=strict,
@@ -1155,7 +1155,7 @@
     """
     def __init__(self, cache=None, timeout=None,
                  proxy_info=proxy_info_from_environment,
-                 ca_certs=None, disable_ssl_certificate_validation=False):
+                 ca_certs=None, disable_ssl_certificate_validation=True):
         """If 'cache' is a string then it is used as a directory name for
         a disk cache. Otherwise it must be an object that supports the
         same interface as FileCache.
