--- a/Makefile
+++ b/Makefile
@@ -150,9 +150,9 @@ clean:
 	$(SWIGS_OBJS) $(SWIGS_LIBS) $(SHARED_LIBS)
 
 install-python:
-	install -d $(D)$(libdir)/python3.8/lib-dynload
-	install -m 644 src/common/crossepg.py $(D)$(libdir)/python3.8
-	install -m 644 bin/_crossepg.so $(D)$(libdir)/python3.8/lib-dynload
+	install -d $(D)$(libdir)/python3.9/lib-dynload
+	install -m 644 src/common/crossepg.py $(D)$(libdir)/python3.9
+	install -m 644 bin/_crossepg.so $(D)$(libdir)/python3.9/lib-dynload
 
 install-python-qboxhd:
 	install -d $(D)/usr/local/lib/python2.6/lib-dynload
@@ -309,8 +309,8 @@ install-py26: install-python-2.6 install-standalone install-plugin
 install-py27: install-python-2.7 install-standalone install-plugin
 
 remote-install:
-	ncftpput -m -u $(FTP_USER) -p $(FTP_PASSWORD) $(FTP_HOST) ${libdir}/python3.8 src/common/crossepg.py
-	ncftpput -m -u $(FTP_USER) -p $(FTP_PASSWORD) $(FTP_HOST) ${libdir}/python3.8/lib-dynload bin/_crossepg.so
+	ncftpput -m -u $(FTP_USER) -p $(FTP_PASSWORD) $(FTP_HOST) ${libdir}/python3.9 src/common/crossepg.py
+	ncftpput -m -u $(FTP_USER) -p $(FTP_PASSWORD) $(FTP_HOST) ${libdir}/python3.9/lib-dynload bin/_crossepg.so
 
 	ncftpput -m -u $(FTP_USER) -p $(FTP_PASSWORD) $(FTP_HOST) /usr/crossepg bin/crossepg_dbconverter
 	ncftpput -m -u $(FTP_USER) -p $(FTP_PASSWORD) $(FTP_HOST) /usr/crossepg bin/crossepg_dbinfo
--- a/make_e2_oealliance.sh
+++ b/make_e2_oealliance.sh
@@ -7,7 +7,7 @@ CROSS=${DEVKIT_ROOT}/sysroots/x86_64-linux/usr/bin/mips32el-oe-linux/mipsel-oe-l
 
 export CFLAGS+="-I${DEVKIT_ROOT}/sysroots/gbquad/usr/include \
  -I${DEVKIT_ROOT}/sysroots/gbquad/usr/include/libxml2 \
- -I${DEVKIT_ROOT}/sysroots/gbquad/usr/include/python3.8"
+ -I${DEVKIT_ROOT}/sysroots/gbquad/usr/include/python3.9"
 export CC=${CROSS}gcc
 export STRIP=${CROSS}strip
 export SWIG=${DEVKIT_ROOT}/sysroots/x86_64-linux/usr/bin/swig
