Copyright (c) 2009 MontaVista Software, Inc.  All rights reserved.

--- a/etc/udev/rules.d/60-persistent-storage.rules	2008-06-12 07:24:30.000000000 +0200
+++ b/etc/udev/rules.d/60-persistent-storage.rules	2009-08-18 11:39:23.595063264 +0200
@@ -10,7 +10,7 @@
 SUBSYSTEM!="block", GOTO="persistent_storage_end"
 
 # skip rules for inappropriate block devices
-KERNEL=="ram*|loop*|fd*|nbd*|gnbd*|dm-*|md*", GOTO="persistent_storage_end"
+KERNEL=="ram*|loop*|fd*|mtd*|nbd*|gnbd*|dm-*|md*|btibm*", GOTO="persistent_storage_end"
 
 # never access non-cdrom removable ide devices, the drivers are causing event loops on open()
 KERNEL=="hd*[!0-9]", ATTR{removable}=="1", DRIVERS=="ide-cs|ide-floppy", GOTO="persistent_storage_end"
