diff --git a/tools/Makefile.am b/tools/Makefile.am
index 828efab..cc65e65 100644
--- a/tools/Makefile.am
+++ b/tools/Makefile.am
@@ -11,8 +11,8 @@ endif
 if HAVE_FFMPEG
 BUILT_SOURCES += tonegen$(BUILD_EXEEXT)
 tonegen$(BUILD_EXEEXT): tonegen.c
-	$(AM_V_GEN)$(CC_FOR_BUILD) $(WARN_CFLAGS) $(CFLAGS) $(LIBM_FOR_BUILD) \
-	    -o $@ $(srcdir)/tonegen.c
+	$(AM_V_GEN)$(CC_FOR_BUILD) $(WARN_CFLAGS) $(CFLAGS) \
+	    -o $@ $(srcdir)/tonegen.c $(LIBM_FOR_BUILD)
 endif
 
 CLEANFILES = $(BUILT_SOURCES)
