--- ~/setup.py.orig	2012-05-23 11:23:03.395503486 +0200
+++ ~/setup.py	2012-05-23 11:23:03.411503488 +0200
@@ -250,7 +250,7 @@
             # figure out target flags here.
             flags = []
             for f in cflags.split():
-                if not f.startswith("-isystem"):
+                if not f.startswith("-isystem") and not f.startswith("-march"):
                     flags.append(f)
             args['compiler_so'] = compiler + ' ' + ccshared + ' ' + ' '.join(flags)
         self.compiler.set_executables(**args)
