From 8426175c13a5c616b4e564eca1db692084a197c2 Mon Sep 17 00:00:00 2001
From: captain <openatv@gmail.com>
Date: Tue, 20 Sep 2016 22:40:31 +0200
Subject: [PATCH] fix build with gcc 6.20

---
 configure    | 2 +-
 configure.ac | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/configure b/configure
index 071e7f7..f11c92d 100755
--- a/configure
+++ b/configure
@@ -17974,7 +17974,7 @@ x$ax_compare_version_B" | sed 's/^ *//' | sort -r | sed "s/x${ax_compare_version
 
 
     if test "$ax_compare_version" = "true" ; then
-    CXXFLAGS="$CXXFLAGS -Wno-c++0x-compat"
+    CXXFLAGS="$CXXFLAGS -Wno-c++0x-compat -std=gnu++98"
       fi
 
 fi
diff --git a/configure.ac b/configure.ac
index e7c28f6..cd17a94 100644
--- a/configure.ac
+++ b/configure.ac
@@ -103,7 +103,7 @@ fi
 # cleanly with that.
 if test "$CXX" = "g++"; then
    CXX_VERSION=`$CXX -dumpversion`
-   AX_COMPARE_VERSION([$CXX_VERSION],[ge],[4.6.0],CXXFLAGS="$CXXFLAGS -Wno-c++0x-compat")
+   AX_COMPARE_VERSION([$CXX_VERSION],[ge],[4.6.0],CXXFLAGS="$CXXFLAGS -Wno-c++0x-compat -std=gnu++98")
 fi
 
 # pthread (not needed on Windows)
-- 
2.9.2.windows.1

