commit 6562b9521098e108c431cd45b8cb691dc4abf8c8
Author: oskwon <oskwon@dev3>
Date:   Sat Feb 4 03:49:59 2017 +0900

    fix!!

diff --git a/Source/WebCore/plugins/PluginView.cpp b/Source/WebCore/plugins/PluginView.cpp
index c21fcd7..d6a3a85 100644
--- a/Source/WebCore/plugins/PluginView.cpp
+++ b/Source/WebCore/plugins/PluginView.cpp
@@ -126,7 +126,11 @@ IntRect PluginView::windowClipRect() const
     IntRect clipRect(m_windowRect);
     
     // Take our element and get the clip rect from the enclosing layer and frame view.
-    RenderLayer* layer = m_element->renderer()->enclosingLayer();
+    RenderLayer* layer = 0;
+    RenderObject* renderer = m_element->renderer();
+    if (renderer) {
+        layer = m_element->renderer()->enclosingLayer();
+    }
     FrameView* parentView = m_element->document()->view();
     clipRect.intersect(parentView->windowClipRectForLayer(layer, true));
 
