--- a/main/Makefile.am
+++ b/main/Makefile.am
@@ -37,7 +37,6 @@ enigma2_LDADD = \
 	@BASE_LIBS@ \
 	@LIBGIF_LIBS@ \
 	@LIBJPEG_LIBS@ \
-	@LIBVUGLES2_LIBS@ \
 	@LIBSDL_LIBS@ \
 	@LIBXINE_LIBS@ \
 	@PTHREAD_LIBS@ \
@@ -49,7 +48,8 @@ enigma2_LDADD = \
 	@DVBCSA_LIBS@ \
 	@AVAHI_LIBS@ \
 	@LIBHIACCEL_LIBS@ \
-	-ltuxtxt32bpp -lcrypt -lrt -ldl
+	-ltuxtxt32bpp -lcrypt -lrt -ldl \
+	-lv3ddriver @LIBVUGLES2_LIBS@
 
 enigma2_LDFLAGS = -Wl,--export-dynamic
 
