diff --git a/pythonwifi/iwlibs.py b/pythonwifi/iwlibs.py
index 495f7ed..2d8de8b 100644
--- a/pythonwifi/iwlibs.py
+++ b/pythonwifi/iwlibs.py
@@ -1478,13 +1478,13 @@ class Iwscan(object):
                 if scanresult:
                     aplist.append(scanresult)
                 scanresult = Iwscanresult(
-                        data[pythonwifi.flags.IW_EV_LCP_PK_LEN:length],
+                        data[pythonwifi.flags.IW_EV_LCP_PK_LEN+4:length],
                         self.range)
             elif scanresult is None:
                 raise RuntimeError("Attempting to add an event without AP data.")
             else:
                 scanresult.addEvent(cmd,
-                        data[pythonwifi.flags.IW_EV_LCP_PK_LEN:length])
+                        data[pythonwifi.flags.IW_EV_LCP_PK_LEN+4:length])
             # We're finished with the previous event
             data = data[length:]
 
-- 
2.26.2.windows.1
