--- a/sysdeps/unix/sysv/linux/sys/ptrace.h
+++ b/sysdeps/unix/sysv/linux/sys/ptrace.h
@@ -197,6 +197,7 @@ enum __ptrace_eventcodes
   PTRACE_EVENT_SECCOMP  = 7
 };
 
+#ifndef PTRACE_PEEKSIGINFO
 /* Arguments for PTRACE_PEEKSIGINFO.  */
 struct __ptrace_peeksiginfo_args
 {
@@ -204,6 +205,7 @@ struct __ptrace_peeksiginfo_args
   __uint32_t flags;	/* Flags for peeksiginfo.  */
   __int32_t nr;		/* How many siginfos to take.  */
 };
+#endif
 
 enum __ptrace_peeksiginfo_flags
 {
