diff --git a/gst/rtsp/gstrtspsrc.c b/gst/rtsp/gstrtspsrc.c
index b6e313c..7ecc04d 100644
--- a/gst/rtsp/gstrtspsrc.c
+++ b/gst/rtsp/gstrtspsrc.c
@@ -5351,12 +5351,13 @@ gst_rtspsrc_setup_streams (GstRTSPSrc * src, gboolean async)
         case GST_RTSP_LOWER_TRANS_UDP_MCAST:
           /* only allow multicast for other streams */
           GST_DEBUG_OBJECT (src, "stream %p as UDP multicast", stream);
-          protocols = GST_RTSP_LOWER_TRANS_UDP_MCAST;
+          //
+          //protocols = GST_RTSP_LOWER_TRANS_UDP_MCAST;
           break;
         case GST_RTSP_LOWER_TRANS_UDP:
           /* only allow unicast for other streams */
           GST_DEBUG_OBJECT (src, "stream %p as UDP unicast", stream);
-          protocols = GST_RTSP_LOWER_TRANS_UDP;
+          //protocols = GST_RTSP_LOWER_TRANS_UDP;
           break;
         default:
           GST_DEBUG_OBJECT (src, "stream %p unknown transport %d", stream,
