--- a/system/addon-manifest.xml
+++ b/system/addon-manifest.xml
@@ -42,5 +42,7 @@
   <addon>xbmc.pvr</addon>
   <addon>xbmc.python</addon>
   <addon>xbmc.webinterface</addon>
+  <addon>visualization.glspectrum</addon>
+  <addon>visualization.waveform</addon> 
   <addon optional="true">peripheral.joystick</addon>
 </addons>
--- a/configure.ac	2017-09-23 22:38:26.000000000 +0300
+++ b/configure.ac	2017-09-23 22:42:05.088757853 +0300
@@ -2295,6 +2295,9 @@
     xbmc/freebsd/Makefile \
     xbmc/linux/Makefile \
     xbmc/filesystem/Makefile \
+    xbmc/visualizations/OpenGLSpectrum/Makefile \
+    xbmc/visualizations/WaveForm/Makefile \
+    xbmc/visualizations/EGLHelpers/Makefile \
     xbmc/windowing/Makefile \
     xbmc/windowing/egl/Makefile \
     lib/addons/library.xbmc.addon/Makefile \
--- a/Makefile.in	2017-09-23 22:38:26.000000000 +0300
+++ b/Makefile.in	2017-09-23 23:04:51.852708197 +0300
@@ -136,6 +136,7 @@
 ifeq (@USE_OPENGLES@,1)
 DIRECTORY_ARCHIVES += xbmc/rendering/gles/rendering_gles.a
 DIRECTORY_ARCHIVES += xbmc/windowing/egl/windowing_egl.a
+DIRECTORY_ARCHIVES += xbmc/visualizations/EGLHelpers/eglhelpers.a
 endif
 
 ifeq (@USE_UPNP@,1)
@@ -216,6 +217,18 @@
 	lib/cpluff \
 	lib/xbmc-dll-symbols
 
+ifneq (@DISABLE_SPECTRUM@,1)
+  VIS_DIRS+= xbmc/visualizations/OpenGLSpectrum
+else
+  INSTALL_FILTER+= .*visualization\.glspectrum.*
+endif
+
+ifneq (@DISABLE_WAVEFORM@,1)
+  VIS_DIRS+= xbmc/visualizations/WaveForm
+else
+  INSTALL_FILTER+= .*visualization\.waveform.*
+endif
+
 LIBADDON_DIRS=\
 	lib/addons/library.xbmc.addon \
 	lib/addons/library.kodi.adsp \
@@ -296,7 +309,7 @@
 
 include Makefile.include
 
-.PHONY : dllloader exports eventclients \
+.PHONY : dllloader exports visualizations eventclients \
 	dvdpcodecs dvdpextcodecs codecs externals force skins libaddon check \
 	testframework testsuite
 
@@ -340,6 +353,18 @@
 	$(MAKE) -C xbmc/cores/DllLoader/exports wrapper.def
 dllloader: exports xbmc/cores/DllLoader/dllloader.a
 
+ifeq (@USE_OPENGLES@,1)
+EGLHELPERS=xbmc/visualizations/EGLHelpers/eglhelpers.a
+endif
+
+$(VIS_DIRS): force exports $(EGLHELPERS)
+	$(MAKE) -C $@
+
+$(SS_DIRS): force exports
+	$(MAKE) -C $@
+
+visualizations: $(VIS_DIRS)
+
 libaddon: exports
 	$(MAKE) -C lib/addons/library.xbmc.addon
 	$(MAKE) -C lib/addons/library.kodi.adsp
@@ -376,10 +401,10 @@
 
 libs: $(LIBSSE4) libexif system/libcpluff-@ARCH@.so
 
-externals: codecs libs libaddon
+externals: codecs libs libaddon visualizations
 
 xcode_depends: \
-	codecs libs eventclients skins libaddon certs
+	codecs libs eventclients visualizations skins libaddon certs
 
 DYNOBJSXBMC= \
 	xbmc/linux/linux.a \
