--- a/rwpng.c	2014-02-09 00:13:20.539665942 +0100
+++ b/rwpng.c	2014-02-09 00:14:03.487665305 +0100
@@ -318,10 +318,8 @@
 
     struct rwpng_write_data write_data;
     if (mainprog_ptr->maximum_file_size) {
-        write_data = (struct rwpng_write_data){
-            .buffer = malloc(mainprog_ptr->maximum_file_size),
-            .bytes_left = mainprog_ptr->maximum_file_size,
-        };
+        write_data.buffer = malloc(mainprog_ptr->maximum_file_size);
+		write_data.bytes_left = mainprog_ptr->maximum_file_size;
         if (!write_data.buffer) return PNG_OUT_OF_MEMORY_ERROR;
         png_set_write_fn(png_ptr, &write_data, user_write_data, user_flush_data);
     } else {
