From 0384f028257826745fbd11a3b64be9e943d0752a Mon Sep 17 00:00:00 2001
From: Captain <captain.onboard@web.de>
Date: Thu, 28 Oct 2021 22:53:53 +0200
Subject: [PATCH 3/3] dont-enable-AES-hwcaps-for-arm

---
 node/AES.hpp   | 3 ---
 node/Utils.cpp | 9 ---------
 2 files changed, 12 deletions(-)

diff --git a/node/AES.hpp b/node/AES.hpp
index dc0130ad..0ff4f3b7 100644
--- a/node/AES.hpp
+++ b/node/AES.hpp
@@ -24,9 +24,6 @@
 #if !defined(ZT_AES_NO_ACCEL) && defined(ZT_ARCH_X64)
 #define ZT_AES_AESNI 1
 #endif
-#if !defined(ZT_AES_NO_ACCEL) && defined(ZT_ARCH_ARM_HAS_NEON) && defined(ZT_ARCH_ARM_HAS_CRYPTO)
-#define ZT_AES_NEON 1
-#endif
 
 #ifndef ZT_INLINE
 #define ZT_INLINE inline
diff --git a/node/Utils.cpp b/node/Utils.cpp
index a0bf0707..d8b49a81 100644
--- a/node/Utils.cpp
+++ b/node/Utils.cpp
@@ -116,15 +116,6 @@ Utils::ARMCapabilities::ARMCapabilities() noexcept
 		this->pmull = (hwcaps2 & HWCAP2_PMULL) != 0;
 		this->sha1 = (hwcaps2 & HWCAP2_SHA1) != 0;
 		this->sha2 = (hwcaps2 & HWCAP2_SHA2) != 0;
-	} else {
-#endif
-		const long hwcaps = getauxval(AT_HWCAP);
-		this->aes = (hwcaps & HWCAP_AES) != 0;
-		this->crc32 = (hwcaps & HWCAP_CRC32) != 0;
-		this->pmull = (hwcaps & HWCAP_PMULL) != 0;
-		this->sha1 = (hwcaps & HWCAP_SHA1) != 0;
-		this->sha2 = (hwcaps & HWCAP_SHA2) != 0;
-#ifdef HWCAP2_AES
 	}
 #endif
 
-- 
2.33.0.windows.2

