diff --git a/gdk/directfb/gdkdisplay-directfb.c b/gdk/directfb/gdkdisplay-directfb.c
index d4dfbef..9e7086c 100644
--- a/gdk/directfb/gdkdisplay-directfb.c
+++ b/gdk/directfb/gdkdisplay-directfb.c
@@ -102,6 +102,10 @@ gdk_display_open (const gchar *display_name)
   ret = directfb->GetInputDevice (directfb, DIDID_KEYBOARD, &keyboard);
   if (ret != DFB_OK)
     {
+      ret = directfb->GetInputDevice (directfb, DIDID_REMOTE, &keyboard);
+    }
+  if (ret != DFB_OK)
+    {
       DirectFBError ("gdk_display_open: GetInputDevice", ret);
       directfb->Release (directfb);
       _gdk_display->directfb = NULL;
diff --git a/gdk/directfb/gdkkeys-directfb.c b/gdk/directfb/gdkkeys-directfb.c
index e775384..7124875 100644
--- a/gdk/directfb/gdkkeys-directfb.c
+++ b/gdk/directfb/gdkkeys-directfb.c
@@ -1385,6 +1385,22 @@ static const struct gdk_key
   { GDK_Hyper_L, "Hyper_L" },
   { GDK_Hyper_R, "Hyper_R" },
   { GDK_Delete, "Delete" },
+  /* Add HBBTV KEY CODES */
+  { GDK_Red, "Red_Key" },
+  { GDK_Green, "Green_Key" },
+  { GDK_Yellow, "Yellow_Key" },
+  { GDK_Blue, "Blue_Key" },
+  { GDK_Play, "Play" },
+  { GDK_Play_Pause, "Play_Pause" },
+  { GDK_Fast_Forward, "Fast_Forward" },
+  { GDK_Rewind, "Rewind" },
+  { GDK_Text, "Text" },
+  { GDK_Exit, "Exit" },
+  { GDK_Channel_Up, "Channel_Up" },
+  { GDK_Channel_Down, "Channel_Down" },
+  { GDK_Volume_Up, "Volume_Up" },
+  { GDK_Volume_Down, "Volume_Down" },
+  { GDK_Record, "Record" },
   { GDK_VoidSymbol, "VoidSymbol" }
 };
 
@@ -1586,6 +1602,29 @@ gdk_directfb_translate_key (DFBInputDeviceKeyIdentifier key_id,
             case DIKS_NEXT:          keyval = GDK_Next;      break;
             case DIKS_BEGIN:         keyval = GDK_Begin;     break;
             case DIKS_BREAK:         keyval = GDK_Break;     break;
+			/* Add HBBTV KEY CODES */
+            case DIKS_OK:            keyval = GDK_Return;    break;
+            case DIKS_RED:           keyval = GDK_Red;       break;
+            case DIKS_GREEN:         keyval = GDK_Green;     break;
+            case DIKS_YELLOW:        keyval = GDK_Yellow;    break;
+            case DIKS_BLUE:          keyval = GDK_Blue;      break;
+            case DIKS_PLAY:          keyval = GDK_Play;      break;
+            case DIKS_PLAYPAUSE:     keyval = GDK_Play_Pause;      break;
+            case DIKS_STOP:          keyval = GDK_Stop;      break;
+            case DIKS_FASTFORWARD:   keyval = GDK_Fast_Forward;    break;
+            case DIKS_REWIND:        keyval = GDK_Rewind;    break;
+            //case DIKS_NEXT:          keyval = GDK_Next;      break;
+            //case DIKS_PREV:          keyval = GDK_Prev;      break;
+            case DIKS_BACK:          keyval = GDK_Back;      break;
+            case DIKS_TEXT:          keyval = GDK_Text;      break;
+            case DIKS_EXIT:          keyval = GDK_Exit;      break;
+            case DIKS_CHANNEL_UP:    keyval = GDK_Channel_Up;      break;
+            case DIKS_CHANNEL_DOWN:  keyval = GDK_Channel_Down;    break;
+            case DIKS_VOLUME_UP:     keyval = GDK_Volume_Up;       break;
+            case DIKS_VOLUME_DOWN:   keyval = GDK_Volume_Down;     break;
+            case DIKS_MUTE:          keyval = GDK_Volume_Mute;     break;
+            case DIKS_RECORD:        keyval = GDK_Record;    break;
+            case DIKS_TV:            keyval = GDK_Tv;        break;
             default:
               break;
             }
diff --git a/gdk/gdkkeysyms-compat.h b/gdk/gdkkeysyms-compat.h
index 4dbec9d..aba0347 100644
--- a/gdk/gdkkeysyms-compat.h
+++ b/gdk/gdkkeysyms-compat.h
@@ -2204,5 +2204,20 @@
 #define GDK_ClearGrab 0x1008fe21
 #define GDK_Next_VMode 0x1008fe22
 #define GDK_Prev_VMode 0x1008fe23
+/* Add HBBTV KEY CODES */
+#define GDK_Play 0x10090000
+#define GDK_Play_Pause 0x10090001
+#define GDK_Fast_Forward 0x10090002
+#define GDK_Rewind 0x10090003
+#define GDK_Prev 0x10090004
+#define GDK_Text 0x10090005
+#define GDK_Exit 0x10090006
+#define GDK_Channel_Up 0x10090007
+#define GDK_Channel_Down 0x10090008
+#define GDK_Volume_Up 0x10090009
+#define GDK_Volume_Down 0x1009000a
+#define GDK_Volume_Mute 0x1009000b
+#define GDK_Record 0x1009000c
+#define GDK_Tv 0x1009000d
 
 #endif /* __GDK_KEYSYMS_COMPAT_H__ */
diff --git a/gdk/gdkkeysyms.h b/gdk/gdkkeysyms.h
index 26fa1c2..8479af7 100644
--- a/gdk/gdkkeysyms.h
+++ b/gdk/gdkkeysyms.h
@@ -2217,5 +2217,20 @@
 #define GDK_KEY_ClearGrab 0x1008fe21
 #define GDK_KEY_Next_VMode 0x1008fe22
 #define GDK_KEY_Prev_VMode 0x1008fe23
+/* Add HBBTV KEY CODES */
+#define GDK_KEY_Play 0x10090000
+#define GDK_KEY_Play_Pause 0x10090001
+#define GDK_KEY_Fast_Forward 0x10090002
+#define GDK_KEY_Rewind 0x10090003
+#define GDK_KEY_Prev 0x10090004
+#define GDK_KEY_Text 0x10090005
+#define GDK_KEY_Exit 0x10090006
+#define GDK_KEY_Channel_Up 0x10090007
+#define GDK_KEY_Channel_Down 0x10090008
+#define GDK_KEY_Volume_Up 0x10090009
+#define GDK_KEY_Volume_Down 0x1009000a
+#define GDK_KEY_Volume_Mute 0x1009000b
+#define GDK_KEY_Record 0x1009000c
+#define GDK_KEY_Tv 0x1009000d
 
 #endif /* __GDK_KEYSYMS_H__ */
