--- a/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESink.cpp	2020-03-01 07:43:27.967726537 +0000
+++ b/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESink.cpp	2020-03-01 07:43:18.995741067 +0000
@@ -678,6 +678,7 @@
   unsigned int c_retry = 4;
   m_sinkInfoList.clear();
   CAESinkFactory::EnumerateEx(m_sinkInfoList, false);
+#if !defined(TARGET_DAGS)
   while (m_sinkInfoList.empty() && c_retry > 0)
   {
     CLog::Log(LOGNOTICE, "No Devices found - retry: %d", c_retry);
@@ -686,6 +687,7 @@
     // retry the enumeration
     CAESinkFactory::EnumerateEx(m_sinkInfoList, true);
   }
+#endif
   CLog::Log(LOGNOTICE, "Found %lu Lists of Devices", m_sinkInfoList.size());
   PrintSinks();
 }
