--- a/parse-base-block.cpp
+++ b/parse-base-block.cpp
@@ -773,7 +773,7 @@ static void detailed_display_range_limits(edid_state &state, const unsigned char
 	int v_max_offset = 0, v_min_offset = 0;
 	int is_cvt = 0;
 	bool has_sec_gtf = false;
-	std::string range_class;
+	const char *range_class = "";
 
 	state.cur_block = "Display Range Limits";
 	printf("%s\n", state.cur_block.c_str());
@@ -833,8 +833,8 @@ static void detailed_display_range_limits(edid_state &state, const unsigned char
 		}
 		break;
 	default: /* invalid */
-		fail("Unknown range class (0x%02x)\n", x[10]);
-		range_class = std::string("Unknown (") + utohex(x[10]) + ")";
+		fail("Invalid range class 0x%02x\n", x[10]);
+		range_class = "invalid";
 		break;
 	}
 
@@ -847,7 +847,7 @@ static void detailed_display_range_limits(edid_state &state, const unsigned char
 	state.min_display_hor_freq_hz = (x[7] + h_min_offset) * 1000;
 	state.max_display_hor_freq_hz = (x[8] + h_max_offset) * 1000;
 	printf("  Monitor ranges (%s): %d-%d Hz V, %d-%d kHz H",
-	       range_class.c_str(),
+	       range_class,
 	       x[5] + v_min_offset, x[6] + v_max_offset,
 	       x[7] + h_min_offset, x[8] + h_max_offset);
 	if (x[9]) {
