--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -432,7 +432,7 @@ if (FMT_INSTALL)
   include(CMakePackageConfigHelpers)
   set_verbose(
     FMT_CMAKE_DIR
-    ${CMAKE_INSTALL_LIBDIR}/cmake/fmt
+    ${BASE_LIB_PATH}/cmake/fmt
     CACHE
     STRING
     "Installation directory for cmake files, a relative path that "
@@ -444,13 +444,13 @@ if (FMT_INSTALL)
   set(targets_export_name fmt-targets)
 
   set_verbose(
-    FMT_LIB_DIR ${CMAKE_INSTALL_LIBDIR} CACHE STRING
+    FMT_LIB_DIR ${BASE_LIB_PATH} CACHE STRING
     "Installation directory for libraries, a relative path that "
     "will be joined to ${CMAKE_INSTALL_PREFIX} or an absolute path.")
 
   set_verbose(
     FMT_PKGCONFIG_DIR
-    ${CMAKE_INSTALL_LIBDIR}/pkgconfig
+    ${BASE_LIB_PATH}/pkgconfig
     CACHE
     STRING
     "Installation directory for pkgconfig (.pc) files, a relative "
