From bc0ae5abdf84e0905d9962ca664c17505747293d Mon Sep 17 00:00:00 2001
From: atvcaptain <openatv@gmail.com>
Date: Sat, 19 Apr 2025 16:32:34 +0200
Subject: [PATCH] fix error: reference to 'size' is ambiguous

---
 tsniv2ni.cpp | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/tsniv2ni.cpp b/tsniv2ni.cpp
index b1c763f..619c76f 100644
--- a/tsniv2ni.cpp
+++ b/tsniv2ni.cpp
@@ -25,7 +25,6 @@ using namespace std;
 
 istream* inbuf;
 ostream* outbuf;
-int size;
 bool active=false;
 const char fsync0[]="\xff\x07\x3a\xb6";
 const char fsync1[]="\xff\xf8\xc5\x49";
@@ -62,7 +61,7 @@ int main(int argc, char** argv) {
 	}
 	inbuf = &cin;
 	outbuf = &cout;
-	size=200;
+	int size=200;
 	char* packets= new char[size*188];
 	frpd= new char[6140];
 	memset(frpd,0x55,6140);
-- 
2.45.2.windows.1

