From db5eb12e8c40b07a501d9f290844b0389843c484 Mon Sep 17 00:00:00 2001
From: Stephan Reichholf <reichi@opendreambox.org>
Date: Wed, 30 May 2012 11:08:18 +0200
Subject: [PATCH] QtGui: add Q_GUI_EXPORT to QUpdateLaterEvent

---
 src/gui/kernel/qevent.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/gui/kernel/qevent.h b/src/gui/kernel/qevent.h
index 74078c7..2a995ca 100644
--- a/src/gui/kernel/qevent.h
+++ b/src/gui/kernel/qevent.h
@@ -319,7 +319,7 @@ protected:
     bool m_erased;
 };
 
-class QUpdateLaterEvent : public QEvent
+class Q_GUI_EXPORT QUpdateLaterEvent : public QEvent
 {
 public:
     QUpdateLaterEvent(const QRegion& paintRegion);
-- 
1.7.9.5

