From e9690354e7975d0ea18daf8b89cb88367f1a7ab2 Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Wed, 1 Aug 2018 10:02:42 +0200
Subject: [PATCH 4/4] makefile disable warnings


diff --git a/Makefile b/Makefile
index 19329844..3ce5dcab 100644
--- a/Makefile
+++ b/Makefile
@@ -578,6 +578,8 @@ endif
 # This warning generated too much noise in a regular build.
 # Use make W=1 to enable this warning (see scripts/Makefile.build)
 KBUILD_CFLAGS += $(call cc-disable-warning, unused-but-set-variable)
+KBUILD_CFLAGS += $(call cc-disable-warning, attribute-alias)
+KBUILD_CFLAGS += $(call cc-disable-warning, packed-not-aligned)
 
 ifdef CONFIG_FRAME_POINTER
 KBUILD_CFLAGS	+= -fno-omit-frame-pointer -fno-optimize-sibling-calls
-- 
2.17.1

