--- a/lib/argv.c
+++ b/lib/argv.c
@@ -61,12 +61,13 @@ void freeargv (char **vector);
 #else	/* !ANSI_PROTOTYPES */
 
 #if !defined _WIN32 || defined __GNUC__
-extern char *memcpy ();		/* Copy memory region */
-extern int strlen ();		/* Count length of string */
-extern char *malloc ();		/* Standard memory allocater */
-extern char *realloc ();	/* Standard memory reallocator */
+extern void *memcpy ();		/* Copy memory region */
+extern unsigned int strlen ();		/* Count length of string */
+extern void *malloc ();		/* Standard memory allocater */
+extern void *realloc ();	/* Standard memory reallocator */
 extern void free ();		/* Free malloc'd memory */
 extern char *strdup ();		/* Duplicate a string */
+extern char *strcpy ();		/* Copy a string */
 #endif
 
 #endif	/* ANSI_PROTOTYPES */
