diff --git a/common/mp4v2/atom_standard.cpp b/common/mp4v2/atom_standard.cpp
index 2bd95f0..f513422 100644
--- a/common/mp4v2/atom_standard.cpp
+++ b/common/mp4v2/atom_standard.cpp
@@ -22,15 +22,15 @@
 #include "mp4common.h"
 #include "atoms.h"
 
-static const char name[5]={0251,'n', 'a', 'm', '\0'};
-static const char art[5]={0251,'A', 'R', 'T', '\0'};
-static const char wrt[5]={0251,'w', 'r', 't', '\0'};
-static const char alb[5]={0251,'a', 'l', 'b', '\0'};
-static const char day[5]={0251,'d', 'a', 'y', '\0'};
-static const char too[5]={0251,'t', 'o', 'o', '\0'};
-static const char cmt[5]={0251,'c', 'm', 't', '\0'};
-static const char gen[5]={0251,'g', 'e', 'n', '\0'};
-static const char grp[5]={0251,'g', 'r', 'p', '\0'};
+static const char name[5]={ (char)0251,'n', 'a', 'm', '\0'};
+static const char art[5]={(char)0251, (char)'A', (char)'R', (char)'T', (char)'\0'};
+static const char wrt[5]={(char)0251, (char)'w', (char)'r', (char)'t', (char)'\0'};
+static const char alb[5]={(char)0251, (char)'a', (char)'l', (char)'b', (char)'\0'};
+static const char day[5]={(char)0251, (char)'d', (char)'a', (char)'y', (char)'\0'};
+static const char too[5]={(char)0251, (char)'t', (char)'o', (char)'o', (char)'\0'};
+static const char cmt[5]={(char)0251, (char)'c', (char)'m', (char)'t', (char)'\0'};
+static const char gen[5]={(char)0251, (char)'g', (char)'e', (char)'n', (char)'\0'};
+static const char grp[5]={(char)0251, (char)'g', (char)'r', (char)'p', (char)'\0'};
 
 MP4StandardAtom::MP4StandardAtom (const char *type) : MP4Atom(type)
 {
diff --git a/common/mp4v2/mp4atom.cpp b/common/mp4v2/mp4atom.cpp
index 97b13e0..2bc169d 100644
--- a/common/mp4v2/mp4atom.cpp
+++ b/common/mp4v2/mp4atom.cpp
@@ -264,11 +264,11 @@ MP4Atom* MP4Atom::CreateAtom(const char* type)
       }
       break;
     case 0251:
-      static const char name[5]={0251,'n', 'a', 'm', '\0'};
-      static const char cmt[5]={0251,'c', 'm', 't', '\0'};
-      static const char cpy[5]={0251,'c', 'p', 'y', '\0'};
-      static const char des[5]={0251,'d', 'e', 's','\0'};
-      static const char prd[5]={0251, 'p', 'r', 'd', '\0'};
+      static const char name[5]={(char)0251, (char)'n', (char)'a', (char)'m', (char) (char)'\0'};
+      static const char cmt[5]={(char)0251, (char)'c', (char)'m', (char)'t', (char)'\0'};
+      static const char cpy[5]={(char)0251, (char)'c', (char)'p', (char)'y', (char)'\0'};
+      static const char des[5]={(char)0251, (char)'d', (char)'e', (char)'s', (char)'\0'};
+      static const char prd[5]={(char)0251, (char)'p', (char)'r', (char)'d', (char)'\0'};
       if (ATOMID(type) == ATOMID(name) ||
 	  ATOMID(type) == ATOMID(cmt) ||
 	  ATOMID(type) == ATOMID(cpy) ||
