Index: linux-3.6.0/drivers/media/dvb/siano/sms-cards.c
===================================================================
--- linux-3.6.0.orig/drivers/media/dvb/siano/sms-cards.c
+++ linux-3.6.0/drivers/media/dvb/siano/sms-cards.c
@@ -294,6 +294,8 @@ EXPORT_SYMBOL_GPL(sms_board_lna_control)
 int sms_board_load_modules(int id)
 {
 	switch (id) {
+	case SMS1XXX_BOARD_SIANO_NOVA_A:
+	case SMS1XXX_BOARD_SIANO_NOVA_B:
 	case SMS1XXX_BOARD_HAUPPAUGE_CATAMOUNT:
 	case SMS1XXX_BOARD_HAUPPAUGE_OKEMO_A:
 	case SMS1XXX_BOARD_HAUPPAUGE_OKEMO_B:
