From b6004430cea2c368e213638e6b42ea7779ffaf0f Mon Sep 17 00:00:00 2001
From: atvcaptain <openatv@gmail.com>
Date: Wed, 30 Apr 2025 11:18:58 +0200
Subject: [PATCH 12/15] avoid duplicate STB window-system destruction

---
 xbmc/windowing/egl/WinSystemstb.cpp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/xbmc/windowing/egl/WinSystemstb.cpp b/xbmc/windowing/egl/WinSystemstb.cpp
index 896ad21d1a..036751e692 100644
--- a/xbmc/windowing/egl/WinSystemstb.cpp
+++ b/xbmc/windowing/egl/WinSystemstb.cpp
@@ -201,8 +201,8 @@ bool CWinSystemstb::DestroyWindowSystem()
   delete m_egl;
   m_egl = NULL;
 
-  CWinSystemBase::DestroyWindowSystem();
-  //return true;
+//  CWinSystemBase::DestroyWindowSystem();
+  return true;
 }
 bool CWinSystemstb::CreateNewWindow(const std::string& name,
                                     bool fullScreen,
-- 
2.45.2.windows.1
