diff --git a/drivers/media/tuners/tda18271-fe.c b/drivers/media/tuners/tda18271-fe.c
index 72c26fd..229584e 100644
--- a/drivers/media/tuners/tda18271-fe.c
+++ b/drivers/media/tuners/tda18271-fe.c
@@ -1337,6 +1337,16 @@ struct dvb_frontend *tda18271_attach(struct dvb_frontend *fe, u8 addr,
 	memcpy(&fe->ops.tuner_ops, &tda18271_tuner_ops,
 	       sizeof(struct dvb_tuner_ops));
 
+	if (fe->ops.delsys[0] == SYS_UNDEFINED) {
+		fe->ops.delsys[0] = SYS_DVBC_ANNEX_A;
+		fe->ops.delsys[1] = SYS_DVBC_ANNEX_B;
+		fe->ops.delsys[2] = SYS_DVBC_ANNEX_C;
+		fe->ops.delsys[3] = SYS_ATSC;
+		fe->ops.delsys[4] = SYS_ISDBT;
+		fe->ops.delsys[5] = SYS_DVBT;
+		fe->ops.delsys[6] = SYS_DVBT2;
+	}
+
 	if (tda18271_debug & (DBG_MAP | DBG_ADV))
 		tda18271_dump_std_map(fe);
 
