--- a/include/ioctl_cfg80211.h	2016-03-30 12:28:34.285626563 +0200
+++ b/include/ioctl_cfg80211.h	2016-03-30 12:28:02.181625540 +0200
@@ -107,7 +107,7 @@
   #define rtw_cfg80211_rx_mgmt(adapter, freq, sig_dbm, buf, len, gfp) cfg80211_rx_mgmt((adapter)->pnetdev, freq, buf, len, gfp)
 #elif (LINUX_VERSION_CODE < KERNEL_VERSION(3,6,0)) && (LINUX_VERSION_CODE >= KERNEL_VERSION(3,4,0))
   #define rtw_cfg80211_rx_mgmt(adapter, freq, sig_dbm, buf, len, gfp) cfg80211_rx_mgmt((adapter)->pnetdev, freq, sig_dbm, buf, len, gfp)
-#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(3,12,0)) && (LINUX_VERSION_CODE <= KERNEL_VERSION(3,17,0))
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(3,12,0)) && (LINUX_VERSION_CODE <= KERNEL_VERSION(3,17,3))
   #define rtw_cfg80211_rx_mgmt(adapter, freq, sig_dbm, buf, len, gfp) cfg80211_rx_mgmt((adapter)->rtw_wdev, freq, sig_dbm, buf, len, 0, gfp)
 #else
   #define rtw_cfg80211_rx_mgmt(adapter, freq, sig_dbm, buf, len, gfp) cfg80211_rx_mgmt((adapter)->rtw_wdev, freq, sig_dbm, buf, len, gfp)
