--- a/makefile.unix	2009-05-03 07:18:45.000000000 +0200
+++ b/makefile.unix	2010-03-15 13:31:31.000000000 +0100
@@ -9,8 +9,8 @@
 #CXX=g++
 #CXXFLAGS=-O2
 DEFINES=-D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE
-STRIP=strip
-DESTDIR=/usr
+#STRIP=strip
+#DESTDIR=/usr
 
 # Linux using LCC
 #CXX=lcc
@@ -136,6 +136,7 @@
 	$(LINK) -shared -o libunrar.so $(LDFLAGS) $(OBJECTS) $(LIB_OBJ)
 
 install-unrar:
+			install -d $(DESTDIR)/bin
 			install unrar $(DESTDIR)/bin
 
 uninstall-unrar:
