diff --git a/gst/flv/gstflvdemux.c b/gst/flv/gstflvdemux.c
index 203cb76..a8d4005 100644
--- a/gst/flv/gstflvdemux.c
+++ b/gst/flv/gstflvdemux.c
@@ -694,6 +694,9 @@ gst_flv_demux_audio_negotiate (GstFlvDemux * demux, guint32 codec_tag,
       caps = gst_caps_new_simple ("audio/mpeg",
           "mpegversion", G_TYPE_INT, 4, "framed", G_TYPE_BOOLEAN, TRUE,
           "stream-format", G_TYPE_STRING, "raw", NULL);
+      if (demux->audio_codec_data &&
+          GST_BUFFER_SIZE (demux->audio_codec_data) >= 2)
+        gst_codec_utils_aac_caps_set_level_and_profile (caps, GST_BUFFER_DATA (demux->audio_codec_data), GST_BUFFER_SIZE (demux->audio_codec_data));
       break;
     }
     case 7:
