diff --git a/oscam.c b/oscam.c
index d014cd1..f948b69 100644
--- a/oscam.c
+++ b/oscam.c
@@ -383,7 +383,19 @@ static void write_versionfile(bool use_stdout)
 				st.tm_hour, st.tm_min, st.tm_sec);
 	}

-	fprintf(fp, "Version:        oscam-%s-r%s\n", CS_VERSION, CS_SVN_VERSION);
+	fprintf(fp, "Version:        oscam-%s-r%s%s\n", CS_VERSION, CS_SVN_VERSION, "- build by kitte888");
+		fprintf(fp, "S3 mit             cc_opts=-ggdb -pipe -ffunction-sections -fdata-sections -funroll-loops -fomit-frame-pointer -fno-schedule-insns\n");
+		fprintf(fp, "compiler_option    -O1   \n");
+		fprintf(fp, "compress           nein   \n");
+		fprintf(fp, "Toolchain          arm_dream  standart aus S3 \n");
+		fprintf(fp, "Bibliothek         SSL_102\n");
+		fprintf(fp, "Bibliothek         LIB_USBB\n");
+		fprintf(fp, "Bibliothek         LIB_PCSC\n");
+		fprintf(fp, "Patch              EMU\n");
+		fprintf(fp, "Patch              ICAM V9\n");
+		fprintf(fp, "Patch Ton          Buffer V3\n");
+		fprintf(fp, "Patch Demuxer-err  nein \n");
+		fprintf(fp, "Neon               ja / 128 \n");
 	fprintf(fp, "Compiler:       %s\n", CS_TARGET);
 	fprintf(fp, "Box type:       %s (%s)\n", boxtype_get(), boxname_get());
 	fprintf(fp, "PID:            %d\n", getppid());

