--- a/avahi-core/iface.c	2016-05-05 19:34:19.000000000 +0200
+++ b/avahi-core/iface.c	2016-05-05 16:58:16.000000000 +0200
@@ -71,7 +71,7 @@
              (a->interface->protocol == AVAHI_PROTO_INET6 && m->server->config.publish_aaaa_on_ipv4) ?
              AVAHI_PROTO_UNSPEC : a->interface->protocol;
 
-        if (avahi_s_entry_group_is_empty(a->entry_mdns_group)) {
+        if (avahi_interface_address_is_relevant_mdns(a) && avahi_s_entry_group_is_empty(a->entry_mdns_group)) {
             char t[AVAHI_ADDRESS_STR_MAX];
 
             avahi_address_snprint(t, sizeof(t), &a->address);
@@ -91,9 +91,9 @@
             char t[AVAHI_ADDRESS_STR_MAX];
 
             avahi_address_snprint(t, sizeof(t), &a->address);
-            avahi_log_info("Registering new LLMNR address record for %s on %s.%s.", t, a->interface->hardware->name, p == AVAHI_PROTO_UNSPEC ? "*" : avahi_proto_to_string(p));
+            avahi_log_info("Registering new LLMNR address record for %s on %s.%s.", t, a->interface->hardware->name, "*");
 
-            if (avahi_server_add_address(m->server, a->entry_llmnr_group, a->interface->hardware->index, p, AVAHI_PUBLISH_USE_LLMNR, NULL, &a->address) < 0) {
+            if (avahi_server_add_address(m->server, a->entry_llmnr_group, a->interface->hardware->index, AVAHI_PROTO_UNSPEC, AVAHI_PUBLISH_USE_LLMNR, NULL, &a->address) < 0) {
                 avahi_log_warn(__FILE__": avahi_server_add_address() failed: %s", avahi_strerror(m->server->error));
                 avahi_s_entry_group_free(a->entry_llmnr_group);
                 a->entry_llmnr_group = NULL;
@@ -897,6 +897,30 @@
     if (a->global_scope && !a->deprecated)
         return 1;
 
+    if (!a->global_scope && !a->deprecated)
+        return 1;
+
+    /* Publish link-local and deprecated IP addresses only if they are
+     * the only ones on the link */
+    for (b = a->interface->addresses; b; b = b->address_next) {
+        if (b == a)
+            continue;
+
+        if (b->global_scope && !b->deprecated)
+            return 0;
+    }
+
+    return 1;
+}
+
+int avahi_interface_address_is_relevant_mdns(AvahiInterfaceAddress *a) {
+    AvahiInterfaceAddress *b;
+    assert(a);
+
+    /* Publish public and non-deprecated IP addresses */
+    if (a->global_scope && !a->deprecated)
+        return 1;
+
     /* Publish link-local and deprecated IP addresses only if they are
      * the only ones on the link */
     for (b = a->interface->addresses; b; b = b->address_next) {
--- a/avahi-core/iface.h	2016-05-05 19:34:19.000000000 +0200
+++ b/avahi-core/iface.h	2016-05-05 15:06:00.000000000 +0200
@@ -218,6 +218,7 @@
 
 void avahi_interface_address_update_rrs(AvahiInterfaceAddress *a, int remove_rrs);
 int avahi_interface_address_is_relevant(AvahiInterfaceAddress *a);
+int avahi_interface_address_is_relevant_mdns(AvahiInterfaceAddress *a);
 
 AvahiInterfaceAddress* avahi_interface_monitor_get_address(AvahiInterfaceMonitor *m, AvahiInterface *i, const AvahiAddress *raddr);


