From cd0daf93605c2843af5c05881b91b11f1a077bbc Mon Sep 17 00:00:00 2001
From: Captain <captain.onboard@web.de>
Date: Thu, 8 Apr 2021 13:49:17 +0200
Subject: [PATCH 2/3] fix build openssl102q

---
 librtmp/hashswf.c | 5 +++++
 librtmp/rtmp.c    | 5 +++++
 2 files changed, 10 insertions(+)

diff --git a/librtmp/hashswf.c b/librtmp/hashswf.c
index 01b97e2..e05d60b 100644
--- a/librtmp/hashswf.c
+++ b/librtmp/hashswf.c
@@ -56,6 +56,11 @@
 #include <openssl/ssl.h>
 #include <openssl/sha.h>
 #include <openssl/hmac.h>
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#ifndef RC4_INT
+#define RC4_INT unsigned int
+#endif
+#endif
 #include <openssl/rc4.h>
 #define HMAC_setup(ctx, key, len)	HMAC_CTX_init(&ctx); HMAC_Init_ex(&ctx, (unsigned char *)key, len, EVP_sha256(), 0)
 #define HMAC_crunch(ctx, buf, len)	HMAC_Update(&ctx, (unsigned char *)buf, len)
diff --git a/librtmp/rtmp.c b/librtmp/rtmp.c
index 7086254..269ac55 100644
--- a/librtmp/rtmp.c
+++ b/librtmp/rtmp.c
@@ -59,6 +59,11 @@ static const char *my_dhm_G = "4";
 #include <nettle/md5.h>
 #else	/* USE_OPENSSL */
 #include <openssl/ssl.h>
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#ifndef RC4_INT
+#define RC4_INT unsigned int
+#endif
+#endif
 #include <openssl/rc4.h>
 #include <openssl/md5.h>
 #include <openssl/bio.h>
-- 
2.31.1.windows.1

