From 7752efd063949d5c822d97b7a5cd3076e1721039 Mon Sep 17 00:00:00 2001
From: Captain <captain.onboard@web.de>
Date: Thu, 8 Apr 2021 13:49:55 +0200
Subject: [PATCH] add-movecast-thx-testi

---
 librtmp/rtmp.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/librtmp/rtmp.c b/librtmp/rtmp.c
index d429877..cb2573a 100644
--- a/librtmp/rtmp.c
+++ b/librtmp/rtmp.c
@@ -3258,6 +3258,11 @@ HandleInvoke(RTMP *r, const char *body, unsigned int nBodySize)
 
               RTMP_SendCreateStream(r);
             }
+         else if (strstr(pageUrl, "movecast.tv"))
+            {
+              SendCommand(r, "trebaDaSmenamSecurity", TRUE);
+              RTMP_SendCreateStream(r);
+            }
           else if ((strstr(host, "highwebmedia.com") || strstr(pageUrl, "chaturbate.com"))
                    && (!strstr(host, "origin")))
             {
