diff --git a/librtmp/hashswf.c b/librtmp/hashswf.c
index 17a8dee..4aad7b9 100644
--- a/librtmp/hashswf.c
+++ b/librtmp/hashswf.c
@@ -56,6 +56,11 @@
 #include <openssl/ssl.h>
 #include <openssl/sha.h>
 #include <openssl/hmac.h>
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#ifndef RC4_INT
+#define RC4_INT unsigned int
+#endif
+#endif
 #include <openssl/rc4.h>
 #define HMAC_setup(ctx, key, len)	HMAC_CTX_init(&ctx); HMAC_Init_ex(&ctx, (unsigned char *)key, len, EVP_sha256(), 0)
 #define HMAC_crunch(ctx, buf, len)	HMAC_Update(&ctx, (unsigned char *)buf, len)
diff --git a/librtmp/rtmp.c b/librtmp/rtmp.c
index 75ffe26..000fe68 100644
--- a/librtmp/rtmp.c
+++ b/librtmp/rtmp.c
@@ -59,6 +59,11 @@ static const char *my_dhm_G = "4";
 #include <nettle/md5.h>
 #else	/* USE_OPENSSL */
 #include <openssl/ssl.h>
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#ifndef RC4_INT
+#define RC4_INT unsigned int
+#endif
+#endif
 #include <openssl/rc4.h>
 #include <openssl/md5.h>
 #include <openssl/bio.h>
