diff -Nurw wget-1.18/doc/sample.wgetrc wget-1.18-patched/doc/sample.wgetrc
--- wget-1.18/doc/sample.wgetrc	2016-06-06 17:36:04.000000000 +0200
+++ wget-1.18-patched/doc/sample.wgetrc	2016-09-06 16:12:05.502959500 +0200
@@ -135,3 +135,9 @@
 
 # Tune HTTPS security (auto, SSLv2, SSLv3, TLSv1, PFS)
 #secureprotocol = auto
+
+# Skip certificate checking
+# Not recommended when transmitting confidential/important data.
+# Command-line equivalent: —no-check-certificate
+#   (or —check-certificate to turn on).
+check_certificate = off
diff -Nurw wget-1.18/doc/sample.wgetrc.munged_for_texi_inclusion wget-1.18-patched/doc/sample.wgetrc.munged_for_texi_inclusion
--- wget-1.18/doc/sample.wgetrc.munged_for_texi_inclusion	2016-06-06 17:36:04.000000000 +0200
+++ wget-1.18-patched/doc/sample.wgetrc.munged_for_texi_inclusion	2016-09-06 16:12:05.502959500 +0200
@@ -135,3 +135,9 @@
 
 # Tune HTTPS security (auto, SSLv2, SSLv3, TLSv1, PFS)
 #secureprotocol = auto
+
+# Skip certificate checking
+# Not recommended when transmitting confidential/important data.
+# Command-line equivalent: —no-check-certificate
+#   (or —check-certificate to turn on).
+check_certificate = off
