--- a/os_dep/linux/rtw_android.c	2018-09-05 09:58:41.549852335 +0200
+++ b/os_dep/linux/rtw_android.c	2018-09-05 10:30:51.993233423 +0200
@@ -592,22 +592,16 @@
 			goto exit;
 	}
 #ifdef CONFIG_COMPAT
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,6,0)
-	if (is_compat_syscall()) {
-#else
-	if (is_compat_task()) {
-#endif
-		/* User space is 32-bit, use compat ioctl */
-		compat_android_wifi_priv_cmd compat_priv_cmd;
+	/* User space is 32-bit, use compat ioctl */
+	compat_android_wifi_priv_cmd compat_priv_cmd;
 
-		if (copy_from_user(&compat_priv_cmd, ifr->ifr_data, sizeof(compat_android_wifi_priv_cmd))) {
-			ret = -EFAULT;
-			goto exit;
-		}
-		priv_cmd.buf = compat_ptr(compat_priv_cmd.buf);
-		priv_cmd.used_len = compat_priv_cmd.used_len;
-		priv_cmd.total_len = compat_priv_cmd.total_len;
-	} else
+	if (copy_from_user(&compat_priv_cmd, ifr->ifr_data, sizeof(compat_android_wifi_priv_cmd))) {
+		ret = -EFAULT;
+		goto exit;
+	}
+	priv_cmd.buf = compat_ptr(compat_priv_cmd.buf);
+	priv_cmd.used_len = compat_priv_cmd.used_len;
+	priv_cmd.total_len = compat_priv_cmd.total_len;
 #endif /* CONFIG_COMPAT */
 	if (copy_from_user(&priv_cmd, ifr->ifr_data, sizeof(android_wifi_priv_cmd))) {
 		ret = -EFAULT;
