diff --git a/Makefile b/Makefile
index 28cefd9..c632ba2 100644
--- a/Makefile
+++ b/Makefile
@@ -32,7 +32,7 @@ ccflags-y += -I$(src)/include
 ccflags-y += -I$(src)/hal
 ccflags-y += -DCONFIG_RTL8723B
 ccflags-y += -DCONFIG_LITTLE_ENDIAN
-ccflags-y += -DCONFIG_IOCTL_CFG80211
+#ccflags-y += -DCONFIG_IOCTL_CFG80211
 ccflags-y += -DRTW_USE_CFG80211_STA_EVENT
 ccflags-$(CONFIG_MP_INCLUDED) += -DCONFIG_MP_INCLUDED
 ccflags-$(CONFIG_POWER_SAVING) += -DCONFIG_POWER_SAVING
