--- a/sys-utils/rfkill.c
+++ b/sys-utils/rfkill.c
@@ -39,19 +39,6 @@
 #include "xalloc.h"
 
 
-/*
- * NFC supported by kernel since v3.10 (year 2013); FM and another types are from
- * year 2009 (2.6.33) or older.
- */
-#ifndef RFKILL_TYPE_NFC
-# ifndef RFKILL_TYPE_FM
-#  define RFKILL_TYPE_FM	RFKILL_TYPE_GPS + 1
-# endif
-# define RFKILL_TYPE_NFC	RFKILL_TYPE_FM + 1
-# undef NUM_RFKILL_TYPES
-# define NUM_RFKILL_TYPES	RFKILL_TYPE_NFC + 1
-#endif
-
 struct rfkill_type_str {
 	enum rfkill_type type;	/* ID */
 	const char *name;	/* generic name */
@@ -67,9 +54,6 @@ static const struct rfkill_type_str rfkill_type_strings[] = {
 	{ .type = RFKILL_TYPE_UWB,       .name = "ultrawideband" }, /* alias */
 	{ .type = RFKILL_TYPE_WIMAX,     .name = "wimax",        .desc = "WiMAX" },
 	{ .type = RFKILL_TYPE_WWAN,      .name = "wwan",         .desc = "Wireless WAN" },
-	{ .type = RFKILL_TYPE_GPS,       .name = "gps",          .desc = "GPS" },
-	{ .type = RFKILL_TYPE_FM,        .name = "fm",           .desc = "FM" },
-	{ .type = RFKILL_TYPE_NFC,       .name = "nfc",          .desc = "NFC" },
 	{ .type = NUM_RFKILL_TYPES,      .name = NULL            }
 };
 
