From 14742354cef0c559bdcbd5a56de64e1223888f59 Mon Sep 17 00:00:00 2001
From: arn354 <arn354@email.de>
Date: Fri, 11 Dec 2015 18:25:29 +0100
Subject: [PATCH] autoconf remove debug

---
 include/autoconf.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/autoconf.h b/include/autoconf.h
index 36199af..e2b36aa 100755
--- a/include/autoconf.h
+++ b/include/autoconf.h
@@ -356,9 +356,9 @@
 /*
  * Debug Related Config
  */
-#define DBG	1
+#define DBG	0
 
-#define CONFIG_DEBUG /* DBG_871X, etc... */
+//#define CONFIG_DEBUG /* DBG_871X, etc... */
 //#define CONFIG_DEBUG_RTL871X /* RT_TRACE, RT_PRINT_DATA, _func_enter_, _func_exit_ */
 
 #define CONFIG_PROC_DEBUG
-- 
2.6.1.windows.1

