From 0aceb7e46c2e50ecfaa26c6cbf00ace9cb914a55 Mon Sep 17 00:00:00 2001
From: arn354 <arn354@email.de>
Date: Fri, 11 Dec 2015 20:35:17 +0100
Subject: [PATCH] Patch makefile * add EXTRA_CFLAGS * disable

---
 Makefile | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 0c800f8..2a6588f 100644
--- a/Makefile
+++ b/Makefile
@@ -7,6 +7,8 @@ EXTRA_CFLAGS += -O1
 #EXTRA_CFLAGS += -pedantic
 #EXTRA_CFLAGS += -Wshadow -Wpointer-arith -Wcast-qual -Wstrict-prototypes -Wmissing-prototypes
 
+EXTRA_CFLAGS += -Wframe-larger-than=1536
+
 EXTRA_CFLAGS += -Wno-unused-variable
 EXTRA_CFLAGS += -Wno-unused-value
 EXTRA_CFLAGS += -Wno-unused-label
@@ -39,7 +41,7 @@ CONFIG_SDIO_HCI = n
 CONFIG_GSPI_HCI = n
 ########################## Features ###########################
 CONFIG_MP_INCLUDED = y
-CONFIG_POWER_SAVING = y
+CONFIG_POWER_SAVING = n
 CONFIG_USB_AUTOSUSPEND = n
 CONFIG_HW_PWRP_DETECTION = n
 CONFIG_WIFI_TEST = n
-- 
2.6.1.windows.1

