From 193d506a31c9e8c1b800c504b65069195643f53b Mon Sep 17 00:00:00 2001
From: Captain <captain.onboard@web.de>
Date: Sun, 13 Jun 2021 15:46:29 +0200
Subject: [PATCH] vuplus-fix-exception-duo4k-20

---
 xbmc/utils/StringUtils.cpp | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/xbmc/utils/StringUtils.cpp b/xbmc/utils/StringUtils.cpp
index c9c5e31916..98a44bb185 100644
--- a/xbmc/utils/StringUtils.cpp
+++ b/xbmc/utils/StringUtils.cpp
@@ -1070,6 +1070,7 @@ int64_t StringUtils::AlphaNumericCompare(const wchar_t* left, const wchar_t* rig
   const wchar_t *ld, *rd;
   wchar_t lc, rc;
   int64_t lnum, rnum;
+try {
   bool lsym, rsym;
   while (*l != 0 && *r != 0)
   {
@@ -1162,6 +1163,9 @@ int64_t StringUtils::AlphaNumericCompare(const wchar_t* left, const wchar_t* rig
     }
     l++; r++;
   }
+} catch (const std::exception &e) {
+  return 0;
+}
   if (*r)
   { // r is longer
     return -1;
-- 
2.32.0.windows.1

