diff --git a/Makefile.am b/Makefile.am
index 9bf6b06..90e9140 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -23,7 +23,6 @@ SUBDIRS = \
 	dvdbackup \
 	easyinfo \
 	easymedia \
-	ecasa \
 	eibox \
 	emailclient \
 	epgrefresh \
@@ -43,8 +42,6 @@ SUBDIRS = \
 	kiddytimer \
 	letterbox \
 	mediadownloader \
-	merlinepg \
-	merlinepgcenter \
 	merlinmusicplayer \
 	mosaic \
 	moviecut \
@@ -58,7 +55,6 @@ SUBDIRS = \
 	ofdb \
 	orfat \
 	orfteletext \
-	partnerbox \
 	permanentclock \
 	pipservicerelation \
 	podcast \
@@ -76,7 +72,6 @@ SUBDIRS = \
 	subsdownloader2 \
 	svdrp \
 	tageditor \
-	teletext \
 	toolkit \
 	tvcharts \
 	unwetterinfo \
