Index: rtorrent-0.8.0/scripts/common.m4
===================================================================
--- rtorrent-0.8.0.orig/scripts/common.m4	2008-01-13 15:02:03.000000000 +0000
+++ rtorrent-0.8.0/scripts/common.m4	2008-02-04 13:12:01.000000000 +0000
@@ -192,6 +192,9 @@
       AC_DEFINE(USE_EXECINFO, 1, Use execinfo.h)
     ], [
       AC_MSG_RESULT(no)
+    ], [
+      AC_MSG_RESULT(yes)
+      AC_DEFINE(USE_EXECINFO, 1, Use execinfo.h)
   ])
 ])
 
