--- linux-sh4/arch/sh/kernel/sh_ksyms_32.c.orig	2011-04-27 19:14:26.577240485 +0200
+++ linux-sh4/arch/sh/kernel/sh_ksyms_32.c	2011-04-27 19:13:50.296079774 +0200
@@ -52,6 +52,9 @@
 EXPORT_SYMBOL(__ndelay);
 EXPORT_SYMBOL(__const_udelay);
 
+/* Dagobert: Need to export this if we use CONFIG_CC_OPTIMIZE_FOR_SIZE */
+EXPORT_SYMBOL(strcpy);
+
 #define DECLARE_EXPORT(name)		\
 	extern void name(void);EXPORT_SYMBOL(name)
 

