--- a/drivers/stm/clk.c	2012-11-03 14:33:19.000000000 +0200
+++ b/drivers/stm/clk.c	2012-11-03 14:43:58.000000000 +0200
@@ -186,9 +186,19 @@
 void _clk_disable(struct clk *clk)
 {
 	int ret;
+	
 
-	if (WARN_ON(clk->usage_counter == 0))
+	if (clk_is_always_enabled(clk)) {
+		/*
+		 * this clock can not be disabled;
+		 * This means this is:
+		 * - an external oscillator
+		 * - a system critical clock
+		 */
+		if (--clk->usage_counter == 0)
+			clk->usage_counter = 1;
 		return;
+	}
 
 	if (--clk->usage_counter == 0) {
 		ret = __clk_disable(clk);
