diff --git a/scripts/mod/modpost.c b/scripts/mod/modpost.c
index 7acbdd8..253c107 100644
--- a/scripts/mod/modpost.c
+++ b/scripts/mod/modpost.c
@@ -528,6 +528,8 @@ static const char *section_white_list[] =
 {
 	".comment*",
 	".debug*",
+	".GCC-command-line", /* mn10300 */
+	".GCC.command.line", /* record-gcc-switches, non mn10300 */
 	".mdebug*",        /* alpha, score, mips etc. */
 	".pdr",            /* alpha, score, mips etc. */
 	".stab*",
