--- linux-sh4-2.6.32.46_stm24_0209.orig/sound/core/oss/pcm_oss.c	2009-12-03 05:51:21.000000000 +0200
+++ linux-sh4/sound/core/oss/pcm_oss.c	2011-11-25 00:15:35.594870488 +0200
@@ -219,6 +219,8 @@
 	return 0;
 }
 
+EXPORT_SYMBOL_GPL(snd_pcm_hw_param_mask);
+
 static int _snd_pcm_hw_param_min(struct snd_pcm_hw_params *params,
 				 snd_pcm_hw_param_t var, unsigned int val,
 				 int dir)
@@ -278,6 +280,8 @@
 	return snd_pcm_hw_param_value_min(params, var, dir);
 }
 
+EXPORT_SYMBOL_GPL(_snd_pcm_hw_param_min);
+
 static int _snd_pcm_hw_param_max(struct snd_pcm_hw_params *params,
 				 snd_pcm_hw_param_t var, unsigned int val,
 				 int dir)
@@ -467,6 +471,8 @@
 	return v;
 }
 
+EXPORT_SYMBOL_GPL(snd_pcm_hw_param_near);
+
 static int _snd_pcm_hw_param_set(struct snd_pcm_hw_params *params,
 				 snd_pcm_hw_param_t var, unsigned int val,
 				 int dir)
@@ -543,6 +549,8 @@
 	return snd_pcm_hw_param_value(params, var, NULL);
 }
 
+EXPORT_SYMBOL_GPL(snd_pcm_hw_param_set);
+
 static int _snd_pcm_hw_param_setinteger(struct snd_pcm_hw_params *params,
 					snd_pcm_hw_param_t var)
 {
@@ -554,7 +562,9 @@
 	}
 	return changed;
 }
-	
+
+EXPORT_SYMBOL_GPL(_snd_pcm_hw_param_setinteger);
+
 /*
  * plugin
  */
--- linux-sh4-2.6.32.46_stm24_0209.orig/sound/core/pcm_lib.c	2009-12-03 05:51:21.000000000 +0200
+++ linux-sh4/sound/core/pcm_lib.c	2011-11-25 00:15:45.402559577 +0200
@@ -434,6 +434,8 @@
 	return snd_pcm_update_hw_ptr_post(substream, runtime);
 }
 
+EXPORT_SYMBOL(snd_pcm_update_hw_ptr);
+
 /**
  * snd_pcm_set_ops - set the PCM operators
  * @pcm: the pcm instance
--- linux-sh4-2.6.32.46_stm24_0209.orig/sound/core/pcm_native.c	2011-11-24 22:22:25.000000000 +0200
+++ linux-sh4/sound/core/pcm_native.c	2011-11-25 00:15:57.291809241 +0200
@@ -885,6 +885,8 @@
 			      SNDRV_PCM_STATE_RUNNING);
 }
 
+EXPORT_SYMBOL(snd_pcm_start);
+
 /*
  * stop callbacks
  */
@@ -1883,6 +1885,8 @@
 	return 0;
 }
 
+EXPORT_SYMBOL_GPL(snd_pcm_hw_constraints_init);
+
 int snd_pcm_hw_constraints_complete(struct snd_pcm_substream *substream)
 {
 	struct snd_pcm_runtime *runtime = substream->runtime;
@@ -1966,6 +1970,8 @@
 	return 0;
 }
 
+EXPORT_SYMBOL_GPL(snd_pcm_hw_constraints_complete);
+
 static void pcm_release_private(struct snd_pcm_substream *substream)
 {
 	snd_pcm_unlink(substream);
--- linux-sh4-2.6.32.46_stm24_0209.orig/sound/core/sound.c	2009-12-03 05:51:21.000000000 +0200
+++ linux-sh4/sound/core/sound.c	2011-11-25 00:16:04.866560277 +0200
@@ -312,6 +312,14 @@
 	return -1;
 }
 
+int snd_get_minor(int type, int card, int dev)
+{
+      return find_snd_minor(type, snd_cards[card], dev);
+}
+
+EXPORT_SYMBOL(snd_get_minor);
+
+
 /**
  * snd_unregister_device - unregister the device on the given card
  * @type: the device type, SNDRV_DEVICE_TYPE_XXX
--- linux-sh4-2.6.32.46_stm24_0209.orig/sound/stm/common.h	2011-11-24 22:22:26.000000000 +0200
+++ linux-sh4/sound/stm/common.h	2011-11-25 00:16:13.881810056 +0200
@@ -173,9 +173,23 @@
 						__FILE__, __LINE__, ##args); \
 		} while (0)
 
+#define snd_stm_assert(expr, args...) \
+		do { \
+			if (unlikely(!(expr))) { \
+				printk(KERN_ERR "snd-stm:%s:%d: BUG? " \
+						"(%s)\n", \
+						__snd_stm_component, \
+						__LINE__, \
+						__stringify(expr)); \
+				dump_stack(); \
+				args; \
+			} \
+		} while (0)
+
 #else
 
 #define snd_stm_printd(...) /* nothing */
+#define snd_stm_assert snd_assert
 
 #endif

*** linux-sh4-2.6.32.46_stm24_0209.orig/include/sound/core.h	2009-12-03 05:51:21.000000000 +0200
--- linux-sh4/include/sound/core.h	2011-11-25 02:39:54.783603673 +0200
***************
*** 362,367 ****
--- 362,384 ----
  	__snd_printk(1, __FILE__, __LINE__, fmt, ##args)
  
  /**
+  * snd_assert - run-time assertion macro
+  * @expr: expression
+  *
+  * This macro checks the expression in run-time and invokes the commands
+  * given in the rest arguments if the assertion is failed.
+  * When CONFIG_SND_DEBUG is not set, the expression is executed but
+  * not checked.
+  */
+ #define snd_assert(expr, args...) do {					\
+ 	if (unlikely(!(expr))) {					\
+ 		snd_printk(KERN_ERR "BUG? (%s)\n", __ASTRING__(expr));	\
+ 		dump_stack();						\
+ 		args;							\
+ 	}								\
+ } while (0)
+ 
+ /**
   * snd_BUG - give a BUG warning message and stack trace
   *
   * Calls WARN() if CONFIG_SND_DEBUG is set.
***************
*** 389,394 ****
--- 406,412 ----
  #else /* !CONFIG_SND_DEBUG */
  
  #define snd_printd(fmt, args...)	do { } while (0)
+ #define snd_assert(expr, args...)	(void)(expr)
  #define snd_BUG()			do { } while (0)
  static inline int __snd_bug_on(int cond)
  {
***************
*** 458,462 ****
--- 476,481 ----
  const struct snd_pci_quirk *
  snd_pci_quirk_lookup(struct pci_dev *pci, const struct snd_pci_quirk *list);
  
+ int snd_get_minor(int type, int card, int dev);
  
  #endif /* __SOUND_CORE_H */
