--- linux-sh4/drivers/char/lirc/lirc_stm.c	2012-01-08 19:13:04.000000000 +0100
+++ linux-sh4/drivers/char/lirc/lirc_stm.c	2012-02-21 12:00:05.695416629 +0100
@@ -137,8 +137,8 @@
 #define LIRC_STM_MAX_SYMBOLS		100
 #define LIRC_STM_SCD_MAX_SYMBOLS        32
 #define LIRC_STM_BUFSIZE		((LIRC_STM_SCD_MAX_SYMBOLS + \
-					LIRC_STM_MAX_SYMBOLS) * \
-					sizeof(lirc_t))
+					(2*LIRC_STM_MAX_SYMBOLS) * \
+					sizeof(lirc_t)))
 
 /* Bit settings */
 #define LIRC_STM_IS_OVERRUN	 	0x04
