--- linux-sh4/sound/stm/Kconfig.orig	2012-03-28 19:35:26.000000000 +0200
+++ linux-sh4/sound/stm/Kconfig	2012-03-31 19:14:30.438704699 +0200
@@ -23,11 +23,6 @@
 	depends on SND_STM
 	default n
 
-config SND_STM_PCM_READER
-	tristate
-	depends on SND_STM
-	default n
-
 config SND_STM_SPDIF_PLAYER
 	tristate
 	depends on SND_STM
@@ -75,7 +70,6 @@
 	depends on SND_STM
 	default y if CPU_SUBTYPE_FLI7510
 	select SND_STM_PCM_PLAYER
-	select SND_STM_PCM_READER
 	select SND_STM_SPDIF_PLAYER
 	help
 		Select this to get ALSA driver for Freeman 510/520/530/540
@@ -89,7 +83,6 @@
 	depends on SND_STM
 	default y if CPU_SUBTYPE_STX7100
 	select SND_STM_PCM_PLAYER
-	select SND_STM_PCM_READER
 	select SND_STM_SPDIF_PLAYER
 	select SND_STM_CONV_DAC_MEMORY_MAPPED
 	select SND_STM_CONV_I2S_TO_SPDIF
@@ -104,7 +97,6 @@
 	depends on SND_STM
 	default y if CPU_SUBTYPE_STX7105
 	select SND_STM_PCM_PLAYER
-	select SND_STM_PCM_READER
 	select SND_STM_SPDIF_PLAYER
 	select SND_STM_CONV_DAC_MEMORY_MAPPED
 	select SND_STM_CONV_I2S_TO_SPDIF
@@ -119,7 +111,6 @@
 	depends on SND_STM
 	default y if CPU_SUBTYPE_STX7108
 	select SND_STM_PCM_PLAYER
-	select SND_STM_PCM_READER
 	select SND_STM_SPDIF_PLAYER
 	select SND_STM_CONV_DAC_SYSCONF
 	select SND_STM_CONV_I2S_TO_SPDIF
@@ -134,7 +125,6 @@
 	depends on SND_STM
 	default y if CPU_SUBTYPE_STX7111
 	select SND_STM_PCM_PLAYER
-	select SND_STM_PCM_READER
 	select SND_STM_SPDIF_PLAYER
 	select SND_STM_CONV_DAC_MEMORY_MAPPED
 	select SND_STM_CONV_I2S_TO_SPDIF
@@ -149,7 +139,6 @@
 	depends on SND_STM
 	default y if CPU_SUBTYPE_STX7141
 	select SND_STM_PCM_PLAYER
-	select SND_STM_PCM_READER
 	select SND_STM_SPDIF_PLAYER
 	select SND_STM_CONV_DAC_MEMORY_MAPPED
 	select SND_STM_CONV_I2S_TO_SPDIF
@@ -164,7 +153,6 @@
 	depends on SND_STM
 	default y if CPU_SUBTYPE_STX7200
 	select SND_STM_PCM_PLAYER
-	select SND_STM_PCM_READER
 	select SND_STM_SPDIF_PLAYER
 	select SND_STM_CONV_DAC_MEMORY_MAPPED
 	select SND_STM_CONV_I2S_TO_SPDIF
