diff --git a/arch/sh/kernel/setup.c b/arch/sh/kernel/setup.c
index e0ce175..92dc212 100644
--- a/arch/sh/kernel/setup.c
+++ b/arch/sh/kernel/setup.c
@@ -578,8 +578,10 @@ static int show_cpuinfo(struct seq_file *m, void *v)
 	if (!cpu_online(cpu))
 		return 0;
 
-	if (cpu == 0)
+	if (cpu == 0) {
 		seq_printf(m, "machine\t\t: %s\n", get_system_type());
+		seq_printf(m, "system type\t: %s STB platform\n", get_cpu_subtype(c));
+	}
 	else
 		seq_printf(m, "\n");
 
@@ -596,6 +598,8 @@ static int show_cpuinfo(struct seq_file *m, void *v)
 
 	show_cpuflags(m, c);
 
+	seq_printf(m, "cpu MHz\t: %lu\n", c->loops_per_jiffy/(500000/HZ)+3);
+
 	seq_printf(m, "cache type\t: ");
 
 	/*
