--- a/sound/stm/pcm_player.c.orig	2012-07-31 16:36:49.000000000 +0200
+++ b/sound/stm/pcm_player.c	2012-11-03 23:15:21.911700974 +0100
@@ -676,10 +676,7 @@
 
 	/* Mute & shutdown DAC */
 
-	if (pcm_player->conv_group) {
-		snd_stm_conv_mute(pcm_player->conv_group);
-		snd_stm_conv_disable(pcm_player->conv_group);
-	}
+	// No we will not
 
 	/* Disable interrupts */
 
@@ -689,7 +686,8 @@
 
 	/* Stop PCM player */
 
-	set__AUD_PCMOUT_CTRL__MODE__OFF(pcm_player);
+	// No, pause it because stopping causes cracking noise
+	set__AUD_PCMOUT_CTRL__MODE__MUTE(pcm_player);
 
 	/* Stop FDMA transfer */
 
@@ -697,8 +695,7 @@
 
 	/* Stop the clock & reset PCM player */
 
-	clk_disable(pcm_player->clock);
-	set__AUD_PCMOUT_RST__SRSTP__RESET(pcm_player);
+	// No we will not
 
 	return 0;
 }
