--- linux-sh4/drivers/video/console/cfbcondecor.c
+++ linux-sh4/drivers/video/console/cfbcondecor.c
@@ -379,7 +379,7 @@
        int height, int width)
 {
  int bgshift = (vc->vc_hi_font_mask) ? 13 : 12;
- int bg_color = attr_bgcol_ec(bgshift, vc);
+ int bg_color = attr_bgcol_ec(bgshift, vc, info);
  int transparent = vc->vc_decor.bg_color == bg_color;
  struct fbcon_ops *ops = info->fbcon_par;
  u8 *dst;
