diff -Naur autofs-4.1.4/samples.orig/auto.net autofs-4.1.4/samples/auto.net
--- autofs-4.1.4/samples.orig/auto.net	2007-11-21 11:29:57.000000000 -0600
+++ autofs-4.1.4/samples/auto.net	2007-11-21 11:30:27.000000000 -0600
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/sh
 
 # $Id: auto.net,v 1.8 2005/04/05 13:02:09 raven Exp $
 
diff -Naur autofs-4.1.4/samples.orig/auto.smb autofs-4.1.4/samples/auto.smb
--- autofs-4.1.4/samples.orig/auto.smb	2007-11-21 11:29:57.000000000 -0600
+++ autofs-4.1.4/samples/auto.smb	2007-11-21 11:30:34.000000000 -0600
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/sh
 
 # $Id: auto.smb,v 1.3 2005/04/05 13:02:09 raven Exp $
 
