diff --git a/src/ac3_descriptor.cpp b/src/ac3_descriptor.cpp
index c7b28b2..fec8c31 100644
--- a/src/ac3_descriptor.cpp
+++ b/src/ac3_descriptor.cpp
@@ -31,6 +31,16 @@ Ac3Descriptor::Ac3Descriptor(const uint8_t * const buffer) : Descriptor(buffer)
 	asvcFlag = (buffer[2] >> 4) & 0x01;
 
 	size_t headerLength = 1 + ac3TypeFlag + bsidFlag + mainidFlag + asvcFlag;
+
+	// broadcasters got it wrong again...
+	if (headerLength > descriptorLength) {
+		ac3TypeFlag = 0;
+		bsidFlag = 0;
+		mainidFlag = 0;
+		asvcFlag = 0;
+		return;
+	}
+
 	ASSERT_MIN_DLEN(headerLength);
 
 	size_t i = 3;
