--- a/lib/service/servicemp3.cpp	Mon Jan 30 21:54:35 2017
+++ b/lib/service/servicemp3.cpp	Mon Jan 30 22:50:35 2017
@@ -1307,9 +1307,9 @@
 		}
 		else
 		{
-			g_signal_emit_by_name(dvb_videosink, "get-decoder-time", &pos);
+			g_signal_emit_by_name(dvb_audiosink, "get-decoder-time", &pos);
 			if (!GST_CLOCK_TIME_IS_VALID(pos) || 0)
-				 g_signal_emit_by_name(dvb_audiosink, "get-decoder-time", &pos);
+				 g_signal_emit_by_name(dvb_videosink, "get-decoder-time", &pos);
 			if(!GST_CLOCK_TIME_IS_VALID(pos))
 				return -1;
 		}
