diff --git a/lib/service/servicedvb.cpp b/lib/service/servicedvb.cpp
index 0b138ee..8216c91 100644
--- a/lib/service/servicedvb.cpp
+++ b/lib/service/servicedvb.cpp
@@ -2901,14 +2901,7 @@ void eDVBServicePlay::updateDecoder(bool sendSeekableStateChanged)
 		if (!m_noaudio)
 		{
 			selectAudioStream();
-#if HAVE_AMLOGIC
 			m_decoder->setSyncPCR(pcrpid);
-#else
-			if (!(m_is_pvr || m_is_stream || m_timeshift_active))
-				m_decoder->setSyncPCR(pcrpid);
-			else
-				m_decoder->setSyncPCR(-1);
-#endif
 		}
 
 		if (m_decoder_index == 0)
-- 
2.10.0.windows.1

