diff a/lib/service/servicedvb.cpp b/lib/service/servicedvb.cpp
--- a/lib/service/servicedvb.cpp	2016-03-15 22:47:00.000000000 +0000
+++ b/lib/service/servicedvb.cpp	2016-03-21 23:32:00.000000000 +0000
@@ -2178,22 +2178,6 @@
 
 	m_current_audio_pid = apid;
 
-	pts_t a_pts, v_pts;
-	a_pts = v_pts = 0;
-	m_decoder->getPTS(2, a_pts);
-	m_decoder->getPTS(1, v_pts);
-	eDebug("a: %lld   v: %lld  %lld",a_pts, v_pts, a_pts-v_pts);
-	bool radio_workaround = false;
-	if(v_pts && a_pts && (abs(a_pts-v_pts) > 2* 90000))
-		radio_workaround = true;
-
-	if(radio_workaround)
-		if (m_decoder->setAudioPID(-1, 0))
-		{
-			eDebug("set audio pid failed");
-			return -4;
-		}
-
 	if (m_decoder->setAudioPID(apid, apidtype))
 	{
 		eDebug("set audio pid failed");
@@ -2224,9 +2208,6 @@
 		}
 	}
 
-	if(radio_workaround)
-		m_decoder->setSyncPCR(-1);
-
 			/* store new pid as default only when:
 				a.) we have an entry in the service db for the current service,
 				b.) we are not playing back something,
@@ -2902,13 +2883,10 @@
 		m_current_video_pid_type = vpidtype;
 		m_have_video_pid = (vpid > 0 && vpid < 0x2000);
 
-		if (!(m_is_pvr || m_is_stream || m_timeshift_active || (pcrpid == 0x1FFF)))
-			m_decoder->setSyncPCR(pcrpid);
-		else
-			m_decoder->setSyncPCR(-1);
-
 		selectAudioStream();
 
+		m_decoder->setSyncPCR(pcrpid);
+
 		if (m_decoder_index == 0)
 		{
 			m_decoder->setTextPID(tpid);
