diff --git a/lib/service/servicedvb.cpp b/lib/service/servicedvb.cpp
index 4bb0c46..8e276f7 100644
--- a/lib/service/servicedvb.cpp
+++ b/lib/service/servicedvb.cpp
@@ -2885,14 +2885,8 @@ void eDVBServicePlay::updateDecoder(bool sendSeekableStateChanged)
 
 		selectAudioStream();
 
-#if HAVE_AMLOGIC
-			m_decoder->setSyncPCR(pcrpid);
-#else
-			if (!(m_is_pvr || m_is_stream || m_timeshift_active))
-				m_decoder->setSyncPCR(pcrpid);
-			else
-				m_decoder->setSyncPCR(-1);
-#endif
+		m_decoder->setSyncPCR(pcrpid);
+
 
 		if (m_decoder_index == 0)
 		{
-- 
2.10.0.windows.1

