--- a/python/plugin/EnigmaLight/EL_HttpServer.py
+++ b/python/plugin/EnigmaLight/EL_HttpServer.py
@@ -13,7 +13,8 @@ from enigma import eEnv
 from Components.config import config
 from Tools.Directories import fileExists
 from twisted.internet import reactor, ssl
-from twisted.web import server, http, static, resource, error, version
+from twisted import version
+from twisted.web import server, http, static, resource, error
 from twisted.internet.error import CannotListenError
 
 from remote.root import RootController
--- a/python/plugin/EnigmaLight/remote/models/info.py
+++ b/python/plugin/EnigmaLight/remote/models/info.py
@@ -15,7 +15,7 @@ from Components.config import config
 
 from Tools.Directories import fileExists, pathExists
 from time import time, localtime, strftime
-from twisted.web import version
+from twisted import version
 
 import os
 import sys
