From 669114edd7e6a0fdb3763f2f5ca3544af80dc6b3 Mon Sep 17 00:00:00 2001
From: Captain <captain.onboard@web.de>
Date: Sun, 25 Jul 2021 05:55:30 +0200
Subject: [PATCH] epgwrapper-20

---
 xbmc/windowing/egl/EGLWrapper.cpp | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/xbmc/windowing/egl/EGLWrapper.cpp b/xbmc/windowing/egl/EGLWrapper.cpp
index 2f6288db84..94ed909ae2 100644
--- a/xbmc/windowing/egl/EGLWrapper.cpp
+++ b/xbmc/windowing/egl/EGLWrapper.cpp
@@ -216,11 +216,14 @@ bool CEGLWrapper::GetNativeResolution(RESOLUTION_INFO *res)
   return m_nativeTypes->GetNativeResolution(res);
 }
 
+bool g_isShow = true;
+
 bool CEGLWrapper::ShowWindow(bool show)
 {
   if (!m_nativeTypes)
     return false;
 
+  g_isShow = show;
   return m_nativeTypes->ShowWindow(show);
 }
 
@@ -419,6 +422,10 @@ void CEGLWrapper::SwapBuffers(EGLDisplay display, EGLSurface surface)
 {
   if ((display == EGL_NO_DISPLAY) || (surface == EGL_NO_SURFACE))
     return;
+
+  if( !g_isShow )
+    return;
+
   eglSwapBuffers(display, surface);
 }
 
-- 
2.32.0.windows.2

