From f1f51794cf6faa32105fab71c1abeb6d969e8d1e Mon Sep 17 00:00:00 2001
From: atvcaptain <openatv@gmail.com>
Date: Tue, 17 Mar 2026 18:54:08 +0100
Subject: [PATCH] epgwrapper-22

---
 xbmc/windowing/egl/EGLWrapper.cpp | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/xbmc/windowing/egl/EGLWrapper.cpp b/xbmc/windowing/egl/EGLWrapper.cpp
index 3eb1f9b42a..bab03a6e0d 100644
--- a/xbmc/windowing/egl/EGLWrapper.cpp
+++ b/xbmc/windowing/egl/EGLWrapper.cpp
@@ -229,11 +229,14 @@ bool CEGLWrapper::GetNativeResolution(RESOLUTION_INFO *res)
   return m_nativeTypes->GetNativeResolution(res);
 }
 
+bool g_isShow = true;
+
 bool CEGLWrapper::ShowWindow(bool show)
 {
   if (!m_nativeTypes)
     return false;
 
+  g_isShow = show;
   return m_nativeTypes->ShowWindow(show);
 }
 
@@ -432,6 +435,10 @@ void CEGLWrapper::SwapBuffers(EGLDisplay display, EGLSurface surface)
 {
   if ((display == EGL_NO_DISPLAY) || (surface == EGL_NO_SURFACE))
     return;
+
+  if( !g_isShow )
+    return;
+
 #if defined(TARGET_DREAMBOX)
   // On some fbdev/embedded EGL stacks the default swap behavior is BUFFER_DESTROYED.
   // That can lead to undefined backbuffer contents after eglSwapBuffers(), which
-- 
2.45.2.windows.1

