diff --git a/browser.py b/browser.py
index cdf990c..1d65d4e 100644
--- a/browser.py
+++ b/browser.py
@@ -19,7 +19,7 @@ class Browser:
             datasocket.onCommandReceived.append(self.onCommandReceived)
             datasocket.onBrowserClosed.append(self.onBrowserClosed)
             container = eConsoleAppContainer()
-            container.execute("/usr/bin/openhbbtvbrowser '%s' --onid %d --tsid %d --sid %d" % (url, onid, tsid, sid))
+            container.execute("export QT_QPA_EGLFS_INTEGRATION=eglfs_mali;/usr/bin/openhbbtvbrowser '%s' --onid %d --tsid %d --sid %d" % (url, onid, tsid, sid))
 
     def stop(self):
         if self.commandserver:
