diff -upr rtl8712/config rtl8712.a/config
--- rtl8712/config	2011-04-01 07:25:21.000000000 +0200
+++ rtl8712.a/config	2013-02-21 13:06:18.434397813 +0100
@@ -13,11 +13,11 @@ CONFIG_SDIO_HCI			= 	n
 
 CONFIG_MP_INCLUDED		=	y
 
-CONFIG_PLATFORM_I386_PC		=	y
+CONFIG_PLATFORM_I386_PC		=	n
 CONFIG_PLATFORM_ANDROID_X86	=	n
 CONFIG_PLATFORM_ARM_S3C		= 	n
 CONFIG_PLATFORM_ARM_PXA		= 	n
-CONFIG_PLATFORM_MIPS_RMI	= 	n
+CONFIG_PLATFORM_MIPS_RMI	= 	y
 CONFIG_PLATFORM_RTK_DMP		= 	n
 CONFIG_PLATFORM_MIPS_PLM	= 	n
 CONFIG_PLATFORM_RTD2880B	=	n
diff -upr rtl8712/ioctl/rtl871x_ioctl_linux.c rtl8712.a/ioctl/rtl871x_ioctl_linux.c
--- rtl8712/ioctl/rtl871x_ioctl_linux.c	2011-04-01 07:25:22.000000000 +0200
+++ rtl8712.a/ioctl/rtl871x_ioctl_linux.c	2013-02-21 13:08:37.843390414 +0100
@@ -86,6 +86,22 @@ static int hex2num_i(char c)
 	return -1;
 }
 
+int strcasecmp(const char *a, const char *b)
+{
+       int ca, cb;
+
+       do {
+               ca = *a++ & 0xff;
+               cb = *b++ & 0xff;
+               if (ca >= 'A' && ca <= 'Z')
+                       ca += 'a' - 'A';
+               if (cb >= 'A' && cb <= 'Z')
+                       cb += 'a' - 'A';
+       } while (ca == cb && ca != '\0');
+
+       return ca - cb;
+}
+
 static int hex2byte_i(const char *hex)
 {
 	int a, b;
