diff -Naur linux-3.2.org/kernel/signal.c linux-3.2/kernel/signal.c
--- linux-3.2.org/kernel/signal.c	2013-03-23 13:54:58.000000000 +0100
+++ linux-3.2/kernel/signal.c	2013-03-23 14:26:57.748005975 +0100
@@ -482,7 +482,7 @@
 			ka->sa.sa_handler = SIG_DFL;
 		ka->sa.sa_flags = 0;
 #ifdef SA_RESTORER
-		ka->sa.sa_restorer = NULL;
+//		ka->sa.sa_restorer = NULL;
 #endif
 		sigemptyset(&ka->sa.sa_mask);
 		ka++;
