diff --git a/drivers/mtd/nand/dreambox_nand.c b/drivers/mtd/nand/dreambox_nand.c
index 3aff2a9..f0adbdc 100644
--- a/drivers/mtd/nand/dreambox_nand.c
+++ b/drivers/mtd/nand/dreambox_nand.c
@@ -89,6 +89,7 @@ static void dreambox_nand_select_chip(struct mtd_info *mtd, int chipnr)
 		PRINTK(KERN_DEBUG "select chip\n");
 		writeb(0xEE, (void __iomem *)0xBF030003);
 	}
+	readb((void __iomem *)0xBF030003);
 }
 
 /*
@@ -118,6 +119,7 @@ static void dreambox_nand_cmd_ctrl(struct mtd_info *mtd, int cmd,
 		ctrl & 0x80 ? "CHANGE" : "");
 
 	writeb(cmd, (void __iomem *)addr);
+	readb((void __iomem *)addr);
 }
 
 static uint8_t dreambox_nand_read_byte(struct mtd_info *mtd)
