--- a/meson.build
+++ b/meson.build
@@ -287,7 +287,9 @@ if freedreno_kmds.length() != 0 and freedreno_kmds != [ 'msm' ] and with_freedre
   endif
 endif
 
-with_dri = false
+dri_drivers = get_option('dri-drivers')
+
+with_dri = get_option('dri') or (dri_drivers.length() != 0 and dri_drivers != [''])
 if with_gallium and system_has_kms_drm
   _glx = get_option('glx')
   _egl = get_option('egl')
--- a/meson.options
+++ b/meson.options
@@ -51,6 +51,19 @@ option(
   description : 'Use Android\'s libbacktrace',
 )
 
+option(
+  'dri',
+  type : 'boolean',
+  value : false,
+  description : 'enable support for dri'
+)
+
+option(
+  'dri-drivers',
+  type : 'array',
+  description : 'List of dri drivers to build. If this is set to auto all drivers applicable to the target OS/architecture will be built'
+)
+
 option(
   'dri-drivers-path',
   type : 'string',
--- a/src/egl/drivers/dri2/egl_dri2.c
+++ b/src/egl/drivers/dri2/egl_dri2.c
@@ -45,7 +45,8 @@
 #include "mesa_interface.h"
 #include <sys/stat.h>
 #include <sys/types.h>
-#include "dri_screen.h"
+//#include "dri_screen.h"
+#include <gallium/frontends/dri/dri_screen.h>
 
 #ifdef HAVE_WAYLAND_PLATFORM
 #include "linux-dmabuf-unstable-v1-client-protocol.h"
--- a/src/egl/main/eglapi.c
+++ b/src/egl/main/eglapi.c
@@ -2816,8 +2816,8 @@ eglGetProcAddress(const char *procname)
          ret = entrypoint->function;
    }
 
-   if (!ret)
-      ret = _mesa_glapi_get_proc_address(procname);
+//  if (!ret)
+//     ret = _mesa_glapi_get_proc_address(procname);
 
    RETURN_EGL_SUCCESS(NULL, ret);
 }
--- a/src/egl/meson.build
+++ b/src/egl/meson.build
@@ -49,6 +49,7 @@ files_egl = files(
   'main/egltypedefs.h',
 )
 
+genCommon_py = files('../mapi/new/genCommon.py')
 g_egldispatchstubs_c = custom_target(
   'g_egldispatchstubs.c',
   input : [
@@ -91,7 +92,7 @@ if with_dri2
   incs_for_egl += inc_gallium
   incs_for_egl += inc_gallium_aux
   incs_for_egl += inc_mesa
-  incs_for_egl += inc_st_dri
+#  incs_for_egl += inc_st_dri
 
   files_egl += files(
     'drivers/dri2/platform_device.c',
diff --git a/src/egl/drivers/dri2/platform_surfaceless.c b/src/egl/drivers/dri2/platform_surfaceless.c
index d981f559..f7705723 100644
--- a/src/egl/drivers/dri2/platform_surfaceless.c
+++ b/src/egl/drivers/dri2/platform_surfaceless.c
@@ -40,7 +40,8 @@
 #include "loader.h"
 #include "loader_dri_helper.h"
 #include "dri_util.h"
-#include "dri_screen.h"
+//#include "dri_screen.h"
+#include <gallium/frontends/dri/dri_screen.h>
 
 static struct dri_image *
 surfaceless_alloc_image(struct dri2_egl_display *dri2_dpy,
