diff --git a/os_dep/osdep_service.c b/os_dep/osdep_service.c
index dcd26b6a15b4..c44b5f593135 100644
--- a/os_dep/osdep_service.c
+++ b/os_dep/osdep_service.c
@@ -2076,7 +2076,13 @@ static int isFileReadable(const char *path, u32 *sz)
 		ret = PTR_ERR(fp);
 	else {
 		oldfs = get_fs();
-		set_fs(get_ds());
+		set_fs(
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,1,0))
+			KERNEL_DS
+#else
+			get_ds()
+#endif
+			);
 
 		if (1 != readFile(fp, &buf, 1))
 			ret = PTR_ERR(fp);
@@ -2114,7 +2120,13 @@ static int retriveFromFile(const char *path, u8 *buf, u32 sz)
 			RTW_INFO("%s openFile path:%s fp=%p\n", __FUNCTION__, path , fp);
 
 			oldfs = get_fs();
-			set_fs(get_ds());
+			set_fs(
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,1,0))
+				KERNEL_DS
+#else
+				get_ds()
+#endif
+				);
 			ret = readFile(fp, buf, sz);
 			set_fs(oldfs);
 			closeFile(fp);
@@ -2149,7 +2161,13 @@ static int storeToFile(const char *path, u8 *buf, u32 sz)
 			RTW_INFO("%s openFile path:%s fp=%p\n", __FUNCTION__, path , fp);
 
 			oldfs = get_fs();
-			set_fs(get_ds());
+			set_fs(
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,1,0))
+				KERNEL_DS
+#else
+				get_ds()
+#endif
+				);
 			ret = writeFile(fp, buf, sz);
 			set_fs(oldfs);
 			closeFile(fp);
