diff --git a/include/wifi.h b/include/wifi.h
index ecb93a364f57..859efa6658ff 100644
--- a/include/wifi.h
+++ b/include/wifi.h
@@ -1011,7 +1011,11 @@ typedef enum _HT_CAP_AMPDU_DENSITY {
  * According to IEEE802.11n spec size varies from 8K to 64K (in powers of 2)
  */
 #define IEEE80211_MIN_AMPDU_BUF 0x8
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,20,0))
+#define IEEE80211_MAX_AMPDU_BUF 0x100
+#else
 #define IEEE80211_MAX_AMPDU_BUF 0x40
+#endif
 
 
 /* Spatial Multiplexing Power Save Modes */
diff --git a/os_dep/linux/ioctl_cfg80211.c b/os_dep/linux/ioctl_cfg80211.c
index 4069d5489873..7c6f1c21fa55 100644
--- a/os_dep/linux/ioctl_cfg80211.c
+++ b/os_dep/linux/ioctl_cfg80211.c
@@ -327,6 +327,13 @@ static const struct ieee80211_txrx_stypes
 };
 #endif
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,20,0))
+static inline void get_monotonic_boottime(struct timespec *ts)
+{
+	*ts = ktime_to_timespec(ktime_get_boottime());
+}
+#endif
+
 static u64 rtw_get_systime_us(void)
 {
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 39))
