From 146abc514fdbff5b0a18e8a051926b9467f05041 Mon Sep 17 00:00:00 2001
From: Captain <captain.onboard@web.de>
Date: Tue, 27 Apr 2021 22:33:17 +0200
Subject: [PATCH] add sh4 support

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index cb02399..aa7570a 100644
--- a/Makefile
+++ b/Makefile
@@ -964,7 +964,7 @@ EXTRA_CFLAGS += -DDM_ODM_SUPPORT_TYPE=0x04
 
 ifeq ($(CONFIG_PLATFORM_I386_PC), y)
 EXTRA_CFLAGS += -DCONFIG_LITTLE_ENDIAN
-EXTRA_CFLAGS += -DCONFIG_IOCTL_CFG80211 -DRTW_USE_CFG80211_STA_EVENT
+EXTRA_CFLAGS += -DRTW_USE_CFG80211_STA_EVENT
 SUBARCH := $(shell uname -m | sed -e s/i.86/i386/)
 ARCH ?= $(SUBARCH)
 CROSS_COMPILE ?=
-- 
2.31.1.windows.1

