--- a/include/uapi/linux/fs.h	2016-12-11 20:17:54.000000000 +0100
+++ b/include/uapi/linux/fs.h	2016-12-12 18:57:20.686857000 +0100
@@ -151,18 +151,6 @@
 #define MS_MGC_MSK 0xffff0000
 
 /*
- * Structure for FS_IOC_FSGETXATTR[A] and FS_IOC_FSSETXATTR.
- */
-struct fsxattr {
-	__u32		fsx_xflags;	/* xflags field value (get/set) */
-	__u32		fsx_extsize;	/* extsize field value (get/set)*/
-	__u32		fsx_nextents;	/* nextents field value (get)	*/
-	__u32		fsx_projid;	/* project identifier (get/set) */
-	__u32		fsx_cowextsize;	/* CoW extsize field value (get/set)*/
-	unsigned char	fsx_pad[8];
-};
-
-/*
  * Flags for the fsx_xflags field
  */
 #define FS_XFLAG_REALTIME	0x00000001	/* data in realtime volume */
