diff --git a/wvmodemscan.cc b/wvmodemscan.cc
index bf59924..ef5ab87 100644
--- a/wvmodemscan.cc
+++ b/wvmodemscan.cc
@@ -485,6 +485,7 @@ const char *WvModemScan::is_isdn() const
 
 static int fileselect(const struct dirent *e)
 {
+#if 0
     return !strncmp(e->d_name, "ttyS", 4)      	// serial
        || !strncmp(e->d_name, "ttyLT", 5)	// Lucent WinModem
        || !strncmp(e->d_name, "ttyACM", 6)      // USB acm Modems
@@ -493,6 +494,13 @@ static int fileselect(const struct dirent *e)
        || !strncmp(e->d_name, "ttySL", 5);	// SmartLink WinModem
 
 	// (no internal ISDN support)   || !strncmp(e->d_name, "ttyI", 4);
+#else
+    return !strncmp(e->d_name, "ttyLT", 5)      // Lucent WinModem
+       || !strncmp(e->d_name, "ttyACM", 6)      // USB acm Modems
+       || !strncmp(e->d_name, "ttyUSB", 6)      // Modems on USB RS232
+       || !strncmp(e->d_name, "ircomm", 6)      // Handys over IrDA
+       || !strncmp(e->d_name, "ttySL", 5);      // SmartLink WinModem
+#endif
 }
 
 
