diff --git a/drivers/mmc/host/sdhci-brcmstb.c b/drivers/mmc/host/sdhci-brcmstb.c
index dea8398..6a1e8f1 100644
--- a/drivers/mmc/host/sdhci-brcmstb.c
+++ b/drivers/mmc/host/sdhci-brcmstb.c
@@ -147,6 +147,10 @@ static int sdhci_brcmstb_probe(struct platform_device *pdev)
 	struct clk *clk;
 	int res;
 
+	if (platform_get_resource(pdev, IORESOURCE_MEM, 0) && (unsigned int)platform_get_resource(pdev, IORESOURCE_MEM, 0)->start == 0xf03e0000) {
+		return -ENODEV;
+	}
+
 	clk = of_clk_get_by_name(dn, "sw_sdio");
 	if (IS_ERR(clk)) {
 		dev_err(&pdev->dev, "Clock not found in Device Tree\n");
