--- a/libavformat/mov.c
+++ b/libavformat/mov.c
@@ -3771,8 +3771,10 @@ static void mov_fix_index(MOVContext *mov, AVStream *st)
 
             if (tts_data_old && tts_index_old < tts_count_old) {
                 curr_ctts = tts_data_old[tts_index_old].offset;
+                /*
                 av_log(mov->fc, AV_LOG_TRACE, "stts: %"PRId64" ctts: %"PRId64", tts_index: %"PRId64", tts_count: %"PRId64"\n",
                        curr_cts, curr_ctts, tts_index_old, tts_count_old);
+                */
                 curr_cts += curr_ctts;
                 tts_sample_old++;
                 if (tts_sample_old == tts_data_old[tts_index_old].count) {
