--- a/os_dep/os_intfs.c	2018-10-23 21:19:56.399974415 +0200
+++ b/os_dep/os_intfs.c	2018-10-23 21:20:34.344185230 +0200
@@ -902,6 +902,11 @@
 	return dscp >> 5;
 }
 
+#if (LINUX_VERSION_CODE>=KERNEL_VERSION(4,19,0))
+static u16 rtw_select_queue(struct net_device *dev, struct sk_buff *skb,
+				struct net_device *sb_dev,
+				select_queue_fallback_t fallback
+#else
 static u16 rtw_select_queue(struct net_device *dev, struct sk_buff *skb
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 13, 0))
 			    , void *unused
@@ -909,6 +914,7 @@
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 14, 0))
 			    , select_queue_fallback_t fallback
 #endif
+#endif
 )
 {
 	struct rtw_adapter	*padapter = rtw_netdev_priv(dev);
