From 12de7bee1d44ac97d159054d1c9068dc1a8f0ee5 Mon Sep 17 00:00:00 2001
From: captain <openatv@gmail.com>
Date: Mon, 13 Mar 2017 07:47:17 +0100
Subject: [PATCH] add extra check for dst_apply

---
 HbbTV/plugin.py | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/HbbTV/plugin.py b/HbbTV/plugin.py
index c461a87..c3e1070 100644
--- a/HbbTV/plugin.py
+++ b/HbbTV/plugin.py
@@ -9,6 +9,7 @@ from Components.ActionMap import ActionMap
 from Components.Language import language
 from Components.ServiceEventTracker import ServiceEventTracker
 from Components.VolumeControl import VolumeControl
+from Tools.Directories import fileExists
 
 from enigma import eTimer, iServiceInformation, iPlayableService
 
@@ -260,18 +261,19 @@ class VBMain(Screen):
 		file = open("/proc/stb/vmpeg/0/dst_height", "r")
 		dst_height = int(file.read().strip(), 16)
 		file.close()
-		
+
 		global g_dst_left, g_dst_width, g_dst_top, g_dst_height
 		if (g_dst_left != dst_left) or (g_dst_width != dst_width) or (g_dst_top != dst_top) or (g_dst_height != dst_height):
 			g_dst_left = dst_left
 			g_dst_width = dst_width
 			g_dst_top = dst_top
 			g_dst_height = dst_height
-			
-			file = open("/proc/stb/vmpeg/0/dst_apply", "w")
-			file.write("1")
-			file.close()
-		
+
+			if fileExists("/proc/stb/vmpeg/0/dst_apply"):
+				file = open("/proc/stb/vmpeg/0/dst_apply", "w")
+				file.write("1")
+				file.close()
+
 	def _cb_hbbtv_activated(self, url=None, app_info=None):
 		self.start_hbbtv_application(url, app_info)
 
-- 
2.12.0.windows.1

