From 3bf32e20ef89e3f781ceb54591b3288b8f3621b4 Mon Sep 17 00:00:00 2001
From: captain <openatv@gmail.com>
Date: Mon, 6 Mar 2017 13:24:51 +0100
Subject: [PATCH] remove model check

---
 src/main.cpp | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/main.cpp b/src/main.cpp
index f620188..ea5384c 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -36,9 +36,9 @@ bool is_hex_str(char* str);
 
 int main(int argc, char *argv[])
 {
-	if (!verify()) {
-		exit(0);
-	}
+//	if (!verify()) {
+//		exit(0);
+//	}
 
 	int param_dmxid   = -1;
 	int param_pmtid = -1;
-- 
2.12.0.windows.1

