diff --git a/source3/lib/messages.c b/source3/lib/messages.c
index b856a28..cdc28c5 100644
--- a/source3/lib/messages.c
+++ b/source3/lib/messages.c
@@ -501,7 +501,7 @@ static NTSTATUS messaging_init_internal(TALLOC_CTX *mem_ctx,
 		return NT_STATUS_ACCESS_DENIED;
 	}
 
-	priv_path = private_path("msg.sock");
+	priv_path = lock_path(talloc_tos(), "msg.sock");
 	if (priv_path == NULL) {
 		return NT_STATUS_NO_MEMORY;
 	}
@@ -664,7 +664,7 @@ NTSTATUS messaging_reinit(struct messaging_context *msg_ctx)
 		msg_ctx->per_process_talloc_ctx,
 		msg_ctx->event_ctx,
 		&msg_ctx->id.unique_id,
-		private_path("msg.sock"),
+		lock_path(talloc_tos(), "msg.sock"),
 		lck_path,
 		messaging_recv_cb,
 		msg_ctx,
