diff --git a/source4/lib/messaging/messaging.c b/source4/lib/messaging/messaging.c
index 8d76f17..6e500d0 100644
--- a/source4/lib/messaging/messaging.c
+++ b/source4/lib/messaging/messaging.c
@@ -528,7 +528,7 @@ static struct imessaging_context *imessaging_init_internal(
 		goto fail;
 	}
 
-	msg->sock_dir = lpcfg_private_path(msg, lp_ctx, "msg.sock");
+	msg->sock_dir = lpcfg_lock_path(msg, lp_ctx, "msg.sock");
 	if (msg->sock_dir == NULL) {
 		goto fail;
 	}
