diff --git a/lib/talloc/wscript b/lib/talloc/wscript
index 8b5e02d..265f937 100644
--- a/lib/talloc/wscript
+++ b/lib/talloc/wscript
@@ -48,7 +48,7 @@ def configure(conf):
         conf.env.PKGCONFIGDIR = '${LIBDIR}/pkgconfig'
         conf.env.TALLOC_VERSION = VERSION
 
-    conf.CHECK_XSLTPROC_MANPAGES()
+    #conf.CHECK_XSLTPROC_MANPAGES()
 
     conf.CHECK_HEADERS('sys/auxv.h')
     conf.CHECK_FUNCS('getauxval')
diff --git a/lib/tdb/wscript b/lib/tdb/wscript
index 812e598..8fbc4dd 100644
--- a/lib/tdb/wscript
+++ b/lib/tdb/wscript
@@ -95,7 +95,7 @@ def configure(conf):
         not conf.env.disable_tdb_mutex_locking):
         conf.define('USE_TDB_MUTEX_LOCKING', 1)
 
-    conf.CHECK_XSLTPROC_MANPAGES()
+    #conf.CHECK_XSLTPROC_MANPAGES()
 
     conf.SAMBA_CHECK_PYTHON()
     conf.SAMBA_CHECK_PYTHON_HEADERS()
diff --git a/wscript b/wscript
index 8dd6599..d1b8f0b 100644
--- a/wscript
+++ b/wscript
@@ -209,7 +209,7 @@ def configure(conf):
     conf.RECURSE('examples/winexe')
 
     conf.SAMBA_CHECK_PERL(mandatory=True)
-    conf.CHECK_XSLTPROC_MANPAGES()
+    #conf.CHECK_XSLTPROC_MANPAGES()
 
     if conf.env.disable_python:
         if not (Options.options.without_ad_dc):
