diff --git a/lib/util/charset/wscript_configure b/lib/util/charset/wscript_configure
index 222ae80..841e6fe 100644
--- a/lib/util/charset/wscript_configure
+++ b/lib/util/charset/wscript_configure
@@ -40,14 +40,14 @@ conf.CHECK_CODE('''
                 headers='errno.h iconv.h')
 """
 
-if conf.CHECK_CFG(package='icu-i18n icu-uc',
-               args='--cflags --libs',
-               msg='Checking for icu-i18n icu-uc',
-               uselib_store='ICUI18N'):
-    conf.env['icu-libs'] = 'icui18n'
-    conf.CHECK_LIB(conf.env['icu-libs'], shlib=True, mandatory=True)
-    if not conf.CHECK_HEADERS('unicode/ustring.h', lib='icui18n'):
-        conf.fatal('Found icui18n, but unicode/ustring.h is missing')
-    conf.DEFINE('HAVE_UTF8_NORMALISATION', 1)
-else:
-    conf.env['icu-libs'] = ''
+#if conf.CHECK_CFG(package='icu-i18n icu-uc',
+               #args='--cflags --libs',
+               #msg='Checking for icu-i18n icu-uc',
+               #uselib_store='ICUI18N'):
+    #conf.env['icu-libs'] = 'icui18n'
+    #conf.CHECK_LIB(conf.env['icu-libs'], shlib=True, mandatory=True)
+    #if not conf.CHECK_HEADERS('unicode/ustring.h', lib='icui18n'):
+        #conf.fatal('Found icui18n, but unicode/ustring.h is missing')
+    #conf.DEFINE('HAVE_UTF8_NORMALISATION', 1)
+#else:
+conf.env['icu-libs'] = ''
