diff --git a/lib/replace/tests/testsuite.c b/lib/replace/tests/testsuite.c
index 58624f1..2400bd2 100644
--- a/lib/replace/tests/testsuite.c
+++ b/lib/replace/tests/testsuite.c
@@ -31,9 +31,9 @@
   we include all the system/ include files here so that libreplace tests
   them in the build farm
 */
-#include "system/capability.h"
 #include "system/dir.h"
 #include "system/filesys.h"
+#include "system/capability.h"
 #include "system/glob.h"
 #include "system/iconv.h"
 #include "system/locale.h"
