diff --git a/wscript_configure_embedded_heimdal b/wscript_configure_embedded_heimdal
index 45f4772..6c5a4bc 100644
--- a/wscript_configure_embedded_heimdal
+++ b/wscript_configure_embedded_heimdal
@@ -13,3 +13,14 @@ conf.RECURSE('third_party/heimdal_build')
 # when this will be available also in
 # system libraries...
 conf.define('HAVE_CLIENT_GSS_C_CHANNEL_BOUND_FLAG', 1)
+
+def check_system_heimdal_binary(name):
+    if conf.LIB_MAY_BE_BUNDLED(name):
+        return False
+    if not conf.find_program(name, var=name.upper()):
+        return False
+    conf.define('USING_SYSTEM_%s' % name.upper(), 1)
+    return True
+
+check_system_heimdal_binary("compile_et")
+check_system_heimdal_binary("asn1_compile")
diff --git a/wscript_configure_system_heimdal b/wscript_configure_system_heimdal
index b6ca9e9..48aa193 100644
--- a/wscript_configure_system_heimdal
+++ b/wscript_configure_system_heimdal
@@ -24,14 +24,6 @@ if krb5_config:
 def check_system_heimdal_lib(name, functions='', headers=''):
     # Only use system library if the user requested the bundled one not be
     # used.
-    if conf.LIB_MAY_BE_BUNDLED(name):
-        return False
-    setattr(conf.env, "CPPPATH_%s" % name.upper(), heimdal_includedirs)
-    setattr(conf.env, "LIBPATH_%s" % name.upper(), heimdal_libdirs)
-    if not conf.CHECK_FUNCS_IN(functions, name, headers=headers, empty_decl=False, set_target=True):
-        return False
-    conf.define('USING_SYSTEM_%s' % name.upper(), 1)
-    return True
 
 check_system_heimdal_lib("com_err", "com_right_r com_err", "com_err.h")
 
