--- a/source3/utils/status.c
+++ b/source3/utils/status.c
@@ -1166,7 +1166,6 @@ int main(int argc, const char *argv[])
 			ret = 0;
 			goto done;
 		} else {
-			TALLOC_FREE(db);
 			TALLOC_FREE(db_path);
 		}
 
@@ -1176,6 +1175,8 @@ int main(int argc, const char *argv[])
 			goto done;
 		}
 
+		TALLOC_FREE(db);
+
 		prepare_share_mode(&state);
 		result = share_entry_forall(print_share_mode, &state);
 
@@ -1194,7 +1195,7 @@ int main(int argc, const char *argv[])
 			brl_forall(print_brl, &state);
 		}
 
-		locking_end();
+		// locking_end();
 	}
 
 	if (show_notify) {
