diff --git a/source3/utils/status.c b/source3/utils/status.c
index 02a5f6d..6e0b609 100644
--- a/source3/utils/status.c
+++ b/source3/utils/status.c
@@ -1259,7 +1259,6 @@ int main(int argc, const char *argv[])
 			ret = 0;
 			goto done;
 		} else {
-			TALLOC_FREE(db);
 			TALLOC_FREE(db_path);
 		}
 
@@ -1269,6 +1268,8 @@ int main(int argc, const char *argv[])
 			goto done;
 		}
 
+		TALLOC_FREE(db);
+
 		prepare_share_mode(&state);
 		result = share_entry_forall(print_share_mode, &state);
 
@@ -1287,7 +1288,7 @@ int main(int argc, const char *argv[])
 			brl_forall(print_brl, &state);
 		}
 
-		locking_end();
+		// locking_end();
 	}
 
 	if (show_notify) {
