From 42506af5003f9e8331da51ce989cd347258a85f5 Mon Sep 17 00:00:00 2001
From: Andy Blackburn <andrew.blackburn644@btinternet.com>
Date: Sun, 4 Aug 2013 14:11:26 +0100

---
 configure.in | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/configure.in b/configure.in
index 235a985..49336af 100644
--- a/configure.in
+++ b/configure.in
@@ -8,9 +8,7 @@ AC_SUBST(VERSION)
 
 AC_CANONICAL_HOST
 
-if test -z "$CC"; then
-	AC_PROG_CC	
-fi
+AC_PROG_CC	
 AC_PROG_RANLIB
 AC_CHECK_PROG(AR,ar,ar)
 AC_CHECK_PROG(INSTALL,install,install)
