From effd7d72b17ab3610e7ccbe460b7d8344a1c48c0 Mon Sep 17 00:00:00 2001
From: Andy Blackburn <andrew.blackburn644@btinternet.com>
Date: Sun, 4 Aug 2013 14:11:26 +0100

---
 configure.in             | 4 ++--
 interface/scan_devices.c | 4 ++++
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/configure.in b/configure.in
index 3ae6989..235a985 100644
--- a/configure.in
+++ b/configure.in
@@ -3,8 +3,8 @@ AC_INIT(interface/interface.c)
 VERSION="10.3pre" 
 AC_SUBST(VERSION) 
 
-cp $srcdir/configure.guess $srcdir/config.guess
-cp $srcdir/configure.sub $srcdir/config.sub
+#cp $srcdir/configure.guess $srcdir/config.guess
+#cp $srcdir/configure.sub $srcdir/config.sub
 
 AC_CANONICAL_HOST
 
diff --git a/interface/scan_devices.c b/interface/scan_devices.c
index fd3355d..b5e0f9f 100644
--- a/interface/scan_devices.c
+++ b/interface/scan_devices.c
@@ -22,6 +22,10 @@
 #include "common_interface.h"
 #include "utils.h"
 
+#ifndef PATH_MAX
+#define PATH_MAX 4096
+#endif
+
 #define MAX_DEV_LEN 20 /* Safe because strings only come from below */
 /* must be absolute paths! */
 static char *scsi_cdrom_prefixes[]={
