From f75ad196be1779e3d45342350089972c4c026637 Mon Sep 17 00:00:00 2001
From: Andy Blackburn <andrew.blackburn644@btinternet.com>
Date: Sun, 4 Aug 2013 14:11:26 +0100

---
 Makefile.in | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/Makefile.in b/Makefile.in
index bfef153..75996f4 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -90,8 +90,9 @@ install:
 	$(INSTALL) -m 0644 $(srcdir)/cdparanoia-3.pc $(PKGCONFIGDIR) 
 
 cdparanoia:	$(OFILES) $(LIBDEP)
-		$(LD) $(CFLAGS) $(LDFLAGS) $(OFILES) \
+		$(LD) $(CFLAGS) \
 		-L$(PWD)/paranoia -L$(PWD)/interface \
+		$(LDFLAGS) $(OFILES) \
 		-o cdparanoia $(LIBS)
 
 .c.o:
