From 618404469a741484ea6581f7edd7b2f672fcf602 Mon Sep 17 00:00:00 2001
From: Andy Blackburn <andrew.blackburn644@btinternet.com>
Date: Sun, 4 Aug 2013 14:11:26 +0100

---
 paranoia/Makefile.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/paranoia/Makefile.in b/paranoia/Makefile.in
index 89d0328..1a2aa07 100644
--- a/paranoia/Makefile.in
+++ b/paranoia/Makefile.in
@@ -44,7 +44,7 @@ libcdda_paranoia.a: 	$(OFILES)
 	$(RANLIB) libcdda_paranoia.a
 
 libcdda_paranoia.so: 	$(OFILES)	
-	$(CC) -fpic -shared -o libcdda_paranoia.so.0.$(VERSION) -Wl,-soname -Wl,libcdda_paranoia.so.0 $(OFILES) -L ../interface -lcdda_interface
+	$(CC) -fpic -shared -o libcdda_paranoia.so.0.$(VERSION) $(LDFLAGS) -Wl,-soname -Wl,libcdda_paranoia.so.0 $(OFILES) -L ../interface -lcdda_interface
 	[ -e libcdda_paranoia.so.0 ] || ln -s libcdda_paranoia.so.0.$(VERSION) libcdda_paranoia.so.0
 	[ -e libcdda_paranoia.so ] || ln -s libcdda_paranoia.so.0.$(VERSION) libcdda_paranoia.so
 
