--- a/genmetaindex.py
+++ b/genmetaindex.py
@@ -1,5 +1,7 @@
 # usage: genmetaindex.py <xml-files>  > index.xml
 import sys, os
+import six
+
 from xml.etree.ElementTree import ElementTree, Element
 
 root = Element("index")
@@ -47,4 +49,7 @@ def indent(elem, level=0):
 
 indent(root)
 
-ElementTree(root).write(sys.stdout)
+if six.PY3:
+	ElementTree(root).write(sys.stdout, encoding='unicode')
+elif six.PY2:
+	ElementTree(root).write(sys.stdout)
