diff -Naur media_build-bst-14.orig/linux/drivers/media/usb/dvb-usb-v2/Makefile media_build-bst-14/linux/drivers/media/usb/dvb-usb-v2/Makefile
--- media_build-bst-14.orig/linux/drivers/media/usb/dvb-usb-v2/Makefile	2014-11-25 18:01:41.198281957 +0100
+++ media_build-bst-14/linux/drivers/media/usb/dvb-usb-v2/Makefile	2014-11-25 18:01:22.578940354 +0100
@@ -1,5 +1,5 @@
-dvb_usb_v2-objs := dvb_usb_core.o dvb_usb_urb.o usb_urb.o
-obj-$(CONFIG_DVB_USB_V2) += dvb_usb_v2.o
+dvb_usb_v2_media_tree-objs := dvb_usb_core.o dvb_usb_urb.o usb_urb.o
+obj-$(CONFIG_DVB_USB_V2) += dvb_usb_v2_media_tree.o
 
 dvb-usb-af9015-objs := af9015.o
 obj-$(CONFIG_DVB_USB_AF9015) += dvb-usb-af9015.o
