diff --git a/builtins/mkbuiltins.c.orig b/builtins/mkbuiltins.c
index b5b2eede..298afb34 100644
--- a/builtins/mkbuiltins.c.orig
+++ b/builtins/mkbuiltins.c
@@ -60,7 +60,8 @@
 extern int errno;
 #endif
 
-static char *xmalloc (), *xrealloc ();
+static void *xmalloc (size_t);
+static void *xrealloc (void *, size_t);
 
 #if !defined (__STDC__) && !defined (strcpy)
 extern char *strcpy ();
@@ -192,36 +193,36 @@ char *arrayvar_builtins[] =
 };
 	
 /* Forward declarations. */
-static int is_special_builtin ();
-static int is_assignment_builtin ();
-static int is_localvar_builtin ();
-static int is_posix_builtin ();
-static int is_arrayvar_builtin ();
+static int is_special_builtin (char *);
+static int is_assignment_builtin (char *);
+static int is_localvar_builtin (char *);
+static int is_posix_builtin (char *);
+static int is_arrayvar_builtin (char *);
 
 #if !defined (HAVE_RENAME)
-static int rename ();
+static int rename (char *, char *);
 #endif
 
-void extract_info ();
+void extract_info (char *, FILE *, FILE *);
 
-void file_error ();
-void line_error ();
+void file_error (char *);
+void line_error (DEF_FILE *, char *, char *, char *);
 
-void write_file_headers ();
-void write_file_footers ();
-void write_ifdefs ();
-void write_endifs ();
-void write_documentation ();
-void write_longdocs ();
-void write_builtins ();
+void write_file_headers (FILE *, FILE *);
+void write_file_footers (FILE *, FILE *);
+void write_ifdefs (FILE *, char **);
+void write_endifs (FILE *, char **);
+void write_documentation (FILE *, char **, int, int);
+void write_longdocs (FILE *, ARRAY *);
+void write_builtins (DEF_FILE *, FILE *, FILE *);
 
-int write_helpfiles ();
+int write_helpfiles (ARRAY *);
 
-void free_defs ();
-void add_documentation ();
+void free_defs (DEF_FILE *);
+void add_documentation (DEF_FILE *, char *);
 
-void must_be_building ();
-void remove_trailing_whitespace ();
+void must_be_building (char *, DEF_FILE *);
+void remove_trailing_whitespace (char *);
 
 #define document_name(b)	((b)->docname ? (b)->docname : (b)->name)
 
@@ -230,9 +231,7 @@ void remove_trailing_whitespace ();
    write the information to STRUCTFILE and EXTERNFILE, while
    creating the production file if necessary. */
 int
-main (argc, argv)
-     int argc;
-     char **argv;
+main (int argc, char **argv)
 {
   int arg_index = 1;
   FILE *structfile, *externfile;
@@ -385,8 +384,7 @@ main (argc, argv)
 /* Make a new array, and return a pointer to it.  The array will
    contain elements of size WIDTH, and is initialized to no elements. */
 ARRAY *
-array_create (width)
-     int width;
+array_create (int width)
 {
   ARRAY *array;
 
@@ -405,8 +403,7 @@ array_create (width)
 
 /* Copy the array of strings in ARRAY. */
 ARRAY *
-copy_string_array (array)
-     ARRAY *array;
+copy_string_array (ARRAY *array)
 {
   register int i;
   ARRAY *copy;
@@ -432,9 +429,7 @@ copy_string_array (array)
 
 /* Add ELEMENT to ARRAY, growing the array if necessary. */
 void
-array_add (element, array)
-     char *element;
-     ARRAY *array;
+array_add (char *element,  ARRAY *array)
 {
   if (array->sindex + 2 > array->size)
     array->array = (char **)xrealloc
@@ -446,8 +441,7 @@ array_add (element, array)
 
 /* Free an allocated array and data pointer. */
 void
-array_free (array)
-     ARRAY *array;
+array_free (ARRAY *array)
 {
   if (array->array)
     free (array->array);
@@ -495,8 +489,7 @@ HANDLER_ENTRY handlers[] = {
 
 /* Return the entry in the table of handlers for NAME. */
 HANDLER_ENTRY *
-find_directive (directive)
-     char *directive;
+find_directive (char *directive)
 {
   register int i;
 
@@ -525,9 +518,7 @@ int output_cpp_line_info = 0;
    builtins found in each $BUILTIN.  Plain text found before the $PRODUCES
    is ignored, as is "$$ comment text". */
 void
-extract_info (filename, structfile, externfile)
-     char *filename;
-     FILE *structfile, *externfile;
+extract_info (char *filename, FILE *structfile, FILE *externfile)
 {
   register int i;
   DEF_FILE *defs;
@@ -610,7 +601,7 @@ extract_info (filename, structfile, externfile)
 
 	  if (!handler)
 	    {
-	      line_error (defs, "Unknown directive `%s'", directive);
+	      line_error (defs, "Unknown directive `%s'", directive, "");
 	      free (directive);
 	      continue;
 	    }
@@ -667,8 +658,7 @@ extract_info (filename, structfile, externfile)
 #define free_safely(x) if (x) free (x)
 
 static void
-free_builtin (builtin)
-     BUILTIN_DESC *builtin;
+free_builtin (BUILTIN_DESC *builtin)
 {
   register int i;
 
@@ -690,8 +680,7 @@ free_builtin (builtin)
 
 /* Free all of the memory allocated to a DEF_FILE. */
 void
-free_defs (defs)
-     DEF_FILE *defs;
+free_defs (DEF_FILE *defs)
 {
   register int i;
   register BUILTIN_DESC *builtin;
@@ -723,8 +712,7 @@ free_defs (defs)
 /* Strip surrounding whitespace from STRING, and
    return a pointer to the start of it. */
 char *
-strip_whitespace (string)
-     char *string;
+strip_whitespace (char *string)
 {
   while (whitespace (*string))
       string++;
@@ -735,8 +723,7 @@ strip_whitespace (string)
 
 /* Remove only the trailing whitespace from STRING. */
 void
-remove_trailing_whitespace (string)
-     char *string;
+remove_trailing_whitespace (char *string)
 {
   register int i;
 
@@ -753,35 +740,29 @@ remove_trailing_whitespace (string)
    DEFS is the DEF_FILE in which the directive is found.
    If there is no argument, produce an error. */
 char *
-get_arg (for_whom, defs, string)
-     char *for_whom, *string;
-     DEF_FILE *defs;
+get_arg (char *for_whom, DEF_FILE *defs, char *string)
 {
   char *new;
 
   new = strip_whitespace (string);
 
   if (!*new)
-    line_error (defs, "%s requires an argument", for_whom);
+    line_error (defs, "%s requires an argument", for_whom, "");
 
   return (savestring (new));
 }
 
 /* Error if not building a builtin. */
 void
-must_be_building (directive, defs)
-     char *directive;
-     DEF_FILE *defs;
+must_be_building (char *directive, DEF_FILE *defs)
 {
   if (!building_builtin)
-    line_error (defs, "%s must be inside of a $BUILTIN block", directive);
+    line_error (defs, "%s must be inside of a $BUILTIN block", directive, "");
 }
 
 /* Return the current builtin. */
 BUILTIN_DESC *
-current_builtin (directive, defs)
-     char *directive;
-     DEF_FILE *defs;
+current_builtin (char *directive, DEF_FILE *defs)
 {
   must_be_building (directive, defs);
   if (defs->builtins)
@@ -793,9 +774,7 @@ current_builtin (directive, defs)
 /* Add LINE to the long documentation for the current builtin.
    Ignore blank lines until the first non-blank line has been seen. */
 void
-add_documentation (defs, line)
-     DEF_FILE *defs;
-     char *line;
+add_documentation (DEF_FILE *defs, char *line)
 {
   register BUILTIN_DESC *builtin;
 
@@ -814,10 +793,7 @@ add_documentation (defs, line)
 
 /* How to handle the $BUILTIN directive. */
 int
-builtin_handler (self, defs, arg)
-     char *self;
-     DEF_FILE *defs;
-     char *arg;
+builtin_handler (char *self, DEF_FILE *defs, char *arg)
 {
   BUILTIN_DESC *new;
   char *name;
@@ -825,7 +801,7 @@ builtin_handler (self, defs, arg)
   /* If we are already building a builtin, we cannot start a new one. */
   if (building_builtin)
     {
-      line_error (defs, "%s found before $END", self);
+      line_error (defs, "%s found before $END", self, "");
       return (-1);
     }
 
@@ -866,10 +842,7 @@ builtin_handler (self, defs, arg)
 
 /* How to handle the $FUNCTION directive. */
 int
-function_handler (self, defs, arg)
-     char *self;
-     DEF_FILE *defs;
-     char *arg;
+function_handler (char *self, DEF_FILE *defs, char *arg)
 {
   register BUILTIN_DESC *builtin;
 
@@ -877,7 +850,7 @@ function_handler (self, defs, arg)
 
   if (builtin == 0)
     {
-      line_error (defs, "syntax error: no current builtin for $FUNCTION directive");
+      line_error (defs, "syntax error: no current builtin for $FUNCTION directive", "", "");
       exit (1);
     }
   if (builtin->function)
@@ -891,10 +864,7 @@ function_handler (self, defs, arg)
 
 /* How to handle the $DOCNAME directive. */
 int
-docname_handler (self, defs, arg)
-     char *self;
-     DEF_FILE *defs;
-     char *arg;
+docname_handler (char *self, DEF_FILE *defs, char *arg)
 {
   register BUILTIN_DESC *builtin;
 
@@ -911,10 +881,7 @@ docname_handler (self, defs, arg)
 
 /* How to handle the $SHORT_DOC directive. */
 int
-short_doc_handler (self, defs, arg)
-     char *self;
-     DEF_FILE *defs;
-     char *arg;
+short_doc_handler (char *self, DEF_FILE *defs, char *arg)
 {
   register BUILTIN_DESC *builtin;
 
@@ -931,20 +898,14 @@ short_doc_handler (self, defs, arg)
 
 /* How to handle the $COMMENT directive. */
 int
-comment_handler (self, defs, arg)
-     char *self;
-     DEF_FILE *defs;
-     char *arg;
+comment_handler (char *self, DEF_FILE *defs, char *arg)
 {
   return (0);
 }
 
 /* How to handle the $DEPENDS_ON directive. */
 int
-depends_on_handler (self, defs, arg)
-     char *self;
-     DEF_FILE *defs;
-     char *arg;
+depends_on_handler (char *self, DEF_FILE *defs, char *arg)
 {
   register BUILTIN_DESC *builtin;
   char *dependent;
@@ -962,10 +923,7 @@ depends_on_handler (self, defs, arg)
 
 /* How to handle the $PRODUCES directive. */
 int
-produces_handler (self, defs, arg)
-     char *self;
-     DEF_FILE *defs;
-     char *arg;
+produces_handler (char *self, DEF_FILE *defs, char *arg)
 {
   /* If just hacking documentation, don't change any of the production
      files. */
@@ -996,10 +954,7 @@ produces_handler (self, defs, arg)
 
 /* How to handle the $END directive. */
 int
-end_handler (self, defs, arg)
-     char *self;
-     DEF_FILE *defs;
-     char *arg;
+end_handler (char *self, DEF_FILE *defs, char *arg)
 {
   must_be_building (self, defs);
   building_builtin = 0;
@@ -1014,9 +969,7 @@ end_handler (self, defs, arg)
 
 /* Produce an error for DEFS with FORMAT and ARGS. */
 void
-line_error (defs, format, arg1, arg2)
-     DEF_FILE *defs;
-     char *format, *arg1, *arg2;
+line_error (DEF_FILE *defs, char *format, char *arg1, char *arg2)
 {
   if (defs->filename[0] != '/')
     fprintf (stderr, "%s", error_directory ? error_directory : "./");
@@ -1028,8 +981,7 @@ line_error (defs, format, arg1, arg2)
 
 /* Print error message for FILENAME. */
 void
-file_error (filename)
-     char *filename;
+file_error (char *filename)
 {
   perror (filename);
   exit (2);
@@ -1041,30 +993,27 @@ file_error (filename)
 /*								    */
 /* **************************************************************** */
 
-static void memory_error_and_abort ();
+static void memory_error_and_abort (void);
 
-static char *
-xmalloc (bytes)
-     int bytes;
+static void *
+xmalloc (size_t bytes)
 {
-  char *temp = (char *)malloc (bytes);
+  void *temp = malloc (bytes);
 
   if (!temp)
     memory_error_and_abort ();
   return (temp);
 }
 
-static char *
-xrealloc (pointer, bytes)
-     char *pointer;
-     int bytes;
+static void *
+xrealloc (void *pointer, size_t bytes)
 {
-  char *temp;
+  void *temp;
 
   if (!pointer)
-    temp = (char *)malloc (bytes);
+    temp = malloc (bytes);
   else
-    temp = (char *)realloc (pointer, bytes);
+    temp = realloc (pointer, bytes);
 
   if (!temp)
     memory_error_and_abort ();
@@ -1073,7 +1022,7 @@ xrealloc (pointer, bytes)
 }
 
 static void
-memory_error_and_abort ()
+memory_error_and_abort (void)
 {
   fprintf (stderr, "mkbuiltins: out of virtual memory\n");
   abort ();
@@ -1088,8 +1037,7 @@ memory_error_and_abort ()
 /* Return a pointer to a newly allocated builtin which is
    an exact copy of BUILTIN. */
 BUILTIN_DESC *
-copy_builtin (builtin)
-     BUILTIN_DESC *builtin;
+copy_builtin (BUILTIN_DESC *builtin)
 {
   BUILTIN_DESC *new;
 
@@ -1110,8 +1058,7 @@ copy_builtin (builtin)
 
 /* How to save away a builtin. */
 void
-save_builtin (builtin)
-     BUILTIN_DESC *builtin;
+save_builtin (BUILTIN_DESC *builtin)
 {
   BUILTIN_DESC *newbuiltin;
 
@@ -1188,8 +1135,7 @@ char *structfile_footer[] = {
 /* Write out any necessary opening information for
    STRUCTFILE and EXTERNFILE. */
 void
-write_file_headers (structfile, externfile)
-     FILE *structfile, *externfile;
+write_file_headers (FILE *structfile, FILE *externfile)
 {
   register int i;
 
@@ -1215,8 +1161,7 @@ write_file_headers (structfile, externfile)
 /* Write out any necessary closing information for
    STRUCTFILE and EXTERNFILE. */
 void
-write_file_footers (structfile, externfile)
-     FILE *structfile, *externfile;
+write_file_footers (FILE *structfile, FILE *externfile)
 {
   register int i;
 
@@ -1231,9 +1176,7 @@ write_file_footers (structfile, externfile)
 /* Write out the information accumulated in DEFS to
    STRUCTFILE and EXTERNFILE. */
 void
-write_builtins (defs, structfile, externfile)
-     DEF_FILE *defs;
-     FILE *structfile, *externfile;
+write_builtins (DEF_FILE *defs, FILE *structfile, FILE *externfile)
 {
   register int i;
 
@@ -1337,9 +1280,7 @@ write_builtins (defs, structfile, externfile)
 
 /* Write out the long documentation strings in BUILTINS to STREAM. */
 void
-write_longdocs (stream, builtins)
-     FILE *stream;
-     ARRAY *builtins;
+write_longdocs (FILE *stream, ARRAY *builtins)
 {
   register int i;
   register BUILTIN_DESC *builtin;
@@ -1376,9 +1317,7 @@ write_longdocs (stream, builtins)
 }
 
 void
-write_dummy_declarations (stream, builtins)
-     FILE *stream;
-     ARRAY *builtins;
+write_dummy_declarations (FILE *stream, ARRAY *builtins)
 {
   register int i;
   BUILTIN_DESC *builtin;
@@ -1402,9 +1341,7 @@ write_dummy_declarations (stream, builtins)
    If a define is preceded by an `!', then the sense of the test is
    reversed. */
 void
-write_ifdefs (stream, defines)
-     FILE *stream;
-     char **defines;
+write_ifdefs (FILE *stream, char **defines)
 {
   register int i;
 
@@ -1433,9 +1370,7 @@ write_ifdefs (stream, defines)
    STREAM is the stream to write the information to.
    DEFINES is a null terminated array of define names. */
 void
-write_endifs (stream, defines)
-     FILE *stream;
-     char **defines;
+write_endifs (FILE *stream, char **defines)
 {
   register int i;
 
@@ -1460,10 +1395,7 @@ write_endifs (stream, defines)
    internationalization (gettext) and the single-string vs. multiple-strings
    issues. */
 void
-write_documentation (stream, documentation, indentation, flags)
-     FILE *stream;
-     char **documentation;
-     int indentation, flags;
+write_documentation (FILE *stream, char **documentation, int indentation, int flags)
 {
   register int i, j;
   register char *line;
@@ -1591,8 +1523,7 @@ write_documentation (stream, documentation, indentation, flags)
 }
 
 int
-write_helpfiles (builtins)
-     ARRAY *builtins;
+write_helpfiles (ARRAY *builtins)
 {
   char *helpfile, *bname;
   FILE *helpfp;
@@ -1633,8 +1564,7 @@ write_helpfiles (builtins)
 }      
       	        
 static int
-_find_in_table (name, name_table)
-     char *name, *name_table[];
+_find_in_table (char *name, char **name_table)
 {
   register int i;
 
@@ -1645,44 +1575,38 @@ _find_in_table (name, name_table)
 }
 
 static int
-is_special_builtin (name)
-     char *name;
+is_special_builtin (char *name)
 {
   return (_find_in_table (name, special_builtins));
 }
 
 static int
-is_assignment_builtin (name)
-     char *name;
+is_assignment_builtin (char *name)
 {
   return (_find_in_table (name, assignment_builtins));
 }
 
 static int
-is_localvar_builtin (name)
-     char *name;
+is_localvar_builtin (char *name)
 {
   return (_find_in_table (name, localvar_builtins));
 }
 
 static int
-is_posix_builtin (name)
-     char *name;
+is_posix_builtin (char *name)
 {
   return (_find_in_table (name, posix_builtins));
 }
 
 static int
-is_arrayvar_builtin (name)
-     char *name;
+is_arrayvar_builtin (char *name)
 {
   return (_find_in_table (name, arrayvar_builtins));
 }
 
 #if !defined (HAVE_RENAME)
 static int
-rename (from, to)
-     char *from, *to;
+rename (char *from, *to)
 {
   unlink (to);
   if (link (from, to) < 0)
