--- /lib/python/enigma_python.i.org	2012-03-24 13:09:04.062875841 +0100
+++ /lib/python/enigma_python.i	2012-03-24 13:01:04.210891966 +0100
@@ -98,6 +98,7 @@
 #include <lib/components/tuxtxtapp.h>
 #include <lib/driver/avswitch.h>
 #include <lib/driver/hdmi_cec.h>
+#include <lib/driver/vfd.h>
 #include <lib/driver/rfmod.h>
 #include <lib/driver/misc_options.h>
 #include <lib/driver/etimezone.h>
@@ -234,6 +235,7 @@
 %include <lib/components/tuxtxtapp.h>
 %include <lib/driver/avswitch.h>
 %include <lib/driver/hdmi_cec.h>
+%include <lib/driver/vfd.h>
 %include <lib/driver/rfmod.h>
 %include <lib/driver/misc_options.h>
 %include <lib/driver/etimezone.h>
