--- a/daemons/Makefile.am
+++ b/daemons/Makefile.am
@@ -83,7 +83,7 @@ irrecord_SOURCES = irrecord.c \
 		transmit.c transmit.h
 
 irrecord_LDADD = libhw_module.a @hw_module_libs@ @receive@
-irrecord_DEPENDENCIES = @receive@
+irrecord_DEPENDENCIES = libhw_module.a @receive@
 
 ## maintainer mode stuff
 EXTRA_PROGRAMS = lircd.simsend lircd.simrec slinke
--- a/daemons/Makefile.in
+++ b/daemons/Makefile.in
@@ -340,7 +340,7 @@ irrecord_SOURCES = irrecord.c \
 		transmit.c transmit.h
 
 irrecord_LDADD = libhw_module.a @hw_module_libs@ @receive@
-irrecord_DEPENDENCIES = @receive@
+irrecord_DEPENDENCIES = libhw_module.a @receive@
 noinst_PROGRAMS = @maintmode_daemons_extra@
 lircd_simsend_SOURCES = lircd.c ir_remote.c config_file.c \
 		lircd.h ir_remote.h ir_remote_types.h config_file.h \
