--- a/daemons/lircd.c
+++ b/daemons/lircd.c
@@ -101,6 +101,8 @@ static __u32 repeat_max = REPEAT_MAX_DEFAULT;
 extern struct hardware hw;
 
 char *progname = "lircd";
+char *uinputname = "remote control";
+
 const char *configfile = NULL;
 #ifndef USE_SYSLOG
 char *logfile = LOGFILE;
@@ -996,7 +998,7 @@ void start_server(mode_t permission, int nodaemon)
 	nolinger(sockfd);
 
 	if (useuinput) {
-		uinputfd = setup_uinputfd(progname);
+		uinputfd = setup_uinputfd(uinputname);
 		setup_repeatrate_uinputfd(uinputfd);
 	}
 	if (listen_tcpip) {
