--- gstreamer/tools/gst-inspect.c.orig	2015-11-17 11:54:36.849000877 +0100
+++ gstreamer/tools/gst-inspect.c	2015-11-17 11:56:23.522978789 +0100
@@ -1511,7 +1511,7 @@
   g_option_context_add_main_entries (ctx, options, GETTEXT_PACKAGE);
   g_option_context_add_group (ctx, gst_init_get_option_group ());
   if (!g_option_context_parse (ctx, &argc, &argv, &err)) {
-    g_printerr ("Error initializing: %s\n", err->message);
+    g_printerr ("Error initializing: %s\n", err ? err->message : "(null)");
     g_clear_error (&err);
     g_option_context_free (ctx);
     return -1;
