From 7c8ca2db3e875a426b3e6c4a6558da23b7b431b1 Mon Sep 17 00:00:00 2001
From: atvcaptain <captain@web.de>
Date: Sun, 10 Nov 2013 11:57:29 +0100
Subject: [PATCH] * Create openssl-old that provides obsolete 0.9.8 versions

Upstream-Status: Backport [debian]

---
 Makefile.org | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile.org b/Makefile.org
index 960cfec..761461d 100644
--- a/Makefile.org
+++ b/Makefile.org
@@ -132,7 +132,7 @@ FIPSCANLIB=
 
 BASEADDR=
 
-DIRS=   crypto fips ssl engines apps test tools
+DIRS=   crypto fips ssl engines apps tools
 SHLIBDIRS= crypto ssl fips
 
 # dirs in crypto to build
