From 801a7183d7dec2ef185b44c92744eacb16c32465 Mon Sep 17 00:00:00 2001
From: atvcaptain <captain@web.de>
Date: Sun, 10 Nov 2013 11:57:29 +0100
Subject: [PATCH] * Create openssl-old that provides obsolete 0.9.8 versions

Upstream-Status: Backport [debian]

---
 Makefile.org | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/Makefile.org b/Makefile.org
index 035c03a..a043225 100644
--- a/Makefile.org
+++ b/Makefile.org
@@ -156,7 +156,8 @@ MAKEFILE= Makefile
 MANDIR=/usr/share/man
 MAN1=1
 MAN3=3
-MANSUFFIX=
+MANSUFFIX=ssl
+MANSECTION=SSL
 SHELL=/bin/sh
 
 TOP=    .
@@ -698,7 +699,7 @@ install_docs:
 		echo "installing man$$sec/$$fn.$${sec}$(MANSUFFIX)"; \
 		(cd `$(PERL) util/dirname.pl $$i`; \
 		sh -c "$$pod2man \
-			--section=$$sec --center=OpenSSL \
+			--section=$${sec}$(MANSECTION) --center=OpenSSL \
 			--release=$(VERSION) `basename $$i`") \
 			>  $(INSTALL_PREFIX)$(MANDIR)/man$$sec/$$fn.$${sec}$(MANSUFFIX); \
 		$(PERL) util/extract-names.pl < $$i | \
@@ -715,7 +716,7 @@ install_docs:
 		echo "installing man$$sec/$$fn.$${sec}$(MANSUFFIX)"; \
 		(cd `$(PERL) util/dirname.pl $$i`; \
 		sh -c "$$pod2man \
-			--section=$$sec --center=OpenSSL \
+			--section=$${sec}$(MANSECTION) --center=OpenSSL \
 			--release=$(VERSION) `basename $$i`") \
 			>  $(INSTALL_PREFIX)$(MANDIR)/man$$sec/$$fn.$${sec}$(MANSUFFIX); \
 		$(PERL) util/extract-names.pl < $$i | \
