From c49f6d51e30a4fe4cf4498e16ccfd727031a7818 Mon Sep 17 00:00:00 2001
From: atvcaptain <captain@web.de>
Date: Sun, 10 Nov 2013 11:57:29 +0100
Subject: [PATCH] * Create openssl-old that provides obsolete 0.9.8 versions

Upstream-Status: Backport [debian]

---
 Makefile.shared | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile.shared b/Makefile.shared
index 6951881..48e2710 100644
--- a/Makefile.shared
+++ b/Makefile.shared
@@ -149,7 +149,7 @@ DO_GNU_SO=$(CALC_VERSIONS); \
 	SHLIB_SUFFIX=; \
 	ALLSYMSFLAGS='-Wl,--whole-archive'; \
 	NOALLSYMSFLAGS='-Wl,--no-whole-archive'; \
-	SHAREDFLAGS="$(CFLAGS) $(SHARED_LDFLAGS) -shared -Wl,-Bsymbolic -Wl,-soname=$$SHLIB$$SHLIB_SOVER$$SHLIB_SUFFIX"
+	SHAREDFLAGS="$(CFLAGS) $(SHARED_LDFLAGS) -shared -Wl,-soname=$$SHLIB$$SHLIB_SOVER$$SHLIB_SUFFIX"
 
 DO_GNU_APP=LDFLAGS="$(CFLAGS)"
 
