--- openssl-0.9.8x.orig/Makefile.org    2010-01-27 17:06:36.000000000 +0100
+++ openssl-0.9.8x/Makefile.org 2014-08-24 13:35:14.335683341 +0200
@@ -615,7 +615,7 @@
 dist_pem_h:
        (cd crypto/pem; $(MAKE) -e $(BUILDENV) pem.h; $(MAKE) clean)

-install: all install_docs install_sw
+install: all install_sw

 install_sw:
        @$(PERL) $(TOP)/util/mkdir-p.pl $(INSTALL_PREFIX)$(INSTALLTOP)/bin \