From cb920a4d4dd4e4f0f50039cb528a0a33c976e83d Mon Sep 17 00:00:00 2001
From: atvcaptain <captain@web.de>
Date: Mon, 1 Sep 2014 15:54:16 +0200

---
 Makefile.org | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile.org b/Makefile.org
index 353cec8..f94a850 100644
--- a/Makefile.org
+++ b/Makefile.org
@@ -619,7 +619,7 @@ dist:
 dist_pem_h:
 	(cd crypto/pem; $(MAKE) -e $(BUILDENV) pem.h; $(MAKE) clean)
 
-install: all install_docs install_sw
+install: all install_sw
 
 install_sw:
 	@$(PERL) $(TOP)/util/mkdir-p.pl $(INSTALL_PREFIX)$(INSTALLTOP)/bin \
