From e98d1ba67a5da9e9775c5199635a0b9d84aaabdf Mon Sep 17 00:00:00 2001
From: atvcaptain <captain@web.de>
Date: Sun, 10 Nov 2013 11:57:29 +0100
Subject: [PATCH] * Create openssl-old that provides obsolete 0.9.8 versions

Upstream-Status: Backport [debian]

---
 Configure | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/Configure b/Configure
index 62ab8d2..621eae3 100755
--- a/Configure
+++ b/Configure
@@ -130,6 +130,9 @@ my $x86_coff_asm="x86cpuid-cof.o:bn86-cof.o co86-cof.o MAYBE-MO86-cof.o:dx86-cof
 my $x86_out_asm="x86cpuid-out.o:bn86-out.o co86-out.o MAYBE-MO86-out.o:dx86-out.o yx86-out.o:ax86-out.o:bx86-out.o:mx86-out.o:sx86-out.o s512sse2-out.o:cx86-out.o:rx86-out.o rc4_skey.o:rm86-out.o:r586-out.o";
 
 my $x86_64_asm="x86_64cpuid.o:x86_64-gcc.o x86_64-mont.o::aes-x86_64.o::md5-x86_64.o:sha1-x86_64.o sha256-x86_64.o sha512-x86_64.o::rc4-x86_64.o::";
+# rc4 asm is disabled on amd64 because we configured it with RC4_CHAR while
+# the assembler only works with int
+my $x86_64_asm_linux="x86_64cpuid.o:x86_64-gcc.o x86_64-mont.o::aes-x86_64.o::md5-x86_64.o:sha1-x86_64.o sha256-x86_64.o sha512-x86_64.o::::";
 my $ia64_asm=":bn-ia64.o::aes_core.o aes_cbc.o aes-ia64.o:::sha1-ia64.o sha256-ia64.o sha512-ia64.o::rc4-ia64.o rc4_skey.o::";
 
 my $no_asm="::::::::::";
