From 8b7f6a6b5f503548f8834e3e95aff028a006c450 Mon Sep 17 00:00:00 2001
From: atvcaptain <captain@web.de>
Date: Sun, 10 Nov 2013 11:57:29 +0100
Subject: [PATCH] * Create openssl-old that provides obsolete 0.9.8 versions

Upstream-Status: Backport [debian]

---
 Configure | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/Configure b/Configure
index 621eae3..31a9f6f 100755
--- a/Configure
+++ b/Configure
@@ -1615,7 +1615,8 @@ while (<IN>)
 	elsif ($shared_extension ne "" && $shared_extension =~ /^\.s([ol])\.[^\.]*\.[^\.]*$/)
 		{
 		my $sotmp = $1;
-		s/^SHARED_LIBS_LINK_EXTS=.*/SHARED_LIBS_LINK_EXTS=.s$sotmp.\$(SHLIB_MAJOR) .s$sotmp/;
+#		s/^SHARED_LIBS_LINK_EXTS=.*/SHARED_LIBS_LINK_EXTS=.s$sotmp.\$(SHLIB_MAJOR) .s$sotmp/;
+		s/^SHARED_LIBS_LINK_EXTS=.*/SHARED_LIBS_LINK_EXTS=.s$sotmp/;
 		}
 	elsif ($shared_extension ne "" && $shared_extension =~ /^\.[^\.]*\.[^\.]*\.dylib$/)
 		{
