From 5abf877c8ab526661f5afb72b63ab3f1075267c3 Mon Sep 17 00:00:00 2001
From: atvcaptain <captain@web.de>
Date: Sun, 10 Nov 2013 11:57:29 +0100
Subject: [PATCH] * Create openssl-old that provides obsolete 0.9.8 versions

Upstream-Status: Backport [debian]

---
 crypto/sha/sha.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/crypto/sha/sha.h b/crypto/sha/sha.h
index 47a2c29..dc5f0fe 100644
--- a/crypto/sha/sha.h
+++ b/crypto/sha/sha.h
@@ -59,6 +59,7 @@
 #ifndef HEADER_SHA_H
 #define HEADER_SHA_H
 
+#include <stddef.h>
 #include <openssl/e_os2.h>
 #include <stddef.h>
 
