--- a/subprojects/FFmpeg/libavcodec/h264_cabac.c
+++ b/subprojects/FFmpeg/libavcodec/h264_cabac.c
@@ -2020,7 +2020,11 @@ decode_intra_mb:
         // In deblocking, the quantizer is 0
         h->cur_pic.qscale_table[mb_xy] = 0;
         // All coeffs are present
-        memset(h->non_zero_count[mb_xy], 16, 48);
+        /*memset(h->non_zero_count[mb_xy], 16, 48);*/
+            /* avoiding this memset because it leads at least with gcc4.9.2 to error: 'asm' operand has impossible constraints */
+        for (size_t i = 0; i < 48; i++) {
+            ( (unsigned char*)(h->non_zero_count[mb_xy]) ) [i] = 16;
+        }
         h->cur_pic.mb_type[mb_xy] = mb_type;
         sl->last_qscale_diff = 0;
         return 0;
