--- a/libnsfb/src/plot/32bpp-xbgr8888.c
+++ b/libnsfb/src/plot/32bpp-xbgr8888.c
@@ -66,7 +66,7 @@
  */
 static inline nsfb_colour_t pixel_to_colour(UNUSED nsfb_t *nsfb, uint32_t pixel)
 {
-        return pixel | 0xFF000000U;
+        return (((pixel & 0xFF) << 16) | ((pixel & 0xFF00)) | ((pixel & 0xFF0000) >> 16))  | 0xff000000;
 }
 
 
@@ -79,7 +79,7 @@
  */
 static inline uint32_t colour_to_pixel(UNUSED nsfb_t *nsfb, nsfb_colour_t c)
 {
-        return c;
+        return (((c >> 16) & 0x000000FF) | ((c << 16)  & 0x00FF0000) | (c  & 0xff00ff00)) | 0xff000000;
 }
 
 #endif

