diff --git a/src/lzma-sdk/C/Common/MyCom.h b/src/lzma-sdk/C/Common/MyCom.h
index b8dbf38..4cb299e 100644
--- a/src/lzma-sdk/C/Common/MyCom.h
+++ b/src/lzma-sdk/C/Common/MyCom.h
@@ -157,7 +157,7 @@ public:
 #define MY_ADDREF_RELEASE \
 STDMETHOD_(ULONG, AddRef)() { return ++__m_RefCount; } \
 STDMETHOD_(ULONG, Release)() { if (--__m_RefCount != 0)  \
-  return __m_RefCount; delete this; return 0; }
+  { return __m_RefCount; } delete this; return 0; }
 
 #define MY_UNKNOWN_IMP_SPEC(i) \
   MY_QUERYINTERFACE_BEGIN \
