diff --git a/src/gstdvbaudiosink.c b/src/gstdvbaudiosink.c
index dc7ace1..3c0bb42 100644
--- a/src/gstdvbaudiosink.c
+++ b/src/gstdvbaudiosink.c
@@ -204,7 +204,6 @@ GST_STATIC_PAD_TEMPLATE (
 		"audio/x-dts; "
 		"audio/x-private1-dts; "
 		"audio/x-private1-lpcm; "
-		"audio/x-wma;"
 		"audio/x-eac3;"
 		"audio/x-private-eac3")
 );
@@ -323,13 +322,7 @@ gst_dvbaudiosink_get_caps (GstBaseSink *basesink)
 
 	int eac3_pos = 12;
 
-	if (hwtype == DM8000) {
-		caps = gst_caps_copy(&hwtemplate->static_caps.caps);
-		gst_caps_remove_structure(caps, 11); // remove WMA!!
-		--eac3_pos;
-	}
-	else
-		caps = gst_static_caps_get(&hwtemplate->static_caps);
+	caps = gst_static_caps_get(&hwtemplate->static_caps);
 
 	if (eac3_support < 0) {
 		if (eac3_pos == 12)
