diff --git a/src/gstdvbaudiosink.c b/src/gstdvbaudiosink.c
index 0275f94..d02a34c 100644
--- a/src/gstdvbaudiosink.c
+++ b/src/gstdvbaudiosink.c
@@ -202,7 +202,6 @@ GST_STATIC_PAD_TEMPLATE (
 		"audio/x-dts; "
 		"audio/x-private1-dts; "
 		"audio/x-private1-lpcm; "
-		"audio/x-wma;"
 		"audio/x-eac3;"
 		"audio/x-private-eac3")
 );
@@ -307,34 +306,7 @@ gst_dvbaudiosink_get_caps (GstBaseSink *basesink)
 //	GstDVBAudioSink *self = GST_DVBAUDIOSINK (basesink);
 //	gchar *strcaps;
 	GstCaps *caps;
-	static int eac3_support;
-
-	if (!eac3_support) {
-		int fd = open("/proc/stb/audio/ac3plus", O_RDONLY);
-		if (fd >= 0) {
-			eac3_support = 1;
-			close(fd);
-		}
-		else
-			eac3_support = -1;
-	}
-
-	int eac3_pos = 12;
-
-	if (hwtype == DM8000) {
-		caps = gst_caps_copy(&hwtemplate->static_caps.caps);
-		gst_caps_remove_structure(caps, 11); // remove WMA!!
-		--eac3_pos;
-	}
-	else
-		caps = gst_static_caps_get(&hwtemplate->static_caps);
-
-	if (eac3_support < 0) {
-		if (eac3_pos == 12)
-			caps = gst_caps_copy(&hwtemplate->static_caps.caps);
-		gst_caps_remove_structure(caps, eac3_pos); // remove x-eac3
-		gst_caps_remove_structure(caps, eac3_pos); // remove x-private-eac3
-	}
+	caps = gst_static_caps_get(&hwtemplate->static_caps);
 
 //	strcaps = gst_caps_to_string(caps);
 //	GST_INFO_OBJECT (self, "dynamic caps for model %d '%s'", hwtype, gst_caps_to_string(caps));
@@ -891,11 +863,11 @@ gst_dvbaudiosink_set_caps (GstBaseSink * basesink, GstCaps * caps)
 	}
 	else if (!strcmp(type, "audio/x-eac3")) {
 		GST_INFO_OBJECT (self, "MIMETYPE %s",type);
-		bypass = 7;
+		bypass = 0x22;
 	}
 	else if (!strcmp(type, "audio/x-private1-eac3")) {
 		GST_INFO_OBJECT (self, "MIMETYPE %s (DVD Audio - 2 byte skipping)",type);
-		bypass = 7;
+		bypass = 0x22;
 		self->skip = 2;
 	}
 	else if (!strcmp(type, "audio/x-private1-lpcm")) {
@@ -1292,7 +1264,7 @@ gst_dvbaudiosink_render (GstBaseSink * sink, GstBuffer * buffer)
 			self->timestamp = timestamp;
 		else
 			timestamp = self->timestamp;
-		if (self->bypass < 0xd)
+		if (self->bypass < 0xd || self->bypass == 0x22)
 			self->timestamp += duration;
 	}
 	else
diff --git a/src/gstdvbvideosink.c b/src/gstdvbvideosink.c
index 29afeee..71014f9 100644
--- a/src/gstdvbvideosink.c
+++ b/src/gstdvbvideosink.c
@@ -324,8 +324,8 @@ gst_dvbvideosink_base_init (gpointer klass)
 				GST_INFO ("model is DM7020HD... set bcm7405 caps");
 			} else if ( !strncasecmp(string, "DM8000", 6) ) {
 				hwtype = DM8000;
-				hwtemplate = &sink_factory_bcm7400;
-				GST_INFO ("model is DM8000... set bcm7400 caps");
+				hwtemplate = &sink_factory_bcm7405;
+				GST_INFO ("model is DM8000... set bcm7405 caps");
 			} else if ( !strncasecmp(string, "DM800", 5) ) {
 				hwtype = DM800;
 				hwtemplate = &sink_factory_bcm7401;
