From 54af4288418da2f8df11de4a87ce4d66e95572c5 Mon Sep 17 00:00:00 2001
From: Huevos <rowland@moralet.com>
Date: Fri, 17 Sep 2021 14:28:31 +0200

---
 Makefile.am | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 09343ff1..addd5090 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -22,7 +22,6 @@ SUBDIRS = \
 	dreammediathek \
 	dvdbackup \
 	easyinfo \
-	easymedia \
 	ecasa \
 	eibox \
 	emailclient \
@@ -59,13 +58,11 @@ SUBDIRS = \
 	ofdb \
 	orfat \
 	orfteletext \
-	partnerbox \
 	permanentclock \
 	pipservicerelation \
 	podcast \
 	porncenter \
 	pushservice \
-	quickbutton \
 	reconstructapsc \
 	reftopiconname \
 	remotetimer \
@@ -128,8 +125,6 @@ if TPM
 SUBDIRS += \
 	mytube \
 	yttrailer \
-	webinterface \
-	menusort \
 	pluginhider \
 	pluginsort 
 endif
