From 54af4288418da2f8df11de4a87ce4d66e95572c5 Mon Sep 17 00:00:00 2001
From: Huevos <rowland@moralet.com>
Date: Fri, 17 Sep 2021 14:28:31 +0200

---
 Makefile.am | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 574603415..7d15059f5 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -21,7 +21,6 @@ SUBDIRS = \
 	dreamirc \
 	dreammediathek \
 	dvdbackup \
-	easyinfo \
 	easymedia \
 	ecasa \
 	eibox \
@@ -59,13 +58,11 @@ SUBDIRS = \
 	ofdb \
 	orfat \
 	orfteletext \
-	partnerbox \
 	permanentclock \
 	pipservicerelation \
 	podcast \
 	porncenter \
 	pushservice \
-	quickbutton \
 	reconstructapsc \
 	reftopiconname \
 	remotetimer \
@@ -127,8 +124,6 @@ if TPM
 SUBDIRS += \
 	mytube \
 	yttrailer \
-	webinterface \
-	menusort \
 	pluginhider \
 	pluginsort 
 endif
