diff --git a/Makefile.am b/Makefile.am
index 4f6461ce..159205a3 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -22,7 +22,6 @@ SUBDIRS = \
 	dreammediathek \
 	dvdbackup \
 	easyinfo \
-	easymedia \
 	ecasa \
 	eibox \
 	emailclient \
@@ -59,13 +58,11 @@ SUBDIRS = \
 	ofdb \
 	orfat \
 	orfteletext \
-	partnerbox \
 	permanentclock \
 	pipservicerelation \
 	podcast \
 	porncenter \
 	pushservice \
-	quickbutton \
 	reconstructapsc \
 	reftopiconname \
 	remotetimer \
@@ -101,8 +98,7 @@ SUBDIRS = \
 	schiffe \
 	sudoku \
 	vierg \
-	bmediacenter \
-	advancedmovieselection
+	bmediacenter
 
 if DREAMBOX
 SUBDIRS +=  fancontrol2 letterbox logomanager movielistpreview genuinedreambox
@@ -129,8 +125,6 @@ if TPM
 SUBDIRS += \
 	mytube \
 	yttrailer \
-	webinterface \
-	menusort \
 	pluginhider \
 	pluginsort 
 endif
