diff --git a/Makefile.am b/Makefile.am
index d0f54e5..e535329 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -2,29 +2,18 @@ ACLOCAL_AMFLAGS = -I m4
 
 SUBDIRS = \
- 	advhdmi \
-	aihdcontroler \
 	antiscrollbar \
-	aspectratioswitch \
 	audiorestart \
 	audiosync \
-	automaticcleanup \
 	automaticvolumeadjustment \
 	autoresolution \
 	autotimer \
-	babelzapper \
 	birthdayreminder \
-	bonjour \
-	cdinfo \
 	channelselectionshorttitle \
 	curlytx \
 	dreamexplorer \
-	dreamirc \
 	dreammediathek \
 	dvdbackup \
-	easyinfo \
-	easymedia \
-	ecasa \
-	eibox \
-	elektro \
 	emailclient \
 	emission \
 	eparted \
@@ -38,74 +27,52 @@ SUBDIRS = \
 	fstabeditor \
 	ftpbrowser \
 	googlemaps \
-	growlee \
 	imdb \
 	infobartunerstate \
-	internetradio \
 	kiddytimer \
-	letterbox \
 	mediadownloader \
-	merlinepg \
-	merlinepgcenter \
 	merlinmusicplayer \
-	mosaic \
 	moviecut \
-	movieretitle \
-	moviesearch \
-	movietagger \
 	mphelp \
-	multirc \
 	ncidclient \
-	netcaster \
 	ofdb \
 	orfat \
 	orfteletext \
-	partnerbox \
 	permanentclock \
 	pipservicerelation \
 	podcast \
 	porncenter \
 	pushservice \
-	quickbutton \
 	reconstructapsc \
 	reftopiconname \
 	remotetimer \
 	rsdownloader \
-	setpasswd \
 	shoutcast \
 	showclock \
 	simplerss \
 	subsdownloader2 \
 	svdrp \
 	tageditor \
-	teletext \
 	toolkit \
 	tvcharts \
 	unwetterinfo \
-	valixdcontrol \
-	videocolorspace \
-	virtualzap \
 	vlcplayer \
 	vps \
 	weatherplugin \
 	weathercomponent \
 	weathercomponenthandler \
 	webadmin \
-	webbouqueteditor \
 	webcamviewer \
-	werbezapper \
-	zaphistorybrowser \
 	zapstatistic \
 	zdfmediathek \
- 	oscaminfo \
 	esame \
 	schiffe \
 	sudoku \
-	vierg \
-	bmediacenter \
-	advancedmovieselection
+	vierg
 
 if DREAMBOX
-SUBDIRS +=  fancontrol2 letterbox logomanager movielistpreview genuinedreambox
+SUBDIRS +=  fancontrol2 logomanager genuinedreambox
 endif
 
 if VUSOLO2
