From 5958bb9d570d0e79882c3de39c172c714b89d59f Mon Sep 17 00:00:00 2001
From: captain <openatv@gmail.com>
Date: Wed, 9 Mar 2016 18:35:47 +0100

---
 src/udev/udev-builtin-input_id.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/udev/udev-builtin-input_id.c b/src/udev/udev-builtin-input_id.c
index ca545be..fc49f5a 100644
--- a/src/udev/udev-builtin-input_id.c
+++ b/src/udev/udev-builtin-input_id.c
@@ -34,6 +34,14 @@
 #include "util.h"
 #include "missing.h"
 
+#ifndef BTN_TRIGGER_HAPPY
+#define BTN_TRIGGER_HAPPY               0x2c0
+#endif
+
+#ifndef INPUT_PROP_MAX
+#define INPUT_PROP_MAX                  0x1f
+#endif
+
 /* we must use this kernel-compatible implementation */
 #define BITS_PER_LONG (sizeof(unsigned long) * 8)
 #define NBITS(x) ((((x)-1)/BITS_PER_LONG)+1)
