From f1f27acf3c73e8c50f8b7a2164c800089988efbd Mon Sep 17 00:00:00 2001
From: Andreas Oberritter <obi@opendreambox.org>
Date: Tue, 5 Aug 2014 15:18:24 +0200
Subject: [PATCH 3/3] replace hardcode, and add EnvironmentFile and Wants

Upstream-Status: Inappropriate [distribution]

Signed-off-by: Roy Li <rongqing.li@windriver.com>
---
 src/connman.service.in | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/connman.service.in b/src/connman.service.in
index ed62add..093e485 100644
--- a/src/connman.service.in
+++ b/src/connman.service.in
@@ -4,12 +4,14 @@ Requires=dbus.socket
 After=dbus.socket
 Before=remote-fs-pre.target
 Wants=remote-fs-pre.target
+Wants=connman-env.service
 
 [Service]
 Type=dbus
 BusName=net.connman
 Restart=on-failure
-ExecStart=@prefix@/sbin/connmand -n
+EnvironmentFile=-@localstatedir@/run/connmand/connmand.env
+ExecStart=@sbindir@/connmand -n $CONNMAND_OPTS
 StandardOutput=null
 
 [Install]
-- 
1.9.1

