From 42f55b5e641a6ac0c44be34399170201bdb4af93 Mon Sep 17 00:00:00 2001
From: Andreas Oberritter <obi@opendreambox.org>
Date: Tue, 5 Aug 2014 15:18:24 +0200
Subject: [PATCH 9/9] replace hardcode, and add EnvironmentFile and Wants

Upstream-Status: Inappropriate [distribution]

Signed-off-by: Roy Li <rongqing.li@windriver.com>
---
 src/connman.service.in | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/connman.service.in b/src/connman.service.in
index 7b6195e..01bdcb4 100644
--- a/src/connman.service.in
+++ b/src/connman.service.in
@@ -3,12 +3,14 @@ Description=Connection service
 Requires=dbus.socket
 After=dbus.socket
 Before=remote-fs.target
+Wants=connman-env.service
 
 [Service]
 Type=dbus
 BusName=net.connman
 Restart=on-failure
-ExecStart=@prefix@/sbin/connmand -n
+EnvironmentFile=-@localstatedir@/run/connmand/connmand.env
+ExecStart=@sbindir@/connmand -n $CONNMAND_OPTS
 StandardOutput=null
 
 [Install]
-- 
1.9.1

