--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -26,6 +26,7 @@
 #include <string.h>
 
 #include <libavformat/avformat.h>
+#include <libavformat/demux.h>
 #include <libavutil/imgutils.h>
 /* #include <ffmpeg/avi.h> */
 #include <gst/gst.h>
@@ -1380,6 +1381,7 @@ gst_ffmpegdemux_type_find (GstTypeFind * tf, gpointer priv)
 {
   const guint8 *data;
   AVInputFormat *in_plugin = (AVInputFormat *) priv;
+  FFInputFormat *in_plugin2 = in_plugin;
   gint res = 0;
   guint64 length;
   GstCaps *sinkcaps;
@@ -1400,7 +1402,7 @@ gst_ffmpegdemux_type_find (GstTypeFind * tf, gpointer priv)
   }
 
   GST_LOG ("typefinding %" G_GUINT64_FORMAT " bytes", length);
-  if (in_plugin->read_probe &&
+  if (in_plugin2->read_probe &&
       (data = gst_type_find_peek (tf, 0, length)) != NULL) {
     AVProbeData probe_data;
 
@@ -1408,7 +1410,7 @@ gst_ffmpegdemux_type_find (GstTypeFind * tf, gpointer priv)
     probe_data.buf = (guint8 *) data;
     probe_data.buf_size = length;
 
-    res = in_plugin->read_probe (&probe_data);
+    res = in_plugin2->read_probe (&probe_data);
     if (res > 0) {
       res = MAX (1, res * GST_TYPE_FIND_MAXIMUM / AVPROBE_SCORE_MAX);
       /* Restrict the probability for MPEG-TS streams, because there is
