--- a/ofgwrite.c
+++ b/ofgwrite.c
@@ -14,6 +14,28 @@
 #include <unistd.h>
 #include <errno.h>
 
+struct stat kernel_file_stat;
+struct stat rootfs_file_stat;
+
+char kernel_device_arg[1000];
+char rootfs_device_arg[1000];
+char kernel_device[1000];
+char rootfs_device[1000];
+char rootfs_sub_dir[1000];
+
+int found_kernel_device;
+int found_rootfs_device;
+int user_kernel;
+int user_rootfs;
+int rootsubdir_check;
+int multiboot_partition;
+char current_rootfs_device[1000];
+char current_kernel_device[1000];
+char current_rootfs_sub_dir[1000];
+
+enum FlashModeTypeEnum kernel_flash_mode;
+enum FlashModeTypeEnum rootfs_flash_mode;
+
 const char ofgwrite_version[] = "4.5.6";
 int flash_kernel  = 0;
 int flash_rootfs  = 0;
--- a/ofgwrite.h
+++ b/ofgwrite.h
@@ -1,23 +1,23 @@
 #include <sys/stat.h>
 
-struct stat kernel_file_stat;
-struct stat rootfs_file_stat;
+extern struct stat kernel_file_stat;
+extern struct stat rootfs_file_stat;
 
-char kernel_device_arg[1000];
-char rootfs_device_arg[1000];
-char kernel_device[1000];
-char rootfs_device[1000];
-char rootfs_sub_dir[1000];
+extern char kernel_device_arg[1000];
+extern char rootfs_device_arg[1000];
+extern char kernel_device[1000];
+extern char rootfs_device[1000];
+extern char rootfs_sub_dir[1000];
 
-int found_kernel_device;
-int found_rootfs_device;
-int user_kernel;
-int user_rootfs;
-int rootsubdir_check;
-int multiboot_partition;
-char current_rootfs_device[1000];
-char current_kernel_device[1000];
-char current_rootfs_sub_dir[1000];
+extern int found_kernel_device;
+extern int found_rootfs_device;
+extern int user_kernel;
+extern int user_rootfs;
+extern int rootsubdir_check;
+extern int multiboot_partition;
+extern char current_rootfs_device[1000];
+extern char current_kernel_device[1000];
+extern char current_rootfs_sub_dir[1000];
 
 void handle_busybox_fatal_error();
 
@@ -31,5 +31,5 @@ enum FlashModeTypeEnum
 	FLASH_MODE_UNKNOWN, MTD, TARBZ2
 };
 
-enum FlashModeTypeEnum kernel_flash_mode;
-enum FlashModeTypeEnum rootfs_flash_mode;
+extern enum FlashModeTypeEnum kernel_flash_mode;
+extern enum FlashModeTypeEnum rootfs_flash_mode;
