diff -urN a/arch/mips/brcmstb/board.c b/arch/mips/brcmstb/board.c
--- a/arch/mips/brcmstb/board.c	2014-07-23 17:45:41.768600409 +0200
+++ b/arch/mips/brcmstb/board.c	2014-07-22 23:27:27.709355017 +0200
@@ -40,6 +40,10 @@
 unsigned long brcm_mtd_rootfs_len;
 unsigned long brcm_mtd_kernel_start;
 unsigned long brcm_mtd_kernel_len;
+unsigned long brcm_mtd_macaddr_start;
+unsigned long brcm_mtd_macaddr_len;
+unsigned long brcm_mtd_macadr_start;
+unsigned long brcm_mtd_macadr_len;
 unsigned long brcm_mtd_ocap_start;
 unsigned long brcm_mtd_ocap_len;
 unsigned long brcm_mtd_flash_size_mb;
@@ -360,7 +364,7 @@
 
 #elif defined(CONFIG_BCM7358) || defined(CONFIG_BCM7552) || \
 	defined(CONFIG_BCM7360)
-
+#if 0
 	PINMUX(11, gpio_89, 1);		/* UARTB TX */
 	PINMUX(11, gpio_90, 1);		/* UARTB RX */
 	PINMUX(11, gpio_91, 1);		/* UARTC TX */
@@ -388,7 +392,7 @@
 	AON_PADCTRL(1, aon_gpio_18_pad_ctrl, 2);
 	AON_PADCTRL(1, aon_gpio_19_pad_ctrl, 2);
 	AON_PADCTRL(1, aon_gpio_20_pad_ctrl, 2);
-
+#endif
 #if defined(CONFIG_BCMGENET_0_GPHY)
 	/* set RGMII lines to 2.5V */
 	BDEV_WR_F(SUN_TOP_CTRL_GENERAL_CTRL_NO_SCAN_0, rgmii_0_pad_mode, 1);
@@ -802,6 +806,10 @@
 	nr_parts = 2;
 	if (brcm_mtd_kernel_len != 0)
 		nr_parts++;
+	if (brcm_mtd_macaddr_len != 0)
+		nr_parts++;
+	if (brcm_mtd_macadr_len != 0)
+		nr_parts++;
 
 	ret = kzalloc(nr_parts * sizeof(struct mtd_partition), GFP_KERNEL);
 	if (!ret)
@@ -811,21 +819,25 @@
 	ret[0].size = brcm_mtd_rootfs_len;
 	ret[0].name = "rootfs";
 
-#if 0
-	ret[1].offset = 0;
-	ret[1].size = MTDPART_SIZ_FULL;
-	ret[1].name = "entire_device";
-#else
         ret[1].offset = brcm_mtd_rootfs_start;
         ret[1].size = brcm_mtd_rootfs_len;
         ret[1].name = "rootfs(redundant)";
-#endif
 
 	if (brcm_mtd_kernel_len != 0) {
 		ret[2].offset = brcm_mtd_kernel_start;
 		ret[2].size = brcm_mtd_kernel_len;
 		ret[2].name = "kernel";
 	}
+	if (brcm_mtd_macaddr_len != 0) {
+		ret[3].offset = brcm_mtd_macaddr_start;
+		ret[3].size = brcm_mtd_macaddr_len;
+		ret[3].name = "mac";
+	}
+	if (brcm_mtd_macadr_len != 0) {
+		ret[3].offset = brcm_mtd_macadr_start;
+		ret[3].size = brcm_mtd_macadr_len;
+		ret[3].name = "mac";
+	}
 
 	*p = ret;
 	return nr_parts;
diff -urN a/arch/mips/brcmstb/prom.c b/arch/mips/brcmstb/prom.c
--- a/arch/mips/brcmstb/prom.c	2014-07-22 23:20:41.061728948 +0200
+++ b/arch/mips/brcmstb/prom.c	2014-07-22 23:24:28.319303124 +0200
@@ -275,6 +275,10 @@
 	FETCH("LINUX_FFS_SIZE", parse_hex, &brcm_mtd_rootfs_len);
 	FETCH("LINUX_PART_STARTAD", parse_hex, &brcm_mtd_kernel_start);
 	FETCH("LINUX_PART_SIZE", parse_hex, &brcm_mtd_kernel_len);
+	FETCH("LINUX_MACADDR_STARTAD", parse_hex, &brcm_mtd_macaddr_start);
+	FETCH("LINUX_MACADDR_SIZE", parse_hex, &brcm_mtd_macaddr_len);
+	FETCH("MACADR_PART_STARTAD", parse_hex, &brcm_mtd_macadr_start);
+	FETCH("MACADR_PART_SIZE", parse_hex, &brcm_mtd_macadr_len);
 	FETCH("OCAP_PART_STARTAD", parse_hex, &brcm_mtd_ocap_start);
 	FETCH("OCAP_PART_SIZE", parse_hex, &brcm_mtd_ocap_len);
 	FETCH("FLASH_SIZE", parse_ulong, &brcm_mtd_flash_size_mb);
diff -urN a/include/linux/brcmstb/brcmstb.h b/include/linux/brcmstb/brcmstb.h
--- a/include/linux/brcmstb/brcmstb.h	2014-07-22 23:19:12.602672381 +0200
+++ b/include/linux/brcmstb/brcmstb.h	2014-07-22 23:28:28.448705294 +0200
@@ -755,6 +755,10 @@
 extern unsigned long brcm_mtd_rootfs_len;
 extern unsigned long brcm_mtd_kernel_start;
 extern unsigned long brcm_mtd_kernel_len;
+extern unsigned long brcm_mtd_macaddr_start;
+extern unsigned long brcm_mtd_macaddr_len;
+extern unsigned long brcm_mtd_macadr_start;
+extern unsigned long brcm_mtd_macadr_len;
 extern unsigned long brcm_mtd_ocap_start;
 extern unsigned long brcm_mtd_ocap_len;
 extern unsigned long brcm_mtd_flash_size_mb;
