--- a/configure.ac
+++ b/configure.ac
@@ -237,7 +237,7 @@ fi
 AC_ARG_ENABLE(obex, AC_HELP_STRING([--disable-obex],
 		[disable OBEX profile support]), [enable_obex=${enableval}])
 if (test "${enable_obex}" != "no"); then
-	PKG_CHECK_MODULES(ICAL, libical, dummy=yes,
+	PKG_CHECK_MODULES(ICAL, [libicalvcal >= 4.0.0], [],
 					AC_MSG_ERROR(libical is required))
 	AC_SUBST(ICAL_CFLAGS)
 	AC_SUBST(ICAL_LIBS)
