From 1d422fd73de04c3f041786bd9fecaee4c28a5817 Mon Sep 17 00:00:00 2001
From: jbleyel <jbleyel@gmx.net>
Date: Wed, 24 Aug 2022 04:27:46 +0000

---
 Makefile.am | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index ab10c1d00..fc69cfab0 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -21,7 +21,6 @@ SUBDIRS = \
 	dreamirc \
 	dreammediathek \
 	dvdbackup \
-	easyinfo \
 	easymedia \
 	ecasa \
 	eibox \
@@ -86,7 +85,6 @@ SUBDIRS = \
 	valixdcontrol \
 	videocolorspace \
 	vlcplayer \
-	vps \
 	weatherplugin \
 	weathercomponent \
 	weathercomponenthandler \
@@ -130,7 +128,6 @@ SUBDIRS += \
 	mytube \
 	yttrailer \
 	webinterface \
-	menusort \
 	pluginhider \
 	pluginsort 
 endif
