diff --git a/src/core/layer_context.c b/src/core/layer_context.c
index 091ed79..a4786fd 100644
--- a/src/core/layer_context.c
+++ b/src/core/layer_context.c
@@ -1880,8 +1880,10 @@ dfb_layer_context_allocate_surface( CoreLayer             *layer,
      else
           surface->rotation = (context->rotation == 180) ? 180 : 0;
 
+#if 0 // vuplus
      if (dfb_config->layers_clear && !dfb_config->surface_clear)
           dfb_surface_clear_buffers( surface );
+#endif
 
      /* Tell the region about its new surface (adds a global reference). */
      ret = dfb_layer_region_set_surface( region, surface );
@@ -2010,8 +2012,10 @@ dfb_layer_context_reallocate_surface( CoreLayer             *layer,
      else
           surface->rotation = (context->rotation == 180) ? 180 : 0;
 
+#if 0 // vuplus
      if (dfb_config->layers_clear && !dfb_config->surface_clear)
           dfb_surface_clear_buffers( surface );
+#endif
 
      dfb_surface_unlock( surface );
      
