--- a/data/tuxtxt2.conf
+++ b/data/tuxtxt2.conf
@@ -7,6 +7,9 @@ NationalSubset 5
 SwapUpDown 0
 ShowHexPages 0
 Transparency 0xA
+TTFScreenResX 720
+TTFBold 1
+CleanAlgo 0
 TTFWidthFactor16 28
 TTFHeightFactor16 16
 TTFShiftX 0
--- a/tuxtxt.c
+++ b/tuxtxt.c
@@ -519,6 +519,12 @@ int Init()
 				renderinfo.showhex = ival & 1;
 			else if (1 == sscanf(line, "Transparency %i", &ival))
 				renderinfo.trans_mode = ival;
+			else if (1 == sscanf(line, "TTFScreenResX %i", &ival))
+				renderinfo.TTFScreenResX = ival;
+			else if (1 == sscanf(line, "TTFBold %i", &ival))
+				renderinfo.TTFBold = ival;
+			else if (1 == sscanf(line, "CleanAlgo %i", &ival))
+				renderinfo.CleanAlgo = ival;
 			else if (1 == sscanf(line, "TTFWidthFactor16 %i", &ival))
 				renderinfo.TTFWidthFactor16 = ival;
 			else if (1 == sscanf(line, "TTFHeightFactor16 %i", &ival))
@@ -686,6 +692,9 @@ void CleanUp()
 			fprintf(conf, "SwapUpDown %d\n", swapupdown);
 			fprintf(conf, "ShowHexPages %d\n", renderinfo.showhex);
 			fprintf(conf, "Transparency 0x%X\n", renderinfo.trans_mode);
+			fprintf(conf, "TTFScreenResX %d\n", renderinfo.TTFScreenResX);
+			fprintf(conf, "TTFBold %d\n", renderinfo.TTFBold);
+			fprintf(conf, "CleanAlgo %d\n", renderinfo.CleanAlgo);
 			fprintf(conf, "TTFWidthFactor16 %d\n", renderinfo.TTFWidthFactor16);
 			fprintf(conf, "TTFHeightFactor16 %d\n", renderinfo.TTFHeightFactor16);
 			fprintf(conf, "TTFShiftX %d\n", renderinfo.TTFShiftX);
