diff -Naurx data a/plugin.py b/plugin.py
--- a/plugin.py	2015-05-15 17:20:17.000000000 +0200
+++ b/plugin.py	2015-05-16 13:47:00.000000000 +0200
@@ -200,13 +200,15 @@
         self.pServerThread.join()
         
         dvbapp_unlock()
+        
+        os.system ('cat /etc/videomode > /proc/stb/video/videomode')
 
         if self.beforeServiceRef is not None:
                 self.session.nav.playService(self.beforeServiceRef)
         self.close()
 
     def _cbCheckXBMCTimer(self):
-        data = os.popen('ps -ef | grep xbmc.bin | grep -v grep').read().strip()
+        data = os.popen('ps | grep xbmc.bin | grep -v grep').read().strip()
 	if data is None or data == "":
 	    self.checkXBMCTimer.stop()
             os.system('xbmc.helper --stop')
@@ -215,8 +217,9 @@
     PServerThread.close()
 
 def plugin_start_xbmc(session, **kwargs):
-    dvbapp_unlock()
-    session.open(XBMCLauncher)
+	os.system ('echo "720p50" > /proc/stb/video/videomode')
+	dvbapp_unlock()
+	session.open(XBMCLauncher)
 
 def Plugins(**kwargs):
     l = []
