From 333cd697c86dfaab175e8543729d3682dd287267 Mon Sep 17 00:00:00 2001
From: captain <openatv@gmail.com>
Date: Tue, 5 Apr 2016 18:33:46 +0200
Subject: [PATCH] [sh4] fix def for non sh4 boxes

---
 gstdvbaudiosink.c | 2 +-
 gstdvbvideosink.c | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/gstdvbaudiosink.c b/gstdvbaudiosink.c
index 3b7c5de..c8d5e7d 100644
--- a/gstdvbaudiosink.c
+++ b/gstdvbaudiosink.c
@@ -64,7 +64,7 @@
 #include <config.h>
 #endif
 
-#if defined(__sh__) || defined(SPARK)
+#ifdef __sh__
 #include <linux/dvb/stm_ioctls.h>
 #endif
 
diff --git a/gstdvbvideosink.c b/gstdvbvideosink.c
index 85330c2..72ed202 100644
--- a/gstdvbvideosink.c
+++ b/gstdvbvideosink.c
@@ -64,7 +64,7 @@
 #include <config.h>
 #endif
 
-#if defined(__sh__) || defined(SPARK)
+#ifdef __sh__
 #include <linux/dvb/stm_ioctls.h>
 #endif
 
-- 
2.6.3.windows.1

