diff --git a/drivers/serial/8250.c b/drivers/serial/8250.c
index d75f8c7..49e0930 100644
--- a/drivers/serial/8250.c
+++ b/drivers/serial/8250.c
@@ -1696,6 +1696,7 @@ static int serial_link_irq_chain(struct uart_8250_port *up)
 		i->head = &up->list;
 		spin_unlock_irq(&i->lock);
 		irq_flags |= up->port.irqflags;
+		printk("request_irq %d %d %p\n", irq_flags, up->port.irq, i);
 		ret = request_irq(up->port.irq, serial8250_interrupt,
 				  irq_flags, "serial", i);
 		if (ret < 0)
