From 27abdd24487df72a7392a0bed44a35735d116d3e Mon Sep 17 00:00:00 2001
From: Koen Kooi <koen@dominion.thruhere.net>
Date: Sat, 6 Feb 2016 15:43:01 +0100
Subject: [PATCH 3/3] configure: don't try to run stuff to find tinyxml

---
 configure.ac | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 4237a34..b439ccd 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1133,7 +1133,7 @@ AC_CHECK_LIB([atomic], [__atomic_add_fetch_8], [LIBS="$LIBS -latomic"],)
 AC_LANG_PUSH([C++])
 PKG_CHECK_MODULES([TINYXML], [tinyxml >= 2.6.2],
   [INCLUDES="$INCLUDES $TINYXML_CFLAGS"; LIBS="$LIBS $TINYXML_LIBS"],
-  [AC_RUN_IFELSE(
+  [AC_COMPILE_IFELSE(
    [AC_LANG_SOURCE([[
 #include <stdlib.h>
 #include <tinyxml.h>
-- 
1.9.1

