--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -556,15 +556,16 @@ if(WITH_EXIV2)
 endif()
 
 if(WITH_MATROSKA)
+    find_package(Matroska REQUIRED)
+
+    target_include_directories(libgerbera PUBLIC ${Matroska_INCLUDE_DIRS})
+    target_link_libraries(libgerbera PUBLIC ${Matroska_LIBRARIES})
+
     find_package(EBML REQUIRED)
 
     target_include_directories(libgerbera PUBLIC ${EBML_INCLUDE_DIRS})
     target_link_libraries(libgerbera PUBLIC ${EBML_LIBRARIES})
 
-    find_package(Matroska REQUIRED)
-    target_include_directories(libgerbera PUBLIC ${Matroska_INCLUDE_DIRS})
-    target_link_libraries(libgerbera PUBLIC ${Matroska_LIBRARIES})
-
     target_compile_definitions(libgerbera PUBLIC HAVE_MATROSKA)
 endif()
 
