From d436a7a0b9c98028662c36fc605d8ba0f8c448cf Mon Sep 17 00:00:00 2001
From: jbleyel <jbleyel@gmx.net>
Date: Mon, 13 Sep 2021 14:47:49 +0200

---
 Makefile.am | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 57460341..0316bb1c 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -21,7 +21,6 @@ SUBDIRS = \
 	dreamirc \
 	dreammediathek \
 	dvdbackup \
-	easyinfo \
 	easymedia \
 	ecasa \
 	eibox \
@@ -59,13 +58,11 @@ SUBDIRS = \
 	ofdb \
 	orfat \
 	orfteletext \
-	partnerbox \
 	permanentclock \
 	pipservicerelation \
 	podcast \
 	porncenter \
 	pushservice \
-	quickbutton \
 	reconstructapsc \
 	reftopiconname \
 	remotetimer \
@@ -128,7 +125,6 @@ SUBDIRS += \
 	mytube \
 	yttrailer \
 	webinterface \
-	menusort \
 	pluginhider \
 	pluginsort 
 endif
