--- nmap-3.77/nmapfe/configure.ac.orig	2004-12-04 14:51:07.000000000 +0100
+++ nmap-3.77/nmapfe/configure.ac	2004-12-04 14:51:21.000000000 +0100
@@ -9,8 +9,8 @@
 #AM_PATH_GTK(1.0.0,
 #            [LIBS="$LIBS $GTK_LIBS" CFLAGS="$CFLAGS $GTK_CFLAGS"],
 #            AC_MSG_ERROR(Cannot find GTK: Is gtk-config in path?))
-AM_PATH_GTK(1.0.0,
-            [LIBS="$LIBS $GTK_LIBS" CFLAGS="$CFLAGS $GTK_CFLAGS"],
+#AM_PATH_GTK(1.0.0,
+#            [LIBS="$LIBS $GTK_LIBS" CFLAGS="$CFLAGS $GTK_CFLAGS"],
             AC_DEFINE(MISSING_GTK) \
 	    AC_MSG_WARN(NMAPFE WILL NOT BE BUILT -- BUT NMAP SHOULD STILL WORK ))
 
