--- a/setup.py
+++ b/setup.py
@@ -275,7 +275,7 @@ class ExtensionConfiguration(object):
         optbuf = ''
         sslhintbuf = ''
         errtext = ''
-        for option in ["--libs", "--static-libs"]:
+        for option in ["--libs"]:
             p = subprocess.Popen((self.curl_config(), option),
                 stdout=subprocess.PIPE, stderr=subprocess.PIPE)
             stdout, stderr = p.communicate()
