diff --git a/configure.ac b/configure.ac
index 3b046de..9a4786b 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1,11 +1,7 @@
 #                                               -*- Autoconf -*-
 # Process this file with autoconf to produce a configure script.
 
-# Get version from git
-m4_define([pkg_version], m4_esyscmd([sh build-aux/git-version-gen]))
-m4_ifset([pkg_version], [], [ m4_define([pkg_version], [UNKNOWN]) ])
-m4_if(pkg_version, [UNKNOWN],
-    [ m4_fatal([unable to obtain package version; please check `git-version-gen' output]) ])
+m4_define([pkg_version], [0.2]) ]
 
 #
 # Autotools setup
