diff --git a/src/main.cpp b/src/main.cpp
index 9996ac2..2853aeb 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -114,7 +114,7 @@ inline int streaming_write(const char *buffer, size_t buffer_len, bool enable_lo
 
 #define DD_LOG(X,...) { \
 		char log_message[128] = {0};\
-		sprintf(log_message, "echo \""X"\" > /tmp/tsp_checker.log", ##__VA_ARGS__);\
+		sprintf(log_message, "echo \"" X "\" > /tmp/tsp_checker.log", ##__VA_ARGS__);\
 		system(log_message);\
 	}
 //----------------------------------------------------------------------
@@ -272,7 +272,7 @@ int main(int argc, char **argv)
 					exit(0);
 				}
 
-				sprintf(update_status_command, "touch "TSP_CHECKER_TEMPLETE, tsp_pid);
+				sprintf(update_status_command, "touch " TSP_CHECKER_TEMPLETE, tsp_pid);
 				system(update_status_command);
 
 				encoder = new Encoder();
