diff --git a/src/ePreDefine.h b/src/ePreDefine.h
index 390431d..83aaa3d 100644
--- a/src/ePreDefine.h
+++ b/src/ePreDefine.h
@@ -11,10 +11,10 @@
 #define BUFFER_SIZE		(188*256)
 #define MAX_LINE_LENGTH	(1024)
 
-#define RETURN_ERR_400(FMT,...) { printf("HTTP/1.0 400 Bad Request\r\n"FMT"\r\n\r\n", ##__VA_ARGS__); return 1; }
-#define RETURN_ERR_401(FMT,...) { printf("HTTP/1.0 401 Unauthorized\r\n"FMT"\r\n\r\n",##__VA_ARGS__); return 1; }
-#define RETURN_ERR_502(FMT,...) { printf("HTTP/1.0 502 Bad Gateway\r\n"FMT"\r\n\r\n", ##__VA_ARGS__); return 1; }
-#define RETURN_ERR_503(FMT,...) { printf("HTTP/1.0 503 Service Unavailable\r\n"FMT"\r\n\r\n", ##__VA_ARGS__); return 1; }
+#define RETURN_ERR_400(FMT,...) { printf("HTTP/1.0 400 Bad Request\r\n" FMT "\r\n\r\n", ##__VA_ARGS__); return 1; }
+#define RETURN_ERR_401(FMT,...) { printf("HTTP/1.0 401 Unauthorized\r\n" FMT "\r\n\r\n", ##__VA_ARGS__); return 1; }
+#define RETURN_ERR_502(FMT,...) { printf("HTTP/1.0 502 Bad Gateway\r\n" FMT "\r\n\r\n", ##__VA_ARGS__); return 1; }
+#define RETURN_ERR_503(FMT,...) { printf("HTTP/1.0 503 Service Unavailable\r\n" FMT "\r\n\r\n", ##__VA_ARGS__); return 1; }
 //-------------------------------------------------------------------------------
 
 char* ReadRequest(char* aRequest);
