--- a/src/SharedMemory.h
+++ b/src/SharedMemory.h
@@ -36,7 +36,7 @@ protected:
 protected:
 	void Close()
 	{
-		if (mShmData > 0) {
+		if (mShmData != 0) {
 			munmap(mShmData, mShmSize);
 		}
 		mShmData = 0;
@@ -45,7 +45,7 @@ protected:
 			//shm_unlink(mShmName.c_str());
 		}
 		mShmFd = 0;
-		if (mSemId > 0) {
+		if (mSemId != 0) {
 			sem_close(mSemId);
 			sem_unlink(mSemName.c_str());
 		}
