diff --git a/debian/src/bootlogd/Makefile b/debian/src/bootlogd/Makefile
index 9741518..3dd292a 100644
--- a/debian/src/bootlogd/Makefile
+++ b/debian/src/bootlogd/Makefile
@@ -1,7 +1,7 @@
 DESTDIR =
 sysconfdir = /etc
 
-INSTALL      = install -m755 -o root -g root
+INSTALL      = install -m755
 
 all:
 
diff --git a/debian/src/initscripts/Makefile b/debian/src/initscripts/Makefile
index 7654e63..8632229 100644
--- a/debian/src/initscripts/Makefile
+++ b/debian/src/initscripts/Makefile
@@ -4,8 +4,8 @@ prefix     = /usr
 docdir     = $(prefix)/share/doc/initscripts
 sbindir    = /sbin
 
-INSTALL      = install -m755 -o root -g root
-INSTALL_DATA = install -m644 -o root -g root
+INSTALL      = install -m755
+INSTALL_DATA = install -m644
 
 all:
 
diff --git a/debian/src/sysv-rc/Makefile b/debian/src/sysv-rc/Makefile
index 54c5cc7..1f3abce 100644
--- a/debian/src/sysv-rc/Makefile
+++ b/debian/src/sysv-rc/Makefile
@@ -5,8 +5,8 @@ docdir     = $(prefix)/share/doc/sysv-rc
 bindir     = $(prefix)/bin
 sbindir    = $(prefix)/sbin
 
-INSTALL      = install -m755 -o root -g root
-INSTALL_DATA = install -m644 -o root -g root
+INSTALL      = install -m755
+INSTALL_DATA = install -m644
 
 all:
 
