From 570035ede66a5e5817995eb9d61c8a4a7683336e Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Mon, 6 Aug 2018 06:51:16 +0200
Subject: [PATCH] makefile disable warnings


diff --git a/Makefile b/Makefile
index 1e5a6683..358b9350 100755
--- a/Makefile
+++ b/Makefile
@@ -605,6 +605,11 @@ endif
 # Use make W=1 to enable this warning (see scripts/Makefile.build)
 KBUILD_CFLAGS += $(call cc-disable-warning, unused-but-set-variable)
 
+KBUILD_CFLAGS += $(call cc-disable-warning, attribute-alias)
+KBUILD_CFLAGS += $(call cc-disable-warning, packed-not-aligned)
+KBUILD_CFLAGS += $(call cc-disable-warning, stringop-truncation)
+KBUILD_CFLAGS += $(call cc-disable-warning, format-truncation)
+
 ifdef CONFIG_FRAME_POINTER
 KBUILD_CFLAGS	+= -fno-omit-frame-pointer -fno-optimize-sibling-calls
 else
-- 
2.17.1

