--- a/arch/arm/include/asm/ftrace.h	2016-03-05 03:35:28.000000000 +0100
+++ b/arch/arm/include/asm/ftrace.h	2016-03-27 20:23:07.763354712 +0200
@@ -45,7 +45,7 @@
 
 #else
 
-extern inline void *return_address(unsigned int level)
+static inline void *return_address(unsigned int level)
 {
 	return NULL;
 }
--- a/arch/arm/kernel/return_address.c	2016-03-05 03:35:28.000000000 +0100
+++ b/arch/arm/kernel/return_address.c	2016-03-27 20:26:02.087351397 +0200
@@ -63,11 +63,6 @@
 #warning "TODO: return_address should use unwind tables"
 #endif
 
-void *return_address(unsigned int level)
-{
-	return NULL;
-}
-
 #endif /* if defined(CONFIG_FRAME_POINTER) && !defined(CONFIG_ARM_UNWIND) / else */
 
 EXPORT_SYMBOL_GPL(return_address);
