diff --git a/os_dep/linux/os_intfs.c b/os_dep/linux/os_intfs.c
index a78e317a8919..900edf77c73c 100644
--- a/os_dep/linux/os_intfs.c
+++ b/os_dep/linux/os_intfs.c
@@ -885,9 +885,11 @@ static u16 rtw_select_queue(struct net_device *dev, struct sk_buff *skb
 #else
 				, void *accel_priv
 #endif
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5,2,0))
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 14, 0) 
 				, select_queue_fallback_t fallback
 #endif
+#endif
 
 #endif
 )
