From 5de3181286a40c63c106acf8490c34ecd60656c5 Mon Sep 17 00:00:00 2001
From: Captain <captain.onboard@web.de>
Date: Mon, 14 May 2018 22:27:50 +0200
Subject: [PATCH] multilib

---
 src/libcec/cmake/CheckPlatformSupport.cmake | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/libcec/cmake/CheckPlatformSupport.cmake b/src/libcec/cmake/CheckPlatformSupport.cmake
index 73612de..062ad10 100644
--- a/src/libcec/cmake/CheckPlatformSupport.cmake
+++ b/src/libcec/cmake/CheckPlatformSupport.cmake
@@ -218,9 +218,9 @@ else()
       endif()
 
       install(TARGETS     ${SWIG_MODULE_cec_REAL_NAME}
-              DESTINATION lib/python${PYTHON_VERSION}/${PYTHON_PKG_DIR}/${PYTHON_LIB_INSTALL_PATH})
+              DESTINATION ${BASE_LIB}/python${PYTHON_VERSION}/${PYTHON_PKG_DIR}/${PYTHON_LIB_INSTALL_PATH})
       install(FILES       ${CMAKE_BINARY_DIR}/src/libcec/cec.py
-              DESTINATION lib/python${PYTHON_VERSION}/${PYTHON_PKG_DIR}/cec
+              DESTINATION ${BASE_LIB}/python${PYTHON_VERSION}/${PYTHON_PKG_DIR}/cec
               RENAME      __init__.py)
     endif()
   endif()
-- 
2.17.0.windows.1

