diff -Nur Twisted-16.4.0/twisted/web/http.py Twisted-16.4.0-fixed/twisted/web/http.py
--- Twisted-16.4.0/twisted/web/http.py	2016-08-26 20:13:26.000000000 +0200
+++ Twisted-16.4.0-fixed/twisted/web/http.py	2016-11-04 13:35:48.000000000 +0100
@@ -1838,7 +1838,7 @@
             self._savedTimeOut = self.setTimeout(None)
 
         # Pause the producer if we can. If we can't, that's ok, we'll buffer.
-        self._producer.pauseProducing()
+        #self._producer.pauseProducing()
         self._handlingRequest = True
 
         req = self.requests[-1]
