diff -Naur linux-2.6.34/drivers/serial/ip3106.c linux-2.6.34.mod/drivers/serial/ip3106.c
--- linux-2.6.34/drivers/serial/ip3106.c	2010-10-13 15:45:12.000000000 -0500
+++ linux-2.6.34.mod/drivers/serial/ip3106.c	2010-10-13 15:30:50.000000000 -0500
@@ -1118,6 +1118,41 @@
 	return ret;
 }
 
+#ifdef CONFIG_CONSOLE_POLL
+static int ip3106_get_poll_char(struct uart_port *port)
+{
+    unsigned int  ch;
+    unsigned int  status;
+
+	/* Read till the FIFO is not empty  */
+	do {
+	   /* Read the status of the Rx FIFO */
+	   status = UART_GET_LSR(port);
+	} while ((status & IP3106_UART_LSR_DR_MSK) == 0); 
+    
+	/* Read the character from Rx FIFO */
+	ch = UART_GET_RBR(port);
+	return (ch & 0x7f);
+}
+
+static void ip3106_put_poll_char(struct uart_port *port,
+			 unsigned char ch)
+{
+	unsigned int status = 0;
+
+	/* Wait till the Tx FIFO is empty */
+	do {
+		status = UART_GET_LSR(port);
+	} while (!(status & IP3106_UART_LSR_THRE_MSK));
+
+	/* Write the character */
+	UART_PUT_THR(port,(ch | 0x80));
+	return;
+
+}
+
+#endif /* CONFIG_CONSOLE_POLL */
+
 /*--------------------------------------------------------------------------*
  * Port Control Functions for UART IP3106                                   *
  *--------------------------------------------------------------------------*/
@@ -1139,6 +1174,10 @@
 	.request_port = ip3106_request_port,
 	.config_port  = ip3106_config_port,
 	.verify_port  = ip3106_verify_port,
+#ifdef CONFIG_CONSOLE_POLL
+	.poll_get_char = ip3106_get_poll_char,
+	.poll_put_char = ip3106_put_poll_char,
+#endif
 };
 
 #if defined(CONFIG_SERIAL_IP3106_CONSOLE)
diff -Naur linux-2.6.34/drivers/serial/Makefile linux-2.6.34.mod/drivers/serial/Makefile
--- linux-2.6.34/drivers/serial/Makefile	2010-10-13 15:45:12.000000000 -0500
+++ linux-2.6.34.mod/drivers/serial/Makefile	2010-10-13 14:56:45.000000000 -0500
@@ -79,8 +79,8 @@
 obj-$(CONFIG_SERIAL_OF_PLATFORM) += of_serial.o
 obj-$(CONFIG_SERIAL_OF_PLATFORM_NWPSERIAL) += nwpserial.o
 obj-$(CONFIG_SERIAL_KS8695) += serial_ks8695.o
-obj-$(CONFIG_KGDB_SERIAL_CONSOLE) += kgdboc.o
 obj-$(CONFIG_SERIAL_QE) += ucc_uart.o
 obj-$(CONFIG_SERIAL_TIMBERDALE)	+= timbuart.o
 obj-$(CONFIG_SERIAL_IP3106_CONSOLE) += ip3106.o
 obj-$(CONFIG_SERIAL_GRLIB_GAISLER_APBUART) += apbuart.o
+obj-$(CONFIG_KGDB_SERIAL_CONSOLE) += kgdboc.o
