--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -23,7 +23,7 @@ include(CheckFunctionExists)
 option(ENABLE_SDL "Enable SDL" ON)
 
 if (ENABLE_SDL)
-	find_package(SDL2)
+	find_package(SDL)
 endif()
 
 find_package(Threads REQUIRED)
--- a/configure.ac
+++ b/configure.ac
@@ -311,8 +311,8 @@ if eval "test x$enable_dec265 = xyes" || eval "test x$enable_sherlock265 = xyes"
 fi
 
 if eval "test x$enable_dec265 = xyes" ; then
-   PKG_CHECK_MODULES([SDL], [sdl2],
-                     [AC_DEFINE([HAVE_SDL], [1], [Whether libsdl2 was found.])
+   PKG_CHECK_MODULES([SDL], [sdl],
+                     [AC_DEFINE([HAVE_SDL], [1], [Whether libsdl was found.])
                       AC_SUBST(SDL_CFLAGS)
                       AC_SUBST(SDL_LIBS)
                       have_sdl="yes"],
@@ -335,7 +335,7 @@ AM_CONDITIONAL([HAVE_SWSCALE], [test "x$have_swscale" = "xyes"])
 AM_CONDITIONAL([HAVE_SDL], [test "x$have_sdl" = "xyes"])
 
 if eval "test $enable_dec265 = yes" && eval "test $have_videogfx != yes" && eval "test $have_sdl != yes" ; then
-  AC_MSG_WARN([Did not find libvideogfx or libsdl2, video output of dec265 will be disabled.])
+  AC_MSG_WARN([Did not find libvideogfx or libsdl, video output of dec265 will be disabled.])
 fi
 
 if eval "test $enable_sherlock265 = yes" && eval "test $have_videogfx != yes" && eval "test $have_swscale != yes" ; then
--- a/dec265/CMakeLists.txt
+++ b/dec265/CMakeLists.txt
@@ -2,15 +2,15 @@ add_executable (dec265 dec265.cc)
 
 target_link_libraries (dec265 PRIVATE de265)
 
-if(SDL2_FOUND)
+if(SDL_FOUND)
   target_sources(dec265 PRIVATE sdl.cc)
   target_compile_definitions(dec265 PRIVATE HAVE_SDL)
-  target_include_directories (dec265 PRIVATE "${SDL2_INCLUDE_DIRS}")
-  target_link_libraries (dec265 PRIVATE ${SDL2_LIBRARIES})
+  target_include_directories (dec265 PRIVATE "${SDL_INCLUDE_DIR}")
+  target_link_libraries (dec265 PRIVATE ${SDL_LIBRARY})
 endif()
 
 if(MSVC)
-  target_sources(dec265 PRIVATE
+  target_sources(dec265 PRIVATE 
     ../extra/getopt.c
     ../extra/getopt_long.c
   )
--- a/dec265/sdl.cc
+++ b/dec265/sdl.cc
@@ -37,49 +37,51 @@ bool SDL_YUV_Display::init(int frame_width, int frame_height, enum SDL_Chroma ch
 
   mChroma = chroma;
 
-  if (SDL_Init(SDL_INIT_VIDEO) < 0 ) {
+  if (SDL_Init(SDL_INIT_VIDEO | SDL_INIT_NOPARACHUTE) < 0 ) {
     printf("SDL_Init() failed: %s\n", SDL_GetError( ) );
     SDL_Quit();
     return false;
   }
 
-  // set window title
-  const char *window_title = "SDL YUV display";
-  mWindow = SDL_CreateWindow(window_title,
-    SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED,
-    frame_width, frame_height, 0);
-  if (!mWindow) {
-    printf("SDL: Couldn't set video mode to %dx%d: %s\n",
-           frame_width, frame_height, SDL_GetError());
+  const SDL_VideoInfo* info = SDL_GetVideoInfo();
+  if( !info ) {
+    printf("SDL_GetVideoInfo() failed: %s\n", SDL_GetError() );
     SDL_Quit();
     return false;
   }
 
-  Uint32 flags = 0;  // Empty flags prioritize SDL_RENDERER_ACCELERATED.
-  mRenderer = SDL_CreateRenderer(mWindow, -1, flags);
-  if (!mRenderer) {
-    printf("SDL: Couldn't create renderer: %s\n", SDL_GetError());
+  Uint8 bpp = info->vfmt->BitsPerPixel;
+
+  Uint32 vflags;
+  if (info->hw_available)
+    vflags = SDL_HWSURFACE;
+  else
+    vflags = SDL_SWSURFACE;
+
+  // set window title
+  const char *window_title = "SDL YUV display";
+  SDL_WM_SetCaption(window_title, NULL);
+
+  mScreen = SDL_SetVideoMode(frame_width, frame_height, bpp, vflags);
+  if (mScreen == NULL) {
+    printf("SDL: Couldn't set video mode to %dx%d,%d bpp: %s",
+           frame_width, frame_height, bpp, SDL_GetError());
     SDL_Quit();
     return false;
   }
 
-  Uint32 pixelFormat = 0;
+  uint32_t pixelFormat;
   switch (mChroma) {
-  case SDL_CHROMA_MONO: pixelFormat = SDL_PIXELFORMAT_YV12; break;
-  case SDL_CHROMA_420:  pixelFormat = SDL_PIXELFORMAT_YV12; break;
-  case SDL_CHROMA_422:  pixelFormat = SDL_PIXELFORMAT_YV12; break;
-  case SDL_CHROMA_444:  pixelFormat = SDL_PIXELFORMAT_YV12; break;
-  //case SDL_CHROMA_444:  pixelFormat = SDL_PIXELFORMAT_YV12; break;
-  default:
-    printf("Unsupported chroma: %d\n", mChroma);
-    SDL_Quit();
-    return false;
+  case SDL_CHROMA_MONO: pixelFormat = SDL_YV12_OVERLAY; break;
+  case SDL_CHROMA_420:  pixelFormat = SDL_YV12_OVERLAY; break;
+  case SDL_CHROMA_422:  pixelFormat = SDL_YUY2_OVERLAY; break;
+    case SDL_CHROMA_444:  pixelFormat = SDL_YV12_OVERLAY; break;
+      //case SDL_CHROMA_444:  pixelFormat = SDL_YUY2_OVERLAY; break;
   }
 
-  mTexture = SDL_CreateTexture(mRenderer, pixelFormat,
-    SDL_TEXTUREACCESS_STREAMING, frame_width, frame_height);
-  if (!mTexture ) {
-    printf("SDL: Couldn't create SDL texture: %s\n", SDL_GetError());
+  mYUVOverlay = SDL_CreateYUVOverlay(frame_width, frame_height, pixelFormat, mScreen);
+  if (mYUVOverlay == NULL ) {
+    printf("SDL: Couldn't create SDL YUV overlay: %s",SDL_GetError());
     SDL_Quit();
     return false;
   }
@@ -100,8 +102,7 @@ void SDL_YUV_Display::display(const unsigned char *Y,
                               int stride, int chroma_stride)
 {
   if (!mWindowOpen) return;
-  if (SDL_LockTexture(mTexture, nullptr,
-    reinterpret_cast<void**>(&mPixels), &mStride) < 0) return;
+  if (SDL_LockYUVOverlay(mYUVOverlay) < 0) return;
 
   if (mChroma == SDL_CHROMA_420) {
     display420(Y,U,V,stride,chroma_stride);
@@ -117,10 +118,9 @@ void SDL_YUV_Display::display(const unsigned char *Y,
     display400(Y,stride);
   }
 
-  SDL_UnlockTexture(mTexture);
+  SDL_UnlockYUVOverlay(mYUVOverlay);
 
-  SDL_RenderCopy(mRenderer, mTexture, nullptr, nullptr);
-  SDL_RenderPresent(mRenderer);
+  SDL_DisplayYUVOverlay(mYUVOverlay, &rect);
 }
 
 
@@ -129,31 +129,26 @@ void SDL_YUV_Display::display420(const unsigned char *Y,
                                  const unsigned char *V,
                                  int stride, int chroma_stride)
 {
-  if (stride == mStride && chroma_stride == mStride/2) {
+  if (stride == rect.w && chroma_stride == rect.w/2) {
 
     // fast copy
 
-    memcpy(mPixels, Y, rect.w * rect.h);
-    memcpy(&mPixels[rect.w * rect.h], V, rect.w * rect.h / 4);
-    memcpy(&mPixels[(rect.w * rect.h) + (rect.w * rect.h / 4)], U, rect.w * rect.h / 4);
+    memcpy(mYUVOverlay->pixels[0], Y, rect.w * rect.h);
+    memcpy(mYUVOverlay->pixels[1], V, rect.w * rect.h / 4);
+    memcpy(mYUVOverlay->pixels[2], U, rect.w * rect.h / 4);
   }
   else {
     // copy line by line, because sizes are different
-    uint8_t *dest = mPixels;
-
-    for (int y=0;y<rect.h;y++,dest+=mStride)
-      {
-        memcpy(dest, Y+stride*y, rect.w);
-      }
 
-    for (int y=0;y<rect.h/2;y++,dest+=mStride/2)
+    for (int y=0;y<rect.h;y++)
       {
-        memcpy(dest, V+chroma_stride*y, rect.w/2);
+        memcpy(mYUVOverlay->pixels[0]+y*rect.w, Y+stride*y, rect.w);
       }
 
-    for (int y=0;y<rect.h/2;y++,dest+=mStride/2)
+    for (int y=0;y<rect.h/2;y++)
       {
-        memcpy(dest, U+chroma_stride*y, rect.w/2);
+        memcpy(mYUVOverlay->pixels[2]+y*rect.w/2, U+chroma_stride*y, rect.w/2);
+        memcpy(mYUVOverlay->pixels[1]+y*rect.w/2, V+chroma_stride*y, rect.w/2);
       }
   }
 }
@@ -161,26 +156,25 @@ void SDL_YUV_Display::display420(const unsigned char *Y,
 
 void SDL_YUV_Display::display400(const unsigned char *Y, int stride)
 {
-  uint8_t *dest = mPixels;
-  if (stride == mStride) {
+  if (stride == rect.w) {
 
     // fast copy
 
-    memcpy(mPixels, Y, rect.w * rect.h);
-    dest += mStride * rect.h;
+    memcpy(mYUVOverlay->pixels[0], Y, rect.w * rect.h);
   }
   else {
     // copy line by line, because sizes are different
 
-    for (int y=0;y<rect.h;y++,dest+=mStride)
+    for (int y=0;y<rect.h;y++)
       {
-        memcpy(dest, Y+stride*y, rect.w);
+        memcpy(mYUVOverlay->pixels[0]+y*rect.w, Y+stride*y, rect.w);
       }
   }
 
   // clear chroma planes
 
-  memset(dest, 0x80, mStride * rect.h / 2);
+  memset(mYUVOverlay->pixels[1], 0x80, rect.w * rect.h / 4);
+  memset(mYUVOverlay->pixels[2], 0x80, rect.w * rect.h / 4);
 }
 
 
@@ -189,9 +183,9 @@ void SDL_YUV_Display::display422(const unsigned char* Y,
                                  const unsigned char* V,
                                  int stride, int chroma_stride)
 {
-  for (int y = 0; y < rect.h; y++) {
-    unsigned char* dstY = mPixels + y * mStride;
-    const unsigned char* Yp = Y + y * stride;
+  for (int y = 0; y <rect.h; y++)
+    {
+      unsigned char* p = mYUVOverlay->pixels[0] + y*rect.w *2;
 
     memcpy(dstY, Yp, rect.w);
   }
@@ -219,7 +213,7 @@ void SDL_YUV_Display::display444as422(const unsigned char *Y,
 {
   for (int y=0;y<rect.h;y++)
     {
-      unsigned char* p = mPixels + y*mStride *2;
+      unsigned char* p = mYUVOverlay->pixels[0] + y*rect.w *2;
 
       const unsigned char* Yp = Y + y*stride;
       const unsigned char* Up = U + y*chroma_stride;
@@ -242,16 +236,14 @@ void SDL_YUV_Display::display444as420(const unsigned char *Y,
 {
   for (int y=0;y<rect.h;y++)
     {
-      unsigned char* p = mPixels + y*mStride;
+      unsigned char* p = mYUVOverlay->pixels[0] + y*rect.w;
       memcpy(p, Y+y*stride, rect.w);
     }
 
-  uint8_t *startV = mPixels + (rect.h*mStride);
-  uint8_t *startU = startV + (rect.h*mStride/2);
   for (int y=0;y<rect.h;y+=2)
     {
-      unsigned char* u = startU + y/2*mStride/2;
-      unsigned char* v = startV + y/2*mStride/2;
+      unsigned char* u = mYUVOverlay->pixels[2] + y/2*rect.w/2;
+      unsigned char* v = mYUVOverlay->pixels[1] + y/2*rect.w/2;
 
       for (int x=0;x<rect.w;x+=2) {
         u[x/2] = (U[ y   *chroma_stride + x] + U[ y   *chroma_stride + x +1] +
@@ -280,18 +272,7 @@ bool SDL_YUV_Display::doQuit() const
 
 void SDL_YUV_Display::close()
 {
-  if (mTexture) {
-    SDL_DestroyTexture(mTexture);
-    mTexture = nullptr;
-  }
-  if (mRenderer) {
-    SDL_DestroyRenderer(mRenderer);
-    mRenderer = nullptr;
-  }
-  if (mWindow) {
-    SDL_DestroyWindow(mWindow);
-    mWindow = nullptr;
-  }
+  SDL_FreeYUVOverlay(mYUVOverlay);
   SDL_Quit();
 
   mWindowOpen=false;
--- a/dec265/sdl.hh
+++ b/dec265/sdl.hh
@@ -48,13 +48,10 @@ public:
   bool isOpen() const { return mWindowOpen; }
 
 private:
-  SDL_Window *mWindow = nullptr;
-  SDL_Renderer *mRenderer = nullptr;
-  SDL_Texture *mTexture = nullptr;
+  SDL_Surface *mScreen;
+  SDL_Overlay *mYUVOverlay;
   SDL_Rect     rect;
   bool         mWindowOpen;
-  uint8_t *mPixels = nullptr;
-  int mStride = 0;
 
   SDL_Chroma mChroma;

--- a/scripts/ci-before-install-linux.sh
+++ b/scripts/ci-before-install-linux.sh
@@ -32,6 +32,6 @@
     INSTALL_PACKAGES="$INSTALL_PACKAGES \
         valgrind \
-        libsdl2-dev \
+        libsdl-dev \
         qtbase5-dev \
         libswscale-dev \
         "
 
