From e9142650006db8740df676397377fac31a5bf4bb Mon Sep 17 00:00:00 2001
From: captain <openatv@gmail.com>
Date: Wed, 9 Mar 2016 18:35:47 +0100

---
 src/shared/missing.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/shared/missing.h b/src/shared/missing.h
index bebbb42..b87258f 100644
--- a/src/shared/missing.h
+++ b/src/shared/missing.h
@@ -57,6 +57,8 @@
 #    define __NR_getrandom 352
 #  elif defined(__s390x__)
 #    define __NR_getrandom 349
+#  elif defined(__sh__)
+#    define __NR_getrandom 337
 #  else
 #    warning "__NR_getrandom unknown for your architecture"
 #    define __NR_getrandom 0xffffffff
@@ -128,6 +130,8 @@ static inline pid_t gettid(void) {
 #    define __NR_name_to_handle_at 370
 #  elif defined(__powerpc__)
 #    define __NR_name_to_handle_at 345
+#  elif defined(__sh__)
+#    define __NR_name_to_handle_at 338
 #  else
 #    error "__NR_name_to_handle_at is not defined"
 #  endif
