--- a/configure.ac
+++ b/configure.ac
@@ -338,7 +338,7 @@ AC_LIBPTHREAD([])
 # rpc/rpc.h can come from the glibc or from libtirpc
 nfsutils_save_CPPFLAGS="${CPPFLAGS}"
 CPPFLAGS="${CPPFLAGS} ${TIRPC_CFLAGS}"
-AC_CHECK_HEADER(rpc/rpc.h, ,
+AC_CHECK_HEADER(tirpc/rpc/rpc.h, ,
                 AC_MSG_ERROR([Header file rpc/rpc.h not found - maybe try building with --enable-tirpc]))
 CPPFLAGS="${nfsutils_save_CPPFLAGS}"
 
