diff --git a/gcc/configure b/gcc/configure
index 8e8987cf4..e28592047 100755
--- a/gcc/configure
+++ b/gcc/configure
@@ -26404,13 +26404,6 @@ $as_echo_n "checking assembler and linker for explicit JALR relocation... " >&6;
         rm -f conftest.*
       fi
     fi
-    if test $gcc_cv_as_ld_jalr_reloc = yes; then
-      if test x$target_cpu_default = x; then
-        target_cpu_default=MASK_RELAX_PIC_CALLS
-      else
-        target_cpu_default="($target_cpu_default)|MASK_RELAX_PIC_CALLS"
-      fi
-    fi
     { $as_echo "$as_me:${as_lineno-$LINENO}: result: $gcc_cv_as_ld_jalr_reloc" >&5
 $as_echo "$gcc_cv_as_ld_jalr_reloc" >&6; }
 
diff --git a/gcc/configure.ac b/gcc/configure.ac
index aa8927279..5391a9d3f 100644
--- a/gcc/configure.ac
+++ b/gcc/configure.ac
@@ -4350,13 +4350,6 @@ x:
         rm -f conftest.*
       fi
     fi
-    if test $gcc_cv_as_ld_jalr_reloc = yes; then
-      if test x$target_cpu_default = x; then
-        target_cpu_default=MASK_RELAX_PIC_CALLS
-      else
-        target_cpu_default="($target_cpu_default)|MASK_RELAX_PIC_CALLS"
-      fi
-    fi
     AC_MSG_RESULT($gcc_cv_as_ld_jalr_reloc)
 
     AC_CACHE_CHECK([linker for .eh_frame personality relaxation],
