diff --git a/libgfortran/configure b/libgfortran/configure
index f90fbc507..6db87d88b 100755
--- a/libgfortran/configure
+++ b/libgfortran/configure
@@ -12715,7 +12715,7 @@ esac
 
 # We need gfortran to compile parts of the library
 #AC_PROG_FC(gfortran)
-FC="$GFORTRAN"
+#FC="$GFORTRAN"
 ac_ext=${ac_fc_srcext-f}
 ac_compile='$FC -c $FCFLAGS $ac_fcflags_srcext conftest.$ac_ext >&5'
 ac_link='$FC -o conftest$ac_exeext $FCFLAGS $LDFLAGS $ac_fcflags_srcext conftest.$ac_ext $LIBS >&5'
diff --git a/libgfortran/configure.ac b/libgfortran/configure.ac
index 438157a98..8bd156391 100644
--- a/libgfortran/configure.ac
+++ b/libgfortran/configure.ac
@@ -227,7 +227,7 @@ AC_SUBST(enable_static)
 
 # We need gfortran to compile parts of the library
 #AC_PROG_FC(gfortran)
-FC="$GFORTRAN"
+#FC="$GFORTRAN"
 AC_PROG_FC(gfortran)
 
 # extra LD Flags which are required for targets
