diff --git a/gcc/toplev.c b/gcc/toplev.c
index 588fa2bd4..d94db6975 100644
--- a/gcc/toplev.c
+++ b/gcc/toplev.c
@@ -556,11 +556,11 @@ compile_file (void)
   if (flag_syntax_only || flag_wpa)
     return;
 
-  ggc_protect_identifiers = false;
-
   /* This must also call finalize_compilation_unit.  */
   lang_hooks.decls.final_write_globals ();
 
+  ggc_protect_identifiers = false;
+
   if (seen_error ())
     return;
 
