diff --git a/gcc/emit-rtl.c b/gcc/emit-rtl.c
index 3041b9e7a..0a9e50dc7 100644
--- a/gcc/emit-rtl.c
+++ b/gcc/emit-rtl.c
@@ -3426,7 +3426,6 @@ try_split (rtx pat, rtx trial, int last)
 {
   rtx before = PREV_INSN (trial);
   rtx after = NEXT_INSN (trial);
-  int has_barrier = 0;
   rtx note, seq, tem;
   int probability;
   rtx insn_last, insn;
@@ -3445,14 +3444,6 @@ try_split (rtx pat, rtx trial, int last)
 
   split_branch_probability = -1;
 
-  /* If we are splitting a JUMP_INSN, it might be followed by a BARRIER.
-     We may need to handle this specially.  */
-  if (after && BARRIER_P (after))
-    {
-      has_barrier = 1;
-      after = NEXT_INSN (after);
-    }
-
   if (!seq)
     return trial;
 
@@ -3598,8 +3589,6 @@ try_split (rtx pat, rtx trial, int last)
   tem = emit_insn_after_setloc (seq, trial, INSN_LOCATION (trial));
 
   delete_insn (trial);
-  if (has_barrier)
-    emit_barrier_after (tem);
 
   /* Recursively call try_split for each new insn created; by the
      time control returns here that insn will be fully split, so
