--- a/configure.ac	2009-12-03 07:24:45.000000000 +0100
+++ b/configure.ac	2014-08-08 19:22:48.388119619 +0200
@@ -6,11 +6,7 @@
 AC_PROG_CC
 AC_PROG_INSTALL
 
-PKG_CHECK_MODULES(XML, libxml-2.0, [HAVE_LIBXML=yes], [HAVE_LIBXML=no])
-if test x$HAVE_LIBXML = "xno"; then
-    AC_MSG_WARN([libxml headers not available. Cannot build evtest-capture.])
-fi
-AM_CONDITIONAL(HAVE_LIBXML, [test "x$HAVE_LIBXML" = "xyes"])
+AM_CONDITIONAL(HAVE_LIBXML, 1, [HAVE_LIBXML])
 
 AC_PATH_PROG(XSLTPROC, [xsltproc])
 AM_CONDITIONAL([HAVE_XSLTPROC], [test "x$XSLTPROC" != "x"])
