--- a/arch/mips/include/asm/brcmstb/brcmstb.h	2021-07-09 06:22:13.723802132 +0200
+++ b/arch/mips/include/asm/brcmstb/brcmstb.h	2021-07-09 06:25:04.542328086 +0200
@@ -1127,8 +1127,8 @@
 #define BRCM_PCI_SLOTS		16
 
 /* board-specific definitions are in arch/mips/brcmstb/board.c */
-extern char irq_tab_brcmstb[BRCM_PCI_SLOTS][4] __initdata;
-extern char irq_tab_brcmstb_docsis[BRCM_PCI_SLOTS][4] __initdata;
+extern char irq_tab_brcmstb[BRCM_PCI_SLOTS][4] __devinitdata;
+extern char irq_tab_brcmstb_docsis[BRCM_PCI_SLOTS][4] __devinitdata;
 
 struct bcmemac_platform_data {
 	int			phy_type;
