---
 src/mainwindow.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/mainwindow.cpp b/src/mainwindow.cpp
index 74c8458..4075cf5 100644
--- a/src/mainwindow.cpp
+++ b/src/mainwindow.cpp
@@ -95,7 +95,7 @@ MainWindow::MainWindow(QWidget *parent)
   keyCtrlF5->setKey(Qt::CTRL + Qt::Key_F5); // Set the key code
   connect(keyCtrlF5, SIGNAL(activated()), this, SLOT(slotShortcutCtrlF5()));
 
-  this->webview->page()->profile()->setHttpUserAgent("Mozilla/5.0 (X11; Linux x86_64; rv:89.0) Gecko/20100101 Firefox/89.0");
+  this->webview->page()->profile()->setHttpUserAgent("Mozilla/5.0 (SMART-TV; Linux; Tizen 4.0.0.2) AppleWebkit/605.1.15 (KHTML, like Gecko) SamsungBrowser/9.2 TV Safari/605.1.15");
 
   // Connect finished loading boolean
   connect(webview, &QWebEngineView::loadFinished, this,
-- 
2.17.1

