--- a/networking/libiproute/ipaddress.c
+++ b/networking/libiproute/ipaddress.c
@@ -572,10 +572,8 @@ int FAST_FUNC ipaddr_list_or_flush(char **argv, int flush)
 	}
 
 	for (l = linfo; l; l = l->next) {
-		if ((oneline && G_filter.family != AF_PACKET)
-		/* ^^^^^^^^^ "ip -oneline a" does not print link info */
-		 || (print_linkinfo(&l->h) == 0)
-		) {
+		if (print_linkinfo(&l->h) == 0)
+		{
 			struct ifinfomsg *ifi = NLMSG_DATA(&l->h);
 			if (G_filter.family != AF_PACKET)
 				print_selected_addrinfo(ifi->ifi_index, ainfo);
