--- a/drivers/brcmstb/board.c	2014-03-07 16:54:09.631223397 +0100
+++ b/drivers/brcmstb/board.c	2014-03-04 22:30:34.459824416 +0100
@@ -642,17 +639,17 @@
 	{
 		.name = "complete",
 		.size = MTDPART_SIZ_FULL,
-		.offset = 0x00000000
+		.offset = 0x00500000,
 	},
 	{
 		.name = "kernel",
 		.size = 0x00600000,
-		.offset = 0x00000000
+		.offset = 0x00500000,
 	},
 	{
 		.name = "rootfs",
 		.size = MTDPART_SIZ_FULL,
-		.offset = 0x00600000
+		.offset = 0x00b00000
 	},
 };
 
