diff --git a/locale/xlocale.h b/locale/xlocale.h
index e4408e2e994..13db0a6b7e5 100644
--- a/locale/xlocale.h
+++ b/locale/xlocale.h
@@ -20,6 +20,8 @@
 #ifndef _XLOCALE_H
 #define _XLOCALE_H	1
 
+#if defined(__GLIBC__) && (__GLIBC__ < 2) || (__GLIBC__ == 2 && __GLIBC_MINOR__ < 26)
+
 /* Structure for reentrant locale using functions.  This is an
    (almost) opaque type for the user level programs.  The file and
    this data structure is not standardized.  Don't rely on it.  It can
@@ -41,4 +43,10 @@ typedef struct __locale_struct
 /* POSIX 2008 makes locale_t official.  */
 typedef __locale_t locale_t;
 
+#else
+
+typedef struct __locale_struct *__locale_t;
+
+#endif
+
 #endif /* xlocale.h */
