From 1033392ab0f3bb5f7f32b1c7f021347a21b45611 Mon Sep 17 00:00:00 2001
From: Captain <captain.onboard@web.de>
Date: Sat, 23 May 2015 19:25:51 +0200
Subject: [PATCH] add more sdl keys

---
 sdl/sdlinput.cpp | 25 ++++++++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

diff --git a/sdl/sdlinput.cpp b/sdl/sdlinput.cpp
index 156d392..627647c 100644
--- a/sdl/sdlinput.cpp
+++ b/sdl/sdlinput.cpp
@@ -273,11 +273,30 @@ void S9xParseInputConfig (ConfigFile &conf, int pass)
 		keymaps.push_back(strpair_t("K00:SDLK_z",            "Joypad1 R"));
 
 		// domaemon: *) GetSDLKeyFromName().
-		name_sdlkeysym["SDLK_s"] = SDLK_s;
+		name_sdlkeysym["SDLK_a"] = SDLK_a;
+		name_sdlkeysym["SDLK_b"] = SDLK_b;
+		name_sdlkeysym["SDLK_c"] = SDLK_c;
 		name_sdlkeysym["SDLK_d"] = SDLK_d;
+		name_sdlkeysym["SDLK_f"] = SDLK_f;
+		name_sdlkeysym["SDLK_g"] = SDLK_g;
+		name_sdlkeysym["SDLK_h"] = SDLK_h;
+		name_sdlkeysym["SDLK_i"] = SDLK_i;
+		name_sdlkeysym["SDLK_j"] = SDLK_j;
+		name_sdlkeysym["SDLK_k"] = SDLK_k;
+		name_sdlkeysym["SDLK_l"] = SDLK_l;
+		name_sdlkeysym["SDLK_m"] = SDLK_m;
+		name_sdlkeysym["SDLK_n"] = SDLK_n;
+		name_sdlkeysym["SDLK_o"] = SDLK_o;
+		name_sdlkeysym["SDLK_p"] = SDLK_p;
+		name_sdlkeysym["SDLK_q"] = SDLK_q;
+		name_sdlkeysym["SDLK_r"] = SDLK_r;
+		name_sdlkeysym["SDLK_s"] = SDLK_s;
+		name_sdlkeysym["SDLK_t"] = SDLK_t;
+		name_sdlkeysym["SDLK_u"] = SDLK_u;
+		name_sdlkeysym["SDLK_v"] = SDLK_v;
+		name_sdlkeysym["SDLK_w"] = SDLK_w;
 		name_sdlkeysym["SDLK_x"] = SDLK_x;
-		name_sdlkeysym["SDLK_c"] = SDLK_c;
-		name_sdlkeysym["SDLK_a"] = SDLK_a;
+		name_sdlkeysym["SDLK_y"] = SDLK_y;
 		name_sdlkeysym["SDLK_z"] = SDLK_z;
 		name_sdlkeysym["SDLK_UP"] = SDLK_UP;
 		name_sdlkeysym["SDLK_DOWN"] = SDLK_DOWN;
-- 
1.9.5.msysgit.0

