--- a/Makefile	2016-12-16 13:46:41.592554304 +0100
+++ b/Makefile	2016-12-16 13:53:23.028375790 +0100
@@ -695,7 +695,7 @@
 KBUILD_CFLAGS   += $(call cc-option,-Werror=strict-prototypes)
 
 # Prohibit date/time macros, which would make the build non-deterministic
-KBUILD_CFLAGS   += $(call cc-option,-Werror=date-time)
+#KBUILD_CFLAGS   += $(call cc-option,-Werror=date-time)
 
 # use the deterministic mode of AR if available
 KBUILD_ARFLAGS := $(call ar-option,D)
