--- a/hfs.c	2015-02-20 21:22:20.363897145 +0100
+++ b/hfs.c	2015-02-20 21:25:14.431899569 +0100
@@ -201,7 +201,11 @@
 
 
 int kcdfsd_add_cdhfs_request(struct file * file, struct page *page){
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  return kcdfsd_add_request(file->f_path.dentry, page, CDHFS_REQUEST);
+#else
   return kcdfsd_add_request(file->f_dentry, page, CDHFS_REQUEST);
+#endif
 }
 
 struct address_space_operations cdfs_cdhfs_aops = {
--- a/cddata.c	2015-02-20 21:14:56.531890965 +0100
+++ b/cddata.c	2015-02-20 21:16:25.099892198 +0100
@@ -97,7 +97,11 @@
 /************************************************************/
 
 int kcdfsd_add_cddata_request(struct file * file, struct page *page){
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  return kcdfsd_add_request(file->f_path.dentry, page, CDDATA_REQUEST);
+#else
   return kcdfsd_add_request(file->f_dentry, page, CDDATA_REQUEST);
+#endif
 }
 
 /***********************************************************/            
--- a/cdXA.c	2015-02-20 21:04:26.195882187 +0100
+++ b/cdXA.c	2015-02-20 21:06:49.275884180 +0100
@@ -38,7 +38,11 @@
 /************************************************************/
 
 int kcdfsd_add_cdXA_request(struct file * file, struct page *page){
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  return kcdfsd_add_request(file->f_path.dentry, page, CDXA_REQUEST);
+#else
   return kcdfsd_add_request(file->f_dentry, page, CDXA_REQUEST);
+#endif
 }
 
 /***********************************************************/            
--- a/audio.c	2015-02-20 20:53:45.359873264 +0100
+++ b/audio.c	2015-02-20 20:56:17.771875386 +0100
@@ -187,12 +187,20 @@
 
 int kcdfsd_add_cdda_request(struct file * file, struct page *page){
   PRINT("kcdfsd_add_cdda_request(%x, %x)\n", file, page);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  return kcdfsd_add_request(file->f_path.dentry, page, CDDA_REQUEST);
+#else
   return kcdfsd_add_request(file->f_dentry, page, CDDA_REQUEST);
+#endif
 }
 
 int kcdfsd_add_cdda_raw_request(struct file * file, struct page *page){
   PRINT("kcdfsd_add_cdda_request(%x, %x)\n", file, page);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  return kcdfsd_add_request(file->f_path.dentry, page, CDDA_RAW_REQUEST);
+#else
   return kcdfsd_add_request(file->f_dentry, page, CDDA_RAW_REQUEST);
+#endif
 }
 
 struct address_space_operations cdfs_cdda_aops = {
--- a/root.c	2015-02-20 20:38:36.815860613 +0100
+++ b/root.c	2015-02-20 20:42:30.931863873 +0100
@@ -456,7 +456,11 @@
 #else
 static int cdfs_readdir(struct file *filp, void *dirent, filldir_t filldir) {
 #endif
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  struct inode *inode = filp->f_path.dentry->d_inode;
+#else
   struct inode *inode = filp->f_dentry->d_inode;
+#endif
   int i;
   cd * this_cd = cdfs_info(inode->i_sb);
 
