--- a/audio.c	2009-12-21 14:54:06.000000000 +0100
+++ b/audio.c	2014-09-11 22:02:58.598344750 +0200
@@ -168,7 +168,11 @@
 
 struct file_operations cdfs_cdda_file_operations = {
     .read = do_sync_read,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,16,0)
+    .read_iter = generic_file_read_iter,
+#else
     .aio_read = generic_file_aio_read,
+#endif
     .mmap = generic_file_mmap,
 };
 
--- a/cdXA.c	2009-12-21 14:54:06.000000000 +0100
+++ b/cdXA.c	2014-09-11 23:30:05.038267254 +0200
@@ -27,7 +27,11 @@
 
 struct file_operations cdfs_cdXA_file_operations = {
     .read = do_sync_read,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,16,0)
+    .read_iter = generic_file_read_iter,
+#else
     .aio_read = generic_file_aio_read,
+#endif
     .mmap = generic_file_mmap,
 };
 
--- a/cddata.c	2009-12-21 15:04:03.000000000 +0100
+++ b/cddata.c	2014-09-11 23:56:49.098243469 +0200
@@ -77,7 +77,11 @@
 
 struct file_operations cdfs_cddata_file_operations = {
   .read = do_sync_read,
-  .aio_read = generic_file_aio_read,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,16,0)
+    .read_iter = generic_file_read_iter,
+#else
+    .aio_read = generic_file_aio_read,
+#endif
 #ifdef OLD_KERNEL
   .mmap             = generic_file_mmap
 #else
--- a/hfs.c	2009-12-21 14:54:06.000000000 +0100
+++ b/hfs.c	2014-09-12 00:03:50.818237216 +0200
@@ -195,7 +195,11 @@
 
 struct file_operations cdfs_cdhfs_file_operations = {
   .read = do_sync_read,
-  .aio_read = generic_file_aio_read,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,16,0)
+    .read_iter = generic_file_read_iter,
+#else
+    .aio_read = generic_file_aio_read,
+#endif
   .mmap             = generic_file_mmap
 };
 
--- a/daemon.c	2014-09-12 00:08:08.466233396 +0200
+++ b/daemon.c	2014-09-12 16:44:12.105347237 +0200
@@ -29,6 +29,9 @@
 
 #include "cdfs.h"
 #include <linux/sched.h>
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,15,0)
+#include <linux/wait.h>
+#endif
 
 int kcdfsd_pid = 0;
 static int kcdfsd_running = 0;
@@ -169,9 +172,18 @@
   strcpy(current->comm, "k"FSNAME"d");
 
   /* Send me a signal to get me die */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,15,0)
+  DEFINE_WAIT(wait);
+#endif
   do {
     kcdfsd_process_request();
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,15,0)
+    prepare_to_wait(&kcdfsd_wait, &wait, TASK_INTERRUPTIBLE);
+    schedule();
+    finish_wait(&kcdfsd_wait, &wait);
+#else
     interruptible_sleep_on(&kcdfsd_wait);
+#endif
   } while (!signal_pending(current));
 
   kcdfsd_running = 0;
 