diff --git a/utils.c b/utils.c
index a145fa3b4e41..7b597276fa6f 100644
--- a/utils.c
+++ b/utils.c
@@ -82,6 +82,17 @@ void cdfs_constructMSFsize(char * result, unsigned length){
 }
 
 /***********************************************************************/
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,8,0))
+int ioctl_by_bdev(struct block_device *bdev, unsigned cmd, unsigned long arg)
+{
+       int res;
+       mm_segment_t old_fs = get_fs();
+       set_fs(KERNEL_DS);
+       res = blkdev_ioctl(bdev, 0, cmd, arg);
+       set_fs(old_fs);
+       return res;
+}
+#endif
 
 int cdfs_ioctl(struct super_block *s, int cmd, unsigned long arg) {
   return ioctl_by_bdev(s->s_bdev, cmd, arg); 
