diff --git a/audio.c b/audio.c
index 67f1664..98f4fd1 100644
--- a/audio.c
+++ b/audio.c
@@ -187,12 +187,20 @@ struct file_operations cdfs_cdda_file_operations = {
 
 int kcdfsd_add_cdda_request(struct file * file, struct page *page){
   PRINT("kcdfsd_add_cdda_request(%x, %x)\n", file, page);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  return kcdfsd_add_request(file->f_path.dentry, page, CDDA_REQUEST);
+#else
   return kcdfsd_add_request(file->f_dentry, page, CDDA_REQUEST);
+#endif
 }
 
 int kcdfsd_add_cdda_raw_request(struct file * file, struct page *page){
   PRINT("kcdfsd_add_cdda_request(%x, %x)\n", file, page);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  return kcdfsd_add_request(file->f_path.dentry, page, CDDA_RAW_REQUEST);
+#else
   return kcdfsd_add_request(file->f_dentry, page, CDDA_RAW_REQUEST);
+#endif
 }
 
 struct address_space_operations cdfs_cdda_aops = {
diff --git a/cdXA.c b/cdXA.c
index 09da159..3649fde 100644
--- a/cdXA.c
+++ b/cdXA.c
@@ -38,7 +38,11 @@ struct address_space_operations cdfs_cdXA_aops = {
 /************************************************************/
 
 int kcdfsd_add_cdXA_request(struct file * file, struct page *page){
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  return kcdfsd_add_request(file->f_path.dentry, page, CDXA_REQUEST);
+#else
   return kcdfsd_add_request(file->f_dentry, page, CDXA_REQUEST);
+#endif
 }
 
 /***********************************************************/            
diff --git a/cddata.c b/cddata.c
index 4a985a7..0d21062 100644
--- a/cddata.c
+++ b/cddata.c
@@ -97,7 +97,11 @@ struct address_space_operations cdfs_cddata_aops = {
 /************************************************************/
 
 int kcdfsd_add_cddata_request(struct file * file, struct page *page){
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  return kcdfsd_add_request(file->f_path.dentry, page, CDDATA_REQUEST);
+#else
   return kcdfsd_add_request(file->f_dentry, page, CDDATA_REQUEST);
+#endif
 }
 
 /***********************************************************/            
diff --git a/hfs.c b/hfs.c
index dcb852b..9513059 100644
--- a/hfs.c
+++ b/hfs.c
@@ -201,7 +201,12 @@ struct file_operations cdfs_cdhfs_file_operations = {
 
 
 int kcdfsd_add_cdhfs_request(struct file * file, struct page *page){
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  return kcdfsd_add_request(file->f_path.dentry, page, CDHFS_REQUEST);
+#else
   return kcdfsd_add_request(file->f_dentry, page, CDHFS_REQUEST);
+#endif
+  
 }
 
 struct address_space_operations cdfs_cdhfs_aops = {
diff --git a/root.c b/root.c
index 353d191..941ac3e 100644
--- a/root.c
+++ b/root.c
@@ -456,7 +456,11 @@ static int cdfs_readdir(struct file *filp, struct dir_context *ctx) {
 #else
 static int cdfs_readdir(struct file *filp, void *dirent, filldir_t filldir) {
 #endif
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0)
+  struct inode *inode = filp->f_path.dentry->d_inode;
+#else
   struct inode *inode = filp->f_dentry->d_inode;
+#endif
   int i;
   cd * this_cd = cdfs_info(inode->i_sb);
 
