diff --git a/e2_proc/e2_proc_main.c b/e2_proc/e2_proc_main.c
index 88ac7e1..4341320 100644
--- a/e2_proc/e2_proc_main.c
+++ b/e2_proc/e2_proc_main.c
@@ -52,6 +52,10 @@
  *  |             |
  *  |             --------- aspect_choices
  *  |             |
+ *  |             --------- hdmi_colorspace
+ *  |             |
+ *  |             --------- hdmi_colorspace_choices
+ *  |             |
  *  |             --------- force_dvi
  *  |             |
  *  |             --------- policy
@@ -186,6 +190,8 @@
  *  |               --------- progressive
  *  |               |
  *  |               --------- aspect
+ *  |               |
+ *  |               --------- pep_apply
  *  |
  *  ---------- fb
  *  |           |
@@ -604,6 +610,8 @@ struct ProcStructure_s e2Proc[] =
 	{cProcEntry, "stb/video/alpha"                                                  , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/video/aspect"                                                 , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/video/aspect_choices"                                         , NULL, NULL, NULL, NULL, ""},
+	{cProcEntry, "stb/video/hdmi_colorspace"                                        , NULL, NULL, NULL, NULL, ""},
+	{cProcEntry, "stb/video/hdmi_colorspace_choices"                                , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/video/force_dvi"                                              , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/video/policy"                                                 , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/video/policy_choices"                                         , NULL, NULL, NULL, NULL, ""},
@@ -677,6 +685,7 @@ struct ProcStructure_s e2Proc[] =
 	{cProcEntry, "stb/vmpeg/0/xres"                                                 , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/vmpeg/0/aspect"                                               , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/vmpeg/0/framerate"                                            , NULL, NULL, NULL, NULL, ""},
+	{cProcEntry, "stb/vmpeg/0/pep_apply"                                            , NULL, NULL, NULL, NULL, ""},
 
 	{cProcDir  , "stb/vmpeg/1"                                                      , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/vmpeg/1/dst_apply"                                            , NULL, NULL, NULL, NULL, ""},
@@ -689,6 +698,7 @@ struct ProcStructure_s e2Proc[] =
 	{cProcEntry, "stb/vmpeg/1/xres"                                                 , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/vmpeg/1/aspect"                                               , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/vmpeg/1/framerate"                                            , NULL, NULL, NULL, NULL, ""},
+	{cProcEntry, "stb/vmpeg/1/pep_apply"                                            , NULL, NULL, NULL, NULL, ""},
 
 	{cProcDir  , "stb/hdmi"                                                         , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/hdmi/bypass_edid_checking"                                    , NULL, NULL, NULL, NULL, ""},
diff --git a/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_avs.c b/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_avs.c
index 965375c..19f03f5 100644
--- a/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_avs.c
+++ b/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_avs.c
@@ -700,3 +700,99 @@ int proc_avs_0_standby_read(char *page, char **start, off_t off, int count,
 		len = sprintf(page, "on\n");
 	return len;
 }
+
+int proc_video_hdmi_colorspace_write(struct file* file, const char __user* buf, unsigned long count, void* data)
+{
+	char* page;
+	char* myString;
+	ssize_t ret = -ENOMEM;
+#ifdef VERY_VERBOSE
+	printk("%s %ld - ", __FUNCTION__, count);
+#endif
+	page = (char*)__get_free_page(GFP_KERNEL);
+	if (page)
+	{
+		struct stmfb_info* info = stmfb_get_fbinfo_ptr();
+		struct stmfbio_output_configuration outputConfig;
+		int err = 0;
+		//int alpha = 0;
+		int hdmi_colour = 0;
+		ret = -EFAULT;
+		if (copy_from_user(page, buf, count))
+			goto out;
+		myString = (char*) kmalloc(count + 1, GFP_KERNEL);
+		strncpy(myString, page, count);
+		myString[count] = '\0';
+#ifdef VERY_VERBOSE
+		printk("%s\n", myString);
+#endif
+		//sscanf(myString, "%d", &alpha);
+		//0rgb 1yuv 2422
+		if (strncmp("hdmi_rgb", page, count - 1) == 0)
+		{
+			hdmi_colour = 0;
+		}
+		else if (strncmp("hdmi_yuv", page, count - 1) == 0)
+		{
+			hdmi_colour = 1;
+		}
+		else if (strncmp("hdmi_422", page, count - 1) == 0)
+		{
+			hdmi_colour = 2;
+		}
+		outputConfig.outputid = 1;
+		stmfb_get_output_configuration(&outputConfig, info);
+		outputConfig.caps = 0;
+		outputConfig.activate = 0; //STMFBIO_ACTIVATE_IMMEDIATE;
+		outputConfig.caps |= STMFBIO_OUTPUT_CAPS_HDMI_CONFIG;
+		outputConfig.hdmi_config &= ~(STMFBIO_OUTPUT_HDMI_YUV | STMFBIO_OUTPUT_HDMI_422);
+		switch (hdmi_colour)
+		{
+			case 1:
+				outputConfig.hdmi_config |= STMFBIO_OUTPUT_HDMI_YUV;
+				break;
+			case 2:
+				outputConfig.hdmi_config |= (STMFBIO_OUTPUT_HDMI_YUV | STMFBIO_OUTPUT_HDMI_422);
+				break;
+			default:
+				break;
+		}
+		err = stmfb_set_output_configuration(&outputConfig, info);
+		//if(ioctl(fbfd, STMFBIO_SET_OUTPUT_CONFIG, &outputConfig)<0)
+		//perror("setting output configuration failed");
+		kfree(myString);
+	}
+	ret = count;
+out:
+	free_page((unsigned long)page);
+	return ret;
+}
+
+int proc_video_hdmi_colorspace_read(char* page, char** start, off_t off, int count, int* eof, void* data_unused)
+{
+	struct stmfb_info* info = stmfb_get_fbinfo_ptr();
+	struct stmfbio_output_configuration outputConfig;
+	int len = 0;
+#ifdef VERY_VERBOSE
+	printk("%s %d\n", __FUNCTION__, count);
+#endif
+	outputConfig.outputid = 1;
+	stmfb_get_output_configuration(&outputConfig, info);
+	if (outputConfig.hdmi_config & STMFBIO_OUTPUT_HDMI_422)
+		len = sprintf(page, "hdmi_422\n");
+	else if (outputConfig.hdmi_config & STMFBIO_OUTPUT_HDMI_YUV)
+		len = sprintf(page, "hdmi_yuv\n");
+	else
+		len = sprintf(page, "hdmi_rgb\n");
+	return len;
+}
+
+int proc_video_hdmi_colorspace_choices_read(char* page, char** start, off_t off, int count, int* eof, void* data_unused)
+{
+	int len = 0;
+#ifdef VERY_VERBOSE
+	printk("%s\n", __FUNCTION__);
+#endif
+	len = sprintf(page, "hdmi_rgb hdmi_yuv hdmi_422\n");
+	return len;
+}
diff --git a/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_main.c b/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_main.c
index 935915c..101bea0 100644
--- a/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_main.c
+++ b/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_main.c
@@ -37,6 +37,10 @@
  * | |
  * | --------- force_dvi
  * | |
+ * | --------- hdmi_colorspace
+ * | |
+ * | --------- hdmi_colorspace_choices
+ * | |
  * | --------- policy
  * | |
  * | --------- policy_choices
@@ -220,6 +224,9 @@ extern int proc_audio_j1_mute_write(struct file *file, const char __user *buf, u
 extern int proc_video_aspect_read(char *page, char **start, off_t off, int count, int *eof, void *data_unused);
 extern int proc_video_aspect_write(struct file *file, const char __user *buf, unsigned long count, void *data);
 extern int proc_video_aspect_choices_read(char *page, char **start, off_t off, int count, int *eof, void *data_unused);
+extern int proc_video_hdmi_colorspace_write(struct file* file, const char __user* buf, unsigned long count, void* data);
+extern int proc_video_hdmi_colorspace_read(char* page, char** start, off_t off, int count, int* eof, void* data_unused);
+extern int proc_video_hdmi_colorspace_choices_read(char* page, char** start, off_t off, int count, int* eof, void* data_unused);
 extern int proc_video_policy_read(char *page, char **start, off_t off, int count, int *eof, void *data_unused);
 extern int proc_video_policy_write(struct file *file, const char __user *buf, unsigned long count, void *data);
 extern int proc_video_policy_choices_read(char *page, char **start, off_t off, int count, int *eof, void *data_unused);
@@ -378,6 +385,8 @@ struct e2_procs
 	/*
 	 {"stb/video/force_dvi", NULL, NULL, 0},
 	*/
+	{"stb/video/hdmi_colorspace", proc_video_hdmi_colorspace_read,        proc_video_hdmi_colorspace_write, 0},
+	{"stb/video/hdmi_colorspace_choices", proc_video_hdmi_colorspace_choices_read,NULL, 0},
 	{"stb/video/policy", proc_video_policy_read, proc_video_policy_write, 0},
 	{"stb/video/policy_choices", proc_video_policy_choices_read, NULL, 0},
 	{"stb/video/videomode", proc_video_videomode_read, proc_video_videomode_write, 0},
