diff --git a/e2_proc/e2_proc_main.c b/e2_proc/e2_proc_main.c
index 4e33b90d..412f5b40 100644
--- a/e2_proc/e2_proc_main.c
+++ b/e2_proc/e2_proc_main.c
@@ -118,8 +118,6 @@
  *  |           |
  *  |           --------- led0_pattern
  *  |           |
- *  |           --------- led1_pattern
- *  |           |
  *  |           --------- led_pattern_speed
  *  |           |
  *  |           |
@@ -673,7 +671,6 @@ struct ProcStructure_s e2Proc[] =
 	{cProcEntry, "stb/fp/lnb_sense1"                                                , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/fp/lnb_sense2"                                                , NULL, NULL, NULL, NULL, ""},
 	{cProcEntry, "stb/fp/led0_pattern"                                              , NULL, NULL, default_write_proc, NULL, ""},
-	{cProcEntry, "stb/fp/led1_pattern"                                              , NULL, NULL, default_write_proc, NULL, ""},
 	{cProcEntry, "stb/fp/led_pattern_speed"                                         , NULL, NULL, default_write_proc, NULL, ""},
 	{cProcEntry, "stb/fp/version"                                                   , NULL, zero_read, NULL, NULL, ""},
 	{cProcEntry, "stb/fp/wakeup_time"                                               , NULL, wakeup_time_read, wakeup_time_write, NULL, ""},
diff --git a/frontcontroller/aotom_spark/aotom_procfs.c b/frontcontroller/aotom_spark/aotom_procfs.c
index 33112464..58ecc9ae 100644
--- a/frontcontroller/aotom_spark/aotom_procfs.c
+++ b/frontcontroller/aotom_spark/aotom_procfs.c
@@ -45,8 +45,6 @@
  *             +--- was_timer_wakeup (rw)
  *             |
  *             +--- led0_pattern (w)
- *             |
- *             +--- led1_pattern (w)
  *  /proc/stb/lcd
  *             |
  *             +--- scroll_delay (w)
@@ -644,29 +642,6 @@ static int led0_pattern_write(struct file *file, const char __user *buf, unsigne
 }
 
 
-static int led1_pattern_write(struct file *file, const char __user *buf, unsigned long count, void *data)
-{
-	char *page;
-	long value;
-	int ret = -ENOMEM;
-	page = (char *)__get_free_page(GFP_KERNEL);
-	if (page)
-	{
-		ret = -EFAULT;
-		if (copy_from_user(page, buf, count) == 0)
-		{
-			page[count] = '\0';
-			value = simple_strtol(page, NULL, 16);
-			if (value==0) aotomSetLed(1,0);
-			else if (value==0xffffffff) aotomSetLed(1,1);
-			ret = count;
-		}
-		free_page((unsigned long)page);
-	}
-	return ret;
-}
-
-
 static int lcd_oled_brightness_write(struct file *file, const char __user *buf, unsigned long count, void *data)
 {
 	char *page;
@@ -743,7 +718,6 @@ struct fp_procs
 	{ "stb/fp/rtc_offset", aotom_read_rtc_offset, aotom_write_rtc_offset },
 	{ "stb/fp/aotom", NULL, aotom_write },
 	{ "stb/fp/led0_pattern", NULL, led0_pattern_write },
-	{ "stb/fp/led1_pattern", NULL, led1_pattern_write },
 	{ "stb/fp/wakeup_time", wakeup_time_read, wakeup_time_write },
 	{ "stb/fp/was_timer_wakeup", was_timer_wakeup_read, null_write },
 	{ "stb/fp/version", fp_version_read, NULL },
