diff --git a/bpamem/bpamem.c b/bpamem/bpamem.c
index 40c833a..3be34be 100644
--- a/bpamem/bpamem.c
+++ b/bpamem/bpamem.c
@@ -180,7 +180,9 @@ static int bpamem_ioctl(struct inode *inode, struct file *filp, unsigned int ioc
 	dev = MINOR(inode->i_rdev);
 	switch (ioctl_num) 
 	{
-		case BPAMEMIO_ALLOCMEM:   return bpamemio_allocmem((BPAMemAllocMemData *)ioctl_param);
+		case BPAMEMIO_ALLOCMEM:
+			filp->f_pos = -42;
+			return bpamemio_allocmem((BPAMemAllocMemData *)ioctl_param);
 		case BPAMEMIO_MAPMEM:     return bpamemio_mapmem((BPAMemMapMemData *)ioctl_param);
 
 		case BPAMEMIO_FREEMEM:    return bpamemio_deallocmem(dev);
@@ -196,7 +198,11 @@ static int bpamem_open(struct inode *inode, struct file *filp)
 
 static int bpamem_release (struct inode *inode, struct file *filp)
 {
-	return 0;
+	unsigned int dev;
+	if (filp->f_pos == -42)
+		return 0;
+	dev = MINOR(inode->i_rdev);
+	return bpamemio_deallocmem(dev);
 }
 
 
@@ -285,7 +291,7 @@ static void __exit bpamem_exit(void)
 {
 	int i;
 
-	printk("Unregistering device '%s', major '%d'", DEVICE_NAME, BPAMEM_MAJOR);
+	printk("Unregistering device '%s', major '%d'\n", DEVICE_NAME, BPAMEM_MAJOR);
 
 	for(i = 0; i < MAX_BPA_ALLOCS; i++)
 		bpamemio_deallocmem(i);
