diff --git a/frontends/spark_dvbapi5/stv090x.c b/frontends/spark_dvbapi5/stv090x.c
index bc881e7..b1cb781 100644
--- a/frontends/spark_dvbapi5/stv090x.c
+++ b/frontends/spark_dvbapi5/stv090x.c
@@ -5336,7 +5336,7 @@ static int stv090x_read_cnr(struct dvb_frontend *fe, u16 *cnr)
 			if (lock_f)
 			{
 				msleep(5);
-				for (i = 0; i < 16; i++)
+				for (i = 0; i < 4; i++)
 				{
 					reg_1 = STV090x_READ_DEMOD(state, NNOSPLHT1);
 					val_1 = STV090x_GETFIELD_Px(reg_1, NOSPLHT_NORMED_FIELD);
@@ -5345,7 +5345,7 @@ static int stv090x_read_cnr(struct dvb_frontend *fe, u16 *cnr)
 					val  += MAKEWORD16(val_1, val_0);
 					msleep(1);
 				}
-				val /= 16;
+				val /= 4;
 				snr = stv090x_table_lookup(stv090x_s2cn_tab, ARRAY_SIZE(stv090x_s2cn_tab) - 1, val);
 				div = stv090x_s2cn_tab[0].read - stv090x_s2cn_tab[ARRAY_SIZE(stv090x_s2cn_tab) - 1].read;
 				*cnr = 0xFFFF - ((val * 0xFFFF) / div);
@@ -5360,7 +5360,7 @@ static int stv090x_read_cnr(struct dvb_frontend *fe, u16 *cnr)
 			if (lock_f)
 			{
 				msleep(5);
-				for (i = 0; i < 16; i++)
+				for (i = 0; i < 4; i++)
 				{
 					reg_1 = STV090x_READ_DEMOD(state, NOSDATAT1);
 					val_1 = STV090x_GETFIELD_Px(reg_1, NOSDATAT_UNNORMED_FIELD);
@@ -5369,7 +5369,7 @@ static int stv090x_read_cnr(struct dvb_frontend *fe, u16 *cnr)
 					val  += MAKEWORD16(val_1, val_0);
 					msleep(1);
 				}
-				val /= 16;
+				val /= 4;
 				snr = stv090x_table_lookup(stv090x_s1cn_tab, ARRAY_SIZE(stv090x_s1cn_tab) - 1, val);
 				div = stv090x_s1cn_tab[0].read - stv090x_s1cn_tab[ARRAY_SIZE(stv090x_s1cn_tab) - 1].read;
 				*cnr = 0xFFFF - ((val * 0xFFFF) / div);
