Index: b/wireless/rt5370sta/sta/sta_cfg.c
===================================================================
--- a/wireless/rt5370sta/sta/sta_cfg.c
+++ b/wireless/rt5370sta/sta/sta_cfg.c
@@ -4984,7 +4984,11 @@ VOID RTMPIoctlShow(
             wrq->u.data.length = strlen(extra) + 1; /* 1: size of '\0' */
             break;
         case SHOW_DRVIER_VERION:
+#ifdef INCLUDE_BUILDDATE
             snprintf(extra, size, "Driver version-%s, %s %s\n", STA_DRIVER_VERSION, __DATE__, __TIME__ );
+#else
+            snprintf(extra, size, "Driver version-%s\n", STA_DRIVER_VERSION);
+#endif
             wrq->u.data.length = strlen(extra) + 1; /* 1: size of '\0' */
             break;
 #ifdef DOT11_N_SUPPORT
