diff --git a/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_video.c b/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_video.c
index f6056c6..6c4bd7b 100644
--- a/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_video.c
+++ b/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_video.c
@@ -143,7 +143,7 @@ static struct Modes
 {
 
 	{
-		25, "1024x768_60", 1024, 768, 1024, 768, 8, 15385, 160, 24, 29, 3, 136, 6,
+		25, "1024x768", 1024, 768, 1024, 768, 8, 15385, 160, 24, 29, 3, 136, 6,
 		FB_VMODE_NONINTERLACED | FB_VMODE_CONUPDATE
 	},
 	{
@@ -164,7 +164,7 @@ static struct Modes
 	},
 
 	{
-		20, "1280x1024_60", 1280, 1024, 1280, 1024, 8, 9260, 248, 48, 38, 1, 112, 3,
+		20, "1280x1024", 1280, 1024, 1280, 1024, 8, 9260, 248, 48, 38, 1, 112, 3,
 		FB_SYNC_HOR_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT, FB_VMODE_NONINTERLACED | FB_VMODE_CONUPDATE
 	},
 	{
@@ -182,7 +182,7 @@ static struct Modes
 	},
 
 	{
-		16, "1080p60", 1920, 1080, 1920, 1080, 16, 6734, 148,  88, 36, 4, 44, 5,
+		16, "1080p", 1920, 1080, 1920, 1080, 16, 6734, 148,  88, 36, 4, 44, 5,
 		FB_SYNC_HOR_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT, FB_VMODE_CONUPDATE
 	},
 	{
@@ -216,7 +216,7 @@ static struct Modes
 	},
 
 	{
-		8, "1080i60", 1920, 1080, 1920, 1080, 16, 13468, 148,  88, 35, 5, 44, 5,
+		8, "1080i", 1920, 1080, 1920, 1080, 16, 13468, 148,  88, 35, 5, 44, 5,
 		FB_SYNC_HOR_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT, FB_VMODE_INTERLACED | FB_VMODE_CONUPDATE
 	},
 	{
@@ -229,7 +229,7 @@ static struct Modes
 	},
 
 	{
-		5, "720p60", 1280, 720, 1280, 720, 16, 13468, 220, 110, 20, 5, 40, 5,
+		5, "720p", 1280, 720, 1280, 720, 16, 13468, 220, 110, 20, 5, 40, 5,
 		FB_SYNC_HOR_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT, FB_VMODE_NONINTERLACED | FB_VMODE_CONUPDATE
 	},
 	{
@@ -242,24 +242,24 @@ static struct Modes
 	 */
 #ifdef alter_tobak
 	{
-		3, "576p50", 720, 576, 720, 576, 16, 37037, 68, 12, 34, 10, 64, 5, 0,
+		3, "576p", 720, 576, 720, 576, 16, 37037, 68, 12, 34, 10, 64, 5, 0,
 		FB_VMODE_NONINTERLACED | FB_VMODE_CONUPDATE
 	},
 #else
 	{
-		3, "576p50", 720, 576, 720, 576, 16, 37037, 68, 12, 39, 5, 64, 5, 0,
+		3, "576p", 720, 576, 720, 576, 16, 37037, 68, 12, 39, 5, 64, 5, 0,
 		FB_VMODE_NONINTERLACED | FB_VMODE_CONUPDATE
 	},
 #endif
 
 #ifdef alter_tobak
 	{
-		2, "576i50", 720, 576, 720, 576, 16, 74074, 68, 12, 38, 5, 64, 6, 0,
+		2, "576i", 720, 576, 720, 576, 16, 74074, 68, 12, 38, 5, 64, 6, 0,
 		FB_VMODE_INTERLACED | FB_VMODE_CONUPDATE
 	},
 #else
 	{
-		2, "576i50", 720, 576, 720, 576, 16, 74074, 69, 12, 41, 5, 63, 3, 0,
+		2, "576i", 720, 576, 720, 576, 16, 74074, 69, 12, 41, 5, 63, 3, 0,
 		FB_VMODE_INTERLACED | FB_VMODE_CONUPDATE
 	},
 #endif
@@ -1390,9 +1390,9 @@ int proc_video_videomode_choices_read(char* page, char** start, off_t off, int c
 #endif
 #if 0
 #if defined(UFS912) || defined(SPARK) || defined (SPARK7162) || defined(ATEVIO7500) || defined(HS7110) || defined(HS7810A) || defined(HS7119) || defined(HS7819) || defined(ATEMIO520) || defined(ATEMIO530) || defined(UFS913) || defined(VITAMIN_HD5000) || defined(SAGEMCOM88)
-	len = sprintf(page, "pal 1080i50 720p50 576p50 576i50 1080i60 720p60 1080p24 1080p25 1080p30 1080p50 1080p59 1080p60 PC\n");
+	len = sprintf(page, "pal 1080i50 720p50 576p50 576i 1080i 720p 1080p24 1080p25 1080p30 1080p50 1080p59 1080p PC\n");
 #else
-	len = sprintf(page, "pal 1080i50 720p50 576p50 576i50 1080i60 720p60 1080p24 1080p25 1080p30 PC\n");
+	len = sprintf(page, "pal 1080i50 720p50 576p50 576i 1080i 720p 1080p24 1080p25 1080p30 PC\n");
 #endif
 #else
 	/* BPanther: shows real resolution support by player (receiver independent) */
