diff -Naur git/player2_191/linux/Makefile git_new/player2_191/linux/Makefile
--- git/player2_191/linux/Makefile	2015-03-13 13:08:06.988671910 +0100
+++ git_new/player2_191/linux/Makefile	2015-03-13 13:07:34.429302675 +0100
@@ -158,7 +158,6 @@
 
 CCFLAGSY += -I$(CONFIG_KERNEL_BUILD)/include2 \
 	-I$(CONFIG_KERNEL_BUILD)/include \
-	-include $(KERNEL_LOCATION)/include/linux/autoconf.h \
 	-I$(CONFIG_KERNEL_PATH)/include -D__KERNEL__ \
 	-I$(DRIVER_TOPDIR)/include \
 	-I$(DRIVER_TOPDIR)/include/stmfb \

