diff --git a/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_video.c b/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_video.c
index 9f8a19d..82beca6 100644
--- a/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_video.c
+++ b/player2_191/linux/drivers/media/dvb/stm/dvb/e2_proc/e2_proc_video.c
@@ -148,7 +148,7 @@ static struct Modes
 {
 
 	{
-		25, "1024x768_60", 1024, 768, 1024, 768, 8, 15385, 160, 24, 29, 3, 136, 6,
+		25, "1024x768", 1024, 768, 1024, 768, 8, 15385, 160, 24, 29, 3, 136, 6,
 		FB_VMODE_NONINTERLACED | FB_VMODE_CONUPDATE
 	},
 	{
@@ -169,7 +169,7 @@ static struct Modes
 	},
 
 	{
-		20, "1280x1024_60", 1280, 1024, 1280, 1024, 8, 9260, 248, 48, 38, 1, 112, 3,
+		20, "1280x1024", 1280, 1024, 1280, 1024, 8, 9260, 248, 48, 38, 1, 112, 3,
 		FB_SYNC_HOR_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT, FB_VMODE_NONINTERLACED | FB_VMODE_CONUPDATE
 	},
 	{
@@ -187,7 +187,7 @@ static struct Modes
 	},
 
 	{
-		16, "1080p60", 1920, 1080, 1920, 1080, 16, 6734, 148, 88, 36, 4, 44, 5,
+		16, "1080p", 1920, 1080, 1920, 1080, 16, 6734, 148, 88, 36, 4, 44, 5,
 		FB_SYNC_HOR_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT, FB_VMODE_CONUPDATE
 	},
 	{
@@ -221,7 +221,7 @@ static struct Modes
 	},
 
 	{
-		8, "1080i60", 1920, 1080, 1920, 1080, 16, 13468, 148, 88, 35, 5, 44, 5,
+		8, "1080i", 1920, 1080, 1920, 1080, 16, 13468, 148, 88, 35, 5, 44, 5,
 		FB_SYNC_HOR_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT, FB_VMODE_INTERLACED | FB_VMODE_CONUPDATE
 	},
 	{
@@ -234,7 +234,7 @@ static struct Modes
 	},
 
 	{
-		5, "720p60", 1280, 720, 1280, 720, 16, 13468, 220, 110, 20, 5, 40, 5,
+		5, "720p", 1280, 720, 1280, 720, 16, 13468, 220, 110, 20, 5, 40, 5,
 		FB_SYNC_HOR_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT, FB_VMODE_NONINTERLACED | FB_VMODE_CONUPDATE
 	},
 	{
@@ -247,24 +247,24 @@ static struct Modes
 	 */
 #ifdef alter_tobak
 	{
-		3, "576p50", 720, 576, 720, 576, 16, 37037, 68, 12, 34, 10, 64, 5, 0,
+		3, "576p", 720, 576, 720, 576, 16, 37037, 68, 12, 34, 10, 64, 5, 0,
 		FB_VMODE_NONINTERLACED | FB_VMODE_CONUPDATE
 	},
 #else
 	{
-		3, "576p50", 720, 576, 720, 576, 16, 37037, 68, 12, 39, 5, 64, 5, 0,
+		3, "576p", 720, 576, 720, 576, 16, 37037, 68, 12, 39, 5, 64, 5, 0,
 		FB_VMODE_NONINTERLACED | FB_VMODE_CONUPDATE
 	},
 #endif
 
 #ifdef alter_tobak
 	{
-		2, "576i50", 720, 576, 720, 576, 16, 74074, 68, 12, 38, 5, 64, 6, 0,
+		2, "576i", 720, 576, 720, 576, 16, 74074, 68, 12, 38, 5, 64, 6, 0,
 		FB_VMODE_INTERLACED | FB_VMODE_CONUPDATE
 	},
 #else
 	{
-		2, "576i50", 720, 576, 720, 576, 16, 74074, 69, 12, 41, 5, 63, 3, 0,
+		2, "576i", 720, 576, 720, 576, 16, 74074, 69, 12, 41, 5, 63, 3, 0,
 		FB_VMODE_INTERLACED | FB_VMODE_CONUPDATE
 	},
 #endif
