*** mt7610u_wifi_sta_v3002_dpo_20130916/include/os/rt_linux.h	2013-09-16 16:03:08.000000000 +0200
--- mt7610u_wifi_sta_v3002_dpo_20130916/include/os/rt_linux.h	2014-05-01 11:31:38.057952000 +0200
*************** typedef struct usb_ctrlrequest devctrlre
*** 127,139 ****
  #ifdef CONFIG_STA_SUPPORT
  
  #ifdef RTMP_MAC_USB
! #define STA_PROFILE_PATH			"/etc/Wireless/RT2870STA/RT2870STA.dat"
  #define STA_DRIVER_VERSION			"3.0.0.2"
  #ifdef SINGLE_SKU_V2
! #define SINGLE_SKU_TABLE_FILE_NAME	"/etc/Wireless/RT2870STA/SingleSKU.dat"
  #endif /* SINGLE_SKU_V2 */
  #ifdef MULTIPLE_CARD_SUPPORT
! #define CARD_INFO_PATH			"/etc/Wireless/RT2870STA/RT2870STACard.dat"
  #endif /* MULTIPLE_CARD_SUPPORT */
  #endif /* RTMP_MAC_USB */
  
--- 127,139 ----
  #ifdef CONFIG_STA_SUPPORT
  
  #ifdef RTMP_MAC_USB
! #define STA_PROFILE_PATH			"/etc/Wireless/mt7610uSTA/mt7610uSTA.dat"
  #define STA_DRIVER_VERSION			"3.0.0.2"
  #ifdef SINGLE_SKU_V2
! #define SINGLE_SKU_TABLE_FILE_NAME	"/etc/Wireless/mt7610uSTA/SingleSKU.dat"
  #endif /* SINGLE_SKU_V2 */
  #ifdef MULTIPLE_CARD_SUPPORT
! #define CARD_INFO_PATH			"/etc/Wireless/mt7610uSTA/mt7610uSTACard.dat"
  #endif /* MULTIPLE_CARD_SUPPORT */
  #endif /* RTMP_MAC_USB */
  
*************** extern int rausb_control_msg(VOID *dev,
*** 1520,1526 ****
  #ifdef RTMP_MAC_USB
  #ifdef CONFIG_STA_SUPPORT
  #undef EEPROM_BIN_FILE_NAME /* Avoid APSTA mode re-define issue */
! #define EEPROM_BIN_FILE_NAME  "/etc/Wireless/RT2870STA/e2p.bin"
  #endif /* CONFIG_STA_SUPPORT */
  #endif /* RTMP_MAC_USB */
  
--- 1520,1526 ----
  #ifdef RTMP_MAC_USB
  #ifdef CONFIG_STA_SUPPORT
  #undef EEPROM_BIN_FILE_NAME /* Avoid APSTA mode re-define issue */
! #define EEPROM_BIN_FILE_NAME  "/etc/Wireless/mt7610uSTA/e2p.bin"
  #endif /* CONFIG_STA_SUPPORT */
  #endif /* RTMP_MAC_USB */
  
*** mt7610u_wifi_sta_v3002_dpo_20130916/include/os/rt_drv.h	2013-09-16 16:03:08.000000000 +0200
--- mt7610u_wifi_sta_v3002_dpo_20130916/include/os/rt_drv.h	2014-05-01 11:31:55.837952000 +0200
*************** typedef VOID	pregs;
*** 56,65 ****
  #ifdef CONFIG_STA_SUPPORT
  
  #ifdef RTMP_MAC_USB
! #define STA_PROFILE_PATH			"/etc/Wireless/RT2870STA/RT2870STA.dat"
  #define STA_DRIVER_VERSION			"3.0.0.2"
  #ifdef MULTIPLE_CARD_SUPPORT
! #define CARD_INFO_PATH			"/etc/Wireless/RT2870STA/RT2870STACard.dat"
  #endif /* MULTIPLE_CARD_SUPPORT */
  #endif /* RTMP_MAC_USB */
  
--- 56,65 ----
  #ifdef CONFIG_STA_SUPPORT
  
  #ifdef RTMP_MAC_USB
! #define STA_PROFILE_PATH			"/etc/Wireless/mt7610uSTA/mt7610uSTA.dat"
  #define STA_DRIVER_VERSION			"3.0.0.2"
  #ifdef MULTIPLE_CARD_SUPPORT
! #define CARD_INFO_PATH			"/etc/Wireless/mt7610uSTA/mt7610uSTACard.dat"
  #endif /* MULTIPLE_CARD_SUPPORT */
  #endif /* RTMP_MAC_USB */
  
*************** extern int rausb_control_msg(VOID *dev,
*** 986,992 ****
  #ifdef RTMP_MAC_USB
  #ifdef CONFIG_STA_SUPPORT
  #undef EEPROM_BIN_FILE_NAME /* Avoid APSTA mode re-define issue */
! #define EEPROM_BIN_FILE_NAME  "/etc/Wireless/RT2870STA/e2p.bin"
  #endif /* CONFIG_STA_SUPPORT */
  #endif /* RTMP_MAC_USB */
  
--- 986,992 ----
  #ifdef RTMP_MAC_USB
  #ifdef CONFIG_STA_SUPPORT
  #undef EEPROM_BIN_FILE_NAME /* Avoid APSTA mode re-define issue */
! #define EEPROM_BIN_FILE_NAME  "/etc/Wireless/mt7610uSTA/e2p.bin"
  #endif /* CONFIG_STA_SUPPORT */
  #endif /* RTMP_MAC_USB */
  
