*** mt7610u_wifi_sta_v3001_dpo_20130725/common/rtusb_dev_id.c	2013-06-27 09:37:05.000000000 +0200
--- mt7610u_wifi_sta_v3001_dpo_20130725/common/rtusb_dev_id.c	2013-07-14 16:38:13.879855365 +0200
*************** USB_DEVICE_ID rtusb_dev_id[] = {
*** 40,45 ****
--- 40,48 ----
  	{USB_DEVICE(0x0E8D,0x7610)}, /* MT7610U */
  	{USB_DEVICE_AND_INTERFACE_INFO(0x0E8D, 0x7630, 0xff, 0x2, 0xff)}, /* MT7630U */
  	{USB_DEVICE_AND_INTERFACE_INFO(0x0E8D, 0x7650, 0xff, 0x2, 0xff)}, /* MT7650U */
+ 	{USB_DEVICE(0x13b1,0x003e)}, /* Cisco Linksys AE6000 */
+	{USB_DEVICE(0x2001,0x3d02)}, /* D-Link DWA-171 B1 */
+	{USB_DEVICE(0x148f,0x761a)}, /* TP-LINK Archer T2U */
  #endif
  	{ }/* Terminating entry */
  };
