--- linux-2.6.18-orig/include/asm-mips/unistd.h	2015-03-29 00:56:05.026515568 +0100
+++ linux-2.6.18-orig/include/asm-mips/unistd.h	2015-03-29 00:56:04.938515570 +0100
@@ -333,6 +333,7 @@
 #define __NR_get_robust_list		(__NR_Linux + 310)
 #define __NR_ioprio_set			(__NR_Linux + 314)
 #define __NR_ioprio_get			(__NR_Linux + 315)
+#define __NR_utimensat			(__NR_Linux + 316)
 
 /*
  * Offset of the last Linux o32 flavoured syscall
@@ -622,6 +623,7 @@
 #define __NR_get_robust_list		(__NR_Linux + 269)
 #define __NR_ioprio_set			(__NR_Linux + 273)
 #define __NR_ioprio_get			(__NR_Linux + 274)
+#define __NR_utimensat			(__NR_Linux + 275)
 
 /*
  * Offset of the last Linux 64-bit flavoured syscall
@@ -915,6 +917,7 @@
 #define __NR_get_robust_list		(__NR_Linux + 273)
 #define __NR_ioprio_set			(__NR_Linux + 277)
 #define __NR_ioprio_get			(__NR_Linux + 278)
+#define __NR_utimensat			(__NR_Linux + 279)
 
 /*
  * Offset of the last N32 flavoured syscall
