--- a/../rc	2013-05-07 15:04:31.964626327 +0100
+++ b/../rc	2013-05-07 15:04:20.388625248 +0100
@@ -26,8 +26,8 @@
         progress=$progress_size
     fi
     #echo "PROGRESS is $progress $runlevel $first_step + ($step of $num_steps) $step_change $progress_size"
-    if [ $progress -gt 0 -a -e /proc/progress ]; then
-        echo $(($progress / 2)) > /proc/progress
+    if [ $progress -gt 0 -a -e /usr/bin/displayvfd ]; then
+        displayvfd -s 18 -t "Booting $progress"
     fi
 }
 
