--- djmount-0.71/libupnp/m4/rt_bool_arg_enable.m4.orig	2006-08-27 22:12:21.000000000 +0200
+++ djmount-0.71/libupnp/m4/rt_bool_arg_enable.m4	2012-04-20 11:15:34.184722609 +0200
@@ -34,7 +34,7 @@
 		               )])
 	test "x$enable_[$1]" != [x]m4_if([$2],[yes],[no],[yes]) dnl
 		&& enable_[]Name=[$2]
-	AC_MSG_RESULT($enable_[]Name)dnl
+	AC_MSG_RESULT($enable_[]Name)
 	AM_CONDITIONAL([ENABLE_]NAME, test x"$enable_[]Name" = xyes)
 	dnl
 	m4_popdef([NAME])dnl
