From fa0a0e83333b5f3baa6ab5a09b23db267970a734 Mon Sep 17 00:00:00 2001
From: skyjet18 <57456827+skyjet18@users.noreply.github.com>
Date: Tue, 2 Apr 2024 11:25:39 +0200
Subject: [PATCH 04/33] added possibility to set decryption keys directly in
 HLS

---
 libavformat/hls.c | 49 ++++++++++++++++++++++++++++++++++-------------
 1 file changed, 36 insertions(+), 13 deletions(-)

diff --git a/libavformat/hls.c b/libavformat/hls.c
index f5f549b24d..ab8772771f 100644
--- a/libavformat/hls.c
+++ b/libavformat/hls.c
@@ -1295,17 +1295,32 @@ static int open_input(HLSContext *c, struct playlist *pls, struct segment *seg,
 
     if (seg->key_type == KEY_AES_128 || seg->key_type == KEY_SAMPLE_AES) {
         if (strcmp(seg->key, pls->key_url)) {
-            AVIOContext *pb = NULL;
-            if (open_url(pls->parent, &pb, seg->key, &c->avio_opts, opts, NULL) == 0) {
-                ret = avio_read(pb, pls->key, sizeof(pls->key));
-                if (ret != sizeof(pls->key)) {
-                    av_log(pls->parent, AV_LOG_ERROR, "Unable to read key file %s\n",
-                           seg->key);
+            if (strncmp(seg->key, "key://", 6) == 0)
+            {
+                if (strlen(seg->key+6) != sizeof(pls->key)*2)
+                {
+                    av_log(pls->parent, AV_LOG_ERROR, "Failed to parse key %s - wrong key length\n",
+                        seg->key+6);
+                }
+                else
+                {
+                    ff_hex_to_data(pls->key, seg->key+6);
+                }
+            }
+            else if (strncmp(seg->key, "keys://", 7) != 0) // "keys://" scheme contains multiple kid=key pairs and it will be handled later
+            {
+                AVIOContext *pb = NULL;
+                if (open_url(pls->parent, &pb, seg->key, &c->avio_opts, opts, NULL) == 0) {
+                    ret = avio_read(pb, pls->key, sizeof(pls->key));
+                    if (ret != sizeof(pls->key)) {
+                        av_log(pls->parent, AV_LOG_ERROR, "Unable to read key file %s\n",
+                            seg->key);
+                    }
+                    ff_format_io_close(pls->parent, &pb);
+                } else {
+                    av_log(pls->parent, AV_LOG_ERROR, "Unable to open key file %s\n",
+                        seg->key);
                 }
-                ff_format_io_close(pls->parent, &pb);
-            } else {
-                av_log(pls->parent, AV_LOG_ERROR, "Unable to open key file %s\n",
-                       seg->key);
             }
             av_strlcpy(pls->key_url, seg->key, sizeof(pls->key_url));
         }
@@ -2121,9 +2136,17 @@ static int hls_read_header(AVFormatContext *s)
 
         if (seg && seg->key_type == KEY_SAMPLE_AES) {
             if (strstr(in_fmt->name, "mov")) {
-                char key[33];
-                ff_data_to_hex(key, pls->key, sizeof(pls->key), 0);
-                av_dict_set(&options, "decryption_key", key, 0);
+                if (strncmp(pls->key_url, "keys://", 7) == 0)
+                {
+                    //we have special URL with encoded kid=key pairs
+                    av_dict_set(&options, "decryption_keys", pls->key_url+7, 0);
+                }
+                else
+                {
+                    char key[33];
+                    ff_data_to_hex(key, pls->key, sizeof(pls->key), 0);
+                    av_dict_set(&options, "decryption_key", key, 0);
+                }
             } else if (!c->crypto_ctx.aes_ctx) {
                 c->crypto_ctx.aes_ctx = av_aes_alloc();
                 if (!c->crypto_ctx.aes_ctx) {
-- 
2.45.2.windows.1

