From 0ef5d56d3d4698f43b338d99190bf3b357cba0b3 Mon Sep 17 00:00:00 2001
From: skyjet18 <57456827+skyjet18@users.noreply.github.com>
Date: Sat, 15 Feb 2025 19:00:00 +0100
Subject: [PATCH 22/33] dashdec: allow setting restart in reopen_demux_*

---
 libavformat/dashdec.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libavformat/dashdec.c b/libavformat/dashdec.c
index 19c4b67342..d1e1612773 100644
--- a/libavformat/dashdec.c
+++ b/libavformat/dashdec.c
@@ -2354,9 +2354,9 @@ static int dash_read_packet(AVFormatContext *s, AVPacket *pkt)
         if (cur->is_restart_needed) {
             cur->cur_seg_offset = 0;
             cur->init_sec_buf_read_offset = 0;
+            cur->is_restart_needed = 0;
             ff_format_io_close(cur->parent, &cur->input);
             ret = reopen_demux_for_component(s, cur);
-            cur->is_restart_needed = 0;
         }
     }
     return AVERROR_EOF;
-- 
2.45.2.windows.1

