From 685078b82f0d029fddac9eafaed7d9c50023b634 Mon Sep 17 00:00:00 2001
From: skyjet18 <57456827+skyjet18@users.noreply.github.com>
Date: Tue, 9 Sep 2025 11:08:08 +0200
Subject: [PATCH 32/33] dashdec: fixed splitted subtitles

---
 libavformat/dashdec.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libavformat/dashdec.c b/libavformat/dashdec.c
index f1ff16764c..97a6f31cf9 100644
--- a/libavformat/dashdec.c
+++ b/libavformat/dashdec.c
@@ -896,7 +896,7 @@ static int parse_manifest_representation(AVFormatContext *s, const char *url,
 
     rep->parent = s;
     rep->type = type;
-    rep->first_pts = INT64_MIN;
+    rep->first_pts = type == AVMEDIA_TYPE_SUBTITLE ? INT64_C(0) : INT64_MIN; // PTS correction is needed only for non subtitle stream
     representation_segmenttemplate_node = find_child_node_by_name(representation_node, "SegmentTemplate");
     representation_baseurl_node = find_child_node_by_name(representation_node, "BaseURL");
     representation_segmentlist_node = find_child_node_by_name(representation_node, "SegmentList");
-- 
2.45.2.windows.1

