--- a/os_dep/linux/rtw_android.c	2015-04-16 15:51:50.593424725 +0200
+++ b/os_dep/linux/rtw_android.c	2015-04-16 15:54:33.789421144 +0200
@@ -230,7 +230,11 @@
 {
 	int cmd_num;
 	for(cmd_num=0 ; cmd_num<ANDROID_WIFI_CMD_MAX; cmd_num++)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 0, 0))
+		if(0 == strncasecmp(cmdstr , android_wifi_cmd_str[cmd_num], strlen(android_wifi_cmd_str[cmd_num])) )
+#else
 		if(0 == strnicmp(cmdstr , android_wifi_cmd_str[cmd_num], strlen(android_wifi_cmd_str[cmd_num])) )
+#endif
 			break;
 		
 	return cmd_num;
