--- a/Source/JavaScriptCore/wtf/gobject/GRefPtr.h
+++ b/Source/JavaScriptCore/wtf/gobject/GRefPtr.h
@@ -30,9 +30,6 @@
 #include "RefPtr.h"
 #include <algorithm>
 
-extern "C" void g_object_unref(gpointer);
-extern "C" gpointer g_object_ref_sink(gpointer);
-
 namespace WTF {
 
 enum GRefPtrAdoptType { GRefPtrAdopt };
--- a/Source/WebCore/accessibility/gtk/AccessibilityObjectWrapperAtk.cpp
+++ b/Source/WebCore/accessibility/gtk/AccessibilityObjectWrapperAtk.cpp
@@ -835,7 +835,7 @@ static void webkit_accessible_class_init(AtkObjectClass* klass)
 GType
 webkit_accessible_get_type(void)
 {
-    static volatile gsize type_volatile = 0;
+    static gsize type_volatile = 0;
 
     if (g_once_init_enter(&type_volatile)) {
         static const GTypeInfo tinfo = {
--- a/Source/WebCore/accessibility/gtk/WebKitAccessibleHyperlink.cpp
+++ b/Source/WebCore/accessibility/gtk/WebKitAccessibleHyperlink.cpp
@@ -363,7 +363,7 @@ static void webkitAccessibleHyperlinkInit(AtkHyperlink* link)
 
 GType webkitAccessibleHyperlinkGetType()
 {
-    static volatile gsize typeVolatile = 0;
+    static gsize typeVolatile = 0;
 
     if (g_once_init_enter(&typeVolatile)) {
         static const GTypeInfo tinfo = {
