--- a/os_dep/linux/os_intfs.c	2018-10-23 20:30:24.385451383 +0200
+++ b/os_dep/linux/os_intfs.c	2018-10-23 20:35:14.283225781 +0200
@@ -786,16 +786,20 @@
 
 	return dscp >> 5;
 }
-
- 
+#if (LINUX_VERSION_CODE>=KERNEL_VERSION(4,19,0))
+static u16 rtw_select_queue(struct net_device *dev, struct sk_buff *skb,
+				struct net_device *sb_dev,
+				select_queue_fallback_t fallback
+#else
 static u16 rtw_select_queue(struct net_device *dev, struct sk_buff *skb
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 13, 0) 	
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 13, 0)
 				, void *accel_priv
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 14, 0) 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 14, 0)
 				, select_queue_fallback_t fallback
 #endif
 
 #endif
+#endif
 )
 {
 	_adapter	*padapter = rtw_netdev_priv(dev);
