diff --git a/os_dep/osdep_service.c b/os_dep/osdep_service.c
index 2f44420ac003..4ad225db6438 100644
--- a/os_dep/osdep_service.c
+++ b/os_dep/osdep_service.c
@@ -1946,7 +1946,13 @@ static int isFileReadable(char *path)
 		ret = PTR_ERR(fp);
 	}
 	else {
-		oldfs = get_fs(); set_fs(get_ds());
+		oldfs = get_fs(); set_fs(
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,1,0))
+				KERNEL_DS
+#else
+				get_ds()
+#endif
+				);
 		
 		if(1!=readFile(fp, &buf, 1))
 			ret = PTR_ERR(fp);
@@ -1974,7 +1980,13 @@ static int retriveFromFile(char *path, u8* buf, u32 sz)
 		if( 0 == (ret=openFile(&fp,path, O_RDONLY, 0)) ){
 			DBG_871X("%s openFile path:%s fp=%p\n",__FUNCTION__, path ,fp);
 
-			oldfs = get_fs(); set_fs(get_ds());
+			oldfs = get_fs(); set_fs(
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,1,0))
+					KERNEL_DS
+#else
+					get_ds()
+#endif
+					);
 			ret=readFile(fp, buf, sz);
 			set_fs(oldfs);
 			closeFile(fp);
@@ -2008,7 +2020,13 @@ static int storeToFile(char *path, u8* buf, u32 sz)
 		if( 0 == (ret=openFile(&fp, path, O_CREAT|O_WRONLY, 0666)) ) {
 			DBG_871X("%s openFile path:%s fp=%p\n",__FUNCTION__, path ,fp);
 
-			oldfs = get_fs(); set_fs(get_ds());
+			oldfs = get_fs(); set_fs(
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,1,0))
+					KERNEL_DS
+#else
+					get_ds()
+#endif
+					);
 			ret=writeFile(fp, buf, sz);
 			set_fs(oldfs);
 			closeFile(fp);
