From f06dfac2ae83c698be59da6bcff982ba4fe2efbd Mon Sep 17 00:00:00 2001
From: captain <openatv@gmail.com>
Date: Fri, 5 May 2017 15:34:11 +0200
Subject: [PATCH] fix secfault w use wrong line_length

---
 tuxtxt_common.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tuxtxt_common.h b/tuxtxt_common.h
index 871e8ce..308d02b 100644
--- a/tuxtxt_common.h
+++ b/tuxtxt_common.h
@@ -4183,12 +4183,12 @@ void tuxtxt_setfontwidth(tstRenderInfo* renderinfo,int newwidth)
 }
 void tuxtxt_ClearBB(tstRenderInfo* renderinfo,int color)
 {
-	tuxtxt_FillRect(renderinfo->lfb,renderinfo->fix_screeninfo.line_length,0, renderinfo->var_screeninfo.yres - renderinfo->var_screeninfo.yoffset, renderinfo->fix_screeninfo.line_length, renderinfo->var_screeninfo.yres, color);
+	tuxtxt_FillRect(renderinfo->lfb,renderinfo->fix_screeninfo.line_length,0, renderinfo->var_screeninfo.yres - renderinfo->var_screeninfo.yoffset, renderinfo->var_screeninfo.xres, renderinfo->var_screeninfo.yres, color);
 }
 
 void tuxtxt_ClearFB(tstRenderInfo* renderinfo,int color)
 {
-	tuxtxt_FillRect(renderinfo->lfb,renderinfo->fix_screeninfo.line_length,0, renderinfo->var_screeninfo.yoffset, renderinfo->fix_screeninfo.line_length, renderinfo->var_screeninfo.yres, color);
+	tuxtxt_FillRect(renderinfo->lfb,renderinfo->fix_screeninfo.line_length,0, renderinfo->var_screeninfo.yoffset, renderinfo->var_screeninfo.xres, renderinfo->var_screeninfo.yres, color);
 }
 
 int  tuxtxt_GetCurFontWidth(tstRenderInfo* renderinfo)
-- 
2.12.0.windows.1

