diff --git a/tuxtxt_common.h b/tuxtxt_common.h
index 7bdf81e..3918d8b 100644
--- a/tuxtxt_common.h
+++ b/tuxtxt_common.h
@@ -4878,7 +4878,7 @@ void tuxtxt_CopyBB2FB(tstRenderInfo* renderinfo)
 		topdst += screenwidth;
 		for (i=0; i < 24*renderinfo->fontheight; i++)
 		{
-			memcpy(topdst, topsrc,renderinfo->ex-screenwidth);
+			memcpy(topdst, topsrc, screenwidth - renderinfo->ex);
 			topdst += renderinfo->fix_screeninfo.line_length;
 			topsrc += renderinfo->fix_screeninfo.line_length;
 		}
