From 9ac81609c9da3094716dd0d7a4027dc98d56b5de Mon Sep 17 00:00:00 2001
From: Nicker <nickersk@gmail.com>
Date: Mon, 30 Jul 2018 07:41:54 +0200
Subject: [PATCH 14/15] makefile-silence packed-not-aligned warn


diff --git a/Makefile b/Makefile
index 3e711514..73b49c98 100644
--- a/Makefile
+++ b/Makefile
@@ -577,6 +577,7 @@ endif
 # This warning generated too much noise in a regular build.
 # Use make W=1 to enable this warning (see scripts/Makefile.build)
 KBUILD_CFLAGS += $(call cc-disable-warning, unused-but-set-variable)
+KBUILD_CFLAGS += $(call cc-disable-warning, packed-not-aligned)
 
 ifdef CONFIG_FRAME_POINTER
 KBUILD_CFLAGS	+= -fno-omit-frame-pointer -fno-optimize-sibling-calls
-- 
2.17.1

