--- a/drivers/usb/storage/isd200.c	2016-09-20 12:33:09.860428026 +0200
+++ b/drivers/usb/storage/isd200.c	2016-09-20 12:33:58.764664894 +0200
@@ -61,7 +61,7 @@
 #include "scsiglue.h"
 
 MODULE_DESCRIPTION("Driver for In-System Design, Inc. ISD200 ASIC");
-MODULE_AUTHOR("Björn Stenberg <bjorn@haxx.se>");
+MODULE_AUTHOR("Bjorn Stenberg <bjorn@haxx.se>");
 MODULE_LICENSE("GPL");
 
 static int isd200_Initialization(struct us_data *us);
