diff --git a/libopkg/opkg_install.c b/libopkg/opkg_install.c
index 70b6710..769c67f 100644
--- a/libopkg/opkg_install.c
+++ b/libopkg/opkg_install.c
@@ -226,7 +226,15 @@ static int prerm_upgrade_old_pkg(pkg_t * pkg, pkg_t * old_pkg)
 
     sprintf_alloc(&script_args, "upgrade %s", new_version);
     free(new_version);
-    err = pkg_run_script(old_pkg, "prerm", script_args);
+    // Busybox needs special care: Do not execute the prerm script
+    // of busybox or all following packages will fail
+    if ((strlen(old_pkg->name) >= 7) && (strncmp(old_pkg->name, "busybox", 7) == 0)) {
+        opkg_msg(NOTICE, "Skipping prerm script for package \"%s\"\n",
+                 old_pkg->name);
+        err = 0;
+    } else {
+        err = pkg_run_script(old_pkg, "prerm", script_args);
+    }
     free(script_args);
     if (err != 0) {
         opkg_msg(NOTICE, "prerm script for package \"%s\" failed\n",
@@ -637,7 +645,15 @@ static int postrm_upgrade_old_pkg(pkg_t * pkg, pkg_t * old_pkg)
 
     sprintf_alloc(&script_args, "upgrade %s", new_version);
     free(new_version);
-    err = pkg_run_script(old_pkg, "postrm", script_args);
+    // Busybox needs special care: Do not execute the postrm script
+    // of busybox or all following packages will fail
+    if ((strlen(old_pkg->name) >= 7) && (strncmp(old_pkg->name, "busybox", 7) == 0)) {
+        opkg_msg(NOTICE, "Skipping postrm script for package \"%s\"\n",
+                 old_pkg->name);
+        err = 0;
+    } else {
+        err = pkg_run_script(old_pkg, "postrm", script_args);
+    }
     free(script_args);
     if (err != 0) {
         opkg_msg(NOTICE, "postrm script for package \"%s\" failed\n",
