diff --git a/libopkg/pkg.c b/libopkg/pkg.c
index a196714..ca4d039 100644
--- a/libopkg/pkg.c
+++ b/libopkg/pkg.c
@@ -706,6 +706,7 @@ static void pkg_formatted_field(FILE * fp, pkg_t * pkg, const char *field, const
             if (pkg->provides_count > 1) {
                 fprintf(fp, "Provides:");
                 for (i = 1; i < pkg->provides_count; i++) {
+                    if (!pkg->provides[i]) break;
                     fprintf(fp, "%s %s", i == 1 ? "" : ",",
                             pkg->provides[i]->name);
                 }
