diff --git a/libopkg/opkg_install.c b/libopkg/opkg_install.c
index 703f29c..6540249 100644
--- a/libopkg/opkg_install.c
+++ b/libopkg/opkg_install.c
@@ -450,9 +450,10 @@ static int check_data_file_clashes(pkg_t * pkg, pkg_t * old_pkg)
             pkg_t *owner;
             pkg_t *obs;
             int existing_is_dir = file_is_dir(filename);
+            int existing_is_lnk = file_is_symlink(filename);
 
             /* OK if both the existing file and new file are directories. */
-            if (existing_is_dir && S_ISDIR(file_info->mode)) {
+            if (existing_is_lnk || (existing_is_dir && S_ISDIR(file_info->mode))) {
                 continue;
             } else if (existing_is_dir || S_ISDIR(file_info->mode)) {
                 /* Can't mix directory and non-directory.  For normal files,
