--- a/libopkg/pkg_depends.c	2015-09-26 21:34:07.324075015 +0200
+++ b/libopkg/pkg_depends.c	2015-09-26 21:35:37.840073464 +0200
@@ -943,15 +943,15 @@
     /* should these be here to support deprecated designations; dpkg does */
     else if (strncmp(*str, "<", 1) == 0) {
         *str += 1;
-        opkg_msg(NOTICE,
+        /*opkg_msg(NOTICE,
                  "Deprecated version constraint '<' was used with the same "
-                 "meaning as '<='. Use '<<' for EARLIER constraint.\n");
+                 "meaning as '<='. Use '<<' for EARLIER constraint.\n");*/
         return EARLIER_EQUAL;
     } else if (strncmp(*str, ">", 1) == 0) {
         *str += 1;
-        opkg_msg(NOTICE,
+        /*opkg_msg(NOTICE,
                  "Deprecated version constraint '>' was used with the same "
-                 "meaning as '>='. Use '>>' for LATER constraint.\n");
+                 "meaning as '>='. Use '>>' for LATER constraint.\n");*/
         return LATER_EQUAL;
     } else {
         return NONE;
