--- a/libopkg/opkg_download.c	2016-04-10 10:41:23.317020992 +0200
+++ b/libopkg/opkg_download.c	2016-04-10 10:41:12.681021194 +0200
@@ -462,6 +462,6 @@
         return opkg_prepare_file_for_install(url, namep);
 
     /* Can't find anything matching the requested URL. */
-    opkg_msg(ERROR, "Couldn't find anything to satisfy '%s'.\n", url);
-    return -1;
+    opkg_msg(NOTICE, "Couldn't find anything to satisfy '%s'.\n", url);
+    return 0;
 }
