diff -Naur linux-2.6.34/include/linux/dvb/audio.h linux-2.6.34.new/include/linux/dvb/audio.h
--- linux-2.6.34/include/linux/dvb/audio.h	2010-05-16 16:17:36.000000000 -0500
+++ linux-2.6.34.new/include/linux/dvb/audio.h	2011-03-16 11:00:26.000000000 -0500
@@ -97,6 +97,29 @@
 #define AUDIO_CAP_OGG   64
 #define AUDIO_CAP_SDDS 128
 #define AUDIO_CAP_AC3  256
+/* extension of trident supported audio decoder capibility*/
+#define AUDIO_CAP_AAC_PLUS       (0x1<<9)
+#define AUDIO_CAP_DD_PLUS        (0x1<<10)
+#define AUDIO_CAP_AVS            (0x1<<11)
+#define AUDIO_CAP_MLP            (0x1<<12)
+#define AUDIO_CAP_WMA            (0x1<<13)
+#define AUDIO_CAP_REALAUDIO      (0x1<<14)
+
+/* extension of trident supported audio stream type*/
+#define AUDIO_STREAM_TYPE_MP1             1
+#define AUDIO_STREAM_TYPE_MP2             2
+#define AUDIO_STREAM_TYPE_MP3             3
+#define AUDIO_STREAM_TYPE_DOLBY_DIGITAL   4
+#define AUDIO_STREAM_TYPE_AAC             5
+#define AUDIO_STREAM_TYPE_AAC_PLUS        6
+#define AUDIO_STREAM_TYPE_DD_PLUS         7
+#define AUDIO_STREAM_TYPE_DTS             8
+#define AUDIO_STREAM_TYPE_AVS             9
+#define AUDIO_STREAM_TYPE_MLP             10
+#define AUDIO_STREAM_TYPE_WMA             11
+#define AUDIO_STREAM_TYPE_REALAUDIO       12
+#define AUDIO_STREAM_TYPE_RAW             13
+
 
 #define AUDIO_STOP                 _IO('o', 1)
 #define AUDIO_PLAY                 _IO('o', 2)
diff -Naur linux-2.6.34/include/linux/dvb/video.h linux-2.6.34.new/include/linux/dvb/video.h
--- linux-2.6.34/include/linux/dvb/video.h	2010-05-16 16:17:36.000000000 -0500
+++ linux-2.6.34.new/include/linux/dvb/video.h	2011-03-16 10:59:48.000000000 -0500
@@ -47,7 +47,25 @@
 	 VIDEO_SYSTEM_PALM,
 	 VIDEO_SYSTEM_NTSC60,
 	 VIDEO_SYSTEM_PAL60,
-	 VIDEO_SYSTEM_PALM60
+	 VIDEO_SYSTEM_PALM60,
+	 VIDEO_SYSTEM_480I,
+	 VIDEO_SYSTEM_480P,
+	 VIDEO_SYSTEM_720P,
+	 VIDEO_SYSTEM_720P_50Hz,
+	 VIDEO_SYSTEM_720P_5994Hz,
+	 VIDEO_SYSTEM_576P,
+	 VIDEO_SYSTEM_SECAM,
+	 VIDEO_SYSTEM_1080I_50Hz,
+	 VIDEO_SYSTEM_1080I_5994Hz,
+	 VIDEO_SYSTEM_1080I,
+	 VIDEO_SYSTEM_1080P_25Hz,
+	 VIDEO_SYSTEM_1080P_24Hz,
+	 VIDEO_SYSTEM_1080P_50Hz,
+	 VIDEO_SYSTEM_1080P_5994Hz,
+	 VIDEO_SYSTEM_1080P_60Hz,
+	 VIDEO_SYSTEM_1080P_30Hz,
+	 VIDEO_SYSTEM_1080P_2997Hz,
+	 VIDEO_SYSTEM_1080P_2397Hz
 } video_system_t;
 
 
@@ -216,6 +234,21 @@
 /* can the hardware decode MPEG1 and/or MPEG2? */
 #define VIDEO_CAP_MPEG1   1
 #define VIDEO_CAP_MPEG2   2
+/* extension of trident supported audio decoder capibility*/
+#define VIDEO_CAP_MPEG4      (0x1<<7)
+#define VIDEO_CAP_MPEG4PART2 (0x1<<8) /* MPEG4 SVH, MPEG4 SP, MPEG4 ASP, DIVX4,5,6 */
+#define VIDEO_CAP_H264       (0x1<<9)
+#define VIDEO_CAP_VC1        (0x1<<10)
+#define VIDEO_CAP_DIVX       (0x1<<11)
+#define VIDEO_CAP_REALVIDEO8 (0x1<<12)
+#define VIDEO_CAP_REALVIDEO9 (0x1<<13)
+#define VIDEO_CAP_ON2_VP6    (0x1<<14)
+#define VIDEO_CAP_ON2_VP8    (0x1<<15)
+#define VIDEO_CAP_SORENSON_SPARK (0x1<<15)
+#define VIDEO_CAP_H263       (0x1<<16)
+#define VIDEO_CAP_AVS        (0x1<<17)
+#define VIDEO_CAP_VIP        (0x1<<18)
+
 /* can you send a system and/or program stream to video device?
    (you still have to open the video and the audio device but only
     send the stream to the video device) */
@@ -227,6 +260,21 @@
 #define VIDEO_CAP_NAVI   32
 #define VIDEO_CAP_CSS    64
 
+/* extension of trident supported video stream type*/
+#define VIDEO_STREAM_TYPE_MPEG1           1
+#define VIDEO_STREAM_TYPE_MPEG2           2
+#define VIDEO_STREAM_TYPE_MPEG4           3
+#define VIDEO_STREAM_TYPE_MPEG4PART2      4
+#define VIDEO_STREAM_TYPE_VC1             6
+#define VIDEO_STREAM_TYPE_DIVX            7
+#define VIDEO_STREAM_TYPE_REALVIDEO8      8
+#define VIDEO_STREAM_TYPE_REALVIDEO9      9
+#define VIDEO_STREAM_TYPE_ON2_VP6         10
+#define VIDEO_STREAM_TYPE_ON2_VP8         11
+#define VIDEO_STREAM_TYPE_SORENSON_SPARK  12
+#define VIDEO_STREAM_TYPE_H263            13
+#define VIDEO_STREAM_TYPE_AVS             14
+#define VIDEO_STREAM_TYPE_VIP             15
 
 #define VIDEO_STOP                 _IO('o', 21)
 #define VIDEO_PLAY                 _IO('o', 22)
